{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38878718-d82a-491d-bbd9-f85cb5d7b9d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *\n",
    "from scripts.sm_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e5c56c1-915a-417e-9a43-2bc353d112f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3402305-1cd9-4d14-abb2-d243850e8dd6",
   "metadata": {},
   "source": [
    "# Goal: get all exons that are detected in one population that aren't in the other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6300f73b-d4ca-43bd-893c-12612f36aa98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get exon / transcript / novelty info from the table that\n",
    "# already ran\n",
    "df = pd.read_csv(proc_cfg(config['lr']['exon_info'],od), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44b48f8c-f269-4df3-a9df-77c9e000d7d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = load_meta()\n",
    "meta = meta.loc[meta.merged_run_mode==True]\n",
    "meta = meta.loc[meta.mixed_samples == False]\n",
    "pops = meta['population'].tolist()\n",
    "\n",
    "mt_df = pd.read_csv('../../data/05_mastertable/poder_master_table_fixed_genics.tsv', sep='\\t')\n",
    "# meta = load_meta()\n",
    "# meta = meta.loc[meta.merged_run_mode==True]\n",
    "# pops = \n",
    "_, pops = get_population_colors()\n",
    "mt_df = mt_df[['isoform']+pops]\n",
    "mt_df.rename({'isoform':'transcript_id'},\n",
    "             inplace=True,\n",
    "             axis=1)\n",
    "\n",
    "# set index and just say t/f\n",
    "mt_df = mt_df.set_index('transcript_id')\n",
    "mt_df = mt_df>0\n",
    "mt_df.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e1c519d6-812b-4ba5-8be7-cca8d6c4ae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge novelty info w/ detection info\n",
    "df = df.merge(mt_df, \n",
    "              how='left',\n",
    "              on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b362af26-0eba-40af-807d-666180872af7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>novelty</th>\n",
       "      <th>AJI</th>\n",
       "      <th>CEU</th>\n",
       "      <th>HAC</th>\n",
       "      <th>ITU</th>\n",
       "      <th>LWK</th>\n",
       "      <th>MPC</th>\n",
       "      <th>PEL</th>\n",
       "      <th>YRI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2_+_135133_135173</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000008.2_+_155429_155531</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000008.2_+_173515_173643</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000008.2_+_83859_84014</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000008.2_+_83926_84014</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          eid novelty   AJI    CEU    HAC    ITU    LWK  \\\n",
       "0  GL000008.2_+_135133_135173   Known  True  False  False  False  False   \n",
       "1  GL000008.2_+_155429_155531   Known  True  False  False   True  False   \n",
       "2  GL000008.2_+_173515_173643   Known  True  False  False   True  False   \n",
       "3    GL000008.2_+_83859_84014   Known  True  False  False  False  False   \n",
       "4    GL000008.2_+_83926_84014   Known  True   True  False   True  False   \n",
       "\n",
       "     MPC    PEL    YRI  \n",
       "0  False   True  False  \n",
       "1  False   True  False  \n",
       "2  False  False  False  \n",
       "3  False  False  False  \n",
       "4  False   True  False  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dedupe eids, take max. of each population to \n",
    "# see if it's det or not\n",
    "df.drop('transcript_id', axis=1, inplace=True)\n",
    "df = df.groupby(['eid', 'novelty']).max().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3db5168-0f53-4ac1-b057-f83fafabf6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each pair of populations, get the exons that are in them\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe24d6db-a8a0-4f98-94fe-6ead6e46d086",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc4f94d-e312-49f8-98c9-20c24916394b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "435aed3a-9fd7-47a7-9365-7d91679ce6b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c63b1d7d-39eb-40f8-96c2-742e75176f6f",
   "metadata": {},
   "source": [
    "## novelty and tid info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfa9c7b5-b615-4f99-8b58-5dcd8e1aaa76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = get_internal_exon_info(proc_cfg(config['lr']['gtf_filt_with_genes'], od),\n",
    "#                             proc_cfg(config['ref']['gtf'], od))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db51fff8-3b5f-48eb-a288-f193365b9ab1",
   "metadata": {},
   "source": [
    "## population detection parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "79bb2223-596a-46d8-918c-ad22e6727263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2267\n"
     ]
    }
   ],
   "source": [
    "# first, get all transcripts that are specific b/w 2 population\n",
    "mt_df = pd.read_csv('../../data/05_mastertable/29102024_PODER_mastertable.tsv', sep='\\t')\n",
    "mt_df['pop_spec_t'] = (mt_df.population_sharing==1)&\\\n",
    "                      (mt_df.sample_sharing>1)\n",
    "print(len(mt_df.loc[mt_df.pop_spec_t==True, 'isoform'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ba9d5bef-d24f-44ff-b77a-10d56fdcf04c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>AJI</th>\n",
       "      <th>CEU</th>\n",
       "      <th>HAC</th>\n",
       "      <th>ITU</th>\n",
       "      <th>LWK</th>\n",
       "      <th>MPC</th>\n",
       "      <th>PEL</th>\n",
       "      <th>YRI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000741425.1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript_0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>transcript_1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>transcript_10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transcript_11</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             isoform  AJI  CEU  HAC  ITU  LWK  MPC  PEL  YRI\n",
       "0  ENST00000741425.1    1    0    0    0    0    0    0    0\n",
       "1       transcript_0    1    0    0    0    0    0    0    0\n",
       "2       transcript_1    1    0    0    0    0    0    1    0\n",
       "3      transcript_10    1    1    0    1    0    0    0    0\n",
       "4      transcript_11    2    1    0    1    0    0    2    0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df = mt_df[['isoform', 'AJI', 'CEU',\n",
    "              'HAC', 'ITU', 'LWK',\n",
    "              'MPC', 'PEL', 'YRI']]\n",
    "t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8e3e5a8e-b2a2-4014-b985-181e004ba720",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>eid</th>\n",
       "      <th>Strand_gc</th>\n",
       "      <th>Start_gc</th>\n",
       "      <th>End_gc</th>\n",
       "      <th>eid_gc</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83926_84014_</td>\n",
       "      <td>+</td>\n",
       "      <td>83859</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83859_84014_</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83926_84014_</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84145</td>\n",
       "      <td>GL000008.2_+_83926_84145_</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83926_84014_</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83926_84014_</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>83926</td>\n",
       "      <td>84014</td>\n",
       "      <td>GL000008.2_+_83926_84014_</td>\n",
       "      <td>+</td>\n",
       "      <td>83952</td>\n",
       "      <td>84145</td>\n",
       "      <td>GL000008.2_+_83952_84145_</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>85456</td>\n",
       "      <td>85477</td>\n",
       "      <td>GL000008.2_+_85456_85477_</td>\n",
       "      <td>+</td>\n",
       "      <td>85442</td>\n",
       "      <td>85477</td>\n",
       "      <td>GL000008.2_+_85442_85477_</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Chromosome Strand  Start    End                        eid  \\\n",
       "0           0  GL000008.2      +  83926  84014  GL000008.2_+_83926_84014_   \n",
       "1           1  GL000008.2      +  83926  84014  GL000008.2_+_83926_84014_   \n",
       "2           2  GL000008.2      +  83926  84014  GL000008.2_+_83926_84014_   \n",
       "3           3  GL000008.2      +  83926  84014  GL000008.2_+_83926_84014_   \n",
       "4           4  GL000008.2      +  85456  85477  GL000008.2_+_85456_85477_   \n",
       "\n",
       "  Strand_gc  Start_gc  End_gc                     eid_gc  Overlap  \n",
       "0         +     83859   84014  GL000008.2_+_83859_84014_       88  \n",
       "1         +     83926   84145  GL000008.2_+_83926_84145_       88  \n",
       "2         +     83926   84014  GL000008.2_+_83926_84014_       88  \n",
       "3         +     83952   84145  GL000008.2_+_83952_84145_       62  \n",
       "4         +     85442   85477  GL000008.2_+_85442_85477_       21  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b309e2-bca1-4ac4-92ee-6ef53b1375ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
