{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b47887af-a739-4dd5-8b66-896d3819b3ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1aca1dc0-5a37-41df-a457-a0ebf6d8e009",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7cb8ee4d-89a8-41b7-afbf-52f76466845d",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../../snakemake/compare_external/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a79fff2c-ffbd-49a1-a3f3-d3c4aa13a898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get ic <-> tid\n",
    "df = pd.DataFrame()\n",
    "gtfs = [od+config['lr']['gtf_filt_with_genes']]\n",
    "sources = ['PODER']\n",
    "\n",
    "for f, s in zip(gtfs, sources):\n",
    "    temp = pr.read_gtf(f)\n",
    "    temp = cerberus.get_ic(temp)\n",
    "    temp['source'] = s\n",
    "    df = pd.concat([df, temp], axis=0)\n",
    "    \n",
    "ics_df = df.copy(deep=True)    \n",
    "ics_df['ic_id'] = ics_df.Chromosome.astype(str)+'_'+\\\n",
    "                  ics_df.Strand.astype(str)+'_'+\\\n",
    "                  ics_df.ic.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf099650-3bd6-4b55-a7d9-bdb08e533ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "ics = [od+config['gtex']['cerb']['ics'],\n",
    "       od+config['chess']['cerb']['ics'],\n",
    "       od+config['enc']['cerb']['ics'],\n",
    "       od+config['lr']['cerb']['ics'],\n",
    "       od+config['ref']['cerb']['ics'],\n",
    "       od+config['ref']['refseq']['cerb']['ics']]\n",
    "sources = ['GTEx', 'CHESS3', 'ENCODE4', 'PODER', 'GENCODE v47', 'RefSeq v110']\n",
    "for ic, source in zip(ics, sources):\n",
    "    temp = pd.read_csv(ic, sep='\\t')\n",
    "    \n",
    "    # remove monoexonic, add source, drop name\n",
    "    temp = temp.loc[temp.Coordinates != '-']\n",
    "    temp['source'] = source\n",
    "    temp = temp.drop('Name', axis=1)\n",
    "    \n",
    "    df = pd.concat([df, temp], axis=0)    \n",
    "temp = df.copy(deep=True)\n",
    "df = df.pivot_table(\n",
    "    index=['Strand', 'Coordinates', 'Chromosome'],\n",
    "    columns='source',\n",
    "    aggfunc=lambda x: True,\n",
    "    fill_value=False).reset_index()\n",
    "# df.columns = ['_'.join(col).strip() if col[1] else col[0] for col in df.columns.values]\n",
    "df.head()\n",
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c44eec8-42f0-4a89-9d20-de0513f6a075",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>source</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>CHESS3</th>\n",
       "      <th>ENCODE4</th>\n",
       "      <th>GENCODE v47</th>\n",
       "      <th>GTEx</th>\n",
       "      <th>PODER</th>\n",
       "      <th>RefSeq v110</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>+</td>\n",
       "      <td>10000-10669-10744-13745-14013-18875-18940-19252</td>\n",
       "      <td>KZ208910.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>+</td>\n",
       "      <td>10000-10669-10744-18875-18940-19252</td>\n",
       "      <td>KZ208910.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>+</td>\n",
       "      <td>10000-10669-10744-19252</td>\n",
       "      <td>KZ208910.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>+</td>\n",
       "      <td>10000-10669-10749-18875-18940-19252</td>\n",
       "      <td>KZ208910.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>+</td>\n",
       "      <td>10000-10669-10749-19252</td>\n",
       "      <td>KZ208910.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "source Strand                                      Coordinates  Chromosome  \\\n",
       "0           +  10000-10669-10744-13745-14013-18875-18940-19252  KZ208910.1   \n",
       "1           +              10000-10669-10744-18875-18940-19252  KZ208910.1   \n",
       "2           +                          10000-10669-10744-19252  KZ208910.1   \n",
       "3           +              10000-10669-10749-18875-18940-19252  KZ208910.1   \n",
       "4           +                          10000-10669-10749-19252  KZ208910.1   \n",
       "\n",
       "source CHESS3 ENCODE4 GENCODE v47   GTEx  PODER RefSeq v110  \n",
       "0       False   False       False  False  False        True  \n",
       "1       False   False       False  False  False        True  \n",
       "2       False   False       False  False  False        True  \n",
       "3       False   False       False  False  False        True  \n",
       "4       False   False       False  False  False        True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "481010b5-ddec-4a41-ad64-a53ee5ec9a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../../supp_tables/16_inter_catalog_overlap.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "202efc6c-6348-4607-aee8-1907a9b89ec3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "631370"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7dfa6b4-404d-43b6-8e6b-a50ecb7d6218",
   "metadata": {},
   "source": [
    "## add the novelty categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c8b9ba85-f16a-44eb-8e0c-f88ddafd9341",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "gtfs = [od+config['gtex']['gtf'],\n",
    "       od+config['enc']['gtf'],\n",
    "       od+config['chess']['gtf']]\n",
    "sources = ['GTEx', 'ENCODE4', 'CHESS3']\n",
    "\n",
    "for f, s in zip(gtfs, sources):\n",
    "    temp = pr.read_gtf(f)\n",
    "    temp = cerberus.get_ic(temp)\n",
    "    temp['source'] = s\n",
    "    df = pd.concat([df, temp], axis=0)\n",
    "    \n",
    "ics_df = df.copy(deep=True)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79eda9c1-80df-4e4e-af70-a22259c579ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93718\n",
      "245884\n",
      "387944\n"
     ]
    }
   ],
   "source": [
    "# get structural categories\n",
    "f_dir = '/Users/fairliereese/Documents/programming/mele_lab/projects/240903_pt/ref/compare_external/'\n",
    "files = [f'{f_dir}gtex_sqanti_classification.txt',\n",
    "     f'{f_dir}enc_sqanti_classification.txt',\n",
    "        f'{f_dir}chess_sqanti_classification.txt']\n",
    "refs = ['GTEx', 'ENCODE4', 'CHESS3']\n",
    "\n",
    "df2 = pd.DataFrame()\n",
    "\n",
    "for f, r in zip(files, refs):\n",
    "    df = pd.read_csv(f, sep='\\t')\n",
    "    print(len(df))\n",
    "    # break\n",
    "    m = {'antisense': 'Antisense',\n",
    "     'full-splice_match': 'FSM',\n",
    "     'fusion': 'Fusion', \n",
    "     'genic': 'Genic', \n",
    "     'genic_intron': 'Genic', \n",
    "     'incomplete-splice_match': 'ISM',\n",
    "     'novel_in_catalog': 'NIC',\n",
    "     'novel_not_in_catalog': 'NNC', \n",
    "     'intergenic': 'Intergenic'}\n",
    "    # print(df.structural_category.unique())\n",
    "    # break\n",
    "    df['structural_category'] = df['structural_category'].map(m)\n",
    "    df2 = pd.concat([df2, df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d2be6e1a-d5f9-4012-b032-18ab50797b5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "508058\n",
      "455963\n"
     ]
    }
   ],
   "source": [
    "# remove monoexonic\n",
    "print(len(ics_df.index))\n",
    "ics_df = ics_df.loc[ics_df.ic!='-']\n",
    "print(len(ics_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "47764a7d-3978-4048-8f74-0d6720d223fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>exons</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>ref_length</th>\n",
       "      <th>ref_exons</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_A_downstream_TTS</th>\n",
       "      <th>dist_to_CAGE_peak</th>\n",
       "      <th>within_CAGE_peak</th>\n",
       "      <th>dist_to_polyA_site</th>\n",
       "      <th>within_polyA_site</th>\n",
       "      <th>polyA_motif</th>\n",
       "      <th>polyA_dist</th>\n",
       "      <th>polyA_motif_found</th>\n",
       "      <th>ORF_seq</th>\n",
       "      <th>ratio_TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1911</td>\n",
       "      <td>11</td>\n",
       "      <td>NIC</td>\n",
       "      <td>ENSG00000160055.20</td>\n",
       "      <td>novel</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CTGTGCCAAATATACCAATT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0016cd97-0137-4194-9821-910931c6e972_ENSG00000...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4492</td>\n",
       "      <td>23</td>\n",
       "      <td>NNC</td>\n",
       "      <td>ENSG00000086015.23</td>\n",
       "      <td>novel</td>\n",
       "      <td>5207.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>TTTATGTCTTTTGCTTGCTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2671</td>\n",
       "      <td>16</td>\n",
       "      <td>FSM</td>\n",
       "      <td>ENSG00000162736.18</td>\n",
       "      <td>ENST00000699545.1</td>\n",
       "      <td>2725.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CTTGTACTGCTAGTTATTCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1345</td>\n",
       "      <td>10</td>\n",
       "      <td>NIC</td>\n",
       "      <td>ENSG00000116885.18</td>\n",
       "      <td>novel</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CCTTTCCCATTTTATTCTTA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3208</td>\n",
       "      <td>8</td>\n",
       "      <td>NIC</td>\n",
       "      <td>ENSG00000143549.22</td>\n",
       "      <td>novel</td>\n",
       "      <td>7064.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CCTCATGCCACCCTGCATTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             isoform chrom strand  length  \\\n",
       "0  00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...  chr1      -    1911   \n",
       "1  0016cd97-0137-4194-9821-910931c6e972_ENSG00000...  chr1      +    4492   \n",
       "2  001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...  chr1      +    2671   \n",
       "3  00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...  chr1      -    1345   \n",
       "4  00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...  chr1      -    3208   \n",
       "\n",
       "   exons structural_category     associated_gene associated_transcript  \\\n",
       "0     11                 NIC  ENSG00000160055.20                 novel   \n",
       "1     23                 NNC  ENSG00000086015.23                 novel   \n",
       "2     16                 FSM  ENSG00000162736.18     ENST00000699545.1   \n",
       "3     10                 NIC  ENSG00000116885.18                 novel   \n",
       "4      8                 NIC  ENSG00000143549.22                 novel   \n",
       "\n",
       "   ref_length  ref_exons  ...  seq_A_downstream_TTS  dist_to_CAGE_peak  \\\n",
       "0      1538.0       11.0  ...  CTGTGCCAAATATACCAATT                NaN   \n",
       "1      5207.0       27.0  ...  TTTATGTCTTTTGCTTGCTT                NaN   \n",
       "2      2725.0       16.0  ...  CTTGTACTGCTAGTTATTCT                NaN   \n",
       "3      1463.0       11.0  ...  CCTTTCCCATTTTATTCTTA                NaN   \n",
       "4      7064.0       10.0  ...  CCTCATGCCACCCTGCATTT                NaN   \n",
       "\n",
       "   within_CAGE_peak  dist_to_polyA_site within_polyA_site  polyA_motif  \\\n",
       "0               NaN                 NaN               NaN          NaN   \n",
       "1               NaN                 NaN               NaN          NaN   \n",
       "2               NaN                 NaN               NaN          NaN   \n",
       "3               NaN                 NaN               NaN          NaN   \n",
       "4               NaN                 NaN               NaN          NaN   \n",
       "\n",
       "  polyA_dist  polyA_motif_found  ORF_seq  ratio_TSS  \n",
       "0        NaN                NaN      NaN        NaN  \n",
       "1        NaN                NaN      NaN        NaN  \n",
       "2        NaN                NaN      NaN        NaN  \n",
       "3        NaN                NaN      NaN        NaN  \n",
       "4        NaN                NaN      NaN        NaN  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "225f4b81-d07d-4dd5-a289-886f8cde2386",
   "metadata": {},
   "outputs": [],
   "source": [
    "ics_df = ics_df.merge(df2[['isoform', 'structural_category']],\n",
    "                      how='left', \n",
    "                      left_on='transcript_id', \n",
    "                      right_on='isoform')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fc8a4f5d-4dd9-4ab1-82a1-ef2d5e6d2a46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>source</th>\n",
       "      <th>isoform</th>\n",
       "      <th>structural_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...</td>\n",
       "      <td>ENSG00000160055.19</td>\n",
       "      <td>32222306-32222018-32221866-32221197-32221130-3...</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...</td>\n",
       "      <td>ENSG00000116885.18</td>\n",
       "      <td>36450257-36444033-36444003-36438910-36438755-3...</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...</td>\n",
       "      <td>ENSG00000143549.19</td>\n",
       "      <td>154191901-154191311-154191185-154176248-154176...</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>004b70dd-1c22-40d9-912d-cdc35d51c816_ENSG00000...</td>\n",
       "      <td>ENSG00000175756.13</td>\n",
       "      <td>1374996-1374790-1374704-1374590-1373999-1373902</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>004b70dd-1c22-40d9-912d-cdc35d51c816_ENSG00000...</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>004f8830-0766-47d5-af4f-11c745414a2d_ENSG00000...</td>\n",
       "      <td>ENSG00000143549.19</td>\n",
       "      <td>154182987-154176248-154176114-154173201-154173...</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>004f8830-0766-47d5-af4f-11c745414a2d_ENSG00000...</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand                                      transcript_id  \\\n",
       "0       chr1      -  00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...   \n",
       "1       chr1      -  00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...   \n",
       "2       chr1      -  00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...   \n",
       "3       chr1      -  004b70dd-1c22-40d9-912d-cdc35d51c816_ENSG00000...   \n",
       "4       chr1      -  004f8830-0766-47d5-af4f-11c745414a2d_ENSG00000...   \n",
       "\n",
       "              gene_id                                                 ic  \\\n",
       "0  ENSG00000160055.19  32222306-32222018-32221866-32221197-32221130-3...   \n",
       "1  ENSG00000116885.18  36450257-36444033-36444003-36438910-36438755-3...   \n",
       "2  ENSG00000143549.19  154191901-154191311-154191185-154176248-154176...   \n",
       "3  ENSG00000175756.13    1374996-1374790-1374704-1374590-1373999-1373902   \n",
       "4  ENSG00000143549.19  154182987-154176248-154176114-154173201-154173...   \n",
       "\n",
       "  source                                            isoform  \\\n",
       "0   GTEx  00146542-c373-4da6-8d48-2242bf43bade_ENSG00000...   \n",
       "1   GTEx  00294b07-572e-422c-a1bc-6f089eff895c_ENSG00000...   \n",
       "2   GTEx  00321d78-2502-4c1c-bbac-bb2e5f99582a_ENSG00000...   \n",
       "3   GTEx  004b70dd-1c22-40d9-912d-cdc35d51c816_ENSG00000...   \n",
       "4   GTEx  004f8830-0766-47d5-af4f-11c745414a2d_ENSG00000...   \n",
       "\n",
       "  structural_category  \n",
       "0                 NIC  \n",
       "1                 NIC  \n",
       "2                 NIC  \n",
       "3                 NIC  \n",
       "4                 NIC  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3b803722-3699-4686-b988-d5942ec4ecfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = ics_df.groupby(['isoform', 'source']).nunique().reset_index().rename({'structural_category':'n_cats'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "79b9a476-1128-415c-aaf1-3acd185ddde2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['50056645-50064076-50064204-50064527-50064607-50068055-50068305-50070843-50070930-50071310-50071518-50071985-50072182-50073915-50074004-50074143-50074280-50074447-50074534-50075458-50075526-50075598-50075735-50076325-50076475-50076583-50076681-50076973-50077121-50077378-50077447-50078045-50078125-50078206-50078284-50078823-50078926-50079075-50079258-50079434-50079557-50080261-50080375-50081309-50081408-50087743-50087869-50089109',\n",
       "       '50056645-50064076-50064204-50064527-50064607-50068055-50068305-50070843-50070930-50071310-50071518-50071985-50072182-50073915-50074004-50074143-50074280-50074447-50074534-50075458-50075526-50075598-50075735-50076325-50076475-50076583-50076681-50076973-50077121-50077378-50077447-50078045-50078125-50078206-50078284-50078823-50078926-50079075-50079258-50079434-50079557-50080261-50080375-50081309-50081408-50087743-50087869-50089109'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ics_df.loc[ics_df.isoform=='ENST00000007722.11'].ic.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "681f80cb-5a5a-45a7-ba89-8e9c59b4d2cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GTEx'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.n_cats>1].source.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1c7ae1-bbc2-4ba4-9e88-e3137db5fd88",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
