{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "31174b8d-66b7-4fe6-89c8-98b76264d72a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "# install.packages(\"car\", verbose=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5130bcb3-22f6-409b-b2da-e35965fcfee4",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans.expr.p.f <- 'tre.df.RData'\n",
    "indexed.geno.f <- 'genotype.tsv.bgz'\n",
    "gene.loc.chunk <- 'nominal_in.596'\n",
    "covariates.f <- 'covariates.df.RData'\n",
    "output.f <- 'test'\n",
    "\n",
    "load(trans.expr.p.f)                                                            # Load tre.df\n",
    "genes.bed <- read.table(gene.loc.chunk, header = FALSE, as.is = TRUE)           # Load chunk\n",
    "colnames(genes.bed) <- c(\"chr\", \"start\", \"end\", \"geneId\")                       # Name chunk\n",
    "genes <- genes.bed$geneId                                                       # Get gene names\n",
    "tre.df <- subset(tre.df, geneId %in% genes)                                     # Subset tre.df\n",
    "load(covariates.f)                                                              # Load covariates.df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "450095cb-78de-4e5a-9c61-1b98917ff422",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <!-- res.df <- sqtl.seeker(tre.df, indexed.geno.f, genes.bed,\n",
    "#                       covariates = covariates.df,\n",
    "#                       genic.window = opt$window,\n",
    "#                       min.nb.ext.scores = opt$min_nb_ext_scores,\n",
    "#                       nb.perm.max = opt$nb_perm_max,\n",
    "#                       nb.perm.max.svQTL = opt$nb_perm_max_svqtl,\n",
    "#                       svQTL = opt$svqtl, asympt = opt$asympt,\n",
    "#                       min.nb.ind.geno = opt$min_nb_ind_geno,\n",
    "#                       ld.filter = LD, verbose = opt$verbose) -->\n",
    "\n",
    "genotype.f = indexed.geno.f\n",
    "gene.loc = genes.bed\n",
    "covariates=covariates.df\n",
    "genic.window = 5000\n",
    "min.nb.ext.scores = 1000\n",
    "nb.perm.max = 1e6\n",
    "nb.perm.max.svQTL = 1e4\n",
    "svQTL = FALSE\n",
    "asympt = TRUE\n",
    "ld.filter = NULL\n",
    "min.nb.ind.geno = 10\n",
    "verbose = TRUE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "215cfe07-3b25-448f-ac1f-5fadb2563e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "read.bedix <- function(file, subset.reg = NULL, header = TRUE)\n",
    "{\n",
    "    if(!is.character(file)){\n",
    "        file <- as.character(file)\n",
    "    }\n",
    "    if(!file.exists(file)){\n",
    "        file <- paste0(file, \".bgz\")\n",
    "    }\n",
    "    if(!file.exists(file)){\n",
    "        stop(file, \"Input file not found (with and without .bgz extension).\")\n",
    "    }\n",
    "    if(!file.exists(paste0(file, \".tbi\"))){\n",
    "        stop(paste0(file, \".tbi\"), \"Index file not found.\")\n",
    "    }\n",
    "    if(is.null(subset.reg)){\n",
    "        return(utils::read.table(file, as.is = TRUE, header = header))\n",
    "    }\n",
    "    if (is.data.frame(subset.reg)) {\n",
    "        if(!all(c(\"chr\", \"start\", \"end\") %in% colnames(subset.reg))){\n",
    "            stop(\"Missing column in 'subset.reg'. 'chr', 'start' and 'end' are required.\")\n",
    "        }\n",
    "        subset.reg <- with(subset.reg,\n",
    "                           GenomicRanges::GRanges(chr, IRanges::IRanges(start, end)))\n",
    "    } else if (class(subset.reg) != \"GRanges\") {\n",
    "        stop(\"'subset.reg' must be a data.frame or a GRanges object.\")\n",
    "    }\n",
    "    print('are we doing anything?')\n",
    "    subset.reg <- subset.reg[order(as.character(GenomicRanges::seqnames(subset.reg)),\n",
    "                                   GenomicRanges::start(subset.reg))]\n",
    "    read.chunk <- function(gr){\n",
    "        print('this is the file name in read bedix')\n",
    "        print(file)\n",
    "        print('\\n')\n",
    "        bed <- unlist(Rsamtools::scanTabix(file))\n",
    "        # bed <- unlist(Rsamtools::scanTabix(file, param = GenomicRanges::reduce(gr)))\n",
    "        print(bed)\n",
    "        print('genomic range thing')\n",
    "        print(gr)\n",
    "        #bed <- tryCatch(unlist(Rsamtools::scanTabix(file, param = GenomicRanges::reduce(gr)),\n",
    "        #                       use.names = FALSE, recursive = FALSE), error = function(e) c())\n",
    "        if (length(bed) == 0) {\n",
    "            print('here in read file thing')\n",
    "            return(NULL)\n",
    "        }\n",
    "        ncol <- length(strsplit(bed[1], \"\\t\")[[1]])\n",
    "        bed <- matrix(unlist(strsplit(bed, \"\\t\", fixed = TRUE), use.names=FALSE, recursive = FALSE),\n",
    "                      length(bed), ncol, byrow = TRUE)\n",
    "        bed <- data.table::data.table(bed)\n",
    "        if (header) {\n",
    "            data.table::setnames(bed,\n",
    "                                 as.character(utils::read.table(file, nrows = 1, as.is = TRUE)))\n",
    "        }\n",
    "        bed <- bed[, lapply(.SD, function(ee)utils::type.convert(as.character(ee), as.is = TRUE))]\n",
    "        bed <- as.data.frame(bed)\n",
    "        return(bed)\n",
    "    }\n",
    "    if (length(subset.reg) > 10000) {\n",
    "        chunks <- cut(1:length(subset.reg), ceiling(length(subset.reg)/10000))\n",
    "        bed.df <- plyr::ldply(levels(chunks), function(ch.id){\n",
    "            read.chunk(subset.reg[which(chunks == ch.id)])\n",
    "        })\n",
    "    } else {\n",
    "        bed.df <- read.chunk(subset.reg)\n",
    "    }\n",
    "    if(!is.null(bed.df)){\n",
    "        bed.df <- bed.df[order(bed.df$chr, bed.df$start), ]\n",
    "    }\n",
    "    return(bed.df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c650e2fe-8587-4642-a3f5-ad5147962f63",
   "metadata": {},
   "outputs": [],
   "source": [
    ". <- nb.groups <- snpId <- NULL ## Uglily appease R checks (dplyr)\n",
    "analyze.gene.f <- function(tre.gene){\n",
    "    if(verbose) message(tre.gene$geneId[1])\n",
    "    if(sum(duplicated(gene.loc$geneId)) > 1){\n",
    "        stop(tre.gene$geneId[1], \" Repeated gene in gene location file.\")\n",
    "    }\n",
    "    gr.gene <- with(gene.loc[which(gene.loc$geneId == tre.gene$geneId[1]), ],\n",
    "                    GenomicRanges::GRanges(chr, IRanges::IRanges(start, end)))\n",
    "    if(genic.window > 0){\n",
    "        gr.gene <- GenomicRanges::resize(gr.gene, GenomicRanges::width(gr.gene) +\n",
    "                                           2 * genic.window, fix = \"center\")\n",
    "    }\n",
    "    if(length(gr.gene) > 0){\n",
    "        if(!is.null(covariates)){\n",
    "            if(!all(rownames(covariates) %in% colnames(tre.gene))){\n",
    "                stop(\"All samples should have covariate information (either a value or NA).\")                  \n",
    "            }\n",
    "            cov.na <- apply(covariates, 1, function(x){any(is.na(x))})\n",
    "            covariates <- covariates[!cov.na,  ,drop = FALSE]                                     \n",
    "            if(sum(cov.na) > 0){\n",
    "                warning(sprintf(\"%s samples with NA values for at least one covariate have been removed.\",\n",
    "                                sum(cov.na)))     \n",
    "            }\n",
    "            if(sum(cov.na) / nrow(covariates) > 0.05){\n",
    "                stop(\"More than 5% of the samples contain NA values for at least one covariate\")  \n",
    "            }\n",
    "        }\n",
    "        tre.gene <- tre.gene[, !is.na(tre.gene[1, ])]\n",
    "        genotype.headers <- as.character(utils::read.table(genotype.f,\n",
    "                                                           as.is = TRUE, nrows = 1))\n",
    "        if(!is.null(covariates)){\n",
    "            com.samples <- Reduce(intersect, list(colnames(tre.gene),\n",
    "                                                  genotype.headers, rownames(covariates)))   \n",
    "            if (length(com.samples) == 0) {\n",
    "                stop(\"No common samples between genotype, covariate and transcript files.\")\n",
    "            }\n",
    "        } else{\n",
    "            com.samples <- intersect(colnames(tre.gene), genotype.headers)\n",
    "            if (length(com.samples) == 0) {\n",
    "                stop(\"No common samples between genotype and transcript files.\")\n",
    "            }\n",
    "        }\n",
    "        tre.gene <- tre.gene[, c(\"trId\", \"geneId\", com.samples)]\n",
    "        allzero <- apply(tre.gene[, com.samples], 1, function(x){ sum(x) == 0 })\n",
    "        if (any(allzero)){\n",
    "            tr.out <- tre.gene$trId[allzero]\n",
    "            tre.gene <- tre.gene[!allzero,]\n",
    "            warning(sprintf(\"Transcript(s) %s is(are) removed due to zero expression in all common samples.\",\n",
    "                            paste(tr.out, collapse = \", \")) )\n",
    "        }\n",
    "        tre.tc <- t(sqrt(tre.gene[, com.samples]))\n",
    "        colnames(tre.tc) <- tre.gene$tr\n",
    "        if(!is.null(covariates)){\n",
    "            covariates <- covariates[com.samples, , drop = FALSE]\n",
    "            multiclass <- apply(covariates, 2, function(x){length(table(x)) > 1})\n",
    "            covariates <- covariates[, multiclass, drop = FALSE]\n",
    "            if (verbose & any(!multiclass)){\n",
    "              message(\"\\t\", \"Covariates removed due to only one value: \",\n",
    "                      paste(names(multiclass)[!multiclass], collapse = \", \"))\n",
    "            }\n",
    "            fit <- stats::lm(tre.tc ~ ., data = covariates)\n",
    "            if (ncol(covariates) > 1){\n",
    "                vifs <- car::vif(stats::lm(tre.tc[, 1] ~ ., data = covariates))\n",
    "                if (verbose){\n",
    "                    message(\"\\t\", \"Covariates VIF - \",\n",
    "                            paste(names(vifs), round(vifs, 2), sep = \": \", collapse = \", \"))\n",
    "                    print(vifs)\n",
    "                }\n",
    "                if (any(vifs > 5)){\n",
    "                    warning(\"Check multicollinearity. VIF > 5 for some covariates:\", \"\\n\",\n",
    "                            paste(names(vifs), round(vifs, 2), sep = \": \", collapse = \", \"))\n",
    "                }  \n",
    "            }\n",
    "            tre.tc <- fit$residual\n",
    "        }\n",
    "        res.df <- data.frame()\n",
    "        if(GenomicRanges::width(gr.gene) > 20000 && is.null(ld.filter)){\n",
    "            pos.breaks <- unique(round(seq(GenomicRanges::start(gr.gene),\n",
    "                                           GenomicRanges::end(gr.gene),\n",
    "                                           length.out = floor(GenomicRanges::width(gr.gene)/10000) + 1)))\n",
    "            gr.gene.spl <- rep(gr.gene, length(pos.breaks) - 1)\n",
    "            GenomicRanges::start(gr.gene.spl) <- pos.breaks[-length(pos.breaks)]\n",
    "            pos.breaks[length(pos.breaks)] <- pos.breaks[length(pos.breaks)] + 1\n",
    "            GenomicRanges::end(gr.gene.spl) <- pos.breaks[-1] - 1\n",
    "        } else {\n",
    "            gr.gene.spl <- gr.gene\n",
    "        }\n",
    "        browser()\n",
    "        res.df <- lapply(1:length(gr.gene.spl), function(ii){\n",
    "            browser()\n",
    "            res.range <- data.frame()\n",
    "            if(verbose) message(\"  Sub-range \", ii)\n",
    "            print(gr.gene.spl[ii])\n",
    "            print('this is the file name in analyze gene')\n",
    "            print(genotype.f)\n",
    "            print('\\n')\n",
    "            genotype.gene <- read.bedix(genotype.f, gr.gene.spl[ii])\n",
    "            print(gr.gene)\n",
    "            print(genotype.gene)\n",
    "            if(verbose && is.null(genotype.gene)) {\n",
    "              message(\"\\tNo SNPs in the genomic range.\")\n",
    "            } else { message(\"\\t oooh I found a snp\")}\n",
    "            if(!is.null(genotype.gene)){\n",
    "                snps.to.keep <- check.genotype(genotype.gene[, com.samples],\n",
    "                                               tre.gene[, com.samples],\n",
    "                                               min.nb.ind.geno = min.nb.ind.geno)\n",
    "                if(verbose){\n",
    "                    snps.to.keep.t <- table(snps.to.keep)\n",
    "                    message(\"\\t\", paste(names(snps.to.keep.t), snps.to.keep.t,\n",
    "                                        sep = \": \", collapse=\", \"))\n",
    "                }\n",
    "                if(any(snps.to.keep == \"PASS\")){\n",
    "                    genotype.gene <- genotype.gene[snps.to.keep == \"PASS\", ]\n",
    "                    if(!is.null(ld.filter)){\n",
    "                        if(verbose) message(\"\\tLD filtering\")\n",
    "                        genotype.gene <- LD.filter(genotype.gene = genotype.gene,\n",
    "                                                   tre.mt = tre.tc, th = ld.filter,\n",
    "                                                   svQTL = svQTL)\n",
    "                    }\n",
    "                    res.range <- dplyr::do(dplyr::group_by(genotype.gene, snpId),\n",
    "                                           compFscore(., tre.tc, svQTL = svQTL,\n",
    "                                                      asympt = asympt,\n",
    "                                                      res = !is.null(covariates)))\n",
    "                }\n",
    "            }\n",
    "            return(res.range)\n",
    "        })\n",
    "        range.done <- which(unlist(lapply(res.df, nrow)) > 0)\n",
    "        if(length(range.done) > 0){\n",
    "            res.df <- res.df[range.done]\n",
    "            res.df <- do.call(rbind, res.df)\n",
    "            if (!is.null(ld.filter)){\n",
    "                ld <- res.df[,c(\"snpId\",\"LD\")]\n",
    "                res.df$LD <- NULL\n",
    "            }\n",
    "            if(!asympt){\n",
    "                res.df <- dplyr::do(dplyr::group_by(res.df, nb.groups),\n",
    "                                    compPvalue(., tre.tc, min.nb.ext.scores = min.nb.ext.scores,\n",
    "                                               nb.perm.max = nb.perm.max))\n",
    "            }\n",
    "            if(svQTL){\n",
    "                res.df <- dplyr::do(dplyr::group_by(res.df, nb.groups),\n",
    "                                    compPvalue(., tre.tc, svQTL = TRUE,\n",
    "                                               min.nb.ext.scores = min.nb.ext.scores,\n",
    "                                               nb.perm.max = nb.perm.max.svQTL))\n",
    "            }\n",
    "            if (!is.null(ld.filter)){\n",
    "                res.df <- merge(res.df, ld, by = \"snpId\")\n",
    "            }\n",
    "            res.df <- dplyr::arrange(res.df, pv)\n",
    "            return(data.frame(done = TRUE, res.df))\n",
    "        }\n",
    "    } else {\n",
    "        if(verbose) warning(\"Issue with the gene location.\")\n",
    "    }\n",
    "    return(data.frame(done = FALSE))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91bc54aa-b7df-4544-aee2-d7ffadfe055c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ret.df <- lapply(c(\"ENSG00000102189.17\"), function(gene.i){\n",
    "#        df <- tre.df[which(tre.df$geneId == gene.i), ]\n",
    "#        data.frame(geneId = gene.i, analyze.gene.f(df))\n",
    "#     })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "44c7aeca-de4a-4715-a9df-1709b77645ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1 range and 0 metadata columns:\n",
       "      seqnames            ranges strand\n",
       "         <Rle>         <IRanges>  <Rle>\n",
       "  [1]    chr12 92770636-92929327      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'c(\"chr\", \"1\")'</li><li>'c(\"start\", \"10583\")'</li><li>'c(\"end\", \"10583\")'</li><li>'c(\"snpId\", \"1:10583:G:A\")'</li><li>'c(\"NA19704\", \"0\")'</li><li>'c(\"NA19332\", \"0\")'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'c(\"chr\", \"1\")'\n",
       "\\item 'c(\"start\", \"10583\")'\n",
       "\\item 'c(\"end\", \"10583\")'\n",
       "\\item 'c(\"snpId\", \"1:10583:G:A\")'\n",
       "\\item 'c(\"NA19704\", \"0\")'\n",
       "\\item 'c(\"NA19332\", \"0\")'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'c(\"chr\", \"1\")'\n",
       "2. 'c(\"start\", \"10583\")'\n",
       "3. 'c(\"end\", \"10583\")'\n",
       "4. 'c(\"snpId\", \"1:10583:G:A\")'\n",
       "5. 'c(\"NA19704\", \"0\")'\n",
       "6. 'c(\"NA19332\", \"0\")'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"c(\\\"chr\\\", \\\"1\\\")\"             \"c(\\\"start\\\", \\\"10583\\\")\"      \n",
       "[3] \"c(\\\"end\\\", \\\"10583\\\")\"         \"c(\\\"snpId\\\", \\\"1:10583:G:A\\\")\"\n",
       "[5] \"c(\\\"NA19704\\\", \\\"0\\\")\"         \"c(\\\"NA19332\\\", \\\"0\\\")\"        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# thing = read.bedix(indexed.geno.f)\n",
    "geneId=\"ENSG00000102189.17\"\n",
    "gr.gene <- with(gene.loc[which(gene.loc$geneId == geneId), ],\n",
    "                GenomicRanges::GRanges(chr, IRanges::IRanges(start, end)))\n",
    "gr.gene\n",
    "\n",
    "temp  <- as.character(utils::read.table(indexed.geno.f,\n",
    "                                                           as.is = TRUE, nrows = 2))\n",
    "head(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c320f9c-32e9-4b26-bc23-d70f88ea3454",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "<!--\n",
    "ret.df <- lapply(unique(tre.df$geneId), function(gene.i){\n",
    "       df <- tre.df[which(tre.df$geneId == gene.i), ]\n",
    "       data.frame(geneId = gene.i, analyze.gene.f(df))\n",
    "    })\n",
    "\n",
    "    done <- which(unlist(lapply(ret.df, ncol)) > 2)\n",
    "    if(length(done) > 0){\n",
    "        ret.df <- ret.df[done]\n",
    "        ret.df <- do.call(rbind, ret.df)\n",
    "        ret.df$done <- NULL\n",
    "        return(ret.df)\n",
    "    } else {\n",
    "        return(NULL)\n",
    "    } -->\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5dd86c9-9ab5-4931-a6f2-27e05a29e79c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
