{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fabf4f07-ebde-47c9-a629-3e1f6b694e6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *\n",
    "from scripts.sm_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1b24dd80-ce1d-4da4-9f0c-6b89488c22d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = ''\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a45cfc2e-1bde-47f4-997c-52d8407fca36",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "    gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d52b553-3c0d-40e5-af29-17f3e8464d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_file = '../config.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2658a33c-1f98-4e40-8d34-7e8f7cd40be4",
   "metadata": {},
   "outputs": [],
   "source": [
    "v47_file = od+config['ref']['gtf']\n",
    "poder_file = od+config['lr']['gtf_filt_with_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ae2d9434-27a6-4ec2-b44e-4ae508592d34",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_df = pr.read_gtf(v47_file)\n",
    "p_df = pr.read_gtf(poder_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9c3df6e3-39f2-4870-90a6-0758075452b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get list of ics from each\n",
    "gc_ics = cerberus.get_ic(gc_df)\n",
    "p_ics = cerberus.get_ic(p_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "72cb66f4-ef03-49dc-a2d6-85705593950b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get uniq ICs (gene+chrom+ic)\n",
    "# to poder that are not in gc\n",
    "gc_ics['id'] = gc_ics['Chromosome']+'_'+\\\n",
    "               gc_ics['gene_id']+'_'+\\\n",
    "               gc_ics['ic']\n",
    "p_ics['id'] = p_ics['Chromosome']+'_'+\\\n",
    "               p_ics['gene_id']+'_'+\\\n",
    "               p_ics['ic']\n",
    "p_ics['in_gc'] = p_ics['id'].isin(gc_ics['id'].tolist())\n",
    "p_ics[['in_gc', 'transcript_id']].groupby('in_gc').count()\n",
    "nov_tids = p_ics.loc[~p_ics['id'].isin(gc_ics['id'])].transcript_id.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "658760c3-1020-4eda-9790-3da485acc121",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter poder gtf based on just novel transcripts\n",
    "p_df = p_df.df\n",
    "p_df = p_df.loc[p_df.transcript_id.isin(nov_tids)]\n",
    "assert len(p_df.loc[p_df.transcript_id.notnull()].transcript_id.unique())==len(nov_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7e7c2676-a5fc-491f-8f05-921d04d23dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict to just transcript + exon entries\n",
    "gc_df = gc_df.df\n",
    "gc_df = gc_df.loc[gc_df.Feature.isin(['transcript', 'exon'])]\n",
    "p_df = p_df.loc[p_df.Feature.isin(['transcript', 'exon'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "da9b66d2-c4ac-42b8-96e1-2c4b5c8669f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate different dfs\n",
    "df = pd.concat([gc_df, p_df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "cf49e16c-883a-43db-a0b5-bf547cbf9989",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop gene name entries cause all they do is screw everything up\n",
    "df.drop('gene_name', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "203c4d7b-77cd-4b0a-8e11-4faa63ddb9c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # make sure that we don't have duplicates from gene name / gene id combos\n",
    "# l1 = len(df[['gene_name', 'gene_id']].drop_duplicates())\n",
    "# l2 = len(df.gene_id.unique())\n",
    "# print(l1)\n",
    "# print(l2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a6bb49a7-791e-4c35-a93e-654f8e400890",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp = df[['gene_name', 'gene_id']].drop_duplicates()\n",
    "# temp.loc[temp.gene_id.duplicated(keep=False)].sort_values(by='gene_id').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4e7152b7-d661-48a4-81c6-f22b03e54d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make new gene entries for everything\n",
    "l1 = len(df.gene_id.unique().tolist())\n",
    "# make gene entry\n",
    "g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "g_df['Source'] = 'v47_poder'\n",
    "g_df['Frame'] = '.'\n",
    "g_df['Score'] = '.'\n",
    "l2 = len(g_df.loc[g_df.Feature=='gene'].index)\n",
    "assert l1 == l2\n",
    "\n",
    "# concat them and then sort gtf\n",
    "df = pd.concat([df, g_df], axis=0)\n",
    "df = cerberus.sort_gtf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8a1571a2-2931-43a3-827d-7d82884b410b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "      <th>samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>v47_poder</td>\n",
       "      <td>gene</td>\n",
       "      <td>100584935</td>\n",
       "      <td>100599885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100584935</td>\n",
       "      <td>100599885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057481.2</td>\n",
       "      <td>HGNC:17757</td>\n",
       "      <td>OTTHUMG00000022001.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000362122.4</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>100584935</td>\n",
       "      <td>100585066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00001459371.5</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057481.2</td>\n",
       "      <td>HGNC:17757</td>\n",
       "      <td>OTTHUMG00000022001.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000362122.4</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>100585230</td>\n",
       "      <td>100585362</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00000401061.1</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057481.2</td>\n",
       "      <td>HGNC:17757</td>\n",
       "      <td>OTTHUMG00000022001.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000362122.4</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>100593894</td>\n",
       "      <td>100594035</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSE00000673400.1</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057481.2</td>\n",
       "      <td>HGNC:17757</td>\n",
       "      <td>OTTHUMG00000022001.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000362122.4</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Source     Feature      Start        End Score Strand Frame  \\\n",
       "0       chrX  v47_poder        gene  100584935  100599885     .      +     .   \n",
       "1       chrX     HAVANA  transcript  100584935  100599885     .      +     .   \n",
       "2       chrX     HAVANA        exon  100584935  100585066     .      +     .   \n",
       "3       chrX     HAVANA        exon  100585230  100585362     .      +     .   \n",
       "4       chrX     HAVANA        exon  100593894  100594035     .      +     .   \n",
       "\n",
       "             gene_id       gene_type  ...            exon_id  \\\n",
       "0  ENSG00000000005.6             NaN  ...                NaN   \n",
       "1  ENSG00000000005.6  protein_coding  ...                NaN   \n",
       "2  ENSG00000000005.6  protein_coding  ...  ENSE00001459371.5   \n",
       "3  ENSG00000000005.6  protein_coding  ...  ENSE00000401061.1   \n",
       "4  ENSG00000000005.6  protein_coding  ...  ENSE00000673400.1   \n",
       "\n",
       "  transcript_support_level     havana_transcript     hgnc_id  \\\n",
       "0                      NaN                   NaN         NaN   \n",
       "1                        1  OTTHUMT00000057481.2  HGNC:17757   \n",
       "2                        1  OTTHUMT00000057481.2  HGNC:17757   \n",
       "3                        1  OTTHUMT00000057481.2  HGNC:17757   \n",
       "4                        1  OTTHUMT00000057481.2  HGNC:17757   \n",
       "\n",
       "            havana_gene  ont         protein_id       ccdsid artif_dupl  \\\n",
       "0                   NaN  NaN                NaN          NaN        NaN   \n",
       "1  OTTHUMG00000022001.2  NaN  ENSP00000362122.4  CCDS14469.1        NaN   \n",
       "2  OTTHUMG00000022001.2  NaN  ENSP00000362122.4  CCDS14469.1        NaN   \n",
       "3  OTTHUMG00000022001.2  NaN  ENSP00000362122.4  CCDS14469.1        NaN   \n",
       "4  OTTHUMG00000022001.2  NaN  ENSP00000362122.4  CCDS14469.1        NaN   \n",
       "\n",
       "  samples  \n",
       "0     NaN  \n",
       "1     NaN  \n",
       "2     NaN  \n",
       "3     NaN  \n",
       "4     NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "69dea178-8945-4078-ac03-a357a4d55019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure we've added the same number of transcripts\n",
    "n_gc_t = len(gc_df.loc[gc_df.transcript_id.notnull()].transcript_id.unique())\n",
    "n_nov_t = len(nov_tids)\n",
    "n_gc_p_t = len(df.loc[df.transcript_id.notnull()].transcript_id.unique())\n",
    "\n",
    "print(n_gc_t)\n",
    "print(n_nov_t)\n",
    "print(n_gc_p_t)\n",
    "assert n_gc_t+n_nov_t == n_gc_p_t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "92a2e5a3-b602-45de-badb-6d857b5e66fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pr.PyRanges(df)\n",
    "df.to_gtf(output.gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "149f0522-83e9-4b16-b6ba-d026abe3424d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
