{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c6cd8346-5277-469b-a34f-53d7216d39f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f9fa4262-6e71-4db1-9185-404636b2452f",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = ''\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4acca558-ec5c-4ae5-a73b-e320d89d5854",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8ef0893-1169-44ed-b455-833d7590278c",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_file = 'filereport_read_run_PRJNA851328_tsv.txt'\n",
    "\n",
    "df = pd.read_csv(meta_file, sep='\\t')\n",
    "df[['cell_line_id', 'batch', 'rep']] = df.experiment_alias.str.split('_', expand=True)\n",
    "len(df.index)\n",
    "len(df.cell_line_id.unique())\n",
    "\n",
    "df[['r1_fq_link', 'r2_fq_link']] = df.fastq_ftp.str.split(';', expand=True)\n",
    "df['r1_verify'] = df.r1_fq_link.str.endswith('_1.fastq.gz')\n",
    "df['r2_verify'] = df.r2_fq_link.str.endswith('_2.fastq.gz')\n",
    "assert len(df.loc[df.r1_verify==False].index)==0\n",
    "assert len(df.loc[df.r2_verify==False].index)==0\n",
    "\n",
    "df['sample'] = df['experiment_alias']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e266b9c1-2b5f-4254-8de8-1cb1f6ebf92f",
   "metadata": {},
   "source": [
    "## Prepare metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43105e6d-1ed8-4b21-ae4c-468a3fec9903",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_file = 'filereport_read_run_PRJNA851328_tsv.txt'\n",
    "df = pd.read_csv(meta_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae9e98b-5604-40fb-9a39-425afaf2705e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3caaaae8-b4a3-4b1d-9e31-88f5b87bfcf4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a8b502-21dc-4b32-b194-775c24bb2bdc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98034206-e6ce-44a7-b738-b6a355b8ae65",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c75ee5-850c-44ee-a63e-927aca408e57",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc46cd7c-2c75-4d90-98fa-3399c356b3fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d30ae04-934d-4994-a1fa-618c1eedb276",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "132d2146-b5b6-43d2-97c3-3a1836f218d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b30263-8174-4c47-af2d-fc7d005b6909",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78b25dc7-89f5-480e-b492-fe2a50d1ce6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c46f0dce-210f-4a37-ba3f-d49423cc355b",
   "metadata": {},
   "source": [
    "## Annotate each t-g pair based on sqtl calling testablilty thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bf9aa3f-ce2e-4659-8799-0075d1748d1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387944\n",
      "315312\n",
      "30599\n"
     ]
    }
   ],
   "source": [
    "def filt_sqtl_transcripts(t_tpm_tsv,\n",
    "                          t2g,\n",
    "                          min_samp_prop,\n",
    "                          min_g_exp,\n",
    "                          min_t_exp,\n",
    "                          min_n_t,\n",
    "                          ofile):\n",
    "\n",
    "    # # g_exp_file = od+config['mage']['v47_kallisto']['gene_tpm_tsv']\n",
    "    # t_tpm_tsv = od+config['mage']['v47_kallisto']['merge_matrix_tpm_tsv']\n",
    "    # t2g = od+config['ref']['v47_kallisto_short']['t2g']\n",
    "\n",
    "    # min_samp_prop = 0.8\n",
    "    # min_g_exp = 1\n",
    "    # min_t_exp = 0.1\n",
    "    # min_n_t = 2\n",
    "\n",
    "    # get melted version of table w/ tid, gid, sample, t_exp, n_t (per gene), passed t exp filt, passed g exp filt\n",
    "    # this is the main output tale\n",
    "    filt_df = pd.read_csv(t_tpm_tsv, sep='\\t')\n",
    "    filt_df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "    filt_df = filt_df.set_index('tid').melt(ignore_index=False, value_name='tpm', var_name='sample').reset_index()\n",
    "    t2g_df = pd.read_csv(t2g, sep='\\t', header=None)\n",
    "    t2g_df = t2g_df[[0,1]]\n",
    "    t2g_df.columns = ['tid', 'gid']\n",
    "    filt_df = filt_df.merge(t2g_df, \n",
    "                            how='left',\n",
    "                            on='tid')\n",
    "\n",
    "    # isoform expressed in at least 1 sample >= 0.1 TPM\n",
    "    df = pd.read_csv(t_tpm_tsv, sep='\\t')\n",
    "    df = df.set_index('transcript_id')\n",
    "    df = df>=min_t_exp\n",
    "    df['n_exp_samples'] = df.sum(axis=1)\n",
    "    print(len(df.index))\n",
    "    df = df.loc[df['n_exp_samples']>=1]\n",
    "    print(len(df.index))\n",
    "    tids = df.index.tolist()\n",
    "    filt_df['t_passed_exp_filt'] = filt_df.tid.isin(tids)\n",
    "\n",
    "    # using these isoform, which genes pass the n transcripts / gene\n",
    "    t2g_df = pd.read_csv(t2g, sep='\\t', header=None)\n",
    "    t2g_df = t2g_df[[0,1]]\n",
    "    t2g_df.columns = ['tid', 'gid']\n",
    "    print(len(t2g_df.index))\n",
    "    t2g_df = t2g_df.loc[(t2g_df.tid.isin(tids))]\n",
    "    print(len(t2g_df.index))\n",
    "    # count # isos / gene and filter one more time\n",
    "    t2g_df = t2g_df.groupby('gid').nunique().reset_index().rename({'tid':'n_tid'}, axis=1)\n",
    "    t2g_df = t2g_df.loc[t2g_df.n_tid >= min_n_t]\n",
    "    print(len(t2g_df.index))\n",
    "    gids2 = t2g_df.gid.tolist()\n",
    "    filt_df['g_passed_n_t_filt'] = filt_df.gid.isin(gids2)\n",
    "\n",
    "    # if sum of remaining transcripts (so gene expression based on filtered trx) is >=1 TPM \n",
    "    # keep transcript, if not, convert to NA (this is per sample)\n",
    "    df = pd.read_csv(t_tpm_tsv, sep='\\t')\n",
    "    df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "    df = df.loc[df.tid.isin(tids)]\n",
    "    t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "    t2g_df = t2g_df[[0,1]]\n",
    "    t2g_df.columns = ['tid', 'gid']\n",
    "    df = df.merge(t2g_df, \n",
    "                  how='left',\n",
    "                  on='tid')\n",
    "    df = df.drop('tid', axis=1)\n",
    "    df = df.groupby('gid').sum().reset_index()\n",
    "    df = df.set_index('gid').melt(ignore_index=False, value_name='gene_tpm', var_name='sample').reset_index()\n",
    "    df['g_passed_exp_filt'] = df.gene_tpm>=min_g_exp\n",
    "    filt_df = filt_df.merge(df,\n",
    "                            how='left',\n",
    "                            on=['gid', 'sample'])\n",
    "\n",
    "    # using just these isoforms, re-compute gene tpm \n",
    "    # genes exp. >= 1 TPM in at least 80% of samples\n",
    "    n_samples = len([c for c in df.columns.tolist() if c not in ['gid']])\n",
    "    n_sample_cutoff = int(n_samples*min_samp_prop)\n",
    "    df = pd.read_csv(t_tpm_tsv, sep='\\t')\n",
    "    df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "    df = df.loc[df.tid.isin(tids)]\n",
    "    t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "    t2g_df = t2g_df[[0,1]]\n",
    "    t2g_df.columns = ['tid', 'gid']\n",
    "    df = df.merge(t2g_df, \n",
    "                  how='left',\n",
    "                  on='tid')\n",
    "    df = df.drop('tid', axis=1)\n",
    "    df = df.groupby('gid').sum().reset_index()\n",
    "    df = df.set_index('gid')\n",
    "    df = df>=min_g_exp\n",
    "    df['n_exp_samples'] = df.sum(axis=1)\n",
    "    print(len(df.index))\n",
    "    df = df.loc[df['n_exp_samples']>=n_sample_cutoff]\n",
    "    print(len(df.index))\n",
    "    gids = df.index.tolist()\n",
    "    filt_df['g_passed_exp_min_samples_filt'] = filt_df.gid.isin(gids)\n",
    "    filt_df.to_csv(ofile, sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b9b32f3e-3409-4910-9951-2432bf3375ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "185c283f-b609-4042-acc3-696a1a58e6d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68777\n",
      "12218\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../data/mage/thing.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-072fb13dc408>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0mfilt_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'g_passed_exp_min_samples_filt'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilt_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgid\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0mfilt_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../data/mage/thing.tsv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[1;32m   3480\u001b[0m             \u001b[0mdoublequote\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdoublequote\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3481\u001b[0m             \u001b[0mescapechar\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mescapechar\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3482\u001b[0;31m             \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3483\u001b[0m         )\n\u001b[1;32m   3484\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/format.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[1;32m   1103\u001b[0m             \u001b[0mformatter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfmt\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1104\u001b[0m         )\n\u001b[0;32m-> 1105\u001b[0;31m         \u001b[0mcsv_formatter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1107\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcreated_buffer\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    241\u001b[0m             \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m             \u001b[0mcompression\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompression\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 243\u001b[0;31m             \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    244\u001b[0m         ) as handles:\n\u001b[1;32m    245\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    705\u001b[0m                 \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    706\u001b[0m                 \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 707\u001b[0;31m                 \u001b[0mnewline\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    708\u001b[0m             )\n\u001b[1;32m    709\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../data/mage/thing.tsv'"
     ]
    }
   ],
   "source": [
    "# using just these isoforms, re-compute gene tpm \n",
    "# genes exp. >= 1 TPM in at least 80% of samples\n",
    "n_samples = len([c for c in df.columns.tolist() if c not in ['gid']])\n",
    "n_sample_cutoff = int(n_samples*min_samp_prop)\n",
    "df = pd.read_csv(t_tpm_tsv, sep='\\t')\n",
    "df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "df = df.loc[df.tid.isin(tids)]\n",
    "t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "t2g_df = t2g_df[[0,1]]\n",
    "t2g_df.columns = ['tid', 'gid']\n",
    "df = df.merge(t2g_df, \n",
    "              how='left',\n",
    "              on='tid')\n",
    "df = df.drop('tid', axis=1)\n",
    "df = df.groupby('gid').sum().reset_index()\n",
    "df = df.set_index('gid')\n",
    "df = df>=min_g_exp\n",
    "df['n_exp_samples'] = df.sum(axis=1)\n",
    "print(len(df.index))\n",
    "df = df.loc[df['n_exp_samples']>=n_sample_cutoff]\n",
    "print(len(df.index))\n",
    "gids = df.index.tolist()\n",
    "filt_df['g_passed_exp_min_samples_filt'] = filt_df.gid.isin(gids)\n",
    "filt_df.to_csv(ofile, sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13f1d307-9724-4f74-bbf9-b0ed7915c8f3",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-b2f293b0da21>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfilt_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../../data/mage/thing.tsv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[1;32m   3480\u001b[0m             \u001b[0mdoublequote\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdoublequote\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3481\u001b[0m             \u001b[0mescapechar\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mescapechar\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3482\u001b[0;31m             \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3483\u001b[0m         )\n\u001b[1;32m   3484\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/format.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[1;32m   1103\u001b[0m             \u001b[0mformatter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfmt\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1104\u001b[0m         )\n\u001b[0;32m-> 1105\u001b[0;31m         \u001b[0mcsv_formatter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1107\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcreated_buffer\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    255\u001b[0m             )\n\u001b[1;32m    256\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 257\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    259\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    260\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_need_to_save_header\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_body\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save_body\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    298\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstart_i\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    314\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    315\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 316\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    317\u001b[0m         )\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/_libs/writers.pyx\u001b[0m in \u001b[0;36mpandas._libs.writers.write_csv_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "filt_df.to_csv('../../data/mage/thing.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c249b8c-f598-445b-a75b-5dd5dbd9ecf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>sample</th>\n",
       "      <th>tpm</th>\n",
       "      <th>gid</th>\n",
       "      <th>t_passed_exp_filt</th>\n",
       "      <th>g_passed_n_t_filt</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>g_passed_exp_filt</th>\n",
       "      <th>g_passed_exp_min_samples_filt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>106.12000</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>115.487841</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.8</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>71.25950</td>\n",
       "      <td>ENSG00000003056.8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>105.136776</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.11</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>6.19072</td>\n",
       "      <td>ENSG00000173153.17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12.247247</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>22.14150</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>31.879552</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.7</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>ENSG00000003137.9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.150489</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tid                sample        tpm                 gid  \\\n",
       "0  ENST00000000233.10  NA19704_batch11_rep1  106.12000  ENSG00000004059.11   \n",
       "1   ENST00000000412.8  NA19704_batch11_rep1   71.25950   ENSG00000003056.8   \n",
       "2  ENST00000000442.11  NA19704_batch11_rep1    6.19072  ENSG00000173153.17   \n",
       "3   ENST00000001008.6  NA19704_batch11_rep1   22.14150   ENSG00000004478.8   \n",
       "4   ENST00000001146.7  NA19704_batch11_rep1    0.00000   ENSG00000003137.9   \n",
       "\n",
       "   t_passed_exp_filt  g_passed_n_t_filt    gene_tpm g_passed_exp_filt  \\\n",
       "0               True               True  115.487841              True   \n",
       "1               True               True  105.136776              True   \n",
       "2               True               True   12.247247              True   \n",
       "3               True               True   31.879552              True   \n",
       "4               True               True    0.150489             False   \n",
       "\n",
       "   g_passed_exp_min_samples_filt  \n",
       "0                           True  \n",
       "1                           True  \n",
       "2                           True  \n",
       "3                           True  \n",
       "4                          False  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b6fc70-c06c-4483-8eff-9dcb254e4ff0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d156077b-c061-4650-9e63-2ba1b40e717c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78932\n",
      "12225\n",
      "387944\n"
     ]
    }
   ],
   "source": [
    "g_exp_file = od+config['mage']['v47_kallisto']['gene_tpm_tsv']\n",
    "exp_file = od+config['mage']['v47_kallisto']['merge_matrix_tpm_tsv']\n",
    "t2g_file = od+config['ref']['v47_kallisto_short']['t2g']\n",
    "\n",
    "min_samp_prop = 0.8\n",
    "min_g_exp = 1\n",
    "min_t_exp = 0.1\n",
    "min_n_t = 2\n",
    "\n",
    "filt_sqtl_transcripts(g_exp_file,\n",
    "                          exp_file,\n",
    "                          t2g_file,\n",
    "                          min_samp_prop,\n",
    "                          min_g_exp,\n",
    "                          min_t_exp,\n",
    "                          min_n_t,\n",
    "                          '../../data/mage/test.tsv')\n",
    "\n",
    "# # get melted version of table w/ tid, gid, sample, t_exp, n_t (per gene), passed t exp filt, passed g exp filt\n",
    "# # this is the main output tale\n",
    "# filt_df = pd.read_csv(exp_file, sep='\\t')\n",
    "# filt_df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "# filt_df = filt_df.set_index('tid').melt(ignore_index=False, value_name='tpm', var_name='sample').reset_index()\n",
    "# t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "# t2g_df = t2g_df[[0,1]]\n",
    "# t2g_df.columns = ['tid', 'gid']\n",
    "# filt_df = filt_df.merge(t2g_df, \n",
    "#                         how='left',\n",
    "#                         on='tid')\n",
    "\n",
    "# # genes exp. >= 1 TPM in at least 80% of samples\n",
    "# df = pd.read_csv(g_exp_file, sep='\\t')\n",
    "# n_samples = len([c for c in df.columns.tolist() if c not in ['gid']])\n",
    "# n_sample_cutoff = int(n_samples*min_samp_prop)\n",
    "# df = df.set_index('gid')\n",
    "# df = df>=min_g_exp\n",
    "# df['n_exp_samples'] = df.sum(axis=1)\n",
    "# print(len(df.index))\n",
    "# df = df.loc[df['n_exp_samples']>=n_sample_cutoff]\n",
    "# print(len(df.index))\n",
    "# gids = df.index.tolist()\n",
    "# filt_df['g_passed_exp_filt'] = filt_df.gid.isin(gids)\n",
    "\n",
    "# # isoform expressed in at least 1 sample >= 0.1 TPM\n",
    "# df = pd.read_csv(exp_file, sep='\\t')\n",
    "# df = df.set_index('transcript_id')\n",
    "# df = df>=min_t_exp\n",
    "# df['n_exp_samples'] = df.sum(axis=1)\n",
    "# print(len(df.index))\n",
    "# df = df.loc[df['n_exp_samples']>=1]\n",
    "# print(len(df.index))\n",
    "# tids = df.index.tolist()\n",
    "# filt_df['t_passed_exp_filt'] = filt_df.tid.isin(tids)\n",
    "\n",
    "# # of these g/isoform pairs, which ones pass the n transcripts / gene\n",
    "# t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "# t2g_df = t2g_df[[0,1]]\n",
    "# t2g_df.columns = ['tid', 'gid']\n",
    "\n",
    "# print(len(t2g_df.index))\n",
    "# t2g_df = t2g_df.loc[(t2g_df.gid.isin(gids))&\\\n",
    "#                     (t2g_df.tid.isin(tids))]\n",
    "# print(len(t2g_df.index))\n",
    "\n",
    "# # then count # isos / gene and filter one more time\n",
    "# t2g_df = t2g_df.groupby('gid').nunique().reset_index().rename({'tid':'n_tid'}, axis=1)\n",
    "# print(len(t2g_df.index))\n",
    "# t2g_df = t2g_df.loc[t2g_df.n_tid >= min_n_t]\n",
    "# print(len(t2g_df.index))\n",
    "# gids2 = t2g_df.gid.tolist()\n",
    "\n",
    "# filt_df['g_passed_n_t_filt'] = filt_df.gid.isin(gids2)\n",
    "# filt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "18dd600d-0528-47f5-b657-c5da19114836",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387944\n",
      "154491\n",
      "12225\n",
      "11199\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>sample</th>\n",
       "      <th>tpm</th>\n",
       "      <th>gid</th>\n",
       "      <th>g_passed_exp_filt</th>\n",
       "      <th>t_passed_exp_filt</th>\n",
       "      <th>g_passed_n_t_filt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>106.12000</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.8</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>71.25950</td>\n",
       "      <td>ENSG00000003056.8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.11</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>6.19072</td>\n",
       "      <td>ENSG00000173153.17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>22.14150</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.7</td>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>ENSG00000003137.9</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tid                sample        tpm                 gid  \\\n",
       "0  ENST00000000233.10  NA19704_batch11_rep1  106.12000  ENSG00000004059.11   \n",
       "1   ENST00000000412.8  NA19704_batch11_rep1   71.25950   ENSG00000003056.8   \n",
       "2  ENST00000000442.11  NA19704_batch11_rep1    6.19072  ENSG00000173153.17   \n",
       "3   ENST00000001008.6  NA19704_batch11_rep1   22.14150   ENSG00000004478.8   \n",
       "4   ENST00000001146.7  NA19704_batch11_rep1    0.00000   ENSG00000003137.9   \n",
       "\n",
       "   g_passed_exp_filt  t_passed_exp_filt  g_passed_n_t_filt  \n",
       "0               True               True               True  \n",
       "1               True               True               True  \n",
       "2               True               True               True  \n",
       "3               True               True               True  \n",
       "4              False               True              False  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c34b23-095e-4372-b746-4781e1ffce5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9734c08-9454-4017-85b6-db2fc6fc340e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51421879-5bda-40ae-be3e-ce0f611d2069",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ffbc36-2394-41a8-8ff4-8cf8aa098edb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3db0d163-fbaf-4f71-8812-879d2554d0fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df9f91e4-fb1b-4343-bdc3-25f3f6464a2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a1d2fa16-92fa-4c02-b0fc-2d3b66d32343",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6f7253f6-e298-4da2-b4ec-b3070eba7906",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3369bc65-182f-40ac-a799-3d8df8f770db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA19704_batch11_rep1</td>\n",
       "      <td>NA19704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA19332_batch14_rep1</td>\n",
       "      <td>NA19332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA19317_batch15_rep1</td>\n",
       "      <td>NA19317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA19312_batch05_rep1</td>\n",
       "      <td>NA19312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA19323_batch11_rep1</td>\n",
       "      <td>NA19323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample cell_line_id\n",
       "0  NA19704_batch11_rep1      NA19704\n",
       "1  NA19332_batch14_rep1      NA19332\n",
       "2  NA19317_batch15_rep1      NA19317\n",
       "3  NA19312_batch05_rep1      NA19312\n",
       "4  NA19323_batch11_rep1      NA19323"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['cell_line_id'] = df['sample'].str.split('_', expand=True)[0]\n",
    "df[['sample', 'cell_line_id']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "074967a8-26c2-4c63-a65a-d0393f806334",
   "metadata": {},
   "source": [
    "## Format the file the way sqtlseeker wants\n",
    "- expression of each transcript in each sample TPM\n",
    "- columns trId and geneId"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ec2cbe16-1525-42bb-90e0-4bf3c523d038",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2g_file = od+config['ref']['v47_kallisto_short']['t2g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "25c609e4-c2aa-48b3-993a-d3f988c61d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('test.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "28681eeb-96b6-4f15-95ec-093e2b471fb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>HG03117</th>\n",
       "      <th>NA19785</th>\n",
       "      <th>HG00956</th>\n",
       "      <th>HG01809</th>\n",
       "      <th>HG01970</th>\n",
       "      <th>NA19308</th>\n",
       "      <th>HG01395</th>\n",
       "      <th>HG02941</th>\n",
       "      <th>HG01170</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02574</th>\n",
       "      <th>NA19670</th>\n",
       "      <th>NA18878</th>\n",
       "      <th>HG00331</th>\n",
       "      <th>HG03091</th>\n",
       "      <th>HG00704</th>\n",
       "      <th>HG00705</th>\n",
       "      <th>NA19704</th>\n",
       "      <th>HG04100</th>\n",
       "      <th>HG01515</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>131.3720</td>\n",
       "      <td>112.57500</td>\n",
       "      <td>86.58550</td>\n",
       "      <td>100.4330</td>\n",
       "      <td>119.743000</td>\n",
       "      <td>129.113000</td>\n",
       "      <td>82.80470</td>\n",
       "      <td>94.68860</td>\n",
       "      <td>74.26770</td>\n",
       "      <td>...</td>\n",
       "      <td>121.22200</td>\n",
       "      <td>91.7148</td>\n",
       "      <td>81.51650</td>\n",
       "      <td>120.0510</td>\n",
       "      <td>139.91300</td>\n",
       "      <td>80.60470</td>\n",
       "      <td>97.962600</td>\n",
       "      <td>106.12000</td>\n",
       "      <td>81.89790</td>\n",
       "      <td>70.50260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.8</td>\n",
       "      <td>84.4788</td>\n",
       "      <td>65.65510</td>\n",
       "      <td>68.64080</td>\n",
       "      <td>62.5138</td>\n",
       "      <td>81.477100</td>\n",
       "      <td>81.621700</td>\n",
       "      <td>68.70510</td>\n",
       "      <td>75.74230</td>\n",
       "      <td>71.85250</td>\n",
       "      <td>...</td>\n",
       "      <td>71.67360</td>\n",
       "      <td>81.4719</td>\n",
       "      <td>68.18530</td>\n",
       "      <td>59.4231</td>\n",
       "      <td>74.79620</td>\n",
       "      <td>67.03900</td>\n",
       "      <td>44.633900</td>\n",
       "      <td>71.25950</td>\n",
       "      <td>57.90720</td>\n",
       "      <td>68.57620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.11</td>\n",
       "      <td>10.0302</td>\n",
       "      <td>6.90736</td>\n",
       "      <td>9.12874</td>\n",
       "      <td>12.6433</td>\n",
       "      <td>16.859600</td>\n",
       "      <td>8.200680</td>\n",
       "      <td>8.68647</td>\n",
       "      <td>7.85214</td>\n",
       "      <td>8.61134</td>\n",
       "      <td>...</td>\n",
       "      <td>9.56501</td>\n",
       "      <td>10.0081</td>\n",
       "      <td>8.45307</td>\n",
       "      <td>11.6978</td>\n",
       "      <td>7.45239</td>\n",
       "      <td>5.43539</td>\n",
       "      <td>6.176470</td>\n",
       "      <td>6.19072</td>\n",
       "      <td>8.32944</td>\n",
       "      <td>6.85148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>29.0699</td>\n",
       "      <td>33.85030</td>\n",
       "      <td>20.27160</td>\n",
       "      <td>43.3731</td>\n",
       "      <td>32.359600</td>\n",
       "      <td>20.762800</td>\n",
       "      <td>38.85590</td>\n",
       "      <td>24.10620</td>\n",
       "      <td>32.58370</td>\n",
       "      <td>...</td>\n",
       "      <td>33.44470</td>\n",
       "      <td>32.5083</td>\n",
       "      <td>18.63250</td>\n",
       "      <td>38.4592</td>\n",
       "      <td>26.09600</td>\n",
       "      <td>17.34890</td>\n",
       "      <td>25.585700</td>\n",
       "      <td>22.14150</td>\n",
       "      <td>23.04920</td>\n",
       "      <td>24.06240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.7</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.238889</td>\n",
       "      <td>0.744964</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.013434</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 732 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id   HG03117    NA19785   HG00956   HG01809     HG01970  \\\n",
       "0  ENST00000000233.10  131.3720  112.57500  86.58550  100.4330  119.743000   \n",
       "1   ENST00000000412.8   84.4788   65.65510  68.64080   62.5138   81.477100   \n",
       "2  ENST00000000442.11   10.0302    6.90736   9.12874   12.6433   16.859600   \n",
       "3   ENST00000001008.6   29.0699   33.85030  20.27160   43.3731   32.359600   \n",
       "4   ENST00000001146.7    0.0000    0.00000   0.00000    0.0000    0.238889   \n",
       "\n",
       "      NA19308   HG01395   HG02941   HG01170  ...    HG02574  NA19670  \\\n",
       "0  129.113000  82.80470  94.68860  74.26770  ...  121.22200  91.7148   \n",
       "1   81.621700  68.70510  75.74230  71.85250  ...   71.67360  81.4719   \n",
       "2    8.200680   8.68647   7.85214   8.61134  ...    9.56501  10.0081   \n",
       "3   20.762800  38.85590  24.10620  32.58370  ...   33.44470  32.5083   \n",
       "4    0.744964   0.00000   0.00000   0.00000  ...    0.00000   0.0000   \n",
       "\n",
       "    NA18878   HG00331    HG03091   HG00704    HG00705    NA19704   HG04100  \\\n",
       "0  81.51650  120.0510  139.91300  80.60470  97.962600  106.12000  81.89790   \n",
       "1  68.18530   59.4231   74.79620  67.03900  44.633900   71.25950  57.90720   \n",
       "2   8.45307   11.6978    7.45239   5.43539   6.176470    6.19072   8.32944   \n",
       "3  18.63250   38.4592   26.09600  17.34890  25.585700   22.14150  23.04920   \n",
       "4   0.00000    0.0000    0.00000   0.00000   0.013434    0.00000   0.00000   \n",
       "\n",
       "    HG01515  \n",
       "0  70.50260  \n",
       "1  68.57620  \n",
       "2   6.85148  \n",
       "3  24.06240  \n",
       "4   0.00000  \n",
       "\n",
       "[5 rows x 732 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "48ed69ee-3953-42e9-bdbb-b7d2adabf3d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename({'transcript_id':'trId'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "10ef0a51-47a2-41dd-a098-a583eb71fc14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>trId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.3</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832823.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneId               trId\n",
       "0  ENST00000456328.3  ENSG00000290825.2\n",
       "1  ENST00000832823.1  ENSG00000290825.2\n",
       "2  ENST00000832824.1  ENSG00000290825.2\n",
       "3  ENST00000832825.1  ENSG00000290825.2\n",
       "4  ENST00000832826.1  ENSG00000290825.2"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2g_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "669b1db6-5dd7-4e14-9390-ddd231fb4b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "t2g_df = t2g_df[[0,1]]\n",
    "t2g_df.columns = ['trId', 'geneId']\n",
    "df = df.merge(t2g_df, how='left', \n",
    "              on='trId')\n",
    "assert len(df.loc[df.geneId.isnull()].index)==0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "58750a78-ef65-40e5-86a7-0fb919dc5f7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trId</th>\n",
       "      <th>HG03117</th>\n",
       "      <th>NA19785</th>\n",
       "      <th>HG00956</th>\n",
       "      <th>HG01809</th>\n",
       "      <th>HG01970</th>\n",
       "      <th>NA19308</th>\n",
       "      <th>HG01395</th>\n",
       "      <th>HG02941</th>\n",
       "      <th>HG01170</th>\n",
       "      <th>...</th>\n",
       "      <th>NA18878</th>\n",
       "      <th>HG00331</th>\n",
       "      <th>HG03091</th>\n",
       "      <th>HG00704</th>\n",
       "      <th>HG00705</th>\n",
       "      <th>NA19704</th>\n",
       "      <th>HG04100</th>\n",
       "      <th>HG01515</th>\n",
       "      <th>geneId_x</th>\n",
       "      <th>geneId_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 734 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [trId, HG03117, NA19785, HG00956, HG01809, HG01970, NA19308, HG01395, HG02941, HG01170, HG01988, HG00285, HG03082, NA21120, HG03129, HG03900, HG01682, HG01870, NA18874, NA12828, HG00337, HG04185, HG02481, NA20532, HG03695, NA19774, HG00473, NA20775, NA19081, HG03696, HG03998, HG02733, NA20763, HG01630, HG00740, HG02731, HG01204, HG01113, NA19116, HG00342, NA18507, NA19323, NA12778, HG03885, NA18504, HG01953, HG03864, HG00589, HG01092, NA19062, NA19143, NA19332, HG02545, HG01593, NA18555, NA20866, HG02136, NA19401, HG02981, HG01844, HG01518, HG00132, HG00631, HG03727, NA18948, HG02232, HG01607, HG02643, NA20342, HG00320, HG03989, NA20294, NA18877, HG02102, HG02715, NA12046, HG03058, NA19449, HG02613, HG01251, NA18881, HG00260, NA18934, HG00556, NA18997, NA18963, HG01610, NA19711, HG02364, NA20334, HG04188, NA12399, HG02611, HG00590, HG00188, NA20878, NA20762, NA18541, NA19129, HG01479, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 734 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_csv(, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac4da8ef-58aa-4c1a-bfb9-9f062f691244",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da24080-aa4f-440a-94ab-ad206d6e312c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4914f369-21d7-44a6-acdf-07f928638d71",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e31fcada-5977-44db-86ea-08442fcf58ba",
   "metadata": {},
   "source": [
    "## Filter samples w/ replicates to one w/ highest read depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e2a475a1-f3cc-43c2-b28d-6afd201fa05a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "779\n",
      "731\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(exp_file, sep='\\t')\n",
    "df = df.set_index('transcript_id')\n",
    "df = df.transpose()\n",
    "\n",
    "# counts / sample\n",
    "df = df.reset_index()\n",
    "df = df.rename({'index':'sample'}, axis=1)\n",
    "df['cell_line_id'] = df['sample'].str.split('_', expand=True)[0]\n",
    "df = df.set_index(['cell_line_id', 'sample'])\n",
    "df['total_counts'] = df.sum(axis=1)\n",
    "df = df[['total_counts']].reset_index()\n",
    "\n",
    "# dedupe on total counts\n",
    "df = df.sort_values(by='total_counts', ascending=False)\n",
    "print(len(df.index))\n",
    "df = df.drop_duplicates(subset=['cell_line_id'], keep='first')\n",
    "print(len(df.index))\n",
    "\n",
    "df = df[['sample']]\n",
    "df.to_csv('keep_samples.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "39d2971e-f792-42dd-9525-153884295301",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_file = od+config['mage']['v47_kallisto']['merge_matrix_tsv']\n",
    "df = pd.read_csv(exp_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d706f9d6-53e9-4b5d-b273-438d09c408f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.set_index('transcript_id')\n",
    "df = df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "79f3db92-2f96-4844-8fda-8a81188c91f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ENST00000000233.10</th>\n",
       "      <th>ENST00000000412.8</th>\n",
       "      <th>ENST00000000442.11</th>\n",
       "      <th>ENST00000001008.6</th>\n",
       "      <th>ENST00000001146.7</th>\n",
       "      <th>ENST00000002125.9</th>\n",
       "      <th>ENST00000002165.11</th>\n",
       "      <th>ENST00000002501.11</th>\n",
       "      <th>ENST00000002596.6</th>\n",
       "      <th>ENST00000002829.8</th>\n",
       "      <th>...</th>\n",
       "      <th>ENST00000850835.1</th>\n",
       "      <th>ENST00000850836.1</th>\n",
       "      <th>ENST00000850837.1</th>\n",
       "      <th>ENST00000850838.1</th>\n",
       "      <th>ENST00000850839.1</th>\n",
       "      <th>ENST00000850840.1</th>\n",
       "      <th>ENST00000850841.1</th>\n",
       "      <th>ENST00000850842.1</th>\n",
       "      <th>ENST00000850843.1</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA19704</th>\n",
       "      <th>NA19704_batch11_rep1</th>\n",
       "      <td>2187.66</td>\n",
       "      <td>4203.78</td>\n",
       "      <td>335.718</td>\n",
       "      <td>2064.24</td>\n",
       "      <td>0.0</td>\n",
       "      <td>322.110</td>\n",
       "      <td>992.572</td>\n",
       "      <td>915.7260</td>\n",
       "      <td>15.0000</td>\n",
       "      <td>20.0654</td>\n",
       "      <td>...</td>\n",
       "      <td>25.9881</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.487025e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19332</th>\n",
       "      <th>NA19332_batch14_rep1</th>\n",
       "      <td>2483.13</td>\n",
       "      <td>5302.25</td>\n",
       "      <td>603.101</td>\n",
       "      <td>1585.80</td>\n",
       "      <td>0.0</td>\n",
       "      <td>176.919</td>\n",
       "      <td>1536.970</td>\n",
       "      <td>158.6100</td>\n",
       "      <td>39.5692</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>11.7222</td>\n",
       "      <td>1.41875</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.93641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.921029</td>\n",
       "      <td>2.944093e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19317</th>\n",
       "      <th>NA19317_batch15_rep1</th>\n",
       "      <td>1081.26</td>\n",
       "      <td>3306.58</td>\n",
       "      <td>307.345</td>\n",
       "      <td>1706.69</td>\n",
       "      <td>0.0</td>\n",
       "      <td>169.047</td>\n",
       "      <td>757.023</td>\n",
       "      <td>614.3550</td>\n",
       "      <td>60.2248</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>12.7007</td>\n",
       "      <td>2.52598</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.17823</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.19724</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.011330e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19312</th>\n",
       "      <th>NA19312_batch05_rep1</th>\n",
       "      <td>4053.85</td>\n",
       "      <td>6273.98</td>\n",
       "      <td>729.524</td>\n",
       "      <td>3544.72</td>\n",
       "      <td>0.0</td>\n",
       "      <td>314.616</td>\n",
       "      <td>1372.780</td>\n",
       "      <td>2229.5300</td>\n",
       "      <td>31.6113</td>\n",
       "      <td>1.7321</td>\n",
       "      <td>...</td>\n",
       "      <td>70.3466</td>\n",
       "      <td>2.18983</td>\n",
       "      <td>4.43783</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.06990</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.75674</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.238001e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19323</th>\n",
       "      <th>NA19323_batch11_rep1</th>\n",
       "      <td>2051.01</td>\n",
       "      <td>3338.97</td>\n",
       "      <td>310.898</td>\n",
       "      <td>2224.32</td>\n",
       "      <td>0.0</td>\n",
       "      <td>218.433</td>\n",
       "      <td>539.937</td>\n",
       "      <td>81.6014</td>\n",
       "      <td>15.0000</td>\n",
       "      <td>18.7386</td>\n",
       "      <td>...</td>\n",
       "      <td>25.7538</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.90098</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.393270e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 387945 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "transcript_id                      ENST00000000233.10  ENST00000000412.8  \\\n",
       "cell_line_id sample                                                        \n",
       "NA19704      NA19704_batch11_rep1             2187.66            4203.78   \n",
       "NA19332      NA19332_batch14_rep1             2483.13            5302.25   \n",
       "NA19317      NA19317_batch15_rep1             1081.26            3306.58   \n",
       "NA19312      NA19312_batch05_rep1             4053.85            6273.98   \n",
       "NA19323      NA19323_batch11_rep1             2051.01            3338.97   \n",
       "\n",
       "transcript_id                      ENST00000000442.11  ENST00000001008.6  \\\n",
       "cell_line_id sample                                                        \n",
       "NA19704      NA19704_batch11_rep1             335.718            2064.24   \n",
       "NA19332      NA19332_batch14_rep1             603.101            1585.80   \n",
       "NA19317      NA19317_batch15_rep1             307.345            1706.69   \n",
       "NA19312      NA19312_batch05_rep1             729.524            3544.72   \n",
       "NA19323      NA19323_batch11_rep1             310.898            2224.32   \n",
       "\n",
       "transcript_id                      ENST00000001146.7  ENST00000002125.9  \\\n",
       "cell_line_id sample                                                       \n",
       "NA19704      NA19704_batch11_rep1                0.0            322.110   \n",
       "NA19332      NA19332_batch14_rep1                0.0            176.919   \n",
       "NA19317      NA19317_batch15_rep1                0.0            169.047   \n",
       "NA19312      NA19312_batch05_rep1                0.0            314.616   \n",
       "NA19323      NA19323_batch11_rep1                0.0            218.433   \n",
       "\n",
       "transcript_id                      ENST00000002165.11  ENST00000002501.11  \\\n",
       "cell_line_id sample                                                         \n",
       "NA19704      NA19704_batch11_rep1             992.572            915.7260   \n",
       "NA19332      NA19332_batch14_rep1            1536.970            158.6100   \n",
       "NA19317      NA19317_batch15_rep1             757.023            614.3550   \n",
       "NA19312      NA19312_batch05_rep1            1372.780           2229.5300   \n",
       "NA19323      NA19323_batch11_rep1             539.937             81.6014   \n",
       "\n",
       "transcript_id                      ENST00000002596.6  ENST00000002829.8  ...  \\\n",
       "cell_line_id sample                                                      ...   \n",
       "NA19704      NA19704_batch11_rep1            15.0000            20.0654  ...   \n",
       "NA19332      NA19332_batch14_rep1            39.5692             0.0000  ...   \n",
       "NA19317      NA19317_batch15_rep1            60.2248             0.0000  ...   \n",
       "NA19312      NA19312_batch05_rep1            31.6113             1.7321  ...   \n",
       "NA19323      NA19323_batch11_rep1            15.0000            18.7386  ...   \n",
       "\n",
       "transcript_id                      ENST00000850835.1  ENST00000850836.1  \\\n",
       "cell_line_id sample                                                       \n",
       "NA19704      NA19704_batch11_rep1            25.9881            0.00000   \n",
       "NA19332      NA19332_batch14_rep1            11.7222            1.41875   \n",
       "NA19317      NA19317_batch15_rep1            12.7007            2.52598   \n",
       "NA19312      NA19312_batch05_rep1            70.3466            2.18983   \n",
       "NA19323      NA19323_batch11_rep1            25.7538            0.00000   \n",
       "\n",
       "transcript_id                      ENST00000850837.1  ENST00000850838.1  \\\n",
       "cell_line_id sample                                                       \n",
       "NA19704      NA19704_batch11_rep1            0.00000                0.0   \n",
       "NA19332      NA19332_batch14_rep1            0.00000                0.0   \n",
       "NA19317      NA19317_batch15_rep1            0.00000                0.0   \n",
       "NA19312      NA19312_batch05_rep1            4.43783                0.0   \n",
       "NA19323      NA19323_batch11_rep1            0.00000                0.0   \n",
       "\n",
       "transcript_id                      ENST00000850839.1  ENST00000850840.1  \\\n",
       "cell_line_id sample                                                       \n",
       "NA19704      NA19704_batch11_rep1            0.00000                0.0   \n",
       "NA19332      NA19332_batch14_rep1            6.93641                0.0   \n",
       "NA19317      NA19317_batch15_rep1            4.17823                0.0   \n",
       "NA19312      NA19312_batch05_rep1           11.06990                0.0   \n",
       "NA19323      NA19323_batch11_rep1            1.90098                0.0   \n",
       "\n",
       "transcript_id                      ENST00000850841.1  ENST00000850842.1  \\\n",
       "cell_line_id sample                                                       \n",
       "NA19704      NA19704_batch11_rep1            0.00000                0.0   \n",
       "NA19332      NA19332_batch14_rep1            0.00000                0.0   \n",
       "NA19317      NA19317_batch15_rep1            1.19724                0.0   \n",
       "NA19312      NA19312_batch05_rep1            8.75674                0.0   \n",
       "NA19323      NA19323_batch11_rep1            0.00000                0.0   \n",
       "\n",
       "transcript_id                      ENST00000850843.1  total_counts  \n",
       "cell_line_id sample                                                 \n",
       "NA19704      NA19704_batch11_rep1           0.000000  2.487025e+07  \n",
       "NA19332      NA19332_batch14_rep1           0.921029  2.944093e+07  \n",
       "NA19317      NA19317_batch15_rep1           0.000000  2.011330e+07  \n",
       "NA19312      NA19312_batch05_rep1           0.000000  4.238001e+07  \n",
       "NA19323      NA19323_batch11_rep1           0.000000  2.393270e+07  \n",
       "\n",
       "[5 rows x 387945 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# counts / sample\n",
    "df = df.reset_index()\n",
    "df = df.rename({'index':'sample'}, axis=1)\n",
    "df['cell_line_id'] = df['sample'].str.split('_', expand=True)[0]\n",
    "df = df.set_index(['cell_line_id', 'sample'])\n",
    "df['total_counts'] = df.sum(axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "03bf312b-15a1-4c4b-b44f-470a3a047492",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['total_counts']].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e237e8c8-aaa5-4d9e-be93-837536f91a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "779\n"
     ]
    }
   ],
   "source": [
    "df = df.sort_values(by='total_counts', ascending=False)\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3edef0dd-0571-4d34-941e-3721ac1514d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>transcript_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NA19143</td>\n",
       "      <td>NA19143_batch15_rep2</td>\n",
       "      <td>2.029975e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>NA19143</td>\n",
       "      <td>NA19143_batch15_rep1</td>\n",
       "      <td>1.846667e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NA19143</td>\n",
       "      <td>NA19143_batch04_rep3</td>\n",
       "      <td>1.812086e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "transcript_id cell_line_id                sample  total_counts\n",
       "6                  NA19143  NA19143_batch15_rep2  2.029975e+07\n",
       "509                NA19143  NA19143_batch15_rep1  1.846667e+07\n",
       "5                  NA19143  NA19143_batch04_rep3  1.812086e+07"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.cell_line_id=='NA19143']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4b0774ad-e1cc-4a9b-9e67-6d38c5d79192",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "731\n"
     ]
    }
   ],
   "source": [
    "df = df.drop_duplicates(subset=['cell_line_id'], keep='first')\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "17812a04-206a-4498-b3e2-bbb8fc5a33b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>transcript_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NA19143</td>\n",
       "      <td>NA19143_batch15_rep2</td>\n",
       "      <td>2.029975e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "transcript_id cell_line_id                sample  total_counts\n",
       "6                  NA19143  NA19143_batch15_rep2  2.029975e+07"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.cell_line_id=='NA19143']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "049a643d-c222-4623-a98c-862f27c2961c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['sample']]\n",
    "df.to_csv('keep_samples.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f66f638b-5e55-43f5-89c5-e1bbfec36936",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now filter\n",
    "s_df = pd.read_csv('keep_samples.tsv', sep='\\t')\n",
    "df = pd.read_csv(exp_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "b5b96651-0306-4477-b0d9-e46e63663b40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HG03117_batch07_rep1',\n",
       " 'NA19785_batch07_rep1',\n",
       " 'HG00956_batch11_rep1',\n",
       " 'HG01809_batch09_rep1',\n",
       " 'HG01970_batch02_rep1']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = s_df['sample'].tolist()\n",
    "samples[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "115f4888-9e87-49d3-900b-16d7019a6db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ind_cols = [c for c in df.columns if 'id' in c]\n",
    "df = df.set_index(ind_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "02791b49-0649-428f-addc-9904d4f7fcee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>HG03117</th>\n",
       "      <th>NA19785</th>\n",
       "      <th>HG00956</th>\n",
       "      <th>HG01809</th>\n",
       "      <th>HG01970</th>\n",
       "      <th>NA19308</th>\n",
       "      <th>HG01395</th>\n",
       "      <th>HG02941</th>\n",
       "      <th>HG01170</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02574</th>\n",
       "      <th>NA19670</th>\n",
       "      <th>NA18878</th>\n",
       "      <th>HG00331</th>\n",
       "      <th>HG03091</th>\n",
       "      <th>HG00704</th>\n",
       "      <th>HG00705</th>\n",
       "      <th>NA19704</th>\n",
       "      <th>HG04100</th>\n",
       "      <th>HG01515</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>131.3720</td>\n",
       "      <td>112.57500</td>\n",
       "      <td>86.58550</td>\n",
       "      <td>100.4330</td>\n",
       "      <td>119.743000</td>\n",
       "      <td>129.113000</td>\n",
       "      <td>82.80470</td>\n",
       "      <td>94.68860</td>\n",
       "      <td>74.26770</td>\n",
       "      <td>...</td>\n",
       "      <td>121.22200</td>\n",
       "      <td>91.7148</td>\n",
       "      <td>81.51650</td>\n",
       "      <td>120.0510</td>\n",
       "      <td>139.91300</td>\n",
       "      <td>80.60470</td>\n",
       "      <td>97.962600</td>\n",
       "      <td>106.12000</td>\n",
       "      <td>81.89790</td>\n",
       "      <td>70.50260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.8</td>\n",
       "      <td>84.4788</td>\n",
       "      <td>65.65510</td>\n",
       "      <td>68.64080</td>\n",
       "      <td>62.5138</td>\n",
       "      <td>81.477100</td>\n",
       "      <td>81.621700</td>\n",
       "      <td>68.70510</td>\n",
       "      <td>75.74230</td>\n",
       "      <td>71.85250</td>\n",
       "      <td>...</td>\n",
       "      <td>71.67360</td>\n",
       "      <td>81.4719</td>\n",
       "      <td>68.18530</td>\n",
       "      <td>59.4231</td>\n",
       "      <td>74.79620</td>\n",
       "      <td>67.03900</td>\n",
       "      <td>44.633900</td>\n",
       "      <td>71.25950</td>\n",
       "      <td>57.90720</td>\n",
       "      <td>68.57620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.11</td>\n",
       "      <td>10.0302</td>\n",
       "      <td>6.90736</td>\n",
       "      <td>9.12874</td>\n",
       "      <td>12.6433</td>\n",
       "      <td>16.859600</td>\n",
       "      <td>8.200680</td>\n",
       "      <td>8.68647</td>\n",
       "      <td>7.85214</td>\n",
       "      <td>8.61134</td>\n",
       "      <td>...</td>\n",
       "      <td>9.56501</td>\n",
       "      <td>10.0081</td>\n",
       "      <td>8.45307</td>\n",
       "      <td>11.6978</td>\n",
       "      <td>7.45239</td>\n",
       "      <td>5.43539</td>\n",
       "      <td>6.176470</td>\n",
       "      <td>6.19072</td>\n",
       "      <td>8.32944</td>\n",
       "      <td>6.85148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>29.0699</td>\n",
       "      <td>33.85030</td>\n",
       "      <td>20.27160</td>\n",
       "      <td>43.3731</td>\n",
       "      <td>32.359600</td>\n",
       "      <td>20.762800</td>\n",
       "      <td>38.85590</td>\n",
       "      <td>24.10620</td>\n",
       "      <td>32.58370</td>\n",
       "      <td>...</td>\n",
       "      <td>33.44470</td>\n",
       "      <td>32.5083</td>\n",
       "      <td>18.63250</td>\n",
       "      <td>38.4592</td>\n",
       "      <td>26.09600</td>\n",
       "      <td>17.34890</td>\n",
       "      <td>25.585700</td>\n",
       "      <td>22.14150</td>\n",
       "      <td>23.04920</td>\n",
       "      <td>24.06240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.7</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.238889</td>\n",
       "      <td>0.744964</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.013434</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 732 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id   HG03117    NA19785   HG00956   HG01809     HG01970  \\\n",
       "0  ENST00000000233.10  131.3720  112.57500  86.58550  100.4330  119.743000   \n",
       "1   ENST00000000412.8   84.4788   65.65510  68.64080   62.5138   81.477100   \n",
       "2  ENST00000000442.11   10.0302    6.90736   9.12874   12.6433   16.859600   \n",
       "3   ENST00000001008.6   29.0699   33.85030  20.27160   43.3731   32.359600   \n",
       "4   ENST00000001146.7    0.0000    0.00000   0.00000    0.0000    0.238889   \n",
       "\n",
       "      NA19308   HG01395   HG02941   HG01170  ...    HG02574  NA19670  \\\n",
       "0  129.113000  82.80470  94.68860  74.26770  ...  121.22200  91.7148   \n",
       "1   81.621700  68.70510  75.74230  71.85250  ...   71.67360  81.4719   \n",
       "2    8.200680   8.68647   7.85214   8.61134  ...    9.56501  10.0081   \n",
       "3   20.762800  38.85590  24.10620  32.58370  ...   33.44470  32.5083   \n",
       "4    0.744964   0.00000   0.00000   0.00000  ...    0.00000   0.0000   \n",
       "\n",
       "    NA18878   HG00331    HG03091   HG00704    HG00705    NA19704   HG04100  \\\n",
       "0  81.51650  120.0510  139.91300  80.60470  97.962600  106.12000  81.89790   \n",
       "1  68.18530   59.4231   74.79620  67.03900  44.633900   71.25950  57.90720   \n",
       "2   8.45307   11.6978    7.45239   5.43539   6.176470    6.19072   8.32944   \n",
       "3  18.63250   38.4592   26.09600  17.34890  25.585700   22.14150  23.04920   \n",
       "4   0.00000    0.0000    0.00000   0.00000   0.013434    0.00000   0.00000   \n",
       "\n",
       "    HG01515  \n",
       "0  70.50260  \n",
       "1  68.57620  \n",
       "2   6.85148  \n",
       "3  24.06240  \n",
       "4   0.00000  \n",
       "\n",
       "[5 rows x 732 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[samples]\n",
    "df.columns = [c.split('_')[0] for c in df.columns]\n",
    "df = df.reset_index()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "193489c4-ad81-49c5-8e8f-1b2673a890c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "732"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1d9c9740-9de7-4d7b-ac54-b1b64cd3aa07",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('test.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758e930c-f187-476a-af1a-d4b2e39740e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "70de2733-6a00-45b0-beaa-d12dfa5975d7",
   "metadata": {},
   "source": [
    "## Filter genes / transcripts used to perform sQTL calling\n",
    "\n",
    "- genes expressed ≥ 1 TPM in at least 80% of the samples\n",
    "- with at least two isoforms and a minimum isoform expression of 0.1 TPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "dc3a1b48-d6ed-4b02-9652-d9061bb1ea54",
   "metadata": {},
   "outputs": [],
   "source": [
    "g_exp_file = od+config['mage']['v47_kallisto']['gene_tpm_tsv']\n",
    "exp_file = od+config['mage']['v47_kallisto']['merge_matrix_tpm_tsv']\n",
    "t2g_file = od+config['ref']['v47_kallisto_short']['t2g']\n",
    "\n",
    "min_samp_prop = 0.8\n",
    "min_g_exp = 1\n",
    "min_t_exp = 0.1\n",
    "min_n_t = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2ebc7dba-04d2-4206-b3aa-1312bfcb0348",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78932\n",
      "12225\n"
     ]
    }
   ],
   "source": [
    "# genes exp. >= 1 TPM in at least 80% of samples\n",
    "df = pd.read_csv(g_exp_file, sep='\\t')\n",
    "n_samples = len([c for c in df.columns.tolist() if c not in ['gid']])\n",
    "n_sample_cutoff = int(n_samples*min_samp_prop)\n",
    "\n",
    "df = df.set_index('gid')\n",
    "df = df>=min_g_exp\n",
    "df['n_exp_samples'] = df.sum(axis=1)\n",
    "print(len(df.index))\n",
    "df = df.loc[df['n_exp_samples']>=n_sample_cutoff]\n",
    "print(len(df.index))\n",
    "gids = df.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8d824723-302e-478d-bf55-253c2807b038",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387944\n",
      "315312\n"
     ]
    }
   ],
   "source": [
    "# with at least two isoforms and\n",
    "# minimum isoform expression of 0.1 TPM\n",
    "# genes exp. >= 1 TPM in at least 80% of samples\n",
    "df = pd.read_csv(exp_file, sep='\\t')\n",
    "df = df.set_index('transcript_id')\n",
    "df = df>=min_t_exp\n",
    "df['n_exp_samples'] = df.sum(axis=1)\n",
    "print(len(df.index))\n",
    "df = df.loc[df['n_exp_samples']>=1]\n",
    "print(len(df.index))\n",
    "tids = df.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "63528ff5-63b7-46b3-9263-82f9c449bb85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387944\n",
      "154491\n"
     ]
    }
   ],
   "source": [
    "# limit to tids and gids\n",
    "t2g_df = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "t2g_df = t2g_df[[0,1]]\n",
    "t2g_df.columns = ['tid', 'gid']\n",
    "\n",
    "print(len(t2g_df.index))\n",
    "t2g_df = t2g_df.loc[(t2g_df.gid.isin(gids))&\\\n",
    "                    (t2g_df.tid.isin(tids))]\n",
    "print(len(t2g_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f8d45531-e1c1-42db-9348-2410d73c746e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12225\n",
      "11199\n"
     ]
    }
   ],
   "source": [
    "# then count # isos / gene and filter one more time\n",
    "t2g_df = t2g_df.groupby('gid').nunique().reset_index().rename({'tid':'n_tid'}, axis=1)\n",
    "print(len(t2g_df.index))\n",
    "t2g_df = t2g_df.loc[t2g_df.n_tid >= min_n_t]\n",
    "print(len(t2g_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cc1e6413-4d00-45fa-b29f-448e899ec1a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# finally, re-read the transcript tpm exp and filt.\n",
    "df = pd.read_csv(exp_file, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc91b8bb-b843-4add-8e12-2f063955f3c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "584e4a45-248f-48cb-9f94-39be97a79893",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e1ad175-0b54-4514-9c7f-ac1d70cbaffc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0f7d6c07-ace6-4d94-82e3-438f9f3c97ba",
   "metadata": {},
   "source": [
    "## Get gene expression values (TPM and counts) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f816f69c-36da-4e70-a0bf-d6fe66fddd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tpm_df = pd.read_csv(od+config['mage']['v47_kallisto']['merge_matrix_tpm_tsv'], \n",
    "#                         sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8651bea0-fc86-4194-9259-fdf7b4e6f315",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_file = od+config['mage']['v47_kallisto']['merge_matrix_tsv']\n",
    "t2g_file = od+config['ref']['v47_kallisto_short']['t2g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "63ecf94b-25d0-4324-b096-af0b5bc411d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.3</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832823.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id                gid\n",
       "0  ENST00000456328.3  ENSG00000290825.2\n",
       "1  ENST00000832823.1  ENSG00000290825.2\n",
       "2  ENST00000832824.1  ENSG00000290825.2\n",
       "3  ENST00000832825.1  ENSG00000290825.2\n",
       "4  ENST00000832826.1  ENSG00000290825.2"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_df = pd.read_csv(exp_file, sep='\\t')\n",
    "t2g = pd.read_csv(t2g_file, sep='\\t', header=None)\n",
    "t2g = t2g[[0,1]]\n",
    "t2g.columns = ['transcript_id', 'gid']\n",
    "t2g.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "9424637a-e192-4541-92a8-ea730ba70999",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene assignments for each transcript using t2g file\n",
    "exp_df = exp_df.merge(t2g[['gid', 'transcript_id']],\n",
    "              how='left', \n",
    "              on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "22e224fa-2a2e-427b-b8fc-d8bd776eeb81",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(exp_df.loc[exp_df.gid.isnull()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f4abe96c-8452-4a0b-8019-68351223fec1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>NA19704_batch11_rep1</th>\n",
       "      <th>NA19332_batch14_rep1</th>\n",
       "      <th>NA19317_batch15_rep1</th>\n",
       "      <th>NA19312_batch05_rep1</th>\n",
       "      <th>NA19323_batch11_rep1</th>\n",
       "      <th>NA19143_batch04_rep3</th>\n",
       "      <th>NA19143_batch15_rep2</th>\n",
       "      <th>HG00326_batch08_rep1</th>\n",
       "      <th>NA18942_batch15_rep1</th>\n",
       "      <th>...</th>\n",
       "      <th>NA19731_batch09_rep1</th>\n",
       "      <th>NA19732_batch14_rep1</th>\n",
       "      <th>HG00525_batch10_rep1</th>\n",
       "      <th>HG00473_batch03_rep1</th>\n",
       "      <th>NA21127_batch04_rep1</th>\n",
       "      <th>NA20342_batch15_rep1</th>\n",
       "      <th>NA20412_batch07_rep1</th>\n",
       "      <th>NA20318_batch07_rep1</th>\n",
       "      <th>NA20298_batch01_rep1</th>\n",
       "      <th>NA20320_batch05_rep1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>5.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>14.000527</td>\n",
       "      <td>45.00000</td>\n",
       "      <td>88.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>27.00000</td>\n",
       "      <td>14.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>12.00002</td>\n",
       "      <td>12.99999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>2334.455390</td>\n",
       "      <td>1795.54352</td>\n",
       "      <td>1400.40445</td>\n",
       "      <td>2132.889530</td>\n",
       "      <td>2103.78422</td>\n",
       "      <td>1360.08493</td>\n",
       "      <td>1276.42764</td>\n",
       "      <td>1472.713011</td>\n",
       "      <td>1397.217400</td>\n",
       "      <td>...</td>\n",
       "      <td>1551.96421</td>\n",
       "      <td>2235.43711</td>\n",
       "      <td>1502.48210</td>\n",
       "      <td>1004.45766</td>\n",
       "      <td>962.10114</td>\n",
       "      <td>1229.19650</td>\n",
       "      <td>1735.522800</td>\n",
       "      <td>2432.45182</td>\n",
       "      <td>1369.26527</td>\n",
       "      <td>2413.30774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>225.561370</td>\n",
       "      <td>257.59018</td>\n",
       "      <td>165.35980</td>\n",
       "      <td>351.460200</td>\n",
       "      <td>188.85600</td>\n",
       "      <td>212.49610</td>\n",
       "      <td>271.21170</td>\n",
       "      <td>215.615020</td>\n",
       "      <td>236.790578</td>\n",
       "      <td>...</td>\n",
       "      <td>340.20660</td>\n",
       "      <td>249.75316</td>\n",
       "      <td>162.36500</td>\n",
       "      <td>112.53656</td>\n",
       "      <td>142.32090</td>\n",
       "      <td>265.54570</td>\n",
       "      <td>160.061090</td>\n",
       "      <td>239.53843</td>\n",
       "      <td>165.29459</td>\n",
       "      <td>344.41416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>309.438419</td>\n",
       "      <td>347.45108</td>\n",
       "      <td>207.64022</td>\n",
       "      <td>458.540230</td>\n",
       "      <td>310.14371</td>\n",
       "      <td>378.50431</td>\n",
       "      <td>267.78767</td>\n",
       "      <td>230.385370</td>\n",
       "      <td>230.208920</td>\n",
       "      <td>...</td>\n",
       "      <td>352.79262</td>\n",
       "      <td>597.24767</td>\n",
       "      <td>300.63499</td>\n",
       "      <td>108.46337</td>\n",
       "      <td>218.67910</td>\n",
       "      <td>150.45407</td>\n",
       "      <td>229.938904</td>\n",
       "      <td>226.46133</td>\n",
       "      <td>294.70550</td>\n",
       "      <td>326.58617</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 780 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gid  NA19704_batch11_rep1  NA19332_batch14_rep1  \\\n",
       "0  ENSG00000000003.16             31.000000               5.00000   \n",
       "1   ENSG00000000005.6              0.000000               0.00000   \n",
       "2  ENSG00000000419.14           2334.455390            1795.54352   \n",
       "3  ENSG00000000457.14            225.561370             257.59018   \n",
       "4  ENSG00000000460.17            309.438419             347.45108   \n",
       "\n",
       "   NA19317_batch15_rep1  NA19312_batch05_rep1  NA19323_batch11_rep1  \\\n",
       "0               0.00000             14.000527              45.00000   \n",
       "1               0.00000              0.000000               0.00000   \n",
       "2            1400.40445           2132.889530            2103.78422   \n",
       "3             165.35980            351.460200             188.85600   \n",
       "4             207.64022            458.540230             310.14371   \n",
       "\n",
       "   NA19143_batch04_rep3  NA19143_batch15_rep2  HG00326_batch08_rep1  \\\n",
       "0              88.00000               0.00000             13.000000   \n",
       "1               0.00000               0.00000              0.000000   \n",
       "2            1360.08493            1276.42764           1472.713011   \n",
       "3             212.49610             271.21170            215.615020   \n",
       "4             378.50431             267.78767            230.385370   \n",
       "\n",
       "   NA18942_batch15_rep1  ...  NA19731_batch09_rep1  NA19732_batch14_rep1  \\\n",
       "0              0.000000  ...               0.00000               2.00000   \n",
       "1              0.000000  ...               0.00000               0.00000   \n",
       "2           1397.217400  ...            1551.96421            2235.43711   \n",
       "3            236.790578  ...             340.20660             249.75316   \n",
       "4            230.208920  ...             352.79262             597.24767   \n",
       "\n",
       "   HG00525_batch10_rep1  HG00473_batch03_rep1  NA21127_batch04_rep1  \\\n",
       "0               1.00000              27.00000              14.00000   \n",
       "1               0.00000               0.00000               0.00000   \n",
       "2            1502.48210            1004.45766             962.10114   \n",
       "3             162.36500             112.53656             142.32090   \n",
       "4             300.63499             108.46337             218.67910   \n",
       "\n",
       "   NA20342_batch15_rep1  NA20412_batch07_rep1  NA20318_batch07_rep1  \\\n",
       "0               1.00000             13.000000              28.00000   \n",
       "1               0.00000              0.000000               0.00000   \n",
       "2            1229.19650           1735.522800            2432.45182   \n",
       "3             265.54570            160.061090             239.53843   \n",
       "4             150.45407            229.938904             226.46133   \n",
       "\n",
       "   NA20298_batch01_rep1  NA20320_batch05_rep1  \n",
       "0              12.00002              12.99999  \n",
       "1               0.00000               0.00000  \n",
       "2            1369.26527            2413.30774  \n",
       "3             165.29459             344.41416  \n",
       "4             294.70550             326.58617  \n",
       "\n",
       "[5 rows x 780 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop tid, gb on gid and sum\n",
    "exp_df.drop('transcript_id', axis=1, inplace=True)\n",
    "exp_df = exp_df.groupby('gid').sum().reset_index()\n",
    "exp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "010e376c-7ed0-4a24-b6d3-8750c7564f64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>NA19704_batch11_rep1</th>\n",
       "      <th>NA19332_batch14_rep1</th>\n",
       "      <th>NA19317_batch15_rep1</th>\n",
       "      <th>NA19312_batch05_rep1</th>\n",
       "      <th>NA19323_batch11_rep1</th>\n",
       "      <th>NA19143_batch04_rep3</th>\n",
       "      <th>NA19143_batch15_rep2</th>\n",
       "      <th>HG00326_batch08_rep1</th>\n",
       "      <th>NA18942_batch15_rep1</th>\n",
       "      <th>...</th>\n",
       "      <th>NA19731_batch09_rep1</th>\n",
       "      <th>NA19732_batch14_rep1</th>\n",
       "      <th>HG00525_batch10_rep1</th>\n",
       "      <th>HG00473_batch03_rep1</th>\n",
       "      <th>NA21127_batch04_rep1</th>\n",
       "      <th>NA20342_batch15_rep1</th>\n",
       "      <th>NA20412_batch07_rep1</th>\n",
       "      <th>NA20318_batch07_rep1</th>\n",
       "      <th>NA20298_batch01_rep1</th>\n",
       "      <th>NA20320_batch05_rep1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>5.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>14.000527</td>\n",
       "      <td>45.00000</td>\n",
       "      <td>88.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>27.00000</td>\n",
       "      <td>14.00000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>12.00002</td>\n",
       "      <td>12.99999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>2334.455390</td>\n",
       "      <td>1795.54352</td>\n",
       "      <td>1400.40445</td>\n",
       "      <td>2132.889530</td>\n",
       "      <td>2103.78422</td>\n",
       "      <td>1360.08493</td>\n",
       "      <td>1276.42764</td>\n",
       "      <td>1472.713011</td>\n",
       "      <td>1397.217400</td>\n",
       "      <td>...</td>\n",
       "      <td>1551.96421</td>\n",
       "      <td>2235.43711</td>\n",
       "      <td>1502.48210</td>\n",
       "      <td>1004.45766</td>\n",
       "      <td>962.10114</td>\n",
       "      <td>1229.19650</td>\n",
       "      <td>1735.522800</td>\n",
       "      <td>2432.45182</td>\n",
       "      <td>1369.26527</td>\n",
       "      <td>2413.30774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>225.561370</td>\n",
       "      <td>257.59018</td>\n",
       "      <td>165.35980</td>\n",
       "      <td>351.460200</td>\n",
       "      <td>188.85600</td>\n",
       "      <td>212.49610</td>\n",
       "      <td>271.21170</td>\n",
       "      <td>215.615020</td>\n",
       "      <td>236.790578</td>\n",
       "      <td>...</td>\n",
       "      <td>340.20660</td>\n",
       "      <td>249.75316</td>\n",
       "      <td>162.36500</td>\n",
       "      <td>112.53656</td>\n",
       "      <td>142.32090</td>\n",
       "      <td>265.54570</td>\n",
       "      <td>160.061090</td>\n",
       "      <td>239.53843</td>\n",
       "      <td>165.29459</td>\n",
       "      <td>344.41416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>309.438419</td>\n",
       "      <td>347.45108</td>\n",
       "      <td>207.64022</td>\n",
       "      <td>458.540230</td>\n",
       "      <td>310.14371</td>\n",
       "      <td>378.50431</td>\n",
       "      <td>267.78767</td>\n",
       "      <td>230.385370</td>\n",
       "      <td>230.208920</td>\n",
       "      <td>...</td>\n",
       "      <td>352.79262</td>\n",
       "      <td>597.24767</td>\n",
       "      <td>300.63499</td>\n",
       "      <td>108.46337</td>\n",
       "      <td>218.67910</td>\n",
       "      <td>150.45407</td>\n",
       "      <td>229.938904</td>\n",
       "      <td>226.46133</td>\n",
       "      <td>294.70550</td>\n",
       "      <td>326.58617</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 780 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gid  NA19704_batch11_rep1  NA19332_batch14_rep1  \\\n",
       "0  ENSG00000000003.16             31.000000               5.00000   \n",
       "1   ENSG00000000005.6              0.000000               0.00000   \n",
       "2  ENSG00000000419.14           2334.455390            1795.54352   \n",
       "3  ENSG00000000457.14            225.561370             257.59018   \n",
       "4  ENSG00000000460.17            309.438419             347.45108   \n",
       "\n",
       "   NA19317_batch15_rep1  NA19312_batch05_rep1  NA19323_batch11_rep1  \\\n",
       "0               0.00000             14.000527              45.00000   \n",
       "1               0.00000              0.000000               0.00000   \n",
       "2            1400.40445           2132.889530            2103.78422   \n",
       "3             165.35980            351.460200             188.85600   \n",
       "4             207.64022            458.540230             310.14371   \n",
       "\n",
       "   NA19143_batch04_rep3  NA19143_batch15_rep2  HG00326_batch08_rep1  \\\n",
       "0              88.00000               0.00000             13.000000   \n",
       "1               0.00000               0.00000              0.000000   \n",
       "2            1360.08493            1276.42764           1472.713011   \n",
       "3             212.49610             271.21170            215.615020   \n",
       "4             378.50431             267.78767            230.385370   \n",
       "\n",
       "   NA18942_batch15_rep1  ...  NA19731_batch09_rep1  NA19732_batch14_rep1  \\\n",
       "0              0.000000  ...               0.00000               2.00000   \n",
       "1              0.000000  ...               0.00000               0.00000   \n",
       "2           1397.217400  ...            1551.96421            2235.43711   \n",
       "3            236.790578  ...             340.20660             249.75316   \n",
       "4            230.208920  ...             352.79262             597.24767   \n",
       "\n",
       "   HG00525_batch10_rep1  HG00473_batch03_rep1  NA21127_batch04_rep1  \\\n",
       "0               1.00000              27.00000              14.00000   \n",
       "1               0.00000               0.00000               0.00000   \n",
       "2            1502.48210            1004.45766             962.10114   \n",
       "3             162.36500             112.53656             142.32090   \n",
       "4             300.63499             108.46337             218.67910   \n",
       "\n",
       "   NA20342_batch15_rep1  NA20412_batch07_rep1  NA20318_batch07_rep1  \\\n",
       "0               1.00000             13.000000              28.00000   \n",
       "1               0.00000              0.000000               0.00000   \n",
       "2            1229.19650           1735.522800            2432.45182   \n",
       "3             265.54570            160.061090             239.53843   \n",
       "4             150.45407            229.938904             226.46133   \n",
       "\n",
       "   NA20298_batch01_rep1  NA20320_batch05_rep1  \n",
       "0              12.00002              12.99999  \n",
       "1               0.00000               0.00000  \n",
       "2            1369.26527            2413.30774  \n",
       "3             165.29459             344.41416  \n",
       "4             294.70550             326.58617  \n",
       "\n",
       "[5 rows x 780 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_csv(ksdljfkl;ajsdkfl;ja)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051979ce-5472-4552-a6f5-3a25fb524469",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7edc2c6-746b-457b-8568-b576ffd7fae5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b46f62-f52f-45b7-80f9-89b771b67aaa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1a7d4e36-fb53-4653-b5f8-34e6be5428c6",
   "metadata": {},
   "source": [
    "## try to intersect mage sample names w/ 1000G sample names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ea5cf075-277c-456a-8542-73c1715f4a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.read_csv('../1000g/1000G_metadata.tsv', sep='\\t', comment='#', header=None)\n",
    "trios_meta = pd.read_csv('../1000g/1000G_trios_metadata.tsv', sep='\\t', comment='#', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2599913d-476e-4fdc-b103-e519020682c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>...</th>\n",
       "      <th>bam_ftp</th>\n",
       "      <th>bam_bytes</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>batch</th>\n",
       "      <th>rep</th>\n",
       "      <th>r1_fq_link</th>\n",
       "      <th>r2_fq_link</th>\n",
       "      <th>r1_verify</th>\n",
       "      <th>r2_verify</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [run_accession, sample_accession, experiment_accession, study_accession, tax_id, scientific_name, instrument_platform, instrument_model, experiment_alias, fastq_bytes, fastq_ftp, submitted_bytes, submitted_ftp, sra_bytes, sra_ftp, sample_alias, bam_ftp, bam_bytes, cell_line_id, batch, rep, r1_fq_link, r2_fq_link, r1_verify, r2_verify, sample]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 26 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check sample correspondence between mage and 1000g\n",
    "df.loc[~(df['cell_line_id'].isin(meta[14].tolist()))&\\\n",
    "       ~(df['cell_line_id'].isin(trios_meta[14].tolist()))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2ab328-1182-492e-8f9d-553958872917",
   "metadata": {},
   "outputs": [],
   "source": [
    "# perfect"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
