import pandas as pd
import os
import sys
import pyranges as pr

p = os.path.dirname(os.path.dirname(os.getcwd()))+'/scripts/'
sys.path.append(p)

from sm_utils import *
from utils import *
# from vcf_utils import *

c_dir = '../common/'

meta_file = '../config.tsv'
configfile: '../config.yml'

include: f'{c_dir}download.smk'
include: f'{c_dir}samtools.smk'
include: f'{c_dir}winnowmap.smk'
include: f'{c_dir}bigwig.smk'
include: f'{c_dir}variant_calling.smk'
include: f'{c_dir}formatting.smk'
include: f'{c_dir}phasing.smk'
include: f'{c_dir}cerberus.smk'
include: f'{c_dir}bedtools.smk'
include: f'{c_dir}transdecoder.smk'
include: f'{c_dir}protein.smk'
include: f'{c_dir}minimap2.smk'
include: f'{c_dir}stats.smk'
include: f'{c_dir}lr-kallisto.smk'
include: f'{c_dir}kallisto.smk'
include: f'{c_dir}suppa.smk'

meta_file = 'filereport_read_run_PRJNA851328_tsv.txt'

df = pd.read_csv(meta_file, sep='\t')
df[['cell_line_id', 'batch', 'rep']] = df.experiment_alias.str.split('_', expand=True)
len(df.index)
len(df.cell_line_id.unique())

df[['r1_fq_link', 'r2_fq_link']] = df.fastq_ftp.str.split(';', expand=True)
df['r1_verify'] = df.r1_fq_link.str.endswith('_1.fastq.gz')
df['r2_verify'] = df.r2_fq_link.str.endswith('_2.fastq.gz')
assert len(df.loc[df.r1_verify==False].index)==0
assert len(df.loc[df.r2_verify==False].index)==0

df['sample'] = df['experiment_alias']
df['cell_line_id'] = df['sample'].str.split('_', expand=True)[0]
autosomes = [str(i) for i in range(1,23)]
chroms = autosomes+['X']

wildcard_constraints:
    sample='|'.join([re.escape(x) for x in df['sample'].tolist()]),
    chrom = '|'.join([re.escape(x) for x in chroms]),

def get_df_val(df, col1, col_dict, uniq_val=True):
    """
    uniq_val (b0ool) needs to return a uniq val rather
        than a list
    """
    temp = df.copy(deep=True)

    for key, item in col_dict.items():
        temp = temp.loc[temp[key] == item]

    if uniq_val:
        val = temp[col1].unique()
        assert len(val) == 1
        return val[0]
    else:
        return temp[col1].tolist()

wildcard_constraints:
    sample='|'.join([re.escape(x) for x in df['sample'].tolist()]),

rule all:
    input:
        # expand(config['mage']['r1_fq'],
        #        sample=df['sample'].tolist()),
        # expand(config['mage']['r2_fq'],
        #        sample=df['sample'].tolist()),
        # config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        # config['mage']['poder_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        # config['mage']['enh_v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        # config['mage']['genotype_pc']['pcs'],
        # config['mage']['v47_kallisto']['sqtl']['genotype'],
        # config['mage']['poder_kallisto']['sqtl']['genotype'],
        # config['mage']['enh_v47_kallisto']['sqtl']['genotype'],
        # config['mage']['v47_kallisto']['sqtl']['t_tpm_tsv'],
        # config['mage']['poder_kallisto']['sqtl']['t_tpm_tsv'],
        # config['mage']['enh_v47_kallisto']['sqtl']['t_tpm_tsv'],
        # # config['mage']['enh_v47_kallisto']['testable']['t_tpm_tsv'],
        # # config['mage']['poder_kallisto']['testable']['t_tpm_tsv'],
        # # config['mage']['v47_kallisto']['testable']['t_tpm_tsv'],
        # config['mage']['v47_kallisto']['sqtl']['gene_loc'],
        # config['mage']['poder_kallisto']['sqtl']['gene_loc'],
        # config['mage']['enh_v47_kallisto']['sqtl']['gene_loc'],
        # # config['mage']['enh_v47_kallisto']['sqtl']['metadata'],
        # # config['mage']['v47_kallisto']['sqtl']['metadata'],
        # # config['mage']['poder_kallisto']['sqtl']['metadata'],
        # config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        # config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        # config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        # config['mage']['poder_kallisto']['sqtl']['genotype_tsv'],
        # config['mage']['v47_kallisto']['sqtl']['genotype_tsv'],
        # config['mage']['enh_v47_kallisto']['sqtl']['genotype_tsv'],
        # config['mage']['poder_kallisto']['sqtlseeker']['results'],
        # config['mage']['enh_v47_kallisto']['sqtlseeker']['results'],
        config['mage']['v47_kallisto']['sqtlseeker']['results']





        # expand(config['mage']['v47_kallisto']['suppa']['psi'],
        #        event=['A3', 'A5', 'AF', 'AL', 'MX', 'RI', 'SE'])

        # expand(config['mage']['genotype_pc']['pcs'],
        #        chrom=chroms)


use rule wget as dl_r1_fq with:
    params:
        link = lambda wc: get_df_val(df, 'r1_fq_link',
                                     {'sample': wc.sample})
    output:
        out = config['mage']['r1_fq']

use rule wget as dl_r2_fq with:
    params:
        link = lambda wc: get_df_val(df, 'r2_fq_link',
                                     {'sample': wc.sample})
    output:
        out = config['mage']['r2_fq']

rule merge_matrices:
    resources:
        nodes = 3,
        threads = 1
    run:
        merge_df = pd.DataFrame()
        samples = list(params.samples)
        i = 0
        for t, s in zip(list(input.tsvs), samples):
            temp = pd.read_csv(t, sep='\t')
            temp.rename({'counts': s}, axis=1, inplace=True)
            if i == 0:
                merge_df = temp.copy(deep=True)
            else:
                merge_df = merge_df.merge(temp, how='outer',
                            on='transcript_id')
            i+=1
        merge_df.to_csv(output.tsv, sep='\t', index=False)

rule compute_gene_level_tsv:
    resources:
        threads = 1,
        nodes = 3
    run:
        exp_df = pd.read_csv(input.tsv, sep='\t')
        t2g = pd.read_csv(input.t2g, sep='\t', header=None)
        t2g = t2g[[0,1]]
        t2g.columns = ['transcript_id', 'gid']

        # get gene assignments for each transcript using t2g file
        exp_df = exp_df.merge(t2g[['gid', 'transcript_id']],
                      how='left',
                      on='transcript_id')
        assert len(exp_df.loc[exp_df.gid.isnull()].index) == 0

        # drop tid, gb on gid and sum
        exp_df.drop('transcript_id', axis=1, inplace=True)
        exp_df = exp_df.groupby('gid').sum().reset_index()
        exp_df.to_csv(output.tsv, sep='\t', index=False)

## v47 kallisto
use rule short_kallisto_build_ind as v47_sr_kallisto_ind with:
    input:
        fa = config['ref']['fa'],
        gtf = config['ref']['gtf']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['ref']['v47_kallisto_short']['ind'].split('.idx')[0]
    output:
        ind = config['ref']['v47_kallisto_short']['ind'],
        fa = config['ref']['v47_kallisto_short']['t_fa'],
        t2g = config['ref']['v47_kallisto_short']['t2g']

use rule short_kb_count as sr_kallisto_count with:
    input:
        ind = config['ref']['v47_kallisto_short']['ind'],
        t2g = config['ref']['v47_kallisto_short']['t2g'],
        r1_fq = config['mage']['r1_fq'],
        r2_fq = config['mage']['r2_fq']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['mage']['v47_kallisto']['odir']
    output:
        mtx = config['mage']['v47_kallisto']['matrix'],
        mtx_tpm = config['mage']['v47_kallisto']['matrix_tpm'],
        transcripts = config['mage']['v47_kallisto']['transcripts']

use rule fmt_mtx_transcripts as get_counts_mtx with:
    input:
        mtx = config['mage']['v47_kallisto']['matrix'],
        ts = config['mage']['v47_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['v47_kallisto']['matrix_tsv']

use rule fmt_mtx_transcripts as get_tpm_mtx with:
    input:
        mtx = config['mage']['v47_kallisto']['matrix_tpm'],
        ts = config['mage']['v47_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['v47_kallisto']['matrix_tpm_tsv']

use rule merge_matrices as merge_matrices_tpm with:
    input:
        tsvs = expand(config['mage']['v47_kallisto']['matrix_tpm_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['v47_kallisto']['merge_matrix_tpm_tsv']

use rule merge_matrices as merge_matrices_counts with:
    input:
        tsvs = expand(config['mage']['v47_kallisto']['matrix_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['v47_kallisto']['merge_matrix_tsv']

use rule compute_gene_level_tsv as gene_level_tpm with:
    input:
        tsv = config['mage']['v47_kallisto']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['v47_kallisto_short']['t2g']
    output:
        tsv = config['mage']['v47_kallisto']['gene_tpm_tsv']

use rule compute_gene_level_tsv as gene_level_counts with:
    input:
        tsv = config['mage']['v47_kallisto']['merge_matrix_tsv'],
        t2g = config['ref']['v47_kallisto_short']['t2g']
    output:
        tsv = config['mage']['v47_kallisto']['gene_tsv']

## v47 enh kallisto
use rule short_kallisto_build_ind as enh_v47_sr_kallisto_ind with:
    input:
        fa = config['ref']['fa'],
        gtf = config['lr']['gtf_gc_poder']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['ref']['enh_v47_kallisto_short']['ind'].split('.idx')[0]
    output:
        ind = config['ref']['enh_v47_kallisto_short']['ind'],
        fa = config['ref']['enh_v47_kallisto_short']['t_fa'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g']

use rule short_kb_count as enh_v47_sr_kallisto_count with:
    input:
        ind = config['ref']['enh_v47_kallisto_short']['ind'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g'],
        r1_fq = config['mage']['r1_fq'],
        r2_fq = config['mage']['r2_fq']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['mage']['enh_v47_kallisto']['odir']
    output:
        mtx = config['mage']['enh_v47_kallisto']['matrix'],
        mtx_tpm = config['mage']['enh_v47_kallisto']['matrix_tpm'],
        transcripts = config['mage']['enh_v47_kallisto']['transcripts']

use rule fmt_mtx_transcripts as enh_v47_get_counts_mtx with:
    input:
        mtx = config['mage']['enh_v47_kallisto']['matrix'],
        ts = config['mage']['enh_v47_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['enh_v47_kallisto']['matrix_tsv']

use rule fmt_mtx_transcripts as enh_v47_get_tpm_mtx with:
    input:
        mtx = config['mage']['enh_v47_kallisto']['matrix_tpm'],
        ts = config['mage']['enh_v47_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['enh_v47_kallisto']['matrix_tpm_tsv']

use rule merge_matrices as enh_v47_merge_matrices_tpm with:
    input:
        tsvs = expand(config['mage']['enh_v47_kallisto']['matrix_tpm_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tpm_tsv']

use rule merge_matrices as enh_v47_merge_matrices_counts with:
    input:
        tsvs = expand(config['mage']['enh_v47_kallisto']['matrix_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tsv']

use rule compute_gene_level_tsv as enh_v47_gene_level_tpm with:
    input:
        tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g']
    output:
        tsv = config['mage']['enh_v47_kallisto']['gene_tpm_tsv']

use rule compute_gene_level_tsv as enh_v47_gene_level_counts with:
    input:
        tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tsv'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g']
    output:
        tsv = config['mage']['enh_v47_kallisto']['gene_tsv']

## poder kallisto
use rule short_kallisto_build_ind as poder_sr_kallisto_ind with:
    input:
        fa = config['ref']['fa'],
        gtf = config['lr']['gtf_filt_with_genes']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['ref']['poder_kallisto_short']['ind'].split('.idx')[0]
    output:
        ind = config['ref']['poder_kallisto_short']['ind'],
        fa = config['ref']['poder_kallisto_short']['t_fa'],
        t2g = config['ref']['poder_kallisto_short']['t2g']

use rule short_kb_count as poder_sr_kallisto_count with:
    input:
        ind = config['ref']['poder_kallisto_short']['ind'],
        t2g = config['ref']['poder_kallisto_short']['t2g'],
        r1_fq = config['mage']['r1_fq'],
        r2_fq = config['mage']['r2_fq']
    params:
        kallisto_path = '/gpfs/home/bsc/bsc083001/miniconda3/envs/kallisto/bin/kallisto',
        odir = config['mage']['poder_kallisto']['odir']
    output:
        mtx = config['mage']['poder_kallisto']['matrix'],
        mtx_tpm = config['mage']['poder_kallisto']['matrix_tpm'],
        transcripts = config['mage']['poder_kallisto']['transcripts']

use rule fmt_mtx_transcripts as poder_get_counts_mtx with:
    input:
        mtx = config['mage']['poder_kallisto']['matrix'],
        ts = config['mage']['poder_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['poder_kallisto']['matrix_tsv']

use rule fmt_mtx_transcripts as poder_get_tpm_mtx with:
    input:
        mtx = config['mage']['poder_kallisto']['matrix_tpm'],
        ts = config['mage']['poder_kallisto']['transcripts']
    params:
        col = 'counts'
    output:
        tsv = config['mage']['poder_kallisto']['matrix_tpm_tsv']

use rule merge_matrices as poder_merge_matrices_tpm with:
    input:
        tsvs = expand(config['mage']['poder_kallisto']['matrix_tpm_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['poder_kallisto']['merge_matrix_tpm_tsv']

use rule merge_matrices as poder_merge_matrices_counts with:
    input:
        tsvs = expand(config['mage']['poder_kallisto']['matrix_tsv'],
                      sample=df['sample'].tolist())
    params:
        samples = df['sample'].tolist()
    output:
        tsv = config['mage']['poder_kallisto']['merge_matrix_tsv']

use rule compute_gene_level_tsv as poder_gene_level_tpm with:
    input:
        tsv = config['mage']['poder_kallisto']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['poder_kallisto_short']['t2g']
    output:
        tsv = config['mage']['poder_kallisto']['gene_tpm_tsv']

use rule compute_gene_level_tsv as poder_gene_level_counts with:
    input:
        tsv = config['mage']['poder_kallisto']['merge_matrix_tsv'],
        t2g = config['ref']['poder_kallisto_short']['t2g']
    output:
        tsv = config['mage']['poder_kallisto']['gene_tsv']

########################
######## Genotype PCs ##
########################

# filter the 1000G VCF for
# 1. samples that are in MAGE
# 2. MAF >= 0.01
# vcftools ref: https://vcftools.github.io/man_latest.html
rule get_samples:
    resources:
        threads = 1,
        nodes = 1
    output:
        samples = config['mage']['samples']
    run:
        temp = df[['cell_line_id']]
        temp.to_csv(output.samples, header=None, index=False)

rule filt_sample_vcf:
    input:
        vcf = config['1000g']['vcf_chr'],
        samples = config['mage']['samples']
    resources:
        threads = 16,
        nodes = 2
    output:
        vcf = config['mage']['genotype_pc']['filt_sample_vcf']
    shell:
        """
        vcftools \
            --gzvcf {input.vcf} \
            --keep {input.samples} \
            --recode \
            --recode-INFO-all \
            --stdout | gzip -c > {output.vcf}
        """

rule filt_maf_vcf:
    input:
        vcf = config['mage']['genotype_pc']['filt_sample_vcf'],
        samples = config['mage']['samples']
    resources:
        threads = 16,
        nodes = 2
    params:
        min_maf = 0.01
    output:
        vcf = config['mage']['genotype_pc']['filt_maf_vcf']
    shell:
        """
        module load samtools
        vcftools \
            --gzvcf {input.vcf} \
            --maf {params.min_maf} \
            --recode \
            --recode-INFO-all \
            --stdout | bgzip -c > {output.vcf}
        """

use rule bcftools_vcf_index as index_genotype_pc_vcf with:
    input:
        vcf = config['mage']['genotype_pc']['filt_maf_vcf']
    output:
        ind = config['mage']['genotype_pc']['filt_maf_vcf_ind']


use rule bcftools_concat as concat_genotype_pc_vcfs with:
    input:
        files = expand(config['mage']['genotype_pc']['filt_maf_vcf'],
                       chrom=autosomes),
        inds = expand(config['mage']['genotype_pc']['filt_maf_vcf_ind'],
                      chrom=autosomes),
    params:
        cli_vcfs = fmt_for_cli(expand(config['mage']['genotype_pc']['filt_maf_vcf'],
                               chrom=autosomes), sep=' ')
    output:
        vcf = config['mage']['genotype_pc']['merge_vcf']

# compute genotype PCs
rule genotype_pcs:
    input:
        vcf = config['mage']['genotype_pc']['merge_vcf']
    resources:
        threads = 112,
        nodes = 2
    params:
        opref = config['mage']['genotype_pc']['pcs'].rsplit('.', maxsplit=1)[0]
    output:
        pcs = config['mage']['genotype_pc']['pcs'],
        eigenvec = config['mage']['genotype_pc']['eigenvec'],
    shell:
        """
        module load plink
        plink2 \
            --vcf {input.vcf} \
            --allow-extra-chr \
            --set-missing-var-ids @:# \
            --make-bed \
            --pca \
            --threads {resources.threads} \
            --out {params.opref}
        """

###########################
###### Transcript / gene filtering
###########################

# sample filtering -- keep rep. / cell line
# w/ highest number of total counts
def get_filt_mage_samples(f, ofile):
    df = pd.read_csv(f, sep='\t')
    df = df.set_index('transcript_id')
    df = df.transpose()

    # counts / sample
    df = df.reset_index()
    df = df.rename({'index':'sample'}, axis=1)
    df['cell_line_id'] = df['sample'].str.split('_', expand=True)[0]
    df = df.set_index(['cell_line_id', 'sample'])
    df['total_counts'] = df.sum(axis=1)
    df = df[['total_counts']].reset_index()

    # dedupe on total counts
    df = df.sort_values(by='total_counts', ascending=False)
    print(len(df.index))
    df = df.drop_duplicates(subset=['cell_line_id'], keep='first')
    print(len(df.index))

    df = df[['sample']]
    df.to_csv(ofile, sep='\t', index=False)

def filt_mage_samples(samples, ab, ofile):
    # now filter
    s_df = pd.read_csv(samples, sep='\t')
    df = pd.read_csv(ab, sep='\t')
    samples = s_df['sample'].tolist()
    ind_cols = [c for c in df.columns if 'id' in c]
    df = df.set_index(ind_cols)
    df = df[samples]

    # convert to cell line id name format
    df.columns = [c.split('_')[0] for c in df.columns]

    df = df.reset_index()
    df.to_csv(ofile, sep='\t', index=False)

rule get_filt_mage_samples:
    input:
        tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tsv'] # we'll use the same sample for everything for consistency
    resources:
        nodes = 2,
        threads = 1
    output:
        tsv = config['mage']['filt_samples']
    run:
        get_filt_mage_samples(input.tsv, output.tsv)

rule filt_mage_samples:
    resources:
        threads = 1,
        nodes = 2
    run:
        filt_mage_samples(input.samples, input.t_tsv, output.t_tsv)
        filt_mage_samples(input.samples, input.t_tpm_tsv, output.t_tpm_tsv)
        filt_mage_samples(input.samples, input.g_tsv, output.g_tsv)
        filt_mage_samples(input.samples, input.g_tpm_tsv, output.g_tpm_tsv)

use rule filt_mage_samples as enh_v47_filt_samples with:
    input:
        samples = config['mage']['filt_samples'],
        t_tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['enh_v47_kallisto']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['enh_v47_kallisto']['gene_tsv'],
        g_tpm_tsv = config['mage']['enh_v47_kallisto']['gene_tpm_tsv']
    output:
        t_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['gene_tsv'],
        g_tpm_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['gene_tpm_tsv']

use rule filt_mage_samples as poder_filt_samples with:
    input:
        samples = config['mage']['filt_samples'],
        t_tsv = config['mage']['poder_kallisto']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['poder_kallisto']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['poder_kallisto']['gene_tsv'],
        g_tpm_tsv = config['mage']['poder_kallisto']['gene_tpm_tsv']
    output:
        t_tsv = config['mage']['poder_kallisto']['samp_filt']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['poder_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['poder_kallisto']['samp_filt']['gene_tsv'],
        g_tpm_tsv = config['mage']['poder_kallisto']['samp_filt']['gene_tpm_tsv']

use rule filt_mage_samples as v47_filt_samples with:
    input:
        samples = config['mage']['filt_samples'],
        t_tsv = config['mage']['v47_kallisto']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['v47_kallisto']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['v47_kallisto']['gene_tsv'],
        g_tpm_tsv = config['mage']['v47_kallisto']['gene_tpm_tsv']
    output:
        t_tsv = config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tsv'],
        t_tpm_tsv = config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        g_tsv = config['mage']['v47_kallisto']['samp_filt']['gene_tsv'],
        g_tpm_tsv = config['mage']['v47_kallisto']['samp_filt']['gene_tpm_tsv']

# format the transcript tpm matrix in the format that sqtlseeker wants
# expression of each transcript in each sample TPM
# columns trId and geneId
rule fmt_sqtlseeker_t_tpm:
    resources:
        threads = 1,
        nodes = 2
    run:
        df = pd.read_csv(input.t_tpm_tsv, sep='\t')
        df.rename({'transcript_id':'trId'}, axis=1, inplace=True)
        t2g_df = pd.read_csv(input.t2g, sep='\t', header=None)
        t2g_df = t2g_df[[0,1]]
        t2g_df.columns = ['trId', 'geneId']
        df = df.merge(t2g_df, how='left',
                      on='trId')
        assert len(df.loc[df.geneId.isnull()].index)==0
        nc1 = len(df.columns)
        id_cols = ['trId', 'geneId']
        col_order = id_cols+[c for c in df.columns if c not in id_cols]
        df = df[col_order]
        assert nc1 == len(df.columns)
        df.to_csv(output.t_tpm_tsv, sep='\t', index=None)

use rule fmt_sqtlseeker_t_tpm as v47_fmt_sqtlseeker_t_tpm with:
    input:
        t_tpm_tsv = config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['v47_kallisto_short']['t2g']
    output:
        t_tpm_tsv = config['mage']['v47_kallisto']['sqtl']['t_tpm_tsv'],
use rule fmt_sqtlseeker_t_tpm as enh_v47_fmt_sqtlseeker_t_tpm with:
    input:
        t_tpm_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g']
    output:
        t_tpm_tsv = config['mage']['enh_v47_kallisto']['sqtl']['t_tpm_tsv'],
use rule fmt_sqtlseeker_t_tpm as poder_fmt_sqtlseeker_t_tpm with:
    input:
        t_tpm_tsv = config['mage']['poder_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['poder_kallisto_short']['t2g']
    output:
        t_tpm_tsv = config['mage']['poder_kallisto']['sqtl']['t_tpm_tsv'],

################
#### Preparing genotype file
################

### get +- 5kb regions for gene bodies
rule get_cis_regions:
    resources:
        threads = 16,
        nodes = 3
    params:
        extension = 5000
    run:
        gtf_df = pr.read_gtf(input.gtf).df
        gtf_df = gtf_df.loc[gtf_df.Feature=='gene']
        gtf_df = pr.PyRanges(gtf_df)
        gtf_df = gtf_df.extend(params.extension)
        gtf_df.to_bed(output.bed)
use rule get_cis_regions as v47_get_cis_regions with:
    input:
        gtf = config['ref']['gtf']
    output:
        bed = config['mage']['v47_kallisto']['sqtl']['cis_bed']
use rule get_cis_regions as enh_v47_get_cis_regions with:
    input:
        gtf = config['lr']['gtf_gc_poder']
    output:
        bed = config['mage']['enh_v47_kallisto']['sqtl']['cis_bed']
use rule get_cis_regions as poder_get_cis_regions with:
    input:
        gtf = config['lr']['gtf_filt_with_genes']
    output:
        bed = config['mage']['poder_kallisto']['sqtl']['cis_bed']

### filter vcfs on these cis regions
# using the maf / sample filtered variants... I think this makes sense
use rule bcftools_filter_on_regions as v47_get_cis_regions_vcf with:
    input:
        bed = config['mage']['v47_kallisto']['sqtl']['cis_bed'],
        vcf = config['mage']['genotype_pc']['filt_maf_vcf'],
        ind = config['mage']['genotype_pc']['filt_maf_vcf_ind']
    output:
        vcf = config['mage']['v47_kallisto']['sqtl']['cis_vcf']
use rule bcftools_filter_on_regions as enh_v47_get_cis_region_vcf with:
    input:
        bed = config['mage']['enh_v47_kallisto']['sqtl']['cis_bed'],
        vcf = config['mage']['genotype_pc']['filt_maf_vcf'],
        ind = config['mage']['genotype_pc']['filt_maf_vcf_ind']
    output:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf']
use rule bcftools_filter_on_regions as poder_get_cis_region_vcf with:
    input:
        bed = config['mage']['poder_kallisto']['sqtl']['cis_bed'],
        vcf = config['mage']['genotype_pc']['filt_maf_vcf'],
        ind = config['mage']['genotype_pc']['filt_maf_vcf_ind']
    output:
        vcf = config['mage']['poder_kallisto']['sqtl']['cis_vcf']

use rule bcftools_vcf_index as v47_vcf_ind with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['cis_vcf']
    output:
        ind = config['mage']['v47_kallisto']['sqtl']['cis_vcf_ind']
use rule bcftools_vcf_index as enh_v47_vcf_ind with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf']
    output:
        ind = config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf_ind']
use rule bcftools_vcf_index as poder_vcf_ind with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['cis_vcf']
    output:
        ind = config['mage']['poder_kallisto']['sqtl']['cis_vcf_ind']

use rule bcftools_concat as v47_concat_sqtl_vcfs with:
    input:
        files = expand(config['mage']['v47_kallisto']['sqtl']['cis_vcf'],
                       chrom=chroms),
        inds = expand(config['mage']['v47_kallisto']['sqtl']['cis_vcf_ind'],
                      chrom=chroms)
    params:
        cli_vcfs = fmt_for_cli(expand(config['mage']['v47_kallisto']['sqtl']['cis_vcf'],
                               chrom=chroms), sep=' ')
    output:
        vcf = temporary(config['mage']['v47_kallisto']['sqtl']['merge_cis_vcf'])
use rule bcftools_concat as enh_v47_concat_sqtl_vcfs with:
    input:
        files = expand(config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf'],
                       chrom=chroms),
        inds = expand(config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf_ind'],
                     chrom=chroms)
    params:
        cli_vcfs = fmt_for_cli(expand(config['mage']['enh_v47_kallisto']['sqtl']['cis_vcf'],
                               chrom=chroms), sep=' ')
    output:
        vcf = temporary(config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_vcf'])
use rule bcftools_concat as poder_concat_sqtl_vcfs with:
    input:
        files = expand(config['mage']['poder_kallisto']['sqtl']['cis_vcf'],
                       chrom=chroms),
        inds = expand(config['mage']['poder_kallisto']['sqtl']['cis_vcf_ind'],
                     chrom=chroms)
    params:
        cli_vcfs = fmt_for_cli(expand(config['mage']['poder_kallisto']['sqtl']['cis_vcf'],
                               chrom=chroms), sep=' ')
    output:
        vcf = temporary(config['mage']['poder_kallisto']['sqtl']['merge_cis_vcf'])
# normalize these guys
use rule vcf_norm as enh_v47_norm with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_vcf'],
        fa = config['ref']['fa']
    output:
        vcf = temporary(config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_vcf'])
use rule vcf_norm as v47_norm with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_vcf'],
        fa = config['ref']['fa']
    output:
        vcf = temporary(config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_vcf'])
use rule vcf_norm as poder_norm with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_vcf'],
        fa = config['ref']['fa']
    output:
        vcf = temporary(config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_vcf'])
# get only snps
use rule rm_indels as enh_v47_rm_indels with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_vcf'],
    output:
        vcf = temporary(config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_snp_vcf'])
use rule rm_indels as v47_rm_indels with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_vcf'],
    output:
        vcf = temporary(config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_snp_vcf'])
use rule rm_indels as poder_rm_indels with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_vcf'],
    output:
        vcf = temporary(config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_snp_vcf'])
# only biallelic
use rule filt_biallelic as enh_v47_biallelic with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_snp_vcf'],
    output:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf']
use rule filt_biallelic as v47_biallelic with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_snp_vcf'],
    output:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf']
use rule filt_biallelic as poder_biallelic with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_snp_vcf'],
    output:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_bial_vcf']

use rule vcftools_012 as v47_get_012 with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf']
    params:
        opref = config['mage']['v47_kallisto']['sqtl']['genotype'].split('.012')[0]
    output:
        genotypes = config['mage']['v47_kallisto']['sqtl']['genotype'],
        indv = config['mage']['v47_kallisto']['sqtl']['indv'],
        pos = config['mage']['v47_kallisto']['sqtl']['pos']

use rule vcftools_012 as enh_v47_get_012 with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf']
    params:
        opref = config['mage']['enh_v47_kallisto']['sqtl']['genotype'].split('.012')[0]
    output:
        genotypes = config['mage']['enh_v47_kallisto']['sqtl']['genotype'],
        indv = config['mage']['enh_v47_kallisto']['sqtl']['indv'],
        pos = config['mage']['enh_v47_kallisto']['sqtl']['pos']

use rule vcftools_012 as poder_get_012 with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_bial_vcf']
    params:
        opref = config['mage']['poder_kallisto']['sqtl']['genotype'].split('.012')[0]
    output:
        genotypes = config['mage']['poder_kallisto']['sqtl']['genotype'],
        indv = config['mage']['poder_kallisto']['sqtl']['indv'],
        pos = config['mage']['poder_kallisto']['sqtl']['pos']

# Metadata file.
# Contains the covariate information for each sample.
# - batch -run -genotype pcs -texp pcs
# In addition, it defines the groups or conditions for which
# sQTL mapping will be performed. T
# the first columns should be: indId, sampleId, group,
# followed by the covariates. This file defines which samples will be tested.
def make_sqtlseeker_meta(dl_metadata,
                         metadata,
                         genotype_pcs,
                         filt_samples,
                         n_genotype_pcs,
                         ofile):
    df = pd.read_csv(dl_metadata, sep='\t')

    # filter for relevant samples
    filt_samples = pd.read_csv(filt_samples, sep='\t')
    df = df.loc[df['experiment_alias'].isin(filt_samples['sample'].tolist())]

    df['sampleId'] = df['indId'] = df.experiment_alias.str.split('_', expand=True)[0]
    df['batch'] = df.experiment_alias.str.split('_', expand=True)[1]

    df2 = pd.read_csv(metadata, sep=' ', header=None)
    df2.columns = ['sampleId', 'pop', 'continent', 'sex', 'sth', 'sth2']
    df = df.merge(df2, how='left',
                  on='sampleId')
    df['group'] = 1 # all the same group because we don't want to call sqtls separately for anything
    df = df[['indId', 'sampleId', 'group', 'batch', 'sex']]

    # add genotype PCs (first 3)
    pcs_df = pd.read_csv(genotype_pcs, sep='\t')
    pcs_df.rename({'#IID':'sampleId'}, axis=1, inplace=True)
    pcs_df = pcs_df[pcs_df.columns[:n_genotype_pcs+1]]
    pcs_df.columns = [f'{c}_genotype' if c != 'sampleId' else c for c in pcs_df.columns]
    df = df.merge(pcs_df,
                  how='left',
                  on='sampleId')

    df.to_csv(ofile, sep='\t', index=False)

rule make_sqtlseeker_meta:
    params:
        n_genotype_pcs = 3
    resources:
        threads = 1,
        nodes = 1
    run:
        make_sqtlseeker_meta(input.dl_metadata,
                             input.metadata,
                             input.genotype_pcs,
                             input.filt_samples,
                             params.n_genotype_pcs,
                             output.metadata)
use rule make_sqtlseeker_meta as v47_make_sqtlseeker_meta with:
    input:
        dl_metadata = config['mage']['dl_metadata'],
        metadata = config['mage']['metadata'],
        genotype_pcs = config['mage']['genotype_pc']['eigenvec'],
        filt_samples = config['mage']['filt_samples'],
    output:
        metadata = config['mage']['v47_kallisto']['sqtl']['metadata']
use rule make_sqtlseeker_meta as poder_make_sqtlseeker_meta with:
    input:
        dl_metadata = config['mage']['dl_metadata'],
        metadata = config['mage']['metadata'],
        genotype_pcs = config['mage']['genotype_pc']['eigenvec'],
        filt_samples = config['mage']['filt_samples'],
    output:
        metadata = config['mage']['poder_kallisto']['sqtl']['metadata']
use rule make_sqtlseeker_meta as enh_v47_make_sqtlseeker_meta with:
    input:
        dl_metadata = config['mage']['dl_metadata'],
        metadata = config['mage']['metadata'],
        genotype_pcs = config['mage']['genotype_pc']['eigenvec'],
        filt_samples = config['mage']['filt_samples'],
    output:
        metadata = config['mage']['enh_v47_kallisto']['sqtl']['metadata']

rule prepare_lineage_genotype:
    resources:
        threads = 1,
        nodes = 2
    params:
        scripts_dir = p
    shell:
        """
        module load R/4.3.2
        Rscript {params.scripts_dir}/prepare_lineage_genotype.R \
            --genotype {input.vcf} \
            --metadata {input.metadata} \
            --in-012 {input.genotypes} \
            --indv-012 {input.indv} \
            --pos-012 {input.pos} \
            --out-file {output.tsv_gz} \
            --verbose
        """
use rule prepare_lineage_genotype as poder_prepare_lineage_genotype with:
    input:
        vcf = config['mage']['poder_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        genotypes = config['mage']['poder_kallisto']['sqtl']['genotype'],
        indv = config['mage']['poder_kallisto']['sqtl']['indv'],
        pos = config['mage']['poder_kallisto']['sqtl']['pos'],
        metadata = config['mage']['poder_kallisto']['sqtl']['metadata'],
    output:
        tsv_gz = config['mage']['poder_kallisto']['sqtl']['genotype_tsv']
use rule prepare_lineage_genotype as v47_prepare_lineage_genotype with:
    input:
        vcf = config['mage']['v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        genotypes = config['mage']['v47_kallisto']['sqtl']['genotype'],
        indv = config['mage']['v47_kallisto']['sqtl']['indv'],
        pos = config['mage']['v47_kallisto']['sqtl']['pos'],
        metadata = config['mage']['v47_kallisto']['sqtl']['metadata'],
    output:
        tsv_gz = config['mage']['v47_kallisto']['sqtl']['genotype_tsv']
use rule prepare_lineage_genotype as enh_v47_prepare_lineage_genotype with:
    input:
        vcf = config['mage']['enh_v47_kallisto']['sqtl']['merge_cis_norm_bial_vcf'],
        genotypes = config['mage']['enh_v47_kallisto']['sqtl']['genotype'],
        indv = config['mage']['enh_v47_kallisto']['sqtl']['indv'],
        pos = config['mage']['enh_v47_kallisto']['sqtl']['pos'],
        metadata = config['mage']['enh_v47_kallisto']['sqtl']['metadata'],
    output:
        tsv_gz = config['mage']['enh_v47_kallisto']['sqtl']['genotype_tsv']

def filt_sqtl_transcripts(t_tpm_tsv,
                          t2g,
                          min_samp_prop,
                          min_g_exp,
                          min_t_exp,
                          min_n_t,
                          ofile):

    # get melted version of table w/ tid, gid, sample, t_exp, n_t (per gene), passed t exp filt, passed g exp filt
    # this is the main output tale
    filt_df = pd.read_csv(t_tpm_tsv, sep='\t')
    filt_df.rename({'transcript_id':'tid'}, axis=1, inplace=True)
    filt_df = filt_df.set_index('tid').melt(ignore_index=False, value_name='tpm', var_name='sample').reset_index()
    t2g_df = pd.read_csv(t2g, sep='\t', header=None)
    t2g_df = t2g_df[[0,1]]
    t2g_df.columns = ['tid', 'gid']
    filt_df = filt_df.merge(t2g_df,
                            how='left',
                            on='tid')

    # isoform expressed in at least 1 sample >= 0.1 TPM
    df = pd.read_csv(t_tpm_tsv, sep='\t')
    df = df.set_index('transcript_id')
    df = df>=min_t_exp
    df['n_exp_samples'] = df.sum(axis=1)
    print(len(df.index))
    df = df.loc[df['n_exp_samples']>=1]
    print(len(df.index))
    tids = df.index.tolist()
    filt_df['t_passed_exp_filt'] = filt_df.tid.isin(tids)

    # using these isoform, which genes pass the n transcripts / gene
    t2g_df = pd.read_csv(t2g, sep='\t', header=None)
    t2g_df = t2g_df[[0,1]]
    t2g_df.columns = ['tid', 'gid']
    print(len(t2g_df.index))
    t2g_df = t2g_df.loc[(t2g_df.tid.isin(tids))]
    print(len(t2g_df.index))
    # count # isos / gene and filter one more time
    t2g_df = t2g_df.groupby('gid').nunique().reset_index().rename({'tid':'n_tid'}, axis=1)
    t2g_df = t2g_df.loc[t2g_df.n_tid >= min_n_t]
    print(len(t2g_df.index))
    gids2 = t2g_df.gid.tolist()
    filt_df['g_passed_n_t_filt'] = filt_df.gid.isin(gids2)

    # if sum of remaining transcripts (so gene expression based on filtered trx) is >=1 TPM
    # keep transcript, if not, convert to NA (this is per sample)
    df = pd.read_csv(t_tpm_tsv, sep='\t')
    df.rename({'transcript_id':'tid'}, axis=1, inplace=True)
    df = df.loc[df.tid.isin(tids)]
    t2g_df = pd.read_csv(t2g, sep='\t', header=None)
    t2g_df = t2g_df[[0,1]]
    t2g_df.columns = ['tid', 'gid']
    df = df.merge(t2g_df,
                  how='left',
                  on='tid')
    df = df.drop('tid', axis=1)
    df = df.groupby('gid').sum().reset_index()
    df = df.set_index('gid').melt(ignore_index=False, value_name='gene_tpm', var_name='sample').reset_index()
    df['g_passed_exp_filt'] = df.gene_tpm>=min_g_exp
    filt_df = filt_df.merge(df,
                            how='left',
                            on=['gid', 'sample'])

    # using just these isoforms, re-compute gene tpm
    # genes exp. >= 1 TPM in at least 80% of samples
    n_samples = len([c for c in df.columns.tolist() if c not in ['gid']])
    n_sample_cutoff = int(n_samples*min_samp_prop)
    df = pd.read_csv(t_tpm_tsv, sep='\t')
    df.rename({'transcript_id':'tid'}, axis=1, inplace=True)
    df = df.loc[df.tid.isin(tids)]
    t2g_df = pd.read_csv(t2g, sep='\t', header=None)
    t2g_df = t2g_df[[0,1]]
    t2g_df.columns = ['tid', 'gid']
    df = df.merge(t2g_df,
                  how='left',
                  on='tid')
    df = df.drop('tid', axis=1)
    df = df.groupby('gid').sum().reset_index()
    df = df.set_index('gid')
    df = df>=min_g_exp
    df['n_exp_samples'] = df.sum(axis=1)
    print(len(df.index))
    df = df.loc[df['n_exp_samples']>=n_sample_cutoff]
    print(len(df.index))
    gids = df.index.tolist()
    filt_df['g_passed_exp_min_samples_filt'] = filt_df.gid.isin(gids)
    filt_df.to_csv(ofile, sep='\t', index=False)




rule filt_sqtl_transcripts:
    resources:
        nodes = 2,
        threads = 1
    params:
        min_samp_prop = 0.8, # min. prop. of samples where gene is exp
        min_g_exp = 1, # min tpm val. of exp. gene
        min_t_exp = 0.1, # min tpm val. of exp. transcript
        min_n_t = 2, # min # exp transcirpts / g
    run:
        filt_sqtl_transcripts(input.t_tpm_tsv,
                              input.t2g,
                              params.min_samp_prop,
                              params.min_g_exp,
                              params.min_t_exp,
                              params.min_n_t,
                              output.t_tpm_testable_tsv)
use rule filt_sqtl_transcripts as v47_filt_sqtl_transcripts with:
    input:
        g_tpm_tsv = config['mage']['v47_kallisto']['samp_filt']['gene_tpm_tsv'],
        t_tpm_tsv = config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['v47_kallisto_short']['t2g'],
    output:
        t_tpm_testable_tsv = config['mage']['v47_kallisto']['testable']['t_tpm_tsv']
use rule filt_sqtl_transcripts as enh_v47_filt_sqtl_transcripts with:
    input:
        g_tpm_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['gene_tpm_tsv'],
        t_tpm_tsv = config['mage']['enh_v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['enh_v47_kallisto_short']['t2g'],
    output:
        t_tpm_testable_tsv = config['mage']['enh_v47_kallisto']['testable']['t_tpm_tsv']
use rule filt_sqtl_transcripts as poder_filt_sqtl_transcripts with:
    input:
        g_tpm_tsv = config['mage']['poder_kallisto']['samp_filt']['gene_tpm_tsv'],
        t_tpm_tsv = config['mage']['poder_kallisto']['samp_filt']['merge_matrix_tpm_tsv'],
        t2g = config['ref']['poder_kallisto_short']['t2g'],
    output:
        t_tpm_testable_tsv = config['mage']['poder_kallisto']['testable']['t_tpm_tsv']

# gene location file. Contains the location of each gene.
# Columns chr, start, end and geneId are required.
# This file defines which genes will be tested.
rule get_gene_bodies:
    resources:
        threads = 16,
        nodes = 3
    run:
        gtf_df = pr.read_gtf(input.gtf).df
        gtf_df = gtf_df.loc[gtf_df.Feature=='gene']
        m = {'Chromosome': 'chr',
             'Start': 'start',
             'End': 'end',
             'gene_id': 'geneId'}
        gtf_df = gtf_df[['Chromosome', 'Start', 'End', 'gene_id']]

        # remove chr
        gtf_df['Chromosome'] = gtf_df['Chromosome'].str.replace('chr', '')
        # inds = gtf_df.loc[gtf_df.Chromosome.str.contains('chr')].index
        # gtf_df.loc[inds, 'Chromosome']= gtf_df.loc[inds].Chromosome.str.split('chr', expand=True)[0]

        gtf_df = gtf_df.rename(m, axis=1)
        gtf_df.to_csv(output.tsv, sep='\t', index=False)
        gtf_df = gtf_df.rename(m, axis=1)
        gtf_df.to_csv(output.tsv, sep='\t', index=False)

use rule get_gene_bodies as v47_get_gene_bodies with:
    input:
        gtf = config['ref']['gtf']
    output:
        tsv = config['mage']['v47_kallisto']['sqtl']['gene_loc']
use rule get_gene_bodies as enh_get_gene_bodies with:
    input:
        gtf = config['lr']['gtf_gc_poder']
    output:
        tsv = config['mage']['enh_v47_kallisto']['sqtl']['gene_loc']
use rule get_gene_bodies as poder_get_gene_bodies with:
    input:
        gtf = config['lr']['gtf_filt_with_genes']
    output:
        tsv = config['mage']['poder_kallisto']['sqtl']['gene_loc']

############################################
######## actually run sqtlseeker
############################################
rule run_sqtlseeker:
    conda:
        "base"
    resources:
        nodes = 5,
        threads = 64
    shell:
        """
        module load nextflow/21.04.0
        module load singularity/3.11.5
        module load R/4.3.2

        nextflow run {params.nf} \
    	  --genotype {input.genotype} \
    	  --trexp {input.tcc} \
    	  --metadata {input.metadata} \
    	  --genes {input.genes} \
    	  --dir {params.odir} \
    	  --mode "nominal" \
    	  --covariates "true" \
          --svqtl "true" \
          --fdr 0.01 \
          --fdr_svqtl 0.01 \
          --min_gene_exp 1 \
          --min_transcript_exp 0.1 \
          --min_md 0.05 \
          --min_proportion 0.8 \
          --with-singularity {params.image}
        """

use rule run_sqtlseeker as poder_run_sqtlseeker with:
    input:
        genotype = config['mage']['poder_kallisto']['sqtl']['genotype_tsv'],
        tcc = config['mage']['poder_kallisto']['sqtl']['t_tpm_tsv'],
        metadata = config['mage']['poder_kallisto']['sqtl']['metadata'],
        genes = config['mage']['poder_kallisto']['sqtl']['gene_loc']
    output:
        results = config['mage']['poder_kallisto']['sqtlseeker']['results']
    params:
        image = config['software']['sqtlseeker_sing_img'],
        nf = config['software']['sqtlseeker_nf'],
        odir = config['mage']['poder_kallisto']['sqtlseeker']['odir']
use rule run_sqtlseeker as v47_run_sqtlseeker with:
    input:
        genotype = config['mage']['v47_kallisto']['sqtl']['genotype_tsv'],
        tcc = config['mage']['v47_kallisto']['sqtl']['t_tpm_tsv'],
        metadata = config['mage']['v47_kallisto']['sqtl']['metadata'],
        genes = config['mage']['v47_kallisto']['sqtl']['gene_loc']
    output:
        results = config['mage']['v47_kallisto']['sqtlseeker']['results']
    params:
        image = config['software']['sqtlseeker_sing_img'],
        nf = config['software']['sqtlseeker_nf'],
        odir = config['mage']['v47_kallisto']['sqtlseeker']['odir']
use rule run_sqtlseeker as enh_v47_run_sqtlseeker with:
    input:
        genotype = config['mage']['enh_v47_kallisto']['sqtl']['genotype_tsv'],
        tcc = config['mage']['enh_v47_kallisto']['sqtl']['t_tpm_tsv'],
        metadata = config['mage']['enh_v47_kallisto']['sqtl']['metadata'],
        genes = config['mage']['enh_v47_kallisto']['sqtl']['gene_loc']
    output:
        results = config['mage']['enh_v47_kallisto']['sqtlseeker']['results']
    params:
        image = config['software']['sqtlseeker_sing_img'],
        nf = config['software']['sqtlseeker_nf'],
        odir = config['mage']['enh_v47_kallisto']['sqtlseeker']['odir']

# ##########################################
# ########### suppa splice event quantification
# ##########################################
# use rule fmt_kallisto_to_suppa_ab as v47_fmt_ab with:
#     input:
#         ab = config['mage']['v47_kallisto']['samp_filt']['merge_matrix_tpm_tsv']
#     output:
#         ab = config['mage']['v47_kallisto']['suppa']['fmt_ab']
#
# use rule suppa_generate_events as v47_suppa_ge with:
#     input:
#         gtf = rules.v47_sr_kallisto_ind.input.gtf
#     params:
#         opref = config['mage']['v47_kallisto']['suppa']['events']['A3'].rsplit('_', maxsplit=2)[0]
#     output:
#         config['mage']['v47_kallisto']['suppa']['events']['A3'],
#         config['mage']['v47_kallisto']['suppa']['events']['A5'],
#         config['mage']['v47_kallisto']['suppa']['events']['AF'],
#         config['mage']['v47_kallisto']['suppa']['events']['AL'],
#         config['mage']['v47_kallisto']['suppa']['events']['MX'],
#         config['mage']['v47_kallisto']['suppa']['events']['RI'],
#         config['mage']['v47_kallisto']['suppa']['events']['SE'],
#
# use rule suppa_psi as v47_suppa_psi with:
#     input:
#         ioe = lambda wc: config['mage']['v47_kallisto']['suppa']['events'][wc.event],
#         filt_ab = config['mage']['v47_kallisto']['suppa']['fmt_ab']
#     params:
#         opref = config['mage']['v47_kallisto']['suppa']['psi'].rsplit('.psi', maxsplit=1)[0]
#     output:
#         config['mage']['v47_kallisto']['suppa']['psi']
