{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "30a3dca7-1563-43a2-b68c-b7a50946a7a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "from pyfaidx import Fasta\n",
    "import upsetplot\n",
    "from pandarallel import pandarallel\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ff010081-73ee-48b2-a540-ac99e4c7165c",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "08bcb503-b109-4a4a-9b26-6a7efb98c1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_df = pd.read_csv('../data/05_mastertable/poder_master_table_fixed_genics.tsv', sep='\\t')\n",
    "# mt_df = mt_df.loc[mt_df['filter']=='pass']\n",
    "mt_df['pop_spec_t'] = (mt_df.population_sharing==1)&\\\n",
    "                      (mt_df.sample_sharing>1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ebbb37bd-629a-4df5-a4ec-41700572de83",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df = pd.read_csv(proc_cfg(config['lr']['poder_protein']['protein']['summary'], od), sep='\\t')\n",
    "p_df = p_df.merge(mt_df[['associated_gene_biotype',\n",
    "                         'structural_category', \n",
    "                         'associated_transcript',\n",
    "                         'isoform',\n",
    "                         'associated_gene',\n",
    "                         'length']],\n",
    "                  how='left',\n",
    "                  left_on='tid', \n",
    "                  right_on='isoform')\n",
    "p_df.rename({'associated_gene':'gid',\n",
    "             'protein_sequence':'seq'}, axis=1, inplace=True)\n",
    "assert len(p_df.loc[p_df.isoform.isnull()]) == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a26fb302-697f-43fe-aa7b-7723450d5e33",
   "metadata": {},
   "source": [
    "## Limit to protein coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "445e74a6-64fc-4b9f-a6a2-1b0fa4c966e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100666\n",
      "100666\n"
     ]
    }
   ],
   "source": [
    "# keep only things from annotated protein coding genes\n",
    "p_df = p_df.loc[p_df.associated_gene_biotype=='Protein Coding']\n",
    "n = len(p_df.index)\n",
    "print(len(p_df.index))\n",
    "\n",
    "# keep only things w/ full ORFs\n",
    "p_df = p_df.loc[(p_df.protein_has_start_codon==True)&\\\n",
    "                (p_df.protein_has_stop_codon==True)]\n",
    "print(len(p_df.index))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0bab944-bc94-4cd1-adc2-d4c3efea628b",
   "metadata": {},
   "source": [
    "## Get reference AA sequences and merge on gid + sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ae0cb705-d73f-410a-a43f-1ade871ed9e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get annotated AA sequencesa\n",
    "fasta_file = proc_cfg(config['ref']['pc'], od)\n",
    "fasta = Fasta(fasta_file)\n",
    "\n",
    "# Extract each entry's name and sequence\n",
    "ref_orfs = {\n",
    "    \"name\": [entry.name for entry in fasta],\n",
    "    \"seq\": [str(entry) for entry in fasta]\n",
    "}\n",
    "\n",
    "ref_orfs = pd.DataFrame(ref_orfs)\n",
    "ref_orfs['gid'] =  ref_orfs.name.str.split('|', expand=True)[2]\n",
    "ref_orfs['tid'] =  ref_orfs.name.str.split('|', expand=True)[1]\n",
    "\n",
    "ref_orfs = ref_orfs[['seq', 'gid']]\n",
    "ref_orfs['annot_aa'] = True\n",
    "ref_orfs.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0eed140b-79fb-4e7d-b0e2-4ec30ebe8b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# also add orf completeness as info!\n",
    "gtf_file = proc_cfg(config['ref']['gtf'], od)\n",
    "gtf_df = pr.read_gtf(gtf_file).df\n",
    "\n",
    "start_codon_tids = gtf_df.loc[gtf_df.Feature=='start_codon', 'transcript_id'].unique().tolist()\n",
    "stop_codon_tids = gtf_df.loc[gtf_df.Feature=='stop_codon', 'transcript_id'].unique().tolist()\n",
    "gtf_df = gtf_df[['transcript_id']].drop_duplicates()\n",
    "\n",
    "gtf_df['has_start_codon'] = gtf_df.transcript_id.isin(start_codon_tids)\n",
    "gtf_df['has_stop_codon'] = gtf_df.transcript_id.isin(stop_codon_tids)\n",
    "\n",
    "gtf_df = gtf_df.loc[gtf_df.transcript_id.notnull()]\n",
    "gtf_df['has_complete_orf'] = (gtf_df.has_start_codon)&(gtf_df.has_stop_codon)\n",
    "\n",
    "assert len(gtf_df.transcript_id.unique()) == len(gtf_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a2604dd9-5c4b-4869-9ef9-933cf4473bea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100666\n",
      "100666\n"
     ]
    }
   ],
   "source": [
    "# just going to call novel aas as those that are not in the \n",
    "# pc translations gencode file\n",
    "l1 = len(p_df.index)\n",
    "p_df = p_df.merge(ref_orfs, how='left', on=['gid', 'seq'])\n",
    "l2 = len(p_df.index)\n",
    "assert l1 == l2\n",
    "p_df['annot_aa'] = p_df['annot_aa'].fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fb8936a5-8a9a-4c1b-9ae8-85730bdc0bea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge in information of if there was a complete annotated orf to begin with\n",
    "p_df = p_df.merge(gtf_df, \n",
    "                  how='left', \n",
    "                  left_on='associated_transcript',\n",
    "                  right_on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c339af0b-7876-47c9-a637-67b4599660d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>annot_aa</th>\n",
       "      <th>has_complete_orf</th>\n",
       "      <th>structural_category</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">False</th>\n",
       "      <th>False</th>\n",
       "      <th>FSM</th>\n",
       "      <td>25834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th>FSM</th>\n",
       "      <td>3315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">True</th>\n",
       "      <th>False</th>\n",
       "      <th>FSM</th>\n",
       "      <td>1315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th>FSM</th>\n",
       "      <td>37704</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               isoform\n",
       "annot_aa has_complete_orf structural_category         \n",
       "False    False            FSM                    25834\n",
       "         True             FSM                     3315\n",
       "True     False            FSM                     1315\n",
       "         True             FSM                    37704"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df[['isoform', 'annot_aa', 'structural_category', 'has_complete_orf']].groupby(['annot_aa', 'has_complete_orf', 'structural_category']).nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97dbeb63-9c01-4df3-bece-a5cf0f93203c",
   "metadata": {},
   "source": [
    "So most FSMs w/o a matching full-length ORF don't have a matching AA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3c90731b-90b7-4171-9a6b-c17814c6e3f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>annot_aa</th>\n",
       "      <th>structural_category</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">False</th>\n",
       "      <th>FSM</th>\n",
       "      <td>29149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>12330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>13897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">True</th>\n",
       "      <th>FSM</th>\n",
       "      <td>39019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>3176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>3095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              isoform\n",
       "annot_aa structural_category         \n",
       "False    FSM                    29149\n",
       "         NIC                    12330\n",
       "         NNC                    13897\n",
       "True     FSM                    39019\n",
       "         NIC                     3176\n",
       "         NNC                     3095"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df[['isoform', 'annot_aa', 'structural_category']].groupby(['annot_aa', 'structural_category']).nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66820cca-4093-4f09-b5f0-aeca5cbf82fd",
   "metadata": {},
   "source": [
    "## Get our own protein sequence novelty categories\n",
    "1. exact AA match\n",
    "2. truncated AA match\n",
    "3. novel, non-truncation AA seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0f60c28b-e5c8-4a9f-92d6-23d173590e7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. truncated AA match\n",
    "def get_aa_seq_truncations(x, ref_orfs):\n",
    "    # limit to just references from same gene\n",
    "    temp = ref_orfs.loc[ref_orfs.gid == x.gid]\n",
    "    for ind, entry in temp.iterrows():\n",
    "        if x.seq in entry.seq:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "# p_df['trunc_annot_aa'] = p_df.apply(lambda x: get_aa_seq_truncations(x, ref_orfs), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "88b9f40e-837b-4473-95dd-152e13a02a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p_df.to_csv('temp.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4f5878da-aa76-4f87-b48d-89680c602a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. elongated AA match\n",
    "def get_aa_seq_elongations(x, ref_orfs):\n",
    "    # limit to just references from same gene\n",
    "    temp = ref_orfs.loc[ref_orfs.gid == x.gid]\n",
    "    for ind, entry in temp.iterrows():\n",
    "        if entry.seq in x.seq:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "# p_df['elong_annot_aa'] = p_df.apply(lambda x: get_aa_seq_elongations(x, ref_orfs), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3d97e231-c104-4715-8203-f01196be699d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p_df.to_csv('temp.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "9db99c80-ce38-4ee2-a0ca-450256c5b9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p_df.to_csv('250223_protein_characterization_mt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "793cfe8a-9a3e-4a1e-a0fc-292f89a39ff5",
   "metadata": {},
   "source": [
    "## What % of novel transcripts, PC-gene transcripts have an ORF predicted?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fe312ff7-183b-4c60-88c1-06308ae7cabb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fmt_perc_print(n_num, n):\n",
    "    return f'{(n_num/n)*100:.2f}% ({n_num}/{n})'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "29af6b05-f9d7-4d97-bfa4-bbacabd5d25d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "96.33% (32498/33737) novel, protein-coding transcripts have a predicted full ORF\n"
     ]
    }
   ],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "n_num = len(p_df.loc[p_df.isoform.isin(tids)].isoform.unique())\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} novel, protein-coding transcripts have a predicted full ORF')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f136473-fed4-4464-a5ea-b8d433118fd2",
   "metadata": {},
   "source": [
    "## What % of novel transcripts, PC-gene transcripts have a non-NMD ORF predicted?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f2e5e18a-ae7e-4797-b0ac-174764a258eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83.39% (28134/33737) novel, protein-coding transcripts have a predicted full ORF w/o predicted NMD\n"
     ]
    }
   ],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "n_num = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)].isoform.unique())\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} novel, protein-coding transcripts have a predicted full ORF w/o predicted NMD')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08df5921-60ff-4ea5-bd5a-f75112c9c7de",
   "metadata": {},
   "source": [
    "## What % of novel, PC-gene transcripts w/ non-NMD ORF encode for a novel AA seq (wrt the sequences from the same gene)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "136b4834-cfb8-4052-9203-1723b01c6a4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80.58% (22671/28134) novel, protein-coding transcripts w/ pred. full ORF and w/o predicted NMD are predicted to have a novel AA seq\n"
     ]
    }
   ],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "n = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)].isoform.unique())\n",
    "\n",
    "n_num = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)&\n",
    "                     (p_df.annot_aa==False)].isoform.unique())\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} novel, protein-coding transcripts w/ pred. full ORF and w/o predicted NMD are predicted to have a novel AA seq')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0097f539-3d9d-4535-81d1-479192892ca2",
   "metadata": {},
   "source": [
    "## What % of novel, PC-gene transcripts w/ non-NMD ORF encode for a novel AA seq (excluding truncations) (wrt the sequences from the same gene)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "886008b1-e543-4dd5-8642-b5cdf0f78d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61.11% (17194/28134) novel, protein-coding transcripts w/ pred. full ORF and w/o predicted NMD are predicted to have a novel AA seq that is not a truncation of an annotated one\n"
     ]
    }
   ],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "n = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)].isoform.unique())\n",
    "\n",
    "n_num = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)&\n",
    "                     (p_df.annot_aa==False)&\n",
    "                     (p_df.trunc_annot_aa==False)].isoform.unique())\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} novel, protein-coding transcripts w/ pred. full ORF and w/o predicted NMD are predicted to have a novel AA seq that is not a truncation of an annotated one')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b816ab0a-baff-4f4b-b500-0f561e8cef1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is a good sanity check because I get the same number as in `explore_orfs_aas_2.ipynb`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7818fdb1-d7d2-4201-8665-9a774946345d",
   "metadata": {},
   "source": [
    "#### And for this analysis, what is the number of unique AA seqs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d4ceb6ed-055b-4be8-9807-3d8e0caa4e2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15553 novel, non-truncation, unique AA seq+gene pairs from novel transcriptsfrom protein coding genes,not predicted to have nmd, not truncations.\n"
     ]
    }
   ],
   "source": [
    "n_num = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)&\n",
    "                     (p_df.annot_aa==False)&\n",
    "                     (p_df.trunc_annot_aa==False)][['gid', 'seq']].drop_duplicates())\n",
    "print(f'{n_num} novel, non-truncation, unique AA seq+gene pairs from novel transcripts'+\\\n",
    "      'from protein coding genes,'+\\\n",
    "      'not predicted to have nmd, not truncations.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46897a0a-48b4-4095-beb3-b6003965e7d6",
   "metadata": {},
   "source": [
    "## What % of novel AA seqs represent extensions of known ones (wrt the sequences from the same gene)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "281c063f-5870-4c96-993f-301e8fb8ad2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21.89% (3405/15553) of uniq novel non-truncation, non-nmd AA seqs from novel transcripts from PC genes are elongations of existing ones from the same gene\n"
     ]
    }
   ],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "n = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)&\n",
    "                     (p_df.annot_aa==False)&\n",
    "                     (p_df.trunc_annot_aa==False)][['gid', 'seq']].drop_duplicates())\n",
    "\n",
    "n_num = len(p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                     (p_df.protein_is_nmd==False)&\n",
    "                     (p_df.annot_aa==False)&\n",
    "                     (p_df.trunc_annot_aa==False)&\n",
    "                     (p_df.elong_annot_aa==True)][['gid', 'seq']].drop_duplicates())\n",
    "\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} of uniq novel non-truncation, non-nmd AA seqs from novel transcripts from PC genes are elongations of existing ones from the same gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf1c949-8714-4361-a0f3-90cf81778152",
   "metadata": {},
   "source": [
    "## For the rest of these novel things, what do the blastp identities look like ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "b64ec251-1168-48ef-80e6-374d876ae274",
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_pc_df = mt_df.loc[(mt_df.associated_gene_biotype=='Protein Coding')&(mt_df.structural_category!='FSM')]\n",
    "n = len(novel_pc_df.index)\n",
    "tids = novel_pc_df.isoform.tolist()\n",
    "\n",
    "assert n == 33737 # number reported in the paper just double checking (2d)\n",
    "\n",
    "temp = p_df.loc[(p_df.isoform.isin(tids))&\n",
    "                 (p_df.protein_is_nmd==False)&\n",
    "                 (p_df.annot_aa==False)&\n",
    "                 (p_df.trunc_annot_aa==False)&\n",
    "                 (p_df.elong_annot_aa==False)]\n",
    "\n",
    "# get only uniq aa seqs / blast ids\n",
    "temp = temp[['gid', 'seq', 'blastp_identity']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "0dcc5103-87fb-4db3-b4bd-d3e018d76f12",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(temp[['gid', 'seq']].drop_duplicates()) == len(temp.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "59649070-26ba-4d17-a0d7-7406dba75b72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x7f804e9d1e90>"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAFgCAYAAACFYaNMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAAsTAAALEwEAmpwYAAAu6ElEQVR4nO3de5zcdX3v8ddnZ3b2vkk2LEnIhQQSwSBIIQQUwSoHDBWNbUFBewSLxVPltLXVHuwFldpTra2XHmmVClXxgkqrjTQVQXqlFBKuEkIkkphsCMkmu8leZndnZ+Zz/vj9ZjMsu5vZzfzmNzt5Px+PfeR3/30yLJ985/P7/r5fc3dERKTy6uIOQETkeKUELCISEyVgEZGYKAGLiMRECVhEJCbJuAMol3Xr1vkPf/jDuMMQEZmITbSxZlrABw4ciDsEEZFpqZkELCIy2ygBi4jERAlYRCQmSsAiIjFRAhYRiYkSsIhITJSARURiogQsIhITJWARkZgoAYuIxCTSBGxm68xsm5ltN7ObJth/sZk9ZmZZM7ty3L5lZvYjM9tqZs+Y2fIoYxURqbTIErCZJYBbgcuB1cA1ZrZ63GG7gOuAb05wia8Bn3b3VwJrgf1RxSoiEocoR0NbC2x39+cBzOwuYD3wTOEAd98Z7ssXnxgm6qS73xceNxBhnCJyHLjrkV3k3HnX+SfHHcqYKEsQi4HdRetd4bZSvAI4ZGb/YGaPm9mnwxb1S5jZDWa22cw2d3d3lyFkEalVX/mvnXzu/ueopomIq/UhXBK4CPgQcB5wCkGp4iXc/TZ3X+Puazo7OysboYjMKnt6h+juH2HnwXTcoYyJMgHvAZYWrS8Jt5WiC3jC3Z939yzwfeCc8oYnIseLw0Oj9I9kAXj4+YMxR3NElAl4E7DKzFaYWQq4GtgwjXPnmlmhWftGimrHIiLT0dV7pNX78I6eGCN5qcgScNhyvRG4F9gKfMfdt5jZLWb2VgAzO8/MuoCrgC+Z2Zbw3BxB+eHHZvYTguk8/jaqWEWktnX1DgGwrKOZR6ooAUc6J5y7bwQ2jtt2c9HyJoLSxETn3gecFWV8InJ8KCTgt519En/1wHa6+0fobGuIOarqfQgnIlI2Xb1pWlIJVi5oA+BQOhNzRAElYBGpeV29QyyZ10x7Y/Clv284G3NEASVgEal5QQJuor2pHoC+4dGYIwooAYtIzevqTQcJOGwB96sFLCISvb7hUfqHsyye10R7Y9gCHlILWEQkcr2DwQO3+S0NtIUJWC1gEZEKKCTbtsYkjfV11CdMNWARkUooJODWxiRmRltjPf1KwCIi0Ssk20L9t70xSd+QShAiIpEbGDlSggj+VAtYRKQixkoQDUECbm9K6iGciEglFFq7hR4QbQ31eggnIlIJ/SNZGpJ1pJJBulMLWESkQvqHs2P1XwhawnoRQ0SkAoIEXD+23t5Yz2AmRzaXn+KsylACFpGaNjA8Oq4FHCwXekfESQlYRGpa/3B2rAcEHEnA1VAHVgIWkZo2MPLSGnBhSMrDVVAHVgIWkZoWtICP1IDVAhYRqZC+cTXgsSEpq6AvsBKwiNQsd2dgJDs2EDscScBqAYuIRGgwk8M9GAmtoL0pnBdONWARkeiMfw0ZoDkVJOCh0VwsMRVTAhaRmjUw/NKR0ADqE0adwbASsIhIdPrGjYQGYGY01ScYyigBi4hEZqISBEBjfYLhrBKwiEhkxg/GXtBYn2AoU+NjQZjZOjPbZmbbzeymCfZfbGaPmVnWzK6cYH+7mXWZ2ReijFNEalP/BDVggMb6utpuAZtZArgVuBxYDVxjZqvHHbYLuA745iSX+RPg36OKUURq28AENWCAplSC4RqvAa8Ftrv78+6eAe4C1hcf4O473f0p4GXfBczsXGAB8KMIYxSRGpYOk2yh61lBYzJR893QFgO7i9a7wm1HZWZ1wF8CHzrKcTeY2WYz29zd3T3jQEWkNqVHs6SSdSTq7CXbm1IJdUObwvuBje7eNdVB7n6bu69x9zWdnZ0VCk1EZovhTI7mVOJl2xuSCYZG438Ilzz6ITO2B1hatL4k3FaK1wAXmdn7gVYgZWYD7v6yB3kiIpNJZ3I01788ATelEoxUQQs4ygS8CVhlZisIEu/VwDtLOdHd31VYNrPrgDVKviIyXenRHI0TtICb6utquwbs7lngRuBeYCvwHXffYma3mNlbAczsPDPrAq4CvmRmW6KKR0SOP0OTlCAa66ujBhxlCxh33whsHLft5qLlTQSliamu8RXgKxGEJyI1Lp3J0lz/8jTXVF/7vSBERGI1NJqnaaKHcPUJhkfzuHsMUR2hBCwiNWsok52wBNEUPpgbycbbE0IJWERqVjqTG0u2xRrrg9QXdx1YCVhEatZQJjdhCaKQlOOuAysBi0jNSk/RCwKIfUxgJWARqUnuztBojqbUy3tBFBLwcMxvwykBi0hNKiTXiWrAhbKEShAiIhFIZ4KhKCcsQSSD1Bf368hKwCJSkwpDUU74EE4tYBGR6BS6mE31EE41YBGRCBwZjF3d0EREKqqQgBsneAjXoBcxRESiMzRaeAg38WA8oAQsIhKJqUoQjUrAIiLRKbzlNlE/4PpEHck6Uw1YRCQKheQ6UTc0CMcEzqgXhIhI2U1VgoBwTOCsWsAiImU31gsiOUkLOFXHsAbjEREpv6FMlqb6BHV1NuH+xqRawCIikQhGQpu49QtBbVjDUYqIRGCy2TAKGpMJvYosIhKFyaakL2hMxT8zshKwiNSkyWbDKGhM1mlSThGRKBytBtxQn2BED+FERMpv6Cg14FSijhHVgEVEyi+dyU44EE9BQ71KECIikZhsSvqChmSdShAiIlFIj079EK4hmajtFrCZrTOzbWa23cxummD/xWb2mJllzezKou1nm9lDZrbFzJ4ys3dEGaeI1J6j1YAbknVksnncvYJRvVRkCdjMEsCtwOXAauAaM1s97rBdwHXAN8dtTwPvdvczgHXA58xsblSxikhtyeWdkWz+KL0ggvSXycXXCp68Qn3s1gLb3f15ADO7C1gPPFM4wN13hvte8gm4+0+Lll8ws/1AJ3AownhFpEYMTTEhZ0FDOEjPSDY/tlxpUZYgFgO7i9a7wm3TYmZrgRTwswn23WBmm81sc3d394wDFZHaks4E0xE1TdELIpUM0l+cXdGq+iGcmS0C7gTe4+4v+5Tc/TZ3X+Puazo7OysfoIhUpcIgO81HqQEDsfaEiDIB7wGWFq0vCbeVxMzagX8C/tDd/7vMsYlIDTvabBhQnIBrswW8CVhlZivMLAVcDWwo5cTw+O8BX3P3uyOMUURqUGEw9qkTcFgDrsUShLtngRuBe4GtwHfcfYuZ3WJmbwUws/PMrAu4CviSmW0JT387cDFwnZk9Ef6cHVWsIlJbSipB1HgvCNx9I7Bx3Labi5Y3EZQmxp/3deDrUcYmIrXryHxwU7yKPPYQrjZrwCIisThSA548xdV6DVhEJBZDJXRDK+4HHBclYBGpOWl1QxMRicdx3wtCRCQuQ5kcdXaklTuRQi8IlSBERMpoaDQYCc3MJj2mkJwzKkGIiJRPOpOb8gEc6CGciEgkhjLZKUdCg6LBeJSARUTK52hT0gMk6oxknakXhIhIOQ2N5micogtaQUMy3pmRlYBFpOYMldACBmioj3deOCVgEak5pZQgIP6ZkZWARaTmDI0evRcEHJmYMy5KwCJSc9KZ7JSvIRfEPTW9ErCI1JyhTG7K15ALUsk6JWARkXIKShCqAYuIVNRoLs9ozksrQdSrG5qISNmUMhJagWrAIiJlNDStBKwShIhI2RSmIyq1H7C6oYmIlEm6MB1RfSn9gFWCEBEpm7Ep6dUNTUSkstLTSMDBYDyqAYuIlEUhAZc0Glq9WsAiImUzPK2HcAmyeSebiycJKwGLSE05UoIobTAegIwSsIjIsRvrBVFiDRjim5o+0gRsZuvMbJuZbTezmybYf7GZPWZmWTO7cty+a83sufDn2ijjFJHaMfYiRkk14OCYmmsBm1kCuBW4HFgNXGNmq8cdtgu4DvjmuHM7gI8C5wNrgY+a2byoYhWR2jE0miNZZ2OTbk4llajdFvBaYLu7P+/uGeAuYH3xAe6+092fAsb/7d8E3OfuPe7eC9wHrIswVhGpEekSh6KEoBcEENvryFEm4MXA7qL1rnBb2c41sxvMbLOZbe7u7p5xoCJSO0qdDw6CXhAQ39T0s/ohnLvf5u5r3H1NZ2dn3OGISBVIj+ZK6gEBRQ/harAFvAdYWrS+JNwW9bkichwbymRLegkDarsXxCZglZmtMLMUcDWwocRz7wUuM7N54cO3y8JtIiJTGhqdRgmivkZLEO6eBW4kSJxbge+4+xYzu8XM3gpgZueZWRdwFfAlM9sSntsD/AlBEt8E3BJuExGZUqlT0kNRL4iYEnBphZIZcveNwMZx224uWt5EUF6Y6Nw7gDuijE9Eas9QJkdna0NJx9ZyLwgRkYqbVje0ZLwtYCVgEakp0ylBqBuaiEgZDY/mSpoNA4pKEDGNCawELCI1w91JZ7LTaAGrBCEiUhYj2Tx5L20kNIi/F4QSsIjUjOmMhAZgFgzaE9fMyErAIlIzpjMlfUFDsk7d0EREjlVhNoxSSxAQ79T0JSVgM/sHM3uzmSlhi0jVGprGdEQFwczIVZyAgb8G3gk8Z2afNLPTIoxJRGRGxqYjKrEGDIWZkau4BOHu97v7u4BzgJ3A/Wb2X2b2HjOrjzJAEZFSpUdrsAQBYGbzCaYPei/wOPB5goR8XySRiYhM03Bmpg/hqngwHjP7HnAacCfwFnffG+76tpltjio4EZHpSM8gAQfd0OIpQZRaqf7bcGSzMWbW4O4j7r4mgrhERKZtrAQxnRpwso6BkWxUIU2p1BLEJybY9lA5AxEROVZDhYdw060Bx9QLYsoWsJktJJgMs8nMfgGwcFc70BxxbCIi05Ke5ptwEG8viKOVIN5E8OBtCfCZou39wB9EFJOIyIykMzka6+tIJkp/ZaFqH8K5+1eBr5rZr7r731coJhGRGRkYydIyjZcwIN5uaEcrQfyau38dWG5mvzt+v7t/ZoLTRERiMTiSpaVhugm4LrbxgI8WaUv4Z2vUgYiIHKvBkdyMEnAmV4UtYHf/UvjnxysTjojIzA2OZGmZRg8IOFIDdnfM7OgnlFGpg/H8uZm1m1m9mf3YzLrN7NeiDk5EZDoGMzMoQdQncIfRnEcU1eRKfVR4mbv3AVcQjAWxEvhwVEGJiMzE4EiW1hmUICCeqelLTcCFv9Gbge+6++GI4hERmbHBkdJnRC6Ic164Uv+puMfMngWGgN80s05gOLqwRESmb2a9IOKbmr7U4ShvAl4LrHH3UWAQWB9lYCIi0+HuDGZmUIKIcWr66UR6OkF/4OJzvlbmeEREZmR4NJgRublheiWIwszIcXRFK3U4yjuBU4EngMI/E44SsIhUicKIZjNvAVdpAgbWAKvdfVr9NMxsHcHA7Qngy+7+yXH7GwiS+LnAQeAd7r4znGXjywQDvieBr7n7n03n3iJyfClMRzSTV5GhimvAwNPAwulc2MwSwK3A5cBq4BozWz3usOuBXndfCXwW+FS4/Sqgwd3PJEjO7zOz5dO5v4gcXwot4JZpliDi7IZW6j8VJwDPmNkjwEhho7u/dYpz1gLb3f15ADO7i+DB3TNFx6wHPhYu3w18wYJXURxoCevNTUAG6CsxVhE5Dg2OBAl0xr0gqrgE8bEZXHsxsLtovQs4f7Jj3D1rZoeB+QTJeD2wl2Dc4Q+6e8/4G5jZDcANAMuWLZtBiCJSKwYLJYiZ1oCrtQTh7v9G8AZcfbi8CXgswrjWEjzsOwlYAfyemZ0yQVy3ufsad1/T2dkZYTgiUu0GR2ZWAy70gqjaN+HM7DcIWqVfCjctBr5/lNP2AEuL1peE2yY8Jiw3zCF4GPdO4IfuPuru+4EHCR4EiohMaHCmNeBqbwEDHwAuJKzDuvtzwIlHOWcTsMrMVphZCrga2DDumA3AteHylcADYU+LXcAbAcysBbgAeLbEWEXkOFSoAU9/LIggYWeqOAGPuHumsBK2VqfskubuWeBG4F5gK/Add99iZreYWeHh3e3AfDPbDvwucFO4/Vag1cy2ECTyv3P3p0r9S4nI8afQAm6edje06u8F8W9m9gcEk3NeCrwf+MHRTgqnst84btvNRcvDBF3Oxp83MNF2EZHJDGSypBJ1pJKlzwcHRQk4hl4QpUZ6E9AN/AR4H0FS/aOoghIRma70SG7a9V+AZKKORJ1V72ho7p43s+8D33f37mhDEhGZvsGR7LTLDwXBrBhV1gvCAh8zswPANmBbOBvGzVOdJyJSaQMzGIy9IBXT1PRHK0F8kKD3w3nu3uHuHQQvU1xoZh+MPDoRkRKlMzMrQUBhZuTqS8D/E7jG3XcUNoSvFv8a8O4oAxMRmY6BGQzGXtCQTMQyHOXREnC9ux8YvzGsA9dHE5KIyPQFMyLXUA2YYBCcmewTEamo/uEsbY0zTMD18ZQgjhbtq81solHIDGiMIB4RkRnpHx6lvWlmX8wbkonq64bm7jOraIuIVFA2l2cwk5t5C7hKSxAiIlWvMBh7e+PMWsDV2g1NRKTq9Q0FCfiYWsBV2A1NRKTq9Q2PAhxTDbgau6GJiFS9sQQ8wxJE0AJWDVhEZNqOuQRRrxqwiMiM9Ict4DmzrBuaErCIzHp9w2V4CKduaCIi01doAR/LaGijOSeXn3Kin7JTAhaRWa9vKBiKMpmYWUqLa144JWARmfX6hkdnXH6AI9MSKQGLiExT//DojLugQfHU9JWtAysBi8is1zeUpb3pWFrAQQmi0j0hlIBFZNbrHxml7VhawDFNTa8ELCKzXt9QlvZjqAEXprIfrvB4EErAIjLrBQ/hytECVgIWESmZu9M/XK4asEoQIiIlS2dy5PJell4Q6oYmIjIN/WOvIasE8RJmts7MtpnZdjO7aYL9DWb27XD/w2a2vGjfWWb2kJltMbOfmJnmoBORlykMRXlsL2LUWDc0M0sAtwKXA6uBa8xs9bjDrgd63X0l8FngU+G5SeDrwP9y9zOAXwRGo4pVRGavQ+kgNcxrTs34GmMt4AqPCRxlC3gtsN3dn3f3DHAXsH7cMeuBr4bLdwOXmJkBlwFPufuTAO5+0N0rP1SRiFS9nsEMAPNaVIIothjYXbTeFW6b8Bh3zwKHgfnAKwA3s3vN7DEz+/2JbmBmN5jZZjPb3N3dXfa/gIhUv950kIA7Wo6lBVxjJYhjlAReB7wr/POXzeyS8Qe5+23uvsbd13R2dlY6RhGpAmMt4GMpQdTgWBB7gKVF60vCbRMeE9Z95wAHCVrL/+7uB9w9DWwEzokwVhGZpXoHMzSnEjTWJ2Z8jVSi9rqhbQJWmdkKM0sBVwMbxh2zAbg2XL4SeMDdHbgXONPMmsPE/HrgmQhjFZFZqiedOabWL0BdnZFKVH5euJn32zgKd8+a2Y0EyTQB3OHuW8zsFmCzu28AbgfuNLPtQA9Bksbde83sMwRJ3IGN7v5PUcUqIrNX72DmmOq/BcHMyDWSgAHcfSNB+aB4281Fy8PAVZOc+3WCrmgiIpPqSY8yrxwJuL7y88JV60M4EZGS9A5m6GieeRe0gjhKEErAIjKr9Q5mmHuMNWCAhvrKT02vBCwis1Ymm6d/JFvGGrBKECIiJTmULrwFV6YErBawiEhpegpvwZWjBJFM1FQ/YBGRSJVjHIgC9YIQEZmG3sFgJLSy1YDVAhYRKU05SxCpZB3DeggnIlKa3rAEUY5uaI31Cc2KLCJSqoMDI7Q1JsemlT8WTfUJtYBFREq1r2+Ehe3lma2sqT7BkBKwiEhp9vUPs6BMCbg5FSTgYEDGylACFpFZa3/fCCe2N5TlWo2pBO6VnRVDCVhEZqV83tnXV74WcFM4oPtQpnJlCCVgEZmVetIZsnlnQVt5WsBjCbiCdWAlYBGZlfb1DQOUrwWcUgIWESnJ/r4RABbMKU8CblQJQkSkNOVuATeHLeBK9gVWAhaRWWlf2ALubC1vDTitFrCIyNT29Q8zvyVVlrfgoKgEoRawiMjU9vcNc2KZyg9w5CGcShAiIkexr2+EBWV6CQPUD1hEpGR7Dw+VbRwIOPIQTiUIEZEpDI5kOTCQYWlHc9mu2aiHcCIiR7e7Nw3AyfPLl4AbknWYqQYsIjKlnx8MEvCyMraAzSwYklItYBGRye3uCVvAHS1lvW6lxwRWAhaRWefnB9O0NyaZ03zssyEXa0rVUAI2s3Vmts3MtpvZTRPsbzCzb4f7Hzaz5eP2LzOzATP7UJRxisjssqsnzcnzy9v6BWqnBGFmCeBW4HJgNXCNma0ed9j1QK+7rwQ+C3xq3P7PAP8cVYwiMjvt6kmXtf5bUEst4LXAdnd/3t0zwF3A+nHHrAe+Gi7fDVxiZgZgZm8DdgBbIoxRRGaZXN7p6k2zrIw9IAoaa6UFDCwGdhetd4XbJjzG3bPAYWC+mbUC/wf4+FQ3MLMbzGyzmW3u7u4uW+AiUr32Hh5iNOecHEULuMIzI1frQ7iPAZ9194GpDnL329x9jbuv6ezsrExkIhKrHQcGASJpAVe6F0QywmvvAZYWrS8Jt010TJeZJYE5wEHgfOBKM/tzYC6QN7Nhd/9ChPGKyCyw7cV+AE5b0Fb2azdXuAYcZQLeBKwysxUEifZq4J3jjtkAXAs8BFwJPODBnNAXFQ4ws48BA0q+IgKwdW8/J7Y1ML9M4wAXa0xVtgYcWQJ296yZ3QjcCySAO9x9i5ndAmx29w3A7cCdZrYd6CFI0iIik9q2r4/TFpa/9QuV74YWZQsYd98IbBy37eai5WHgqqNc42ORBCcis042l+en+wa47rXLI7l+oQbs7oQdsiJVrQ/hREReZufBNJlsPpL6LwT9gPMOmVw+kuuPpwQsIrPGsy/2AXD6ouhKEADDGSVgEZGXeHZvP4k6Y+WJrZFcvzAtUXo0G8n1x1MCFpFZ44ndhzhtQRsNyUQk16/0tERKwCIyK2RzeR7b1cua5fMiu0elZ8VQAhaRWWHr3n7SmRxrlndEdo+2xqBj2OCIShAiImM2/7wHgDUnR9cCbm0IEvCAErCIyBGbd/Zy0pxGTprbFNk9WpSARUReyt3ZtLOHcyMsP8CREoQSsIhIaMsLfezvH+GiVSdEep+xEsSwErCICAAPPLsfM3jDaSdGep/mVAIztYBFRMb8+Nn9vHrJXDrbyj8CWjEzo7UhSb9awCIi0N0/wpO7D3HJ6dG2fgvaGpJqAYuIAPzz03sBuPSMBRW5X2tjUjVgERGAv3+0i1cuauf0he0VuV+rWsAiIvDcvn6e7DrMlecuqdg9Wxvr6VcCFpHj3Xcf7SJZZ6w/+6SK3bOtIalXkUXk+DY4kuWuR3Zx2RkLOCGC+d8m09qgGrCIHOe+u3k3fcNZ3nvRKRW9b2ujasAichwbzeW5/cEdnHvyPM5ZFt3gOxNpCR/C5fMe+b2UgEWk6tz9aBe7e4Z4/y+eWvF7t4WvIw9mom8FKwGLSFUZHs3x+fuf45xlc3ljhV6+KNZawQF5lIBFpKrc/p87eLFvmA+/6fSKTA0/XiUH5FECFpGqsefQEP/vgedYd8ZCXnPq/FhiKLSAK9EXWAlYRKqCu/PRf9wCwB+/ZXVscbSpBSwix5vvPb6H+7fu4/cuPY3FEc56cTSqAYvIcWV3T5qPbtjCmpPn8euvWxFrLJWcFy7SBGxm68xsm5ltN7ObJtjfYGbfDvc/bGbLw+2XmtmjZvaT8M83RhmniMRneDTHb37jUQz4zNvPJlFX+Qdvxdoa6oFZXoIwswRwK3A5sBq4xszGF3auB3rdfSXwWeBT4fYDwFvc/UzgWuDOqOIUkfjk886H736Kp/f08ZdvP5tl85vjDomWhgQw+1vAa4Ht7v68u2eAu4D1445ZD3w1XL4buMTMzN0fd/cXwu1bgCYzq9zL4CISOXfnTzdu5QdPvsDvrzuNS1dXZrzfo0km6miqT9A3NBr5vaJMwIuB3UXrXeG2CY9x9yxwGBjf9+RXgcfcfWT8DczsBjPbbGabu7u7yxa4iEQrn3f+5J6t3P6fO7jutcv5zddX/o23qXS0pOhJZyK/T1U/hDOzMwjKEu+baL+73+bua9x9TWdnZ2WDE5EZGRzJ8oFvPsYdD+7g1y9cwUffsjqWFy6m0tGSomcw+gScjPDae4ClRetLwm0THdNlZklgDnAQwMyWAN8D3u3uP4swThGpkEd29PCh7z5JV2+aP3rzK7n+dSuqLvlC5RJwlC3gTcAqM1thZingamDDuGM2EDxkA7gSeMDd3czmAv8E3OTuD0YYo4hUwPBojk/c8wzvuO0hHOdbv3EB773olKpMvgDzW1IcHJjFLWB3z5rZjcC9QAK4w923mNktwGZ33wDcDtxpZtuBHoIkDXAjsBK42cxuDrdd5u77o4pXRKLx+K5efu+7T/J89yC/dsEyPnL5K2lpiPLL97Gb15KitwI14Eg/BXffCGwct+3mouVh4KoJzvsE8IkoYxORaI1kg1HNvvhvP2NheyNfv/58XrfqhLjDKklHS4p0JsfwaI7G+kRk96nuf4ZEZFb66b5+futbj/Psi/28fc0S/uiK1bQ31scdVsnmt6QAODiYifS1aCVgESkbd+euTbv5+A+20NqQ5PZr13DJK6ujf+90dIQJuGdACVhEZoG+4VE+8g8/4Z+e2stFq07gL9/+ak5sa4w7rBmZ31poAb/s9YOyUgIWkWP25O5D3Pitx3jh0DC/v+40/tfFp1IX85gOx6KjJXjxNuquaErAIjJj7s4dD+7kk/+8lRPbGvnO+y7g3JM74g7rmHU0hyUIJWARqUaH0hk+9N2nuH/rPi5dvYBPX3kWc8PENdu1NyVJ1pkSsIhUn0d/3sP//ubjdA+McPMVq3nPhcur9qWKmTAz5lXgbTglYBEpWTaX52/+9Wd87sfPsXhuE3//m6/lrCVz4w4rEvNbUhxUAhaRavDTff186LtP8lTXYa44axH/91fOnFV9e6erEuNBKAGLyJSyuTy3/cfzfO6+52htTHLrO8/hzWctijusyHW0pNjyQl+k91ACFpFJbd7Zw0c3bGHLC31c/qqF/MnbXsUJrcfH3AgL2hu5f+s+3D2y+rYSsIi8TFdvmr+4dxvff+IFFs1p5NZ3nsMvnbmwph60Hc2yjmaGR/N0D4xE9kKJErCIjNnXN8yt/7Kdbz2yCzPjxjes5P1vOJXm1PGXKpZ1BPPT7e5JKwGLSHSe3nOYv3twJz948gXy7rz9vKXc+IaVnBThOAjVbmmYgHf1pCN7uUQJWOQ49eLhYe556gV+8OQLPNl1mOZUgqvXLuW9rzulKmYnjtuSeU2Ywa6DQ5HdQwlY5DjSM5jhn5/ey4YnXuCRnT24w6sWt/PHV6zmqjVLarpb2XQ11idY2N7Iz3sGI7uHErBIjdtzaIj7trzIvVv28cjOHnJ559TOFn7nkldwxasXcWpna9whVq2lHc3s7klHdn0lYJEac2BghM07e9m0s4f/fv7gWF/WVSe28r6LT+HNZy1i9aL246pHw0wt62jmP57rjuz6SsAis9wLh4Z4ZEcPD+84yMM7eni+O/jK3JCs4+ylc/nI5adz6eoFnKKW7rQt62hmX99IZFMTKQGLzCLuzu6eIf57x0Eefr6HR3YeZHdP8JCorTHJecs7ePuapZy3fB6vWjyHhmR085kdD4q7oq1a0Fb26ysBi1Sx/uFRnn2xn2de6OPxXb08vKOHvYeHgeBV2bXLO3jPa1dw/ikdnL6wncQsHgS9Gq08MfjW8MzePiVgkVqUzzv7+ofZ3TPErp40u3rSbHuxj617+9lV9ADohNYGzj+lgwtWdHD+KfNZ2dk6q2edmA1euaid1oYkj+zoYf3Zi8t+fSVgkYi4O3sPD/NsmExfODREbzpDz2CGvqEs/SOj9A9n6R/Oksv72HlmsHx+C2cunsM7zlvKKxe1sXrRHBa0N+jBWYUl6oxzT57Hpp09kVxfCVikjDLZPI/s6OH+rfv48bP7xuqzAPOa6+loSdHRkmLRnEZe0dhKW2M97U1JFs1pYmlHM0vnNbF4XpNqt1Vk7YoOPn3vNnoHM8xrKe+MH0rAIseoqzfNQz87yL9u6+bff9pN/0iWhmQdF648gesvXMEZi+dw2sI2veQwS523PHgNedPOHi47Y2FZr60ELFKCwZEs3f0jdA+MsPfwMNv3D/Cz/QM82XWIrt6glXtiWwNXvHoRl5y+gAtXnkBTSq3YWnDWkjmkknX85/YDSsAiM5XLOy8cGuLnB9PsPDjI7t40XT1D7OsbZjCTI53JksnmyeWdvDu5fPCTyeUZHs2/5Fp1FnRROnPxHH7jolO44JT5rDpRD8VqUWN9gjefuYi7H+3idy99RVknHlUCllnL3ekbynJgcIQD/SMcHMxwcGCEnsFRDg1lOJwe5dDQKIfSGXrTo3T1phnNHXnYlUrUsXheEwvbG1kyL0VzKkEqUUeizqirMxJmJOqMZJ1xQlsDna0NdLY1sKC9kZPnN0fSMV+q0/tefwrfe3wPX3vo5/zWJavKdt1IE7CZrQM+DySAL7v7J8ftbwC+BpwLHATe4e47w30fAa4HcsBvufu9UcYq5TOayzM4EjzZz7mTz4PjGIYZGMGss8XLhYZjOpNjcCTLwEjQO6BnMMOBgeCr/4H+DAcHRzhQtFycUIu1NSSZ01zP3OZ65jalWDS3iTedsZDl85s5eX4LJ89vZmF7o1qsUpLTF7Zzyekn8ncP7uC9F60o2/jIkSVgM0sAtwKXAl3AJjPb4O7PFB12PdDr7ivN7GrgU8A7zGw1cDVwBnAScL+ZvcLdc1HFO5u4O3mHvAdflfP5SZZ93HL+yHIu72PXKXzVHs3nGc3myYZfu7M5ZzQXrOfywXou72TzzuBIkBwLPwcHM2NdrPqHs2X/O9cnjBNaGzihNWiJnr6wPVxPjW2f35pifmuKec0p6hN1ZY9Bjm8fvPQVdPeP0FTGbz5RtoDXAtvd/XkAM7sLWA8UJ+D1wMfC5buBL1jQ0XE9cJe7jwA7zGx7eL2HyhngVV/8L/b1jeAErSj34KfAPdhT2OZ4cMzY/nDr2P4j5xy53pF1/MgxR47npfcv7Jjwfi+NJ26pRN1Yt6qOlhTLOprpaAkSYGtjkmTRV3mzI39XL/ocij8jd2hOJWhpSNLakKS1Mcm85hSdrQ20NyXVB1Zi9arFc8p+zSgT8GJgd9F6F3D+ZMe4e9bMDgPzw+3/Pe7cl72GYmY3ADeEqwNmtq08oY85AThQ5mvOlGKZXDXFo1gmVk2xQOXj+aG7rxu/cVY/hHP324Dborq+mW129zVRXX86FMvkqikexTKxaooFqieeKAtle4ClRetLwm0THmNmSWAOwcO4Us4VEZnVokzAm4BVZrbCzFIED9U2jDtmA3BtuHwl8IAHBdINwNVm1mBmK4BVwCMRxioiUnGRlSDCmu6NwL0E3dDucPctZnYLsNndNwC3A3eGD9l6CJI04XHfIXhglwU+EFMPiMjKGzOgWCZXTfEololVUyxQJfGYV8sjdRGR44w6S4qIxEQJWEQkJkrAgJk1mtkjZvakmW0xs4+H21eY2cNmtt3Mvh0+TKxUTAkze9zM7qmCWHaa2U/M7Akz2xxu6zCz+8zsufDPeRWKZa6Z3W1mz5rZVjN7TRyxmNlp4edR+Okzs9+J63MJY/pg+Pv7tJl9K/y9juX3xsx+O4xji5n9TritIp+Nmd1hZvvN7OmibRPe2wJ/FX4+T5nZOVHENBkl4MAI8EZ3fzVwNrDOzC4geDX6s+6+EugleHW6Un4b2Fq0HmcsAG9w97OL+k7eBPzY3VcBPw7XK+HzBJ3aTwdeTfAZVTwWd98Wfh5nE4xlkga+F0csAGa2GPgtYI27v4rgwXfh9f6K/t6Y2auA3yB4e/XVwBVmtpLKfTZfAca/9DDZvS8n6GW1iuClrr+JKKaJBa+D6qfwAzQDjxG8tXcASIbbXwPcW6EYlhD8krwRuIdgzJpYYgnvtxM4Ydy2bcCicHkRsK0CccwBdhA+PI4zlnH3vwx4MM5YOPJWaQdB76Z7gDfF8XsDXAXcXrT+x8DvV/KzAZYDTx/tdwT4EnDNRMdV4kct4FD4lf8JYD9wH/Az4JC7F0aWmfB16Ih8juAXtjAI7fwYY4Fg6IYfmdmj4evfAAvcfW+4/CKwoAJxrAC6gb8LyzNfNrOWmGIpdjXwrXA5lljcfQ/wF8AuYC9wGHiUeH5vngYuMrP5ZtYM/BLBi1Vx/nea7N4TDZlQsf+3lIBD7p7z4OvkEoKvTqfHEYeZXQHsd/dH47j/JF7n7ucQfF37gJldXLzTg6ZDJfozJoFzgL9x918ABhn3NbaCsQAQ1lTfCnx3/L5KxhLWNNcT/CN1EtDCy7+GV4S7byUoffwI+CHwBMGwssXHVPS/U7Xcezwl4HHc/RDwLwRf1+aGr0hD5V6HvhB4q5ntBO4iKEN8PqZYgLHWFe6+n6DOuRbYZ2aLAMI/91cglC6gy90fDtfvJkjIccRScDnwmLvvC9fjiuV/ADvcvdvdR4F/IPhdiuX3xt1vd/dz3f1igtrzT4n3v9Nk94512AMlYMDMOs1sbrjcRDCG8VaCRHxleNi1wD9GHYu7f8Tdl7j7coKvtg+4+7viiAXAzFrMrK2wTFDvfJqXvkZeqc/mRWC3mZ0WbrqE4G3JisdS5BqOlB+IMZZdwAVm1mxmxpHPJq7fmxPDP5cBvwJ8k3j/O0127w3Au8PeEBcAh4tKFdGrVLG5mn+As4DHgacIksvN4fZTCMag2E7wFbOhwnH9InBPnLGE930y/NkC/GG4fT7Bg8LngPuBjgrFczawOfxv9X1gXoyxtBAMHjWnaFsssYT3/jjwbPg7fCfQEOPvzX8Q/APwJHBJJT8bgn8Q9wKjBN+arp/s3gQPuG8leObzE4JeJBX57+XuehVZRCQuKkGIiMRECVhEJCZKwCIiMVECFhGJiRKwiEhMlIBFRGKiBCxVwcyWFw8fWLT9X81s2rPXmtnbzGz1DGN5q5lNOFKXmQ3M5JrhudeZ2UlF618uxGhmfzDT68rspQQsteptwIwSsLtvcPdPljccAK4jGKehcJ/3uvsz4aoS8HFICViqSdLMvhEOtH53OJLWGDP7GzPbbEWD5ofbP2lmz4QDav+Fmb2WYICcT4eDpZ8atqQ/H64/bWZrJwsibKl+IVxeYWYPWTAg/SfGHfdhM9sU3rcwiP/yMP6/DeP8kZk1mdmVwBrgG2EMTYXWvZl9EmgKt3/DzG6xcBDz8Jp/ama/fewfr1QbJWCpJqcBf+3urwT6gPeP2/+HHgwIfxbwejM7y8zmA78MnOHuZwGfcPf/InjH/8MeDJr+s/D8Zg9GvHs/cEeJMX2eYPS1MwlebwXAzC4jGMR7LcHr0ecWjRK3CrjV3c8ADgG/6u53E7xC/a4wpqHCtdz9JmAo3P6uMLZ3h/epIxgT5OslxiuziBKwVJPd7v5guPx14HXj9r/dzB4jGLfjDIISw2FgGLjdzH6FYGaKyXwLwN3/HWgvDMB0FBdyZLCdO4u2Xxb+PE4wgP/pBIkXglHJngiXHyUYHLxk7r4TOGhmv1C4h7sfnM41ZHZIHv0QkYoZPzDJ2LqZrQA+BJzn7r1m9hWg0d2zYTnhEoJRv24kGMJzWtefZlwQDOLyZ+7+pZdsNFtOMMVVQQ5oKvE+xb5MUDNeSOmtdZll1AKWarLMzF4TLr8T+M+ife0EA7AfNrMFBOPwYmatBKORbQQ+SDAHGUA/0Dbu+u8Iz3kdwbCDh0uI6UGCEgDAu4q23wv8enh/zGxxYQjGKUwUU8GomdUXrX+PYED188J7SQ1SApZqso1gxo2tBMNMjk2Q6O5PEnzdf5ZgbNlCqaINuMfMniJI2L8bbr8L+HA4ddGp4bZhM3sc+CKlT07522FMP6Foqhp3/1EYx0PhvruZPLkWfAX4YuEh3Lh9twFPmdk3wutnCMby/Y6755CapOEo5bhgZv8KfMjdN8cdSynCh2+PAVe5+3NxxyPRUAtYpMqEL2dsJ5hGXcm3hqkFLMctM3sPQYmh2IPu/oE44pHjjxKwiEhMVIIQEYmJErCISEyUgEVEYqIELCISk/8PlUSM2h50JNAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(temp,\n",
    "            kind='kde',\n",
    "            x='blastp_identity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "121564f2-ff73-4149-84ce-8ca5c5cf5a05",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_blastp_thresh = 90\n",
    "temp['high_blastp'] = temp.blastp_identity>=high_blastp_thresh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ee0abd0a-4ec3-4d24-90d5-c276d60b7e90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>high_blastp</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>3715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>8433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seq\n",
       "high_blastp      \n",
       "False        3715\n",
       "True         8433"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[['seq', 'high_blastp']].groupby(['high_blastp']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "540966f3-2383-4655-86eb-5bd9628de40b",
   "metadata": {},
   "source": [
    "## Ok now switching gears -- with the categorization of non-nmd ORFs only, what % of CDS FSMs that don't match actually had an incomplete orf in gencode?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "0e794438-eb0f-4d82-90f0-88eab64d4ab1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>gid</th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000493376.2|ENST00000641515.2|ENSG000001...</td>\n",
       "      <td>MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000409316.1|ENST00000426406.4|ENSG000002...</td>\n",
       "      <td>MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000329982.2|ENST00000332831.5|ENSG000002...</td>\n",
       "      <td>MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000478421.2|ENST00000616016.5|ENSG000001...</td>\n",
       "      <td>MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000480678.2|ENST00000618323.5|ENSG000001...</td>\n",
       "      <td>MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>ENST00000618323.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  \\\n",
       "0  ENSP00000493376.2|ENST00000641515.2|ENSG000001...   \n",
       "1  ENSP00000409316.1|ENST00000426406.4|ENSG000002...   \n",
       "2  ENSP00000329982.2|ENST00000332831.5|ENSG000002...   \n",
       "3  ENSP00000478421.2|ENST00000616016.5|ENSG000001...   \n",
       "4  ENSP00000480678.2|ENST00000618323.5|ENSG000001...   \n",
       "\n",
       "                                                 seq                 gid  \\\n",
       "0  MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...   ENSG00000186092.7   \n",
       "1  MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...   ENSG00000284733.2   \n",
       "2  MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...   ENSG00000284662.2   \n",
       "3  MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...  ENSG00000187634.13   \n",
       "4  MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...  ENSG00000187634.13   \n",
       "\n",
       "                 tid  \n",
       "0  ENST00000641515.2  \n",
       "1  ENST00000426406.4  \n",
       "2  ENST00000332831.5  \n",
       "3  ENST00000616016.5  \n",
       "4  ENST00000618323.5  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add status of fsm cdss\n",
    "\n",
    "# get annotated AA sequencesa\n",
    "fasta_file = proc_cfg(config['ref']['pc'], od)\n",
    "fasta = Fasta(fasta_file)\n",
    "\n",
    "# Extract each entry's name and sequence\n",
    "ref_orfs = {\n",
    "    \"name\": [entry.name for entry in fasta],\n",
    "    \"seq\": [str(entry) for entry in fasta]\n",
    "}\n",
    "\n",
    "ref_orfs = pd.DataFrame(ref_orfs)\n",
    "ref_orfs['gid'] =  ref_orfs.name.str.split('|', expand=True)[2]\n",
    "ref_orfs['tid'] = ref_orfs.name.str.split('|', expand=True)[1]\n",
    "ref_orfs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "c08c4699-7fdb-460f-a2db-c7a1ec8a9b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = p_df.loc[p_df.protein_is_nmd==False]\n",
    "\n",
    "temp['aa_seq_novelty'] = 'Novel'\n",
    "temp.loc[(temp.annot_aa==True), 'aa_seq_novelty'] = 'Known'\n",
    "temp.loc[(temp.annot_aa==False)&\\\n",
    "         (temp.trunc_annot_aa==True), 'aa_seq_novelty'] = 'Truncation'\n",
    "\n",
    "temp.loc[~(temp.isoform.isin(ref_orfs.tid.tolist()))&\\\n",
    "         (temp.structural_category=='FSM')]\n",
    "\n",
    "temp.loc[~(temp.associated_transcript.isin(ref_orfs.tid.tolist()))&\\\n",
    "         (temp.structural_category=='FSM'),'structural_category'] = 'FSM w/o CDS'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "3d827d34-3a55-4daa-858b-c63242ef00a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_seq_novelty</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>n_t</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Known</td>\n",
       "      <td>FSM</td>\n",
       "      <td>34094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Known</td>\n",
       "      <td>FSM w/o CDS</td>\n",
       "      <td>842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Known</td>\n",
       "      <td>NIC</td>\n",
       "      <td>2647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Known</td>\n",
       "      <td>NNC</td>\n",
       "      <td>2816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Novel</td>\n",
       "      <td>FSM</td>\n",
       "      <td>3039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Novel</td>\n",
       "      <td>FSM w/o CDS</td>\n",
       "      <td>11109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Novel</td>\n",
       "      <td>NIC</td>\n",
       "      <td>7325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Novel</td>\n",
       "      <td>NNC</td>\n",
       "      <td>9869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Truncation</td>\n",
       "      <td>FSM</td>\n",
       "      <td>3870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Truncation</td>\n",
       "      <td>FSM w/o CDS</td>\n",
       "      <td>5218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Truncation</td>\n",
       "      <td>NIC</td>\n",
       "      <td>3044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Truncation</td>\n",
       "      <td>NNC</td>\n",
       "      <td>2433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aa_seq_novelty structural_category    n_t\n",
       "0           Known                 FSM  34094\n",
       "1           Known         FSM w/o CDS    842\n",
       "2           Known                 NIC   2647\n",
       "3           Known                 NNC   2816\n",
       "4           Novel                 FSM   3039\n",
       "5           Novel         FSM w/o CDS  11109\n",
       "6           Novel                 NIC   7325\n",
       "7           Novel                 NNC   9869\n",
       "8      Truncation                 FSM   3870\n",
       "9      Truncation         FSM w/o CDS   5218\n",
       "10     Truncation                 NIC   3044\n",
       "11     Truncation                 NNC   2433"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp2 = temp[['isoform', 'aa_seq_novelty', 'structural_category']].groupby(['aa_seq_novelty', 'structural_category']).nunique().rename({'isoform':'n_t'}, axis=1)\n",
    "temp2 = temp2.reset_index()\n",
    "temp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "98a0fe01-8273-4212-89c1-18fd3c9666d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54.72% (1663/3039) of FSM (with CDS) transcripts that we predict novel AA seq do not have a complete ORF annotated in GENCODE\n"
     ]
    }
   ],
   "source": [
    "temp = temp.loc[(temp.structural_category=='FSM')&\n",
    "                (temp.aa_seq_novelty=='Novel')]\n",
    "n = len(temp)\n",
    "\n",
    "n_num = len(temp.loc[temp.has_complete_orf==False])\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} of FSM (with CDS) transcripts that we predict novel AA seq do not have a complete ORF annotated in GENCODE')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8308c891-b02d-402a-b989-e6829b3830af",
   "metadata": {},
   "source": [
    "## Similarly, maybe some of these can also be explained by just being elongations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "f73d36c8-fea3-45bd-95d5-56f0f374a0f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23.96% (728/3039) of FSM (with CDS) transcripts that we predict novel AA seq are elongations of annot. seqs from the same gene\n"
     ]
    }
   ],
   "source": [
    "temp = temp.loc[(temp.structural_category=='FSM')&\n",
    "                (temp.aa_seq_novelty=='Novel')]\n",
    "n = len(temp)\n",
    "\n",
    "n_num = len(temp.loc[temp.elong_annot_aa==True])\n",
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} of FSM (with CDS) transcripts that we predict novel AA seq are elongations of annot. seqs from the same gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e5f6a7-af5d-4d25-9eae-cad9f1e81eff",
   "metadata": {},
   "source": [
    "## Get a summary table for sankey plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3b5b1dba-89e7-4248-aa3b-0010acb7b30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "p_df = pd.read_csv('250223_protein_characterization_mt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "361cf1c1-b6c8-4db2-bb7d-817c6d10811f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Source</th>\n",
       "      <th>CDS_Source</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>...</th>\n",
       "      <th>isoform</th>\n",
       "      <th>gid</th>\n",
       "      <th>length</th>\n",
       "      <th>annot_aa</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>has_start_codon</th>\n",
       "      <th>has_stop_codon</th>\n",
       "      <th>has_complete_orf</th>\n",
       "      <th>trunc_annot_aa</th>\n",
       "      <th>elong_annot_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17626006</td>\n",
       "      <td>17697868</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>17627332</td>\n",
       "      <td>17691151</td>\n",
       "      <td>ENST00000167825.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000167825.5</td>\n",
       "      <td>ENSG00000074964.17</td>\n",
       "      <td>3018</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000167825.5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5784560</td>\n",
       "      <td>5825285</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>5787146</td>\n",
       "      <td>5825187</td>\n",
       "      <td>ENST00000214893.10</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000214893.10</td>\n",
       "      <td>ENSG00000099219.15</td>\n",
       "      <td>4724</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000214893.10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>chrX</td>\n",
       "      <td>53584210</td>\n",
       "      <td>53686728</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>53584280</td>\n",
       "      <td>53654107</td>\n",
       "      <td>ENST00000218328.12</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000218328.12</td>\n",
       "      <td>ENSG00000086758.17</td>\n",
       "      <td>5488</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000218328.12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>chr16</td>\n",
       "      <td>53235136</td>\n",
       "      <td>53327451</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>53238367</td>\n",
       "      <td>53324892</td>\n",
       "      <td>ENST00000219084.10</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000219084.10</td>\n",
       "      <td>ENSG00000177200.18</td>\n",
       "      <td>8739</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000219084.10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>chr4</td>\n",
       "      <td>153555895</td>\n",
       "      <td>153636711</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>153556977</td>\n",
       "      <td>153636573</td>\n",
       "      <td>ENST00000240487.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000240487.5</td>\n",
       "      <td>ENSG00000121210.16</td>\n",
       "      <td>4299</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000240487.5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100658</th>\n",
       "      <td>100658</td>\n",
       "      <td>chr2</td>\n",
       "      <td>181457369</td>\n",
       "      <td>181487876</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>181457654</td>\n",
       "      <td>181485997</td>\n",
       "      <td>transcript_99938</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript_99938</td>\n",
       "      <td>ENSG00000115232.14</td>\n",
       "      <td>3322</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000465522.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100660</th>\n",
       "      <td>100660</td>\n",
       "      <td>chr2</td>\n",
       "      <td>181457427</td>\n",
       "      <td>181509805</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>CPAT</td>\n",
       "      <td>181485977</td>\n",
       "      <td>181494752</td>\n",
       "      <td>transcript_99940</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript_99940</td>\n",
       "      <td>ENSG00000115232.14</td>\n",
       "      <td>2070</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000233573.6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100663</th>\n",
       "      <td>100663</td>\n",
       "      <td>chr2</td>\n",
       "      <td>181493033</td>\n",
       "      <td>181495832</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>CPAT</td>\n",
       "      <td>181493324</td>\n",
       "      <td>181494752</td>\n",
       "      <td>transcript_99983</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript_99983</td>\n",
       "      <td>ENSG00000115232.14</td>\n",
       "      <td>573</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000473002.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100664</th>\n",
       "      <td>100664</td>\n",
       "      <td>chr2</td>\n",
       "      <td>181498280</td>\n",
       "      <td>181511771</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>CPAT</td>\n",
       "      <td>181509661</td>\n",
       "      <td>181509793</td>\n",
       "      <td>transcript_99989</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript_99989</td>\n",
       "      <td>ENSG00000115232.14</td>\n",
       "      <td>720</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000490435.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100665</th>\n",
       "      <td>100665</td>\n",
       "      <td>chr2</td>\n",
       "      <td>181523334</td>\n",
       "      <td>181524365</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>CPAT</td>\n",
       "      <td>181523344</td>\n",
       "      <td>181523449</td>\n",
       "      <td>transcript_99992</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript_99992</td>\n",
       "      <td>ENSG00000115232.14</td>\n",
       "      <td>393</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000468948.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27149 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0 Chromosome      Start       Stop Strand    Source  \\\n",
       "6                6       chr1   17626006   17697868      +    HAVANA   \n",
       "11              11       chr9    5784560    5825285      -    HAVANA   \n",
       "14              14       chrX   53584210   53686728      -    HAVANA   \n",
       "16              16      chr16   53235136   53327451      +    HAVANA   \n",
       "35              35       chr4  153555895  153636711      +    HAVANA   \n",
       "...            ...        ...        ...        ...    ...       ...   \n",
       "100658      100658       chr2  181457369  181487876      +  Cerberus   \n",
       "100660      100660       chr2  181457427  181509805      +  Cerberus   \n",
       "100663      100663       chr2  181493033  181495832      +  Cerberus   \n",
       "100664      100664       chr2  181498280  181511771      +  Cerberus   \n",
       "100665      100665       chr2  181523334  181524365      +  Cerberus   \n",
       "\n",
       "       CDS_Source  CDS_Start   CDS_Stop                 tid  ...  \\\n",
       "6        ORFanage   17627332   17691151   ENST00000167825.5  ...   \n",
       "11       ORFanage    5787146    5825187  ENST00000214893.10  ...   \n",
       "14       ORFanage   53584280   53654107  ENST00000218328.12  ...   \n",
       "16       ORFanage   53238367   53324892  ENST00000219084.10  ...   \n",
       "35       ORFanage  153556977  153636573   ENST00000240487.5  ...   \n",
       "...           ...        ...        ...                 ...  ...   \n",
       "100658   ORFanage  181457654  181485997    transcript_99938  ...   \n",
       "100660       CPAT  181485977  181494752    transcript_99940  ...   \n",
       "100663       CPAT  181493324  181494752    transcript_99983  ...   \n",
       "100664       CPAT  181509661  181509793    transcript_99989  ...   \n",
       "100665       CPAT  181523344  181523449    transcript_99992  ...   \n",
       "\n",
       "                   isoform                 gid  length  annot_aa  \\\n",
       "6        ENST00000167825.5  ENSG00000074964.17    3018     False   \n",
       "11      ENST00000214893.10  ENSG00000099219.15    4724     False   \n",
       "14      ENST00000218328.12  ENSG00000086758.17    5488     False   \n",
       "16      ENST00000219084.10  ENSG00000177200.18    8739     False   \n",
       "35       ENST00000240487.5  ENSG00000121210.16    4299     False   \n",
       "...                    ...                 ...     ...       ...   \n",
       "100658    transcript_99938  ENSG00000115232.14    3322     False   \n",
       "100660    transcript_99940  ENSG00000115232.14    2070     False   \n",
       "100663    transcript_99983  ENSG00000115232.14     573     False   \n",
       "100664    transcript_99989  ENSG00000115232.14     720     False   \n",
       "100665    transcript_99992  ENSG00000115232.14     393     False   \n",
       "\n",
       "             transcript_id  has_start_codon  has_stop_codon has_complete_orf  \\\n",
       "6        ENST00000167825.5            False            True            False   \n",
       "11      ENST00000214893.10            False           False            False   \n",
       "14      ENST00000218328.12            False           False            False   \n",
       "16      ENST00000219084.10            False           False            False   \n",
       "35       ENST00000240487.5            False            True            False   \n",
       "...                    ...              ...             ...              ...   \n",
       "100658   ENST00000465522.5            False           False            False   \n",
       "100660   ENST00000233573.6             True           False            False   \n",
       "100663   ENST00000473002.1            False           False            False   \n",
       "100664   ENST00000490435.5            False           False            False   \n",
       "100665   ENST00000468948.1            False           False            False   \n",
       "\n",
       "       trunc_annot_aa  elong_annot_aa  \n",
       "6                True           False  \n",
       "11               True           False  \n",
       "14              False            True  \n",
       "16               True           False  \n",
       "35               True           False  \n",
       "...               ...             ...  \n",
       "100658          False           False  \n",
       "100660          False           False  \n",
       "100663          False           False  \n",
       "100664          False           False  \n",
       "100665          False           False  \n",
       "\n",
       "[27149 rows x 35 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# just for novel transcripts\n",
    "# AA categories are NMD, known, known truncation, \n",
    "# known elongation, novel\n",
    "p_df.loc[p_df.has_complete_orf==False]\n",
    "\n",
    "print(len(p_df.index))\n",
    "df = p_df.loc[(p_df.structural_category!='FSM')&\\\n",
    "              (p_df.structural_category!=\"FSM w/o CDS\")]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bf60367-7393-448c-8c21-d5dcbb48fb7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100666\n",
      "32498\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c6bf143e-67b7-45f5-b5f5-6c8cca03a24d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    }
   ],
   "source": [
    "print(len(p_df.index))\n",
    "df = p_df.loc[(p_df.structural_category!='FSM')&\\\n",
    "              (p_df.structural_category!=\"FSM w/o CDS\")]\n",
    "print(len(df.index))\n",
    "\n",
    "# completely novel\n",
    "df['aa_novelty_2'] = ''\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==False)&\\\n",
    "       (df.elong_annot_aa==False), 'aa_novelty_2'] = 'Novel'\n",
    "\n",
    "# known elongation\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==False)&\\\n",
    "       (df.elong_annot_aa==True), 'aa_novelty_2'] = 'Known elongation'\n",
    "\n",
    "# known truncation\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==True), 'aa_novelty_2'] = 'Known truncation'\n",
    "\n",
    "# known\n",
    "df.loc[(df.annot_aa==True), 'aa_novelty_2'] = 'Known'\n",
    "\n",
    "# NMD\n",
    "df.loc[df.protein_is_nmd==True, 'aa_novelty_2'] = 'NMD'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "09039ba0-ec33-48d6-841b-000f83ff7537",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structural_category</th>\n",
       "      <th>aa_novelty_2</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">NIC</th>\n",
       "      <th>Known</th>\n",
       "      <td>2647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known elongation</th>\n",
       "      <td>2111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known truncation</th>\n",
       "      <td>3044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMD</th>\n",
       "      <td>2490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>5214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">NNC</th>\n",
       "      <th>Known</th>\n",
       "      <td>2816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known elongation</th>\n",
       "      <td>1682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known truncation</th>\n",
       "      <td>2433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMD</th>\n",
       "      <td>1874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>8187</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      isoform\n",
       "structural_category aa_novelty_2             \n",
       "NIC                 Known                2647\n",
       "                    Known elongation     2111\n",
       "                    Known truncation     3044\n",
       "                    NMD                  2490\n",
       "                    Novel                5214\n",
       "NNC                 Known                2816\n",
       "                    Known elongation     1682\n",
       "                    Known truncation     2433\n",
       "                    NMD                  1874\n",
       "                    Novel                8187"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df[['structural_category', 'aa_novelty_2', 'isoform']].groupby(['structural_category', 'aa_novelty_2']).nunique()\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef8cb482-fad9-4172-a596-f2ff6550c984",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.to_csv('250224_protein_table_for_alluvial.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "54303836-4fb7-4e54-afb4-5692d7d37ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[(df.annot_aa==True)&(df.protein_is_nmd==True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "8ba51485-1e94-475a-ab45-34a9b13d21e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Source</th>\n",
       "      <th>CDS_Source</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>pid</th>\n",
       "      <th>...</th>\n",
       "      <th>gid</th>\n",
       "      <th>length</th>\n",
       "      <th>annot_aa</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>has_start_codon</th>\n",
       "      <th>has_stop_codon</th>\n",
       "      <th>has_complete_orf</th>\n",
       "      <th>trunc_annot_aa</th>\n",
       "      <th>elong_annot_aa</th>\n",
       "      <th>aa_seq_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr14</td>\n",
       "      <td>54842016</td>\n",
       "      <td>54902811</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>54844019</td>\n",
       "      <td>54902663</td>\n",
       "      <td>ENST00000254299.8</td>\n",
       "      <td>ENSP00000419045.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000131979.20</td>\n",
       "      <td>2901</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000254299.8</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chr5</td>\n",
       "      <td>66158126</td>\n",
       "      <td>66180811</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>66159271</td>\n",
       "      <td>66178865</td>\n",
       "      <td>ENST00000284041.7</td>\n",
       "      <td>ENSP00000370305.3</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000153914.16</td>\n",
       "      <td>3751</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000284041.7</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>chr16</td>\n",
       "      <td>87707811</td>\n",
       "      <td>87750942</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>87707916</td>\n",
       "      <td>87730607</td>\n",
       "      <td>ENST00000316853.10</td>\n",
       "      <td>ENSP00000482433.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000104731.14</td>\n",
       "      <td>1287</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000316853.10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>chr5</td>\n",
       "      <td>96728308</td>\n",
       "      <td>96772819</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>96730794</td>\n",
       "      <td>96771676</td>\n",
       "      <td>ENST00000348386.7</td>\n",
       "      <td>ENSP00000422677.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000310517.1</td>\n",
       "      <td>2272</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000348386.7</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>chr10</td>\n",
       "      <td>16593312</td>\n",
       "      <td>16817463</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>16593396</td>\n",
       "      <td>16782034</td>\n",
       "      <td>ENST00000377911.1</td>\n",
       "      <td>ENSP00000339521.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000148484.18</td>\n",
       "      <td>955</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000377911.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100613</th>\n",
       "      <td>chr1</td>\n",
       "      <td>226063475</td>\n",
       "      <td>226072002</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>226064351</td>\n",
       "      <td>226071476</td>\n",
       "      <td>transcript_9954</td>\n",
       "      <td>ENSP00000466020.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000163041.13</td>\n",
       "      <td>1263</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100627</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178112409</td>\n",
       "      <td>178129656</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>178112546</td>\n",
       "      <td>178126173</td>\n",
       "      <td>transcript_99666</td>\n",
       "      <td>ENSP00000286070.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000155636.16</td>\n",
       "      <td>1868</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100633</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178450464</td>\n",
       "      <td>178461380</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>178456079</td>\n",
       "      <td>178461271</td>\n",
       "      <td>transcript_99766</td>\n",
       "      <td>ENSP00000494225.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000204311.15</td>\n",
       "      <td>1098</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100634</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178450492</td>\n",
       "      <td>178461287</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>178456079</td>\n",
       "      <td>178461271</td>\n",
       "      <td>transcript_99767</td>\n",
       "      <td>ENSP00000494225.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000204311.15</td>\n",
       "      <td>1081</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100639</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178453250</td>\n",
       "      <td>178461287</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>178456079</td>\n",
       "      <td>178461271</td>\n",
       "      <td>transcript_99772</td>\n",
       "      <td>ENSP00000494225.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000204311.15</td>\n",
       "      <td>1942</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000437056.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8384 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start       Stop Strand    Source CDS_Source  \\\n",
       "58          chr14   54842016   54902811      -    HAVANA   ORFanage   \n",
       "153          chr5   66158126   66180811      +    HAVANA   ORFanage   \n",
       "255         chr16   87707811   87750942      -    HAVANA   ORFanage   \n",
       "408          chr5   96728308   96772819      +    HAVANA   ORFanage   \n",
       "754         chr10   16593312   16817463      -    HAVANA   ORFanage   \n",
       "...           ...        ...        ...    ...       ...        ...   \n",
       "100613       chr1  226063475  226072002      +  Cerberus   ORFanage   \n",
       "100627       chr2  178112409  178129656      +  Cerberus   ORFanage   \n",
       "100633       chr2  178450464  178461380      +  Cerberus   ORFanage   \n",
       "100634       chr2  178450492  178461287      +  Cerberus   ORFanage   \n",
       "100639       chr2  178453250  178461287      +  Cerberus   ORFanage   \n",
       "\n",
       "        CDS_Start   CDS_Stop                 tid                pid  ...  \\\n",
       "58       54844019   54902663   ENST00000254299.8  ENSP00000419045.2  ...   \n",
       "153      66159271   66178865   ENST00000284041.7  ENSP00000370305.3  ...   \n",
       "255      87707916   87730607  ENST00000316853.10  ENSP00000482433.1  ...   \n",
       "408      96730794   96771676   ENST00000348386.7  ENSP00000422677.2  ...   \n",
       "754      16593396   16782034   ENST00000377911.1  ENSP00000339521.5  ...   \n",
       "...           ...        ...                 ...                ...  ...   \n",
       "100613  226064351  226071476     transcript_9954  ENSP00000466020.1  ...   \n",
       "100627  178112546  178126173    transcript_99666  ENSP00000286070.5  ...   \n",
       "100633  178456079  178461271    transcript_99766  ENSP00000494225.1  ...   \n",
       "100634  178456079  178461271    transcript_99767  ENSP00000494225.1  ...   \n",
       "100639  178456079  178461271    transcript_99772  ENSP00000494225.1  ...   \n",
       "\n",
       "                       gid  length  annot_aa       transcript_id  \\\n",
       "58      ENSG00000131979.20    2901      True   ENST00000254299.8   \n",
       "153     ENSG00000153914.16    3751      True   ENST00000284041.7   \n",
       "255     ENSG00000104731.14    1287      True  ENST00000316853.10   \n",
       "408      ENSG00000310517.1    2272     False   ENST00000348386.7   \n",
       "754     ENSG00000148484.18     955      True   ENST00000377911.1   \n",
       "...                    ...     ...       ...                 ...   \n",
       "100613  ENSG00000163041.13    1263      True                 NaN   \n",
       "100627  ENSG00000155636.16    1868      True                 NaN   \n",
       "100633  ENSG00000204311.15    1098      True                 NaN   \n",
       "100634  ENSG00000204311.15    1081      True                 NaN   \n",
       "100639  ENSG00000204311.15    1942      True   ENST00000437056.5   \n",
       "\n",
       "        has_start_codon  has_stop_codon has_complete_orf trunc_annot_aa  \\\n",
       "58                False           False            False           True   \n",
       "153               False           False            False           True   \n",
       "255               False           False            False           True   \n",
       "408               False           False            False           True   \n",
       "754               False           False            False           True   \n",
       "...                 ...             ...              ...            ...   \n",
       "100613              NaN             NaN              NaN           True   \n",
       "100627              NaN             NaN              NaN           True   \n",
       "100633              NaN             NaN              NaN           True   \n",
       "100634              NaN             NaN              NaN           True   \n",
       "100639            False           False            False           True   \n",
       "\n",
       "        elong_annot_aa  aa_seq_novelty  \n",
       "58                True           Known  \n",
       "153               True           Known  \n",
       "255               True           Known  \n",
       "408               True      Truncation  \n",
       "754               True           Known  \n",
       "...                ...             ...  \n",
       "100613            True           Known  \n",
       "100627            True           Known  \n",
       "100633            True           Known  \n",
       "100634            True           Known  \n",
       "100639            True           Known  \n",
       "\n",
       "[8384 rows x 35 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df.loc[(df.trunc_annot_aa==True)&\n",
    "#        (df.elong_annot_aa==True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "20f7e940-df6b-4b7d-b19a-c739bb78170d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Source</th>\n",
       "      <th>CDS_Source</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>pid</th>\n",
       "      <th>...</th>\n",
       "      <th>gid</th>\n",
       "      <th>length</th>\n",
       "      <th>annot_aa</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>has_start_codon</th>\n",
       "      <th>has_stop_codon</th>\n",
       "      <th>has_complete_orf</th>\n",
       "      <th>trunc_annot_aa</th>\n",
       "      <th>elong_annot_aa</th>\n",
       "      <th>aa_seq_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>5784560</td>\n",
       "      <td>5825285</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>5787146</td>\n",
       "      <td>5825187</td>\n",
       "      <td>ENST00000214893.10</td>\n",
       "      <td>ENSP00000340427.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000099219.15</td>\n",
       "      <td>4724</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000214893.10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrX</td>\n",
       "      <td>53584210</td>\n",
       "      <td>53686728</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>53584280</td>\n",
       "      <td>53654107</td>\n",
       "      <td>ENST00000218328.12</td>\n",
       "      <td>ENSP00000262854.6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000086758.17</td>\n",
       "      <td>5488</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000218328.12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr16</td>\n",
       "      <td>53235136</td>\n",
       "      <td>53327451</td>\n",
       "      <td>+</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>53238367</td>\n",
       "      <td>53324892</td>\n",
       "      <td>ENST00000219084.10</td>\n",
       "      <td>ENSP00000455307.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000177200.18</td>\n",
       "      <td>8739</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000219084.10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr9</td>\n",
       "      <td>131258075</td>\n",
       "      <td>131264715</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>131260824</td>\n",
       "      <td>131261202</td>\n",
       "      <td>ENST00000247295.4</td>\n",
       "      <td>ENSP00000361343.3</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000126882.14</td>\n",
       "      <td>4560</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000247295.4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>chr4</td>\n",
       "      <td>112605559</td>\n",
       "      <td>112636995</td>\n",
       "      <td>-</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>112607874</td>\n",
       "      <td>112633176</td>\n",
       "      <td>ENST00000264370.9</td>\n",
       "      <td>ENSP00000426938.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000138658.16</td>\n",
       "      <td>3583</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000264370.9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100587</th>\n",
       "      <td>chr2</td>\n",
       "      <td>176269403</td>\n",
       "      <td>176338025</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>176269629</td>\n",
       "      <td>176337498</td>\n",
       "      <td>transcript_99427</td>\n",
       "      <td>ENSP00000398757.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000128654.14</td>\n",
       "      <td>1176</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100588</th>\n",
       "      <td>chr2</td>\n",
       "      <td>176269403</td>\n",
       "      <td>176338034</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>176328337</td>\n",
       "      <td>176337661</td>\n",
       "      <td>transcript_99428</td>\n",
       "      <td>ENSP00000414176.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000128654.14</td>\n",
       "      <td>1315</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100615</th>\n",
       "      <td>chr2</td>\n",
       "      <td>177216721</td>\n",
       "      <td>177223336</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>177219110</td>\n",
       "      <td>177219296</td>\n",
       "      <td>transcript_99590</td>\n",
       "      <td>ENSP00000400688.2</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000170144.23</td>\n",
       "      <td>1828</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100635</th>\n",
       "      <td>chr2</td>\n",
       "      <td>178451310</td>\n",
       "      <td>178461394</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>178451757</td>\n",
       "      <td>178461271</td>\n",
       "      <td>transcript_99768</td>\n",
       "      <td>ENSP00000364271.4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000204311.15</td>\n",
       "      <td>1617</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100650</th>\n",
       "      <td>chr2</td>\n",
       "      <td>180980679</td>\n",
       "      <td>181063425</td>\n",
       "      <td>+</td>\n",
       "      <td>Cerberus</td>\n",
       "      <td>ORFanage</td>\n",
       "      <td>180982042</td>\n",
       "      <td>181062811</td>\n",
       "      <td>transcript_99842</td>\n",
       "      <td>ENSP00000473591.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000170035.16</td>\n",
       "      <td>1332</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16203 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start       Stop Strand    Source CDS_Source  \\\n",
       "11           chr9    5784560    5825285      -    HAVANA   ORFanage   \n",
       "14           chrX   53584210   53686728      -    HAVANA   ORFanage   \n",
       "16          chr16   53235136   53327451      +    HAVANA   ORFanage   \n",
       "44           chr9  131258075  131264715      -    HAVANA   ORFanage   \n",
       "100          chr4  112605559  112636995      -    HAVANA   ORFanage   \n",
       "...           ...        ...        ...    ...       ...        ...   \n",
       "100587       chr2  176269403  176338025      +  Cerberus   ORFanage   \n",
       "100588       chr2  176269403  176338034      +  Cerberus   ORFanage   \n",
       "100615       chr2  177216721  177223336      +  Cerberus   ORFanage   \n",
       "100635       chr2  178451310  178461394      +  Cerberus   ORFanage   \n",
       "100650       chr2  180980679  181063425      +  Cerberus   ORFanage   \n",
       "\n",
       "        CDS_Start   CDS_Stop                 tid                pid  ...  \\\n",
       "11        5787146    5825187  ENST00000214893.10  ENSP00000340427.5  ...   \n",
       "14       53584280   53654107  ENST00000218328.12  ENSP00000262854.6  ...   \n",
       "16       53238367   53324892  ENST00000219084.10  ENSP00000455307.1  ...   \n",
       "44      131260824  131261202   ENST00000247295.4  ENSP00000361343.3  ...   \n",
       "100     112607874  112633176   ENST00000264370.9  ENSP00000426938.1  ...   \n",
       "...           ...        ...                 ...                ...  ...   \n",
       "100587  176269629  176337498    transcript_99427  ENSP00000398757.1  ...   \n",
       "100588  176328337  176337661    transcript_99428  ENSP00000414176.1  ...   \n",
       "100615  177219110  177219296    transcript_99590  ENSP00000400688.2  ...   \n",
       "100635  178451757  178461271    transcript_99768  ENSP00000364271.4  ...   \n",
       "100650  180982042  181062811    transcript_99842  ENSP00000473591.1  ...   \n",
       "\n",
       "                       gid  length  annot_aa       transcript_id  \\\n",
       "11      ENSG00000099219.15    4724     False  ENST00000214893.10   \n",
       "14      ENSG00000086758.17    5488     False  ENST00000218328.12   \n",
       "16      ENSG00000177200.18    8739     False  ENST00000219084.10   \n",
       "44      ENSG00000126882.14    4560     False   ENST00000247295.4   \n",
       "100     ENSG00000138658.16    3583     False   ENST00000264370.9   \n",
       "...                    ...     ...       ...                 ...   \n",
       "100587  ENSG00000128654.14    1176     False                 NaN   \n",
       "100588  ENSG00000128654.14    1315     False                 NaN   \n",
       "100615  ENSG00000170144.23    1828     False                 NaN   \n",
       "100635  ENSG00000204311.15    1617     False                 NaN   \n",
       "100650  ENSG00000170035.16    1332     False                 NaN   \n",
       "\n",
       "        has_start_codon  has_stop_codon has_complete_orf trunc_annot_aa  \\\n",
       "11                False           False            False           True   \n",
       "14                False           False            False          False   \n",
       "16                False           False            False           True   \n",
       "44                False           False            False           True   \n",
       "100               False           False            False          False   \n",
       "...                 ...             ...              ...            ...   \n",
       "100587              NaN             NaN              NaN          False   \n",
       "100588              NaN             NaN              NaN           True   \n",
       "100615              NaN             NaN              NaN           True   \n",
       "100635              NaN             NaN              NaN          False   \n",
       "100650              NaN             NaN              NaN          False   \n",
       "\n",
       "        elong_annot_aa  aa_seq_novelty  \n",
       "11               False      Truncation  \n",
       "14                True           Novel  \n",
       "16               False      Truncation  \n",
       "44               False      Truncation  \n",
       "100               True           Novel  \n",
       "...                ...             ...  \n",
       "100587            True           Novel  \n",
       "100588           False      Truncation  \n",
       "100615           False      Truncation  \n",
       "100635            True           Novel  \n",
       "100650            True           Novel  \n",
       "\n",
       "[16203 rows x 35 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.trunc_annot_aa!=df.elong_annot_aa]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9acc22f-558f-4f77-a74a-d1dcaaa2ecf9",
   "metadata": {},
   "source": [
    "## number of truly novel AA seqs from NIC / NNC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "80a0bf6f-3467-430f-a214-e8689dc38eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12148"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[df.aa_novelty_2 == 'Novel'][['seq', 'gid']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f319fc8a-1dcb-43e9-9f91-02e9b43b4878",
   "metadata": {},
   "source": [
    "## BLAST vals. of truly novel AA seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e66f95d0-fadf-4cdc-9498-473645ec5a8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>high_blastp</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>3715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>8433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seq\n",
       "high_blastp      \n",
       "False        3715\n",
       "True         8433"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.aa_novelty_2 == 'Novel']\n",
    "\n",
    "# get only uniq aa seqs / blast ids\n",
    "temp = temp[['gid', 'seq', 'blastp_identity']].drop_duplicates()\n",
    "\n",
    "high_blastp_thresh = 90\n",
    "temp['high_blastp'] = temp.blastp_identity>=high_blastp_thresh\n",
    "temp[['seq', 'high_blastp']].groupby(['high_blastp']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64011509-050b-41e9-bdc5-615da6f79679",
   "metadata": {},
   "source": [
    "## Perc. NIC / NNC that encode for truly novel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "128e2ca5-09b1-4efa-bd6b-cad5ab914ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = len(df.loc[df.structural_category.isin(['NIC', 'NNC'])].isoform.unique())\n",
    "n_num  = len(df.loc[(df.structural_category.isin(['NIC', 'NNC']))&\n",
    "                (df.aa_novelty_2=='Novel')].isoform.unique())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "757bc7b3-69e2-4700-a372-90277080e067",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41.24% (13401/32498) of NNC/NIC transcripts from PC genes are predicted to encode for a completely novel AA seq (not nmd, not elongation or truncation or known)\n"
     ]
    }
   ],
   "source": [
    "s = fmt_perc_print(n_num, n)\n",
    "print(f'{s} of NNC/NIC transcripts from PC genes are predicted to encode for a completely novel AA seq (not nmd, not elongation or truncation or known)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "282709b2-abdb-467d-848a-12ac4716813a",
   "metadata": {},
   "source": [
    "## Add some of this extra protein info to the master table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "62aee085-0920-4eea-ac21-6b0b226a4467",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_df = pd.read_csv('../supp_tables/04_poder_mt.tsv', sep='\\t')\n",
    "p_df = pd.read_csv('250223_protein_characterization_mt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d15e4559-e7a5-47ef-9949-46441060081b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100666\n",
      "32498\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  import sys\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    }
   ],
   "source": [
    "print(len(p_df.index))\n",
    "df = p_df.loc[(p_df.structural_category!='FSM')&\\\n",
    "              (p_df.structural_category!=\"FSM w/o CDS\")]\n",
    "print(len(df.index))\n",
    "\n",
    "# completely novel\n",
    "df['aa_novelty_2'] = ''\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==False)&\\\n",
    "       (df.elong_annot_aa==False), 'aa_novelty_2'] = 'Novel'\n",
    "\n",
    "# known elongation\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==False)&\\\n",
    "       (df.elong_annot_aa==True), 'aa_novelty_2'] = 'Known elongation'\n",
    "\n",
    "# known truncation\n",
    "df.loc[(df.annot_aa==False)&\\\n",
    "       (df.protein_is_nmd==False)&\\\n",
    "       (df.trunc_annot_aa==True), 'aa_novelty_2'] = 'Known truncation'\n",
    "\n",
    "# known\n",
    "df.loc[(df.annot_aa==True), 'aa_novelty_2'] = 'Known'\n",
    "\n",
    "# NMD\n",
    "df.loc[df.protein_is_nmd==True, 'aa_novelty_2'] = 'NMD'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3c19270c-4216-4848-ac19-fcbc23b2a2c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py:5047: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    }
   ],
   "source": [
    "df.rename({'aa_novelty_2':'aa_novelty'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cc1d7d47-5452-42fc-8752-105680a100dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>trunc_annot_aa</th>\n",
       "      <th>elong_annot_aa</th>\n",
       "      <th>annot_aa</th>\n",
       "      <th>aa_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7251</th>\n",
       "      <td>transcript_100122</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7255</th>\n",
       "      <td>transcript_100131</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Known truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7257</th>\n",
       "      <td>transcript_100134</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7258</th>\n",
       "      <td>transcript_100135</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Known truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7276</th>\n",
       "      <td>transcript_100180</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Known truncation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                isoform  trunc_annot_aa  elong_annot_aa  annot_aa  \\\n",
       "7251  transcript_100122           False           False     False   \n",
       "7255  transcript_100131            True           False     False   \n",
       "7257  transcript_100134           False           False     False   \n",
       "7258  transcript_100135            True           False     False   \n",
       "7276  transcript_100180            True           False     False   \n",
       "\n",
       "            aa_novelty  \n",
       "7251             Novel  \n",
       "7255  Known truncation  \n",
       "7257             Novel  \n",
       "7258  Known truncation  \n",
       "7276  Known truncation  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[['isoform',\n",
    "         'trunc_annot_aa', \n",
    "         'elong_annot_aa',\n",
    "         'annot_aa', \n",
    "         'aa_novelty']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8cf6f294-c23b-4ad1-abde-ef6c857c84ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_df = mt_df.merge(df, how='left', \n",
    "                    on='isoform')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6a3f38ca-bb43-457a-a652-38607878bc76",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_df.to_csv('../supp_tables/04_poder_mt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "920bd27c-1711-4626-8564-d7eb507579f0",
   "metadata": {},
   "source": [
    "## Get the number of novel transcripts going into this analysis in the first place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "221bc38f-3c15-4a9f-97a8-57c843192475",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df = pd.read_csv('250223_protein_characterization_mt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "78643fa2-ca39-4cd3-b5a5-18f0140e962f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100666\n",
      "32498\n"
     ]
    }
   ],
   "source": [
    "# just for novel transcripts\n",
    "# AA categories are NMD, known, known truncation, \n",
    "# known elongation, novel\n",
    "p_df.loc[p_df.has_complete_orf==False]\n",
    "\n",
    "print(len(p_df.index))\n",
    "df = p_df.loc[(p_df.structural_category!='FSM')&\\\n",
    "              (p_df.structural_category!=\"FSM w/o CDS\")]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "4b5e8536-1e12-4180-8cf7-7a6746b3979e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Source</th>\n",
       "      <th>CDS_Source</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>...</th>\n",
       "      <th>isoform</th>\n",
       "      <th>gid</th>\n",
       "      <th>length</th>\n",
       "      <th>annot_aa</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>has_start_codon</th>\n",
       "      <th>has_stop_codon</th>\n",
       "      <th>has_complete_orf</th>\n",
       "      <th>trunc_annot_aa</th>\n",
       "      <th>elong_annot_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Unnamed: 0, Chromosome, Start, Stop, Strand, Source, CDS_Source, CDS_Start, CDS_Stop, tid, pid, blastp_identity, blastp_bitscore, transcript_length_nt, orf_length_nt, protein_length_cd, protein_is_nmd, protein_splice_category, protein_splice_subcategory, protein_has_stop_codon, protein_has_start_codon, seq, associated_gene_biotype, structural_category, associated_transcript, isoform, gid, length, annot_aa, transcript_id, has_start_codon, has_stop_codon, has_complete_orf, trunc_annot_aa, elong_annot_aa]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 35 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.tid.str.startswith('ENS')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3e5b25-b9a9-4e00-a009-53dd4164364e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
