{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "249c9609-bb6f-4150-9c12-0e1103c6371e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1fec790c-25de-4a89-9fb1-be58ba42266b",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "14740e95-46f6-457a-b3e3-b71fa26369ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3084b468-be6e-4c0e-a972-7200a620607b",
   "metadata": {},
   "source": [
    "We want to find an example of a gene that is \n",
    "* PODER exclusive sig ASTU\n",
    "* GWAS hit thing\n",
    "* needs to have some novel transcripst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4c41bfca-980d-43a9-beae-0d67947ec7c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# browser models with the variants; "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fa7ce8da-7982-4128-aba0-0fbe27e209f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>population</th>\n",
       "      <th>map_reads_assemblymap</th>\n",
       "      <th>gene_variant</th>\n",
       "      <th>statistic</th>\n",
       "      <th>p.value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>variant</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>totalCount</th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>significant</th>\n",
       "      <th>tested_genes</th>\n",
       "      <th>significant_genes</th>\n",
       "      <th>gid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000116251.11_chr3_169483680_G_T</td>\n",
       "      <td>21.160000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000116251.11</td>\n",
       "      <td>chr3_169483680_G_T</td>\n",
       "      <td>transcript_207329</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>transcript_207329_chr3_169483680_G_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000116251.11_chr3_169483724_A_G</td>\n",
       "      <td>22.153846</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000116251.11</td>\n",
       "      <td>chr3_169483724_A_G</td>\n",
       "      <td>transcript_207329</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>transcript_207329_chr3_169483724_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000130770.18_chr1_28236165_A_G</td>\n",
       "      <td>25.717736</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000130770.18</td>\n",
       "      <td>chr1_28236165_A_G</td>\n",
       "      <td>transcript_12375</td>\n",
       "      <td>5250</td>\n",
       "      <td>5472</td>\n",
       "      <td>10722</td>\n",
       "      <td>transcript_12375_chr1_28236165_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000130770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000130770.18_chr1_28236165_A_G</td>\n",
       "      <td>25.717736</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000130770.18</td>\n",
       "      <td>chr1_28236165_A_G</td>\n",
       "      <td>transcript_12372</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>transcript_12372_chr1_28236165_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000130770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000130770.18_chr1_28236165_A_G</td>\n",
       "      <td>25.717736</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000130770.18</td>\n",
       "      <td>chr1_28236165_A_G</td>\n",
       "      <td>transcript_12377</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>46</td>\n",
       "      <td>transcript_12377_chr1_28236165_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000130770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cell_line_id sample population  map_reads_assemblymap  \\\n",
       "10      GM19117   YRI5        YRI                9522208   \n",
       "11      GM19117   YRI5        YRI                9522208   \n",
       "28      GM19117   YRI5        YRI                9522208   \n",
       "29      GM19117   YRI5        YRI                9522208   \n",
       "30      GM19117   YRI5        YRI                9522208   \n",
       "\n",
       "                             gene_variant  statistic  p.value       FDR  \\\n",
       "10  ENSG00000116251.11_chr3_169483680_G_T  21.160000   0.0005  0.002662   \n",
       "11  ENSG00000116251.11_chr3_169483724_A_G  22.153846   0.0005  0.002662   \n",
       "28   ENSG00000130770.18_chr1_28236165_A_G  25.717736   0.0005  0.002662   \n",
       "29   ENSG00000130770.18_chr1_28236165_A_G  25.717736   0.0005  0.002662   \n",
       "30   ENSG00000130770.18_chr1_28236165_A_G  25.717736   0.0005  0.002662   \n",
       "\n",
       "              geneid.v             variant     transcriptid.v  refCount  \\\n",
       "10  ENSG00000116251.11  chr3_169483680_G_T  transcript_207329        24   \n",
       "11  ENSG00000116251.11  chr3_169483724_A_G  transcript_207329        25   \n",
       "28  ENSG00000130770.18   chr1_28236165_A_G   transcript_12375      5250   \n",
       "29  ENSG00000130770.18   chr1_28236165_A_G   transcript_12372        24   \n",
       "30  ENSG00000130770.18   chr1_28236165_A_G   transcript_12377        24   \n",
       "\n",
       "    altCount  totalCount                    transcript_variant significant  \\\n",
       "10         1          25  transcript_207329_chr3_169483680_G_T    FDR<0.05   \n",
       "11         1          26  transcript_207329_chr3_169483724_A_G    FDR<0.05   \n",
       "28      5472       10722    transcript_12375_chr1_28236165_A_G    FDR<0.05   \n",
       "29         0          24    transcript_12372_chr1_28236165_A_G    FDR<0.05   \n",
       "30        22          46    transcript_12377_chr1_28236165_A_G    FDR<0.05   \n",
       "\n",
       "    tested_genes  significant_genes              gid  \n",
       "10           287                 38  ENSG00000116251  \n",
       "11           287                 38  ENSG00000116251  \n",
       "28           287                 38  ENSG00000130770  \n",
       "29           287                 38  ENSG00000130770  \n",
       "30           287                 38  ENSG00000130770  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# astu sig things\n",
    "astu_df = pd.read_csv('/Users/fairliereese/Documents/programming/mele_lab/projects/240903_pt/data/08_allele_specifics/ASTS_results_pantrx.tsv', sep='\\t')\n",
    "astu_df = astu_df.loc[astu_df.significant!='ns']\n",
    "astu_df['gid'] = astu_df['geneid.v'].str.split('.', expand=True)[0]\n",
    "astu_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "45ce80c4-6012-4513-86b2-a202f68dc564",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14665\n",
      "6836\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>population</th>\n",
       "      <th>map_reads_assemblymap</th>\n",
       "      <th>gene_variant</th>\n",
       "      <th>statistic</th>\n",
       "      <th>p.value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>variant</th>\n",
       "      <th>...</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>totalCount</th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>significant</th>\n",
       "      <th>tested_genes</th>\n",
       "      <th>significant_genes</th>\n",
       "      <th>gid</th>\n",
       "      <th>ID</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000139428.12_chr12_109556836_A_G</td>\n",
       "      <td>58.995771</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000139428.12</td>\n",
       "      <td>chr12_109556836_A_G</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>11</td>\n",
       "      <td>59</td>\n",
       "      <td>transcript_226867_chr12_109556836_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "      <td>Total cholesterol levels (MTAG)</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000139428.12_chr12_109556836_A_G</td>\n",
       "      <td>58.995771</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000139428.12</td>\n",
       "      <td>chr12_109556836_A_G</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>11</td>\n",
       "      <td>59</td>\n",
       "      <td>transcript_226867_chr12_109556836_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "      <td>Concentration of very large HDL particles</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000139428.12_chr12_109556836_A_G</td>\n",
       "      <td>58.995771</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000139428.12</td>\n",
       "      <td>chr12_109556836_A_G</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>116</td>\n",
       "      <td>152</td>\n",
       "      <td>transcript_226865_chr12_109556836_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "      <td>Total cholesterol levels (MTAG)</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000139428.12_chr12_109556836_A_G</td>\n",
       "      <td>58.995771</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.002662</td>\n",
       "      <td>ENSG00000139428.12</td>\n",
       "      <td>chr12_109556836_A_G</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>116</td>\n",
       "      <td>152</td>\n",
       "      <td>transcript_226865_chr12_109556836_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "      <td>Concentration of very large HDL particles</td>\n",
       "      <td>ENSG00000139428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000181991.16_chr15_88467745_C_T</td>\n",
       "      <td>20.659419</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>0.014766</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15_88467745_C_T</td>\n",
       "      <td>...</td>\n",
       "      <td>173</td>\n",
       "      <td>160</td>\n",
       "      <td>333</td>\n",
       "      <td>transcript_61885_chr15_88467745_C_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "      <td>Hip circumference adjusted for BMI</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_line_id sample population  map_reads_assemblymap  \\\n",
       "0      GM19117   YRI5        YRI                9522208   \n",
       "1      GM19117   YRI5        YRI                9522208   \n",
       "2      GM19117   YRI5        YRI                9522208   \n",
       "3      GM19117   YRI5        YRI                9522208   \n",
       "4      GM19117   YRI5        YRI                9522208   \n",
       "\n",
       "                             gene_variant  statistic   p.value       FDR  \\\n",
       "0  ENSG00000139428.12_chr12_109556836_A_G  58.995771  0.000500  0.002662   \n",
       "1  ENSG00000139428.12_chr12_109556836_A_G  58.995771  0.000500  0.002662   \n",
       "2  ENSG00000139428.12_chr12_109556836_A_G  58.995771  0.000500  0.002662   \n",
       "3  ENSG00000139428.12_chr12_109556836_A_G  58.995771  0.000500  0.002662   \n",
       "4   ENSG00000181991.16_chr15_88467745_C_T  20.659419  0.002999  0.014766   \n",
       "\n",
       "             geneid.v              variant  ... refCount  altCount  \\\n",
       "0  ENSG00000139428.12  chr12_109556836_A_G  ...       48        11   \n",
       "1  ENSG00000139428.12  chr12_109556836_A_G  ...       48        11   \n",
       "2  ENSG00000139428.12  chr12_109556836_A_G  ...       36       116   \n",
       "3  ENSG00000139428.12  chr12_109556836_A_G  ...       36       116   \n",
       "4  ENSG00000181991.16   chr15_88467745_C_T  ...      173       160   \n",
       "\n",
       "   totalCount                     transcript_variant significant tested_genes  \\\n",
       "0          59  transcript_226867_chr12_109556836_A_G    FDR<0.05          287   \n",
       "1          59  transcript_226867_chr12_109556836_A_G    FDR<0.05          287   \n",
       "2         152  transcript_226865_chr12_109556836_A_G    FDR<0.05          287   \n",
       "3         152  transcript_226865_chr12_109556836_A_G    FDR<0.05          287   \n",
       "4         333    transcript_61885_chr15_88467745_C_T    FDR<0.05          287   \n",
       "\n",
       "   significant_genes              gid  \\\n",
       "0                 38  ENSG00000139428   \n",
       "1                 38  ENSG00000139428   \n",
       "2                 38  ENSG00000139428   \n",
       "3                 38  ENSG00000139428   \n",
       "4                 38  ENSG00000181991   \n",
       "\n",
       "                                          ID           geneID  \n",
       "0            Total cholesterol levels (MTAG)  ENSG00000139428  \n",
       "1  Concentration of very large HDL particles  ENSG00000139428  \n",
       "2            Total cholesterol levels (MTAG)  ENSG00000139428  \n",
       "3  Concentration of very large HDL particles  ENSG00000139428  \n",
       "4         Hip circumference adjusted for BMI  ENSG00000181991  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# limit to astu gwas thing\n",
    "df = pd.read_csv('/Users/fairliereese/Documents/programming/mele_lab/projects/240903_pt/data/08_allele_specifics/ASTU_GWAS_PODER.tsv', sep='\\t')\n",
    "df.head()\n",
    "\n",
    "# explode gene id\n",
    "df['geneID'] = df.geneID.str.split('/')\n",
    "df = df.explode('geneID')\n",
    "\n",
    "print(len(astu_df.index))\n",
    "astu_df = astu_df.loc[astu_df.gid.isin(df.geneID.tolist())]\n",
    "print(len(astu_df.index))\n",
    "\n",
    "# add in trait\n",
    "astu_df = astu_df.merge(df[['ID', 'geneID']],\n",
    "                        how='left',\n",
    "                        left_on='gid',\n",
    "                        right_on='geneID')\n",
    "astu_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1024ca20-1ec0-41a8-a646-1c8311cf67bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2267\n",
      "70447\n",
      "47840\n"
     ]
    }
   ],
   "source": [
    "# limit to novel transcripts\n",
    "mt_df = pd.read_csv('../data/05_mastertable/poder_master_table_fixed_genics.tsv', sep='\\t')\n",
    "mt_df['pop_spec_t'] = (mt_df.population_sharing==1)&\\\n",
    "                      (mt_df.sample_sharing>1)\n",
    "print(len(mt_df.loc[mt_df.pop_spec_t==True, 'isoform'].unique()))\n",
    "# temp = mt_df[['geneid.v', 'structural_category']]\n",
    "# temp['novel'] = temp.structural_category!='FSM'\n",
    "# temp = temp[['geneid.v', 'novel']]\n",
    "# temp = temp.groupby('geneid.v').max().reset_index().rename({'novel':'has_novel_t'},axis=1)\n",
    "# temp['gid'] = temp['geneid.v'].str.split('.', expand=True)[0]\n",
    "\n",
    "print(len(astu_df.index))\n",
    "astu_df = astu_df.loc[astu_df['transcriptid.v'].isin(mt_df.loc[mt_df.structural_category!='FSM', 'isoform'].tolist())]\n",
    "print(len(astu_df.index))\n",
    "\n",
    "# merge in struct stuff\n",
    "astu_df = astu_df.merge(mt_df[['isoform', 'structural_category']],\n",
    "                        how='left',\n",
    "                        left_on='transcriptid.v',\n",
    "                        right_on='isoform')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3aa144dc-a06d-4e17-9c7e-4b30a52b2cf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>population</th>\n",
       "      <th>map_reads_assemblymap</th>\n",
       "      <th>gene_variant</th>\n",
       "      <th>statistic</th>\n",
       "      <th>p.value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>variant</th>\n",
       "      <th>...</th>\n",
       "      <th>totalCount</th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>significant</th>\n",
       "      <th>tested_genes</th>\n",
       "      <th>significant_genes</th>\n",
       "      <th>gid</th>\n",
       "      <th>ID</th>\n",
       "      <th>geneID</th>\n",
       "      <th>isoform</th>\n",
       "      <th>structural_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000181991.16_chr15_88467745_C_T</td>\n",
       "      <td>20.659419</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>0.014766</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15_88467745_C_T</td>\n",
       "      <td>...</td>\n",
       "      <td>137</td>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "      <td>Hip circumference adjusted for BMI</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000181991.16_chr15_88467745_C_T</td>\n",
       "      <td>20.659419</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>0.014766</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15_88467745_C_T</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>transcript_61916_chr15_88467745_C_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "      <td>Hip circumference adjusted for BMI</td>\n",
       "      <td>ENSG00000181991</td>\n",
       "      <td>transcript_61916</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000171863.15_chr17_28468275_C_T</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.010122</td>\n",
       "      <td>ENSG00000171863.15</td>\n",
       "      <td>chr17_28468275_C_T</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>transcript_106030_chr17_28468275_C_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000171863</td>\n",
       "      <td>Triglyceride levels in very large VLDL</td>\n",
       "      <td>ENSG00000171863</td>\n",
       "      <td>transcript_106030</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000171863.15_chr17_28468275_C_T</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>0.001999</td>\n",
       "      <td>0.010122</td>\n",
       "      <td>ENSG00000171863.15</td>\n",
       "      <td>chr17_28468275_C_T</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>transcript_106030_chr17_28468275_C_T</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000171863</td>\n",
       "      <td>Ratio of triglycerides to phosphoglycerides</td>\n",
       "      <td>ENSG00000171863</td>\n",
       "      <td>transcript_106030</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>YRI</td>\n",
       "      <td>9522208</td>\n",
       "      <td>ENSG00000145287.12_chr4_83090803_A_G</td>\n",
       "      <td>19.577926</td>\n",
       "      <td>0.001499</td>\n",
       "      <td>0.007784</td>\n",
       "      <td>ENSG00000145287.12</td>\n",
       "      <td>chr4_83090803_A_G</td>\n",
       "      <td>...</td>\n",
       "      <td>65</td>\n",
       "      <td>transcript_331391_chr4_83090803_A_G</td>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "      <td>287</td>\n",
       "      <td>38</td>\n",
       "      <td>ENSG00000145287</td>\n",
       "      <td>Systemic lupus erythematosus</td>\n",
       "      <td>ENSG00000145287</td>\n",
       "      <td>transcript_331391</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_line_id sample population  map_reads_assemblymap  \\\n",
       "0      GM19117   YRI5        YRI                9522208   \n",
       "1      GM19117   YRI5        YRI                9522208   \n",
       "2      GM19117   YRI5        YRI                9522208   \n",
       "3      GM19117   YRI5        YRI                9522208   \n",
       "4      GM19117   YRI5        YRI                9522208   \n",
       "\n",
       "                            gene_variant  statistic   p.value       FDR  \\\n",
       "0  ENSG00000181991.16_chr15_88467745_C_T  20.659419  0.002999  0.014766   \n",
       "1  ENSG00000181991.16_chr15_88467745_C_T  20.659419  0.002999  0.014766   \n",
       "2  ENSG00000171863.15_chr17_28468275_C_T  11.000000  0.001999  0.010122   \n",
       "3  ENSG00000171863.15_chr17_28468275_C_T  11.000000  0.001999  0.010122   \n",
       "4   ENSG00000145287.12_chr4_83090803_A_G  19.577926  0.001499  0.007784   \n",
       "\n",
       "             geneid.v             variant  ... totalCount  \\\n",
       "0  ENSG00000181991.16  chr15_88467745_C_T  ...        137   \n",
       "1  ENSG00000181991.16  chr15_88467745_C_T  ...         10   \n",
       "2  ENSG00000171863.15  chr17_28468275_C_T  ...         33   \n",
       "3  ENSG00000171863.15  chr17_28468275_C_T  ...         33   \n",
       "4  ENSG00000145287.12   chr4_83090803_A_G  ...         65   \n",
       "\n",
       "                     transcript_variant  significant  tested_genes  \\\n",
       "0   transcript_61914_chr15_88467745_C_T     FDR<0.05           287   \n",
       "1   transcript_61916_chr15_88467745_C_T     FDR<0.05           287   \n",
       "2  transcript_106030_chr17_28468275_C_T     FDR<0.05           287   \n",
       "3  transcript_106030_chr17_28468275_C_T     FDR<0.05           287   \n",
       "4   transcript_331391_chr4_83090803_A_G     FDR<0.05           287   \n",
       "\n",
       "  significant_genes              gid  \\\n",
       "0                38  ENSG00000181991   \n",
       "1                38  ENSG00000181991   \n",
       "2                38  ENSG00000171863   \n",
       "3                38  ENSG00000171863   \n",
       "4                38  ENSG00000145287   \n",
       "\n",
       "                                            ID           geneID  \\\n",
       "0           Hip circumference adjusted for BMI  ENSG00000181991   \n",
       "1           Hip circumference adjusted for BMI  ENSG00000181991   \n",
       "2       Triglyceride levels in very large VLDL  ENSG00000171863   \n",
       "3  Ratio of triglycerides to phosphoglycerides  ENSG00000171863   \n",
       "4                 Systemic lupus erythematosus  ENSG00000145287   \n",
       "\n",
       "             isoform structural_category  \n",
       "0   transcript_61914                 NIC  \n",
       "1   transcript_61916                 NIC  \n",
       "2  transcript_106030                 NNC  \n",
       "3  transcript_106030                 NNC  \n",
       "4  transcript_331391                 NNC  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "astu_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aea54e61-c4c4-4f40-9aba-25f641e9eb57",
   "metadata": {},
   "outputs": [],
   "source": [
    "astu_df.to_csv('astu_gwas_novel_table.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eef5ecf6-dce1-4c81-b592-6569cf73f28e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a parseabe table for variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "70e2fee2-4292-471f-9a3e-0061cc5b9c6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47840\n",
      "4566\n"
     ]
    }
   ],
   "source": [
    "temp = astu_df[['transcript_variant', 'sample', 'cell_line_id', 'population', \n",
    "         'refCount', 'altCount', 'transcriptid.v', 'geneid.v']]\n",
    "print(len(temp.index))\n",
    "temp = temp.drop_duplicates()\n",
    "print(len(temp.index))     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e4384a0a-d2e9-41a5-b3db-e7384e66d6ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>84</td>\n",
       "      <td>53</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript_61916_chr15_88467745_C_T</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>transcript_61916</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>transcript_106030_chr17_28468275_C_T</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>transcript_106030</td>\n",
       "      <td>ENSG00000171863.15</td>\n",
       "      <td>chr17</td>\n",
       "      <td>28468275</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transcript_331391_chr4_83090803_A_G</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>38</td>\n",
       "      <td>27</td>\n",
       "      <td>transcript_331391</td>\n",
       "      <td>ENSG00000145287.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>83090803</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>transcript_345632_chr6_31353908_A_C</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>transcript_345632</td>\n",
       "      <td>ENSG00000234745.14</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31353908</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     transcript_variant sample cell_line_id population  \\\n",
       "0   transcript_61914_chr15_88467745_C_T   YRI5      GM19117        YRI   \n",
       "1   transcript_61916_chr15_88467745_C_T   YRI5      GM19117        YRI   \n",
       "2  transcript_106030_chr17_28468275_C_T   YRI5      GM19117        YRI   \n",
       "4   transcript_331391_chr4_83090803_A_G   YRI5      GM19117        YRI   \n",
       "5   transcript_345632_chr6_31353908_A_C   YRI5      GM19117        YRI   \n",
       "\n",
       "   refCount  altCount     transcriptid.v            geneid.v    chr       pos  \\\n",
       "0        84        53   transcript_61914  ENSG00000181991.16  chr15  88467745   \n",
       "1         2         8   transcript_61916  ENSG00000181991.16  chr15  88467745   \n",
       "2        21        12  transcript_106030  ENSG00000171863.15  chr17  28468275   \n",
       "4        38        27  transcript_331391  ENSG00000145287.12   chr4  83090803   \n",
       "5         8        35  transcript_345632  ENSG00000234745.14   chr6  31353908   \n",
       "\n",
       "  ref alt  \n",
       "0   C   T  \n",
       "1   C   T  \n",
       "2   C   T  \n",
       "4   A   G  \n",
       "5   A   C  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[['chr', 'pos', 'ref', 'alt']] = temp.transcript_variant.str.rsplit('_', expand=True, n=3)\n",
    "temp['chr'] = temp['chr'].str.rsplit('_', expand=True, n=1)[1]\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "936f1ee8-910c-44f2-882f-ee8b4645c675",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = temp.loc[temp.transcript_variant.str.contains('chr15_88467745')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3d2b56eb-d458-4f46-8c3a-5d2aca649677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>84</td>\n",
       "      <td>53</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript_61916_chr15_88467745_C_T</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>transcript_61916</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14194</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>LWK2</td>\n",
       "      <td>GM19328</td>\n",
       "      <td>LWK</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14195</th>\n",
       "      <td>transcript_61911_chr15_88467745_C_T</td>\n",
       "      <td>LWK2</td>\n",
       "      <td>GM19328</td>\n",
       "      <td>LWK</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_61911</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18112</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>LWK5</td>\n",
       "      <td>GM19446</td>\n",
       "      <td>LWK</td>\n",
       "      <td>84</td>\n",
       "      <td>56</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28511</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>ITU2</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>ITU</td>\n",
       "      <td>82</td>\n",
       "      <td>81</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28512</th>\n",
       "      <td>transcript_61911_chr15_88467745_C_T</td>\n",
       "      <td>ITU2</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>ITU</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_61911</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43568</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>YRI2</td>\n",
       "      <td>GM18489</td>\n",
       "      <td>YRI</td>\n",
       "      <td>105</td>\n",
       "      <td>67</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43569</th>\n",
       "      <td>transcript_61911_chr15_88467745_C_T</td>\n",
       "      <td>YRI2</td>\n",
       "      <td>GM18489</td>\n",
       "      <td>YRI</td>\n",
       "      <td>49</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_61911</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43570</th>\n",
       "      <td>transcript_61916_chr15_88467745_C_T</td>\n",
       "      <td>YRI2</td>\n",
       "      <td>GM18489</td>\n",
       "      <td>YRI</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>transcript_61916</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43571</th>\n",
       "      <td>transcript_61921_chr15_88467745_C_T</td>\n",
       "      <td>YRI2</td>\n",
       "      <td>GM18489</td>\n",
       "      <td>YRI</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>transcript_61921</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45747</th>\n",
       "      <td>transcript_61914_chr15_88467745_C_T</td>\n",
       "      <td>YRI3</td>\n",
       "      <td>GM18906</td>\n",
       "      <td>YRI</td>\n",
       "      <td>101</td>\n",
       "      <td>95</td>\n",
       "      <td>transcript_61914</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45748</th>\n",
       "      <td>transcript_61911_chr15_88467745_C_T</td>\n",
       "      <td>YRI3</td>\n",
       "      <td>GM18906</td>\n",
       "      <td>YRI</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_61911</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45749</th>\n",
       "      <td>transcript_61921_chr15_88467745_C_T</td>\n",
       "      <td>YRI3</td>\n",
       "      <td>GM18906</td>\n",
       "      <td>YRI</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>transcript_61921</td>\n",
       "      <td>ENSG00000181991.16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>88467745</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        transcript_variant sample cell_line_id population  \\\n",
       "0      transcript_61914_chr15_88467745_C_T   YRI5      GM19117        YRI   \n",
       "1      transcript_61916_chr15_88467745_C_T   YRI5      GM19117        YRI   \n",
       "14194  transcript_61914_chr15_88467745_C_T   LWK2      GM19328        LWK   \n",
       "14195  transcript_61911_chr15_88467745_C_T   LWK2      GM19328        LWK   \n",
       "18112  transcript_61914_chr15_88467745_C_T   LWK5      GM19446        LWK   \n",
       "28511  transcript_61914_chr15_88467745_C_T   ITU2      HG03729        ITU   \n",
       "28512  transcript_61911_chr15_88467745_C_T   ITU2      HG03729        ITU   \n",
       "43568  transcript_61914_chr15_88467745_C_T   YRI2      GM18489        YRI   \n",
       "43569  transcript_61911_chr15_88467745_C_T   YRI2      GM18489        YRI   \n",
       "43570  transcript_61916_chr15_88467745_C_T   YRI2      GM18489        YRI   \n",
       "43571  transcript_61921_chr15_88467745_C_T   YRI2      GM18489        YRI   \n",
       "45747  transcript_61914_chr15_88467745_C_T   YRI3      GM18906        YRI   \n",
       "45748  transcript_61911_chr15_88467745_C_T   YRI3      GM18906        YRI   \n",
       "45749  transcript_61921_chr15_88467745_C_T   YRI3      GM18906        YRI   \n",
       "\n",
       "       refCount  altCount    transcriptid.v            geneid.v    chr  \\\n",
       "0            84        53  transcript_61914  ENSG00000181991.16  chr15   \n",
       "1             2         8  transcript_61916  ENSG00000181991.16  chr15   \n",
       "14194        60        65  transcript_61914  ENSG00000181991.16  chr15   \n",
       "14195        18         0  transcript_61911  ENSG00000181991.16  chr15   \n",
       "18112        84        56  transcript_61914  ENSG00000181991.16  chr15   \n",
       "28511        82        81  transcript_61914  ENSG00000181991.16  chr15   \n",
       "28512        10         1  transcript_61911  ENSG00000181991.16  chr15   \n",
       "43568       105        67  transcript_61914  ENSG00000181991.16  chr15   \n",
       "43569        49         1  transcript_61911  ENSG00000181991.16  chr15   \n",
       "43570         6         5  transcript_61916  ENSG00000181991.16  chr15   \n",
       "43571         6         7  transcript_61921  ENSG00000181991.16  chr15   \n",
       "45747       101        95  transcript_61914  ENSG00000181991.16  chr15   \n",
       "45748        12         1  transcript_61911  ENSG00000181991.16  chr15   \n",
       "45749        10         7  transcript_61921  ENSG00000181991.16  chr15   \n",
       "\n",
       "            pos ref alt  \n",
       "0      88467745   C   T  \n",
       "1      88467745   C   T  \n",
       "14194  88467745   C   T  \n",
       "14195  88467745   C   T  \n",
       "18112  88467745   C   T  \n",
       "28511  88467745   C   T  \n",
       "28512  88467745   C   T  \n",
       "43568  88467745   C   T  \n",
       "43569  88467745   C   T  \n",
       "43570  88467745   C   T  \n",
       "43571  88467745   C   T  \n",
       "45747  88467745   C   T  \n",
       "45748  88467745   C   T  \n",
       "45749  88467745   C   T  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6571a84c-8f00-48d5-b694-c40b7b7daba9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14194</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14195</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18112</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28511</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28512</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43568</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43569</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43570</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43571</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45747</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45748</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45749</th>\n",
       "      <td>FDR&lt;0.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      significant\n",
       "0        FDR<0.05\n",
       "1        FDR<0.05\n",
       "14194    FDR<0.05\n",
       "14195    FDR<0.05\n",
       "18112    FDR<0.05\n",
       "28511    FDR<0.05\n",
       "28512    FDR<0.05\n",
       "43568    FDR<0.05\n",
       "43569    FDR<0.05\n",
       "43570    FDR<0.05\n",
       "43571    FDR<0.05\n",
       "45747    FDR<0.05\n",
       "45748    FDR<0.05\n",
       "45749    FDR<0.05"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "astu_df.loc[astu_df.transcript_variant.str.contains('chr15_88467745')][['significant']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "179f934c-5e33-4fbc-88e7-1c3c34274c05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "813\n",
      "813\n"
     ]
    }
   ],
   "source": [
    "temp2 = temp2[['sample',\t'cell_line_id',\t'population',\t'geneid.v',\t'chr', 'pos', 'ref', 'alt']]\n",
    "print(len(temp2.index))\n",
    "temp2 = temp2.drop_duplicates()\n",
    "print(len(temp2.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "dcb30a74-5f32-4a91-ab66-57471e263fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2.to_csv('astu_example_config.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9be5b9f-9bfb-4815-b1de-2dc4c750b2a4",
   "metadata": {},
   "source": [
    "## What examples of these are highly-reproducible between samples?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5b0db0db-7ae6-4c2b-a56f-f231109ff7a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = temp[['transcript_variant', 'sample']].groupby('transcript_variant').nunique().reset_index().rename({'sample':'n_sample'}, axis=1)\n",
    "temp2 = temp2.sort_values(by='n_sample', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3816313e-d378-4a14-99c2-a5eb45bb6767",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2.to_csv('astu_example_high_sharing.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "05473d56-c56c-468f-aa6a-edab0c5c506f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(temp['sample'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f787fa24-39b6-4505-a704-b59566384ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp['total_count'] = temp['refCount']+temp['altCount']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f06d9c70-ef69-4225-842d-f1232df9dcf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HAC1', 'HAC6', 'LWK1', 'LWK4', 'LWK5', 'PEL4', 'PEL5', 'ITU1',\n",
       "       'ITU2', 'CEU3', 'CEU4', 'CEU5', 'YRI1'], dtype=object)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_347509_chr6_33065128_T_C'\n",
    "temp.loc[temp.transcript_variant == curr_var, 'sample'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "4b34fc50-db71-4233-a094-ceb914302760",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43354</th>\n",
       "      <td>transcript_347509_chr6_33065128_T_C</td>\n",
       "      <td>YRI1</td>\n",
       "      <td>GM18486</td>\n",
       "      <td>YRI</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>transcript_347509</td>\n",
       "      <td>ENSG00000231389.8</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33065128</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        transcript_variant sample cell_line_id population  \\\n",
       "43354  transcript_347509_chr6_33065128_T_C   YRI1      GM18486        YRI   \n",
       "\n",
       "       refCount  altCount     transcriptid.v           geneid.v   chr  \\\n",
       "43354         3        10  transcript_347509  ENSG00000231389.8  chr6   \n",
       "\n",
       "            pos ref alt  \n",
       "43354  33065128   T   C  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_347509_chr6_33065128_T_C'\n",
    "curr_sample = 'YRI1'\n",
    "\n",
    "temp.loc[(temp.transcript_variant == curr_var)&\\\n",
    "         (temp['sample'] == curr_sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "0fc95e1d-7f33-4953-86d4-d7cc22534201",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['YRI5', 'YRI7', 'LWK1', 'LWK2', 'LWK5', 'PEL1', 'PEL5', 'ITU3',\n",
       "       'ITU5', 'CEU1', 'CEU5', 'YRI2', 'YRI3'], dtype=object)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_345640_chr6_31353908_A_C'\n",
    "temp.loc[temp.transcript_variant == curr_var, 'sample'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "e27b2c09-48df-4bdf-ad76-c8fd1fd3d3b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>transcript_345640_chr6_31353908_A_C</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>GM19117</td>\n",
       "      <td>YRI</td>\n",
       "      <td>25</td>\n",
       "      <td>34</td>\n",
       "      <td>transcript_345640</td>\n",
       "      <td>ENSG00000234745.14</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31353908</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     transcript_variant sample cell_line_id population  \\\n",
       "35  transcript_345640_chr6_31353908_A_C   YRI5      GM19117        YRI   \n",
       "\n",
       "    refCount  altCount     transcriptid.v            geneid.v   chr       pos  \\\n",
       "35        25        34  transcript_345640  ENSG00000234745.14  chr6  31353908   \n",
       "\n",
       "   ref alt  \n",
       "35   A   C  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_345640_chr6_31353908_A_C'\n",
    "curr_sample = 'YRI5'\n",
    "\n",
    "temp.loc[(temp.transcript_variant == curr_var)&\\\n",
    "         (temp['sample'] == curr_sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "b45c2523-9f6e-471d-b060-adbd10412172",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get high-count things w/ high sample sharing\n",
    "temp3 = temp.loc[temp.transcript_variant.isin(temp2.loc[temp2.n_sample>=10, 'transcript_variant'].tolist())]\n",
    "temp3 = temp3.loc[temp3.total_count >= 100]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2d89c829-8ac1-426e-a2b8-5371b19e00f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>total_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13128</th>\n",
       "      <td>transcript_346703_chr6_32578927_G_A</td>\n",
       "      <td>LWK1</td>\n",
       "      <td>GM19307</td>\n",
       "      <td>LWK</td>\n",
       "      <td>211</td>\n",
       "      <td>25</td>\n",
       "      <td>transcript_346703</td>\n",
       "      <td>ENSG00000196126.12</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578927</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4908</th>\n",
       "      <td>transcript_346723_chr6_32578981_C_T</td>\n",
       "      <td>YRI7</td>\n",
       "      <td>GM19240</td>\n",
       "      <td>YRI</td>\n",
       "      <td>151</td>\n",
       "      <td>72</td>\n",
       "      <td>transcript_346723</td>\n",
       "      <td>ENSG00000196126.12</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578981</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4728</th>\n",
       "      <td>transcript_346723_chr6_32578927_G_A</td>\n",
       "      <td>YRI7</td>\n",
       "      <td>GM19240</td>\n",
       "      <td>YRI</td>\n",
       "      <td>146</td>\n",
       "      <td>73</td>\n",
       "      <td>transcript_346723</td>\n",
       "      <td>ENSG00000196126.12</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578927</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21161</th>\n",
       "      <td>transcript_345640_chr6_31354139_A_G</td>\n",
       "      <td>PEL1</td>\n",
       "      <td>HG01567</td>\n",
       "      <td>PEL</td>\n",
       "      <td>153</td>\n",
       "      <td>57</td>\n",
       "      <td>transcript_345640</td>\n",
       "      <td>ENSG00000234745.14</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31354139</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20936</th>\n",
       "      <td>transcript_345640_chr6_31354138_A_G</td>\n",
       "      <td>PEL1</td>\n",
       "      <td>HG01567</td>\n",
       "      <td>PEL</td>\n",
       "      <td>157</td>\n",
       "      <td>51</td>\n",
       "      <td>transcript_345640</td>\n",
       "      <td>ENSG00000234745.14</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31354138</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        transcript_variant sample cell_line_id population  \\\n",
       "13128  transcript_346703_chr6_32578927_G_A   LWK1      GM19307        LWK   \n",
       "4908   transcript_346723_chr6_32578981_C_T   YRI7      GM19240        YRI   \n",
       "4728   transcript_346723_chr6_32578927_G_A   YRI7      GM19240        YRI   \n",
       "21161  transcript_345640_chr6_31354139_A_G   PEL1      HG01567        PEL   \n",
       "20936  transcript_345640_chr6_31354138_A_G   PEL1      HG01567        PEL   \n",
       "\n",
       "       refCount  altCount     transcriptid.v            geneid.v   chr  \\\n",
       "13128       211        25  transcript_346703  ENSG00000196126.12  chr6   \n",
       "4908        151        72  transcript_346723  ENSG00000196126.12  chr6   \n",
       "4728        146        73  transcript_346723  ENSG00000196126.12  chr6   \n",
       "21161       153        57  transcript_345640  ENSG00000234745.14  chr6   \n",
       "20936       157        51  transcript_345640  ENSG00000234745.14  chr6   \n",
       "\n",
       "            pos ref alt  total_count  \n",
       "13128  32578927   G   A          236  \n",
       "4908   32578981   C   T          223  \n",
       "4728   32578927   G   A          219  \n",
       "21161  31354139   A   G          210  \n",
       "20936  31354138   A   G          208  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp3 = temp3.sort_values(by='total_count', ascending=False)\n",
    "temp3.head(30).tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "1fccd205-58a7-4455-b280-8817156d19ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>total_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11550</th>\n",
       "      <td>transcript_345590_chr6_31354139_A_G</td>\n",
       "      <td>HAC6</td>\n",
       "      <td>HG00621</td>\n",
       "      <td>HAC</td>\n",
       "      <td>79</td>\n",
       "      <td>30</td>\n",
       "      <td>transcript_345590</td>\n",
       "      <td>ENSG00000234745.14</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31354139</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        transcript_variant sample cell_line_id population  \\\n",
       "11550  transcript_345590_chr6_31354139_A_G   HAC6      HG00621        HAC   \n",
       "\n",
       "       refCount  altCount     transcriptid.v            geneid.v   chr  \\\n",
       "11550        79        30  transcript_345590  ENSG00000234745.14  chr6   \n",
       "\n",
       "            pos ref alt  total_count  \n",
       "11550  31354139   A   G          109  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_345590_chr6_31354139_A_G'\n",
    "curr_sample = 'HAC6'\n",
    "\n",
    "temp.loc[(temp.transcript_variant == curr_var)&\\\n",
    "         (temp['sample'] == curr_sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "36db926d-4619-427a-97a5-95f0b2b0f49a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>total_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_variant, sample, cell_line_id, population, refCount, altCount, transcriptid.v, geneid.v, chr, pos, ref, alt, total_count]\n",
       "Index: []"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_var = 'transcript_345590_chr6_31354139_A_G'\n",
    "curr_sample = 'YRI7'\n",
    "\n",
    "temp.loc[(temp.transcript_variant == curr_var)&\\\n",
    "         (temp['sample'] == curr_sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fb60db5a-3ee3-4d0a-b1f9-27bacd962a00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_variant</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>population</th>\n",
       "      <th>refCount</th>\n",
       "      <th>altCount</th>\n",
       "      <th>transcriptid.v</th>\n",
       "      <th>geneid.v</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>total_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4656</th>\n",
       "      <td>transcript_346703_chr6_32578927_G_A</td>\n",
       "      <td>YRI7</td>\n",
       "      <td>GM19240</td>\n",
       "      <td>YRI</td>\n",
       "      <td>517</td>\n",
       "      <td>45</td>\n",
       "      <td>transcript_346703</td>\n",
       "      <td>ENSG00000196126.12</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578927</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       transcript_variant sample cell_line_id population  \\\n",
       "4656  transcript_346703_chr6_32578927_G_A   YRI7      GM19240        YRI   \n",
       "\n",
       "      refCount  altCount     transcriptid.v            geneid.v   chr  \\\n",
       "4656       517        45  transcript_346703  ENSG00000196126.12  chr6   \n",
       "\n",
       "           pos ref alt  total_count  \n",
       "4656  32578927   G   A          562  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maybe but transcript is hard to say if it is expressed...\n",
    "curr_var = 'transcript_346703_chr6_32578927_G_A'\n",
    "curr_sample = 'YRI7'\n",
    "\n",
    "temp.loc[(temp.transcript_variant == curr_var)&\\\n",
    "         (temp['sample'] == curr_sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "1f95a478-5d98-4aab-84c2-e3166a86e52f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       tid sample           tpm population\n",
      "55175    transcript_345590   YRI5  1.067800e+02        YRI\n",
      "211050   transcript_345590   YRI6  1.099900e+00        YRI\n",
      "366925   transcript_345590   YRI7  7.256610e+01        YRI\n",
      "522800   transcript_345590   HAC1  2.527890e-01        HAC\n",
      "678675   transcript_345590   HAC2  1.624330e-04        HAC\n",
      "834550   transcript_345590   HAC3  0.000000e+00        HAC\n",
      "990425   transcript_345590   HAC4  2.836090e+00        HAC\n",
      "1146300  transcript_345590   HAC5  1.677830e-03        HAC\n",
      "1302175  transcript_345590   HAC6  3.893740e+01        HAC\n",
      "1458050  transcript_345590   LWK1  0.000000e+00        LWK\n",
      "1613925  transcript_345590   MPC1  0.000000e+00        MPC\n",
      "1769800  transcript_345590   LWK2  1.861920e-08        LWK\n",
      "1925675  transcript_345590   LWK3  1.226370e+00        LWK\n",
      "2081550  transcript_345590   LWK4  2.658300e+01        LWK\n",
      "2237425  transcript_345590   LWK5  1.764090e+00        LWK\n",
      "2393300  transcript_345590   AJI1  2.098540e+01        AJI\n",
      "2549175  transcript_345590   AJI2  7.350030e-08        AJI\n",
      "2705050  transcript_345590   AJI3  3.316860e-05        AJI\n",
      "2860925  transcript_345590   AJI4  1.554120e+02        AJI\n",
      "3016800  transcript_345590   AJI5  7.802460e+00        AJI\n",
      "3172675  transcript_345590   PEL1  2.677770e+00        PEL\n",
      "3328550  transcript_345590   MPC2  0.000000e+00        MPC\n",
      "3484425  transcript_345590   PEL2  1.220160e-07        PEL\n",
      "3640300  transcript_345590   PEL3  1.195370e-05        PEL\n",
      "3796175  transcript_345590   PEL4  1.257180e+02        PEL\n",
      "3952050  transcript_345590   PEL5  1.874310e+01        PEL\n",
      "4107925  transcript_345590   PEL6  0.000000e+00        PEL\n",
      "4263800  transcript_345590   ITU1  0.000000e+00        ITU\n",
      "4419675  transcript_345590   ITU2  7.412680e-01        ITU\n",
      "4575550  transcript_345590   ITU3  8.207490e-01        ITU\n",
      "4731425  transcript_345590   ITU4  7.503240e-01        ITU\n",
      "4887300  transcript_345590   ITU5  2.279880e+00        ITU\n",
      "5043175  transcript_345590   MPC3  2.679290e+00        MPC\n",
      "5199050  transcript_345590   AJI6  8.307870e-01        AJI\n",
      "5354925  transcript_345590   CEU1  9.898890e-01        CEU\n",
      "5510800  transcript_345590   CEU2  2.831340e-03        CEU\n",
      "5666675  transcript_345590   CEU3  6.547650e+00        CEU\n",
      "5822550  transcript_345590   CEU4  5.275610e-01        CEU\n",
      "5978425  transcript_345590   CEU5  1.759200e-05        CEU\n",
      "6134300  transcript_345590   MPC4  1.053170e+00        MPC\n",
      "6290175  transcript_345590   YRI1  2.216570e+00        YRI\n",
      "6446050  transcript_345590   YRI2  8.997750e-01        YRI\n",
      "6601925  transcript_345590   YRI3  8.018250e+00        YRI\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdoUlEQVR4nO3de3hddZ3v8fenCa2tcksIiClQtOBlwGsO43UUaWfCyGM5cxyhXoIzjD06VFv1iOg5Z1DncYYZ5wj1qTpPkQ7NjIeLqICC0dYLqCNouLVAVSIU25xC09SWS2vLTr7nj7Wy2ITsZDfZe6+9k8/refJkr9/67b2+2cnOZ//WWnv9FBGYmZkBzMq7ADMzqx8OBTMzyzgUzMws41AwM7OMQ8HMzDLNeRcwFUcddVQsWLAg7zLMzBrKHXfcsTMi2sZaV7VQkLQWOAvYERGnFLV/CLgAGAJuiogL0/ZPAuen7R+OiO9NtI0FCxbQ29tbjfLNzKYtSQ+XWlfNkcKVwGqgu6iQ04ElwCsiYr+ko9P2lwHnAn8EvADYIOnkiBiqYn1mZjZK1Y4pRMStwK5RzR8ELomI/WmfHWn7EuDqiNgfEQ8BfcBp1arNzMzGVusDzScDb5J0u6RbJP2XtL0d2FrUb1va9iySlknqldQ7MDBQ5XLNzGaWWodCM9ACvBb4OHCtJB3MA0TEmojoiIiOtrYxj5OYmdkk1ToUtgHfjMQvgGHgKKAfOK6o3/y0zcxSg4ODrFixgl27Ru+VNaucWofC9cDpAJJOBmYDO4EbgXMlzZF0InAS8Isa12ZW17q7u9m0aRPd3d0TdzabpKqFgqSrgJ8DL5a0TdL5wFrghZLuBa4GzktHDfcB1wL3Az3ABT7zyOxpg4OD9PT0EBH09PR4tGBVU7VTUiNiaYlV7ynR/3PA56pVj1kj6+7uZnh4GIChoSG6u7tZuXJlvkXZtOTLXJg1gA0bNlAoFAAoFAqsX78+54psunIomDWARYsW0dycDOybm5tZvHhxzhXZdOVQMGsAXV1dzJqVvFybmpro6urKuSKbrhwKZg2gtbWVzs5OJNHZ2UlLS0veJdk01dBXSTWbSbq6utiyZYtHCVZVDgWzBtHa2sqqVavyLsOmOe8+MjOzjEPBzMwyDgUzM8s4FMzMLONQMDOzjEPBzMwyDgUzM8s4FMzMLONQMDOzjEPBzMwyDgUzM8tUczrOtZJ2pFNvjl73MUkh6ah0WZK+KKlP0kZJr65WXWZmVlo1RwpXAp2jGyUdB/wp8Lui5jOBk9KvZcBXqliXmZmVULVQiIhbgbFmF78UuBCIorYlQHckbgOOkHRstWozM7Ox1fSYgqQlQH9E3DNqVTuwtWh5W9o21mMsk9QrqXdgYKBKlZqZzUw1CwVJ84BPAX83lceJiDUR0RERHW1tbZUpzszMgNpOsvMi4ETgHkkA84E7JZ0G9APHFfWdn7aZmVkN1WykEBGbIuLoiFgQEQtIdhG9OiIeAW4EutKzkF4L7ImI7bWqzczMEtU8JfUq4OfAiyVtk3T+ON1vBh4E+oDLgb+tVl1mZlZa1XYfRcTSCdYvKLodwAXVqsXMzMrjTzSbmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYWcUMDg6yYsUKdu0a66r51ggcCmZWMd3d3WzatInu7u68S7FJciiYWUUMDg7S09NDRNDT0+PRQoNyKJhZRXR3dzM8PAzA0NCQRwsNyqFgZhWxYcMGCoUCAIVCgfXr1+dckU2GQ8HMKmLRokU0NycXXm5ubmbx4sU5V2ST4VAws4ro6upi1qzkX0pTUxNdXV05V2ST4VAws4pobW2ls7MTSXR2dtLS0pJ3STYJ1Zx5ba2kHZLuLWr7vKRfSdoo6VuSjiha90lJfZJ+LenPqlWXmVVPV1cXp556qkcJDayaI4Urgc5RbeuBUyLi5cBvgE8CSHoZcC7wR+l9viypqYq1mVkVtLa2smrVKo8SGljVQiEibgV2jWr7fkQU0sXbgPnp7SXA1RGxPyIeIpmr+bRq1WZmZmPL85jCXwPfTW+3A1uL1m1L255F0jJJvZJ6BwYGqlyimdnMkksoSPqfQAH42sHeNyLWRERHRHS0tbVVvjgzsxmsudYblPQ+4CzgjIiItLkfOK6o2/y0zczMaqimIwVJncCFwNsjYm/RqhuBcyXNkXQicBLwi1rWZmZmVRwpSLoKeAtwlKRtwMUkZxvNAdZLArgtIj4QEfdJuha4n2S30gURMVSt2szMbGx6eg9O4+no6Ije3t68yzAzayiS7oiIjrHW+RPNZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVmm5ldJzdPq1avp6+srub6/P7kwa3v7mFM5ALBw4UKWL19e8drMzOrBjAqFiezbty/vEszMcjWjQmGid/grV64E4LLLLqt+MWZmdcjHFMzMLONQMDOzjEPBzMwyDgUzM8tULRQkrZW0Q9K9RW0tktZLeiD9fmTaLklflNQnaaOkV1erLjMzK62aI4Urgc5RbRcBP4iIk4AfpMsAZwInpV/LgK9UsS4zMyuhaqEQEbcCu0Y1LwHWpbfXAWcXtXdH4jbgCEnHVqs2MzMbW62PKRwTEdvT248Ax6S324GtRf22pW3PImmZpF5JvQMDA9Wr1MxsBsrtQHNEBBCTuN+aiOiIiI62trYqVGZmNnPVOhQeHdktlH7fkbb3A8cV9ZuftpmZWQ3VOhRuBM5Lb58H3FDU3pWehfRaYE/RbiYzM6uRql37SNJVwFuAoyRtAy4GLgGulXQ+8DDwzrT7zcCfA33AXuCvqlWXmZmVVrVQiIilJVadMUbfAC6oVi1mZlYef6LZzMwyDgUzM8s4FMzMLONQMDOzjEPBzMwyDgUzM8s4FMzMLONQMDOzjEPBzMwyDgUzM8s4FMzMLONQMDOzjEPBzMwyDgUzM8s4FMzMLONQMDOzzIShIOksSXdJ2iXpMUmPS3psKhuV9BFJ90m6V9JVkp4j6URJt0vqk3SNpNlT2YaZmR28ckYKl5HMp9waEYdFxKERcdhkNyipHfgw0BERpwBNwLnAPwGXRsRC4PfA+ZPdhpmZTU45obAVuDedMrNSmoG5kpqBecB24K3Aden6dcDZFdyemZmVoZw5mi8EbpZ0C7B/pDEivjCZDUZEv6R/AX4H7AO+D9wB7I6IQtptG9A+1v0lLQOWARx//PGTKcHMzEooZ6TwOWAv8Bzg0KKvSZF0JLAEOBF4AfBcoLPc+0fEmojoiIiOtra2yZZhZmZjKGek8IJ033+lLAIeiogBAEnfBN4AHCGpOR0tzAf6K7hNMzMrQzkjhZsl/WkFt/k74LWS5kkScAZwP/Aj4B1pn/OAGyq4TTMzK0M5ofBBoEfSH9LTUad0SmpE3E5yQPlOYFNawxrgE8BHJfUBrcAVk92GmZlNzoS7jyJi0scPxnnMi4GLRzU/CJxW6W2ZmVn5yjmmgKS/AN4IBPCTiLi+mkWZmVk+yvlE85eBD5Ds6rkX+ICkL1W7MDMzq71yRgpvBV468uE1SeuA+6palZmZ5aKcA819QPGnxI5L28zMbJopZ6RwKLBZ0i9IjimcBvxS0o0AEfH2KtZnZmY1VE4ozAXOLFoWycXrRp89ZGZmDa6cUGiOiFuKGyTNHd1mZmaNr2QoSPog8LfACyVtLFp1KPCzahdmZma1N95I4f8C3wX+EbioqP3xiNhV1arMzCwXJUMhIvYAe4CltSvHzMzy5Dmazcws41AwM7OMQ8HMzDIOBTMzyzgUzMws41AwM7NMLqEg6QhJ10n6laTNkl4nqUXSekkPpN+PzKM2M7OZLK+RwiqgJyJeArwC2EzyAbkfRMRJwA945gfmzMysBmoeCpIOB/6EdA7miDgQEbuBJcC6tNs64Oxa12ZmNtPlMVI4ERgA/k3SXZK+Kum5wDERsT3t8whwTA61mZnNaHmEQjPwauArEfEq4ElG7SpKZ3mLse4saZmkXkm9AwMDVS/WzGwmySMUtgHbIuL2dPk6kpB4VNKxAOn3HWPdOSLWRERHRHS0tbXVpGAzs5mi5qEQEY8AWyW9OG06A7gfuBE4L207D7ih1rWZmc105UyyUw0fAr4maTbwIPBXJAF1raTzgYeBdx7MA65evZq+vqlNHT1y/5UrV07pcRYuXMjy5cun9BhmZnnIJRQi4m6gY4xVZ0z2Mfv6+vj15jt5/lEHJl3XLA4BYM/AbZN+jEd2zp70fc3M8pbXSKEqnn/UAc4/uz/XGq64vj3X7ZuZTYUvc2FmZhmHgpmZZRwKZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVnGoWBmZhmHgpmZZRwKZmaWcSiYmVkmt1CQ1CTpLknfSZdPlHS7pD5J16RTddoYBgcHWbFiBbt27cq7FDObZvIcKawANhct/xNwaUQsBH4PnJ9LVQ2gu7ubTZs20d3dnXcpZjbN5BIKkuYDbwO+mi4LeCtwXdplHXB2HrXVu8HBQXp6eogIenp6PFows4rKa6RwGXAhMJwutwK7I6KQLm8DxpzsWNIySb2SegcGBqpeaL3p7u5meDh52oaGhjxaMLOKqnkoSDoL2BERd0zm/hGxJiI6IqKjra2twtXVvw0bNlAoJNlZKBRYv359zhWZ2XSSx0jhDcDbJW0BribZbbQKOEJSc9pnPtCfQ211b9GiRTQ3J09Tc3MzixcvzrkiM5tOah4KEfHJiJgfEQuAc4EfRsS7gR8B70i7nQfcUOvaGkFXVxezZiW/tqamJrq6unKuyMymk3r6nMIngI9K6iM5xnBFzvXUpdbWVjo7O5FEZ2cnLS0teZdkZtNI88Rdqicifgz8OL39IHBanvU0iq6uLrZs2eJRgplVXK6hYJPT2trKqlWr8i7DzKahetp9ZGZmOXMomJlZxqFgZmYZh4KZmWUcCmZmlnEomJlZxqFgVeN5H8waj0PBqsbzPpg1HoeCVYXnfTBrTA4FqwrP+2DWmBwKVhWe98GsMTkUrCo874NZY3IoWFV43gezxuRQsKrwvA9mjcmXzraq8bwPZo2n5iMFScdJ+pGk+yXdJ2lF2t4iab2kB9LvR9a6NquskXkfPEowaxx5jBQKwMci4k5JhwJ3SFoPvA/4QURcIuki4CKSKTrNpr3Vq1fT19c3bp/+/n4A2tvbS/ZZuHAhy5cvr2htNrPUfKQQEdsj4s709uPAZqAdWAKsS7utA86udW1m9Wzfvn3s27cv7zJsmsv1mIKkBcCrgNuBYyJie7rqEeCYEvdZBiwDOP7442tQpVn1lfPufuXKlQBcdtll1S3GZrTcQkHS84BvACsj4jFJ2bqICEkx1v0iYg2wBqCjo2PMPmZWHRPt5ipnFxd4N1c9yyUUJB1CEghfi4hvps2PSjo2IrZLOhbYkUdtZtVQzjGDiYzcf2TEMBnV/mfs3VuNr+ahoGRIcAWwOSK+ULTqRuA84JL0+w21rs2sWvr6+tj4q99wyNHHTfoxCjoEgM27JveP96kdWye97RETBYp3cTW+PEYKbwDeC2ySdHfa9imSMLhW0vnAw8A7c6jNrGoOOfo42pZ+PLftD1z1+dy2bY2j5qEQET8FVGL1GbWsxczMnsmfaDazzFSPfVTiuAf4QHSeHApmlunr6+OBjfdwQtPkPsI0eyiZQ+PAfZsmXcPD6WNYPhwKZvYMJzTN4n8fNi+37f/9Y3tz27b5KqlmZlbEoWBmZhmHgpmZZRwKZmaWcSiYmVnGoWBmVqcGBwdZsWIFu3btqtk2HQpmZnWqu7ubTZs20d3dXbNtOhTMzOrQ4OAgPT09RAQ9PT01Gy04FMzM6lB3dzfDw8mnu4eGhmo2Wpg2n2ju7+/n8cdmc8X140/uUW3bd87miQP9udZgZo1vw4YNFAoFAAqFAuvXr5/yNaXKMW1Cwaye9ff389TjT+Z6+eqndmylf99zc9t+PRkcHOSzn/0sF198MS0tLXmXM6ZFixZx8803UygUaG5uZvHixTXZ7rQJhfb2dvbM3sr5Z+f7Lv2K69s5vC3f0YqZja/4AG4t3n1PRldXFz09PQA0NTXR1dVVk+1Om1CYTioxD64vPVxf2tvbeWzXvtwn2WlvmZvb9uvF6AO4XV1ddTlaaG1tpbOzk29/+9t0dnbWrEaHQo2Vc736/v7+cee6HVk3Xp/+/v4JtzMTgqPc5xscsjPFWAdw63m0sGXLlpqNEqAOQ0FSJ7AKaAK+GhGXlHvfR3ZO7UDz4J5kDtzWw5+a9GM8snM2h7eVXt/X18fGzffQ3NJUulMzcOg4G4kA4MChfyjZ5QB/YM+jvy+5vrBraJwNwPvf/362b99ecv3+/fuzF9ZUzJo1izlz5ozb59hjj+Xyyy8fc105o6qJJpOvRMhOl9Do7+9nb2E418tXP1wYZl5/9XYD53UAdzJaW1tZtWpVTbdZV6EgqQn4ErAY2Ab8UtKNEXH/RPdduHDhhI9f7jvwYQ4v2Wfu3LnjvqM8vG3iWppbmmg5K7/r1QPs+s74L/rdu3ez98knaSrxFzI0lGXTlEQMs39/oeT6oUJSSym33HILg4MDzGmefDFKZ4cdOvB4yT5PHHicX+/ZMea6/QXR398/LUKhEaxevTrb117K3r17iTL/QPfu3cvpp58+5jpJzJtX+rXa2dlZ8vc+0RsrqMybq6m+sRqtrkIBOA3oi4gHASRdDSwBJgyFcl6Q9bCvvr+/n6cGh9ixrvQ/oBgCpvoPV6BxBiNRgP5C6Xdj7e3t7Nm9k8NLjFie2AuF8QcbZWlugueNk497Hh//9zGRp4bEcJnP5f5CqanDYZbgkKap/VKe2rG15NlHhd/vIJ7aP6XHB9Ahc2g+8uiS26fl5HHv397ezgODO0uuf2RomP0VeDMwR/D8UrO7aWq/83qxe/dunnzyyapvZ3h4OBv5jFdLueotFNqBrUXL24A/Lu4gaRmwDOD4448/qAevh3dyRxxxxIS7Myr27mH2OO8eZie1lDLRaKec3TLlmGjk1T5BLW9+85unvPuoHBPVOdHzNeHzua+ZfTH53ZYj5j6nufTB5JaTp1xnU38/qsDz2TR3LrNLPJ8nTVDH8uXLp/Raftvb3sbevU+PlOfNm8dNN9006ccrZaK/TajM3+dEf5tQ3p6UESp3iFULkt4BdEbE36TL7wX+OCLG/Avo6OiI3t7eWpZoZg3u0ksv5aabbmJoaIimpibOOuusuj2mUC2S7oiIjrHW1dtlLvqB44qW56dtZmYV0dXVRVNTsm+1ubm5pmf2NIJ6C4VfAidJOlHSbOBc4MacazKzaWTk/H9JNT3/v1HU1TGFiChIWg58j+SU1LURcV/OZZnZNJPH+f+Noq6OKRwsH1MwMzt4jXRMwczMcuRQMDOzjEPBzMwyDgUzM8s09IFmSQPAwxV+2KOA0p/zrx+us7JcZ+U0Qo0ws+s8ISLGvHRnQ4dCNUjqLXVUvp64zspynZXTCDWC6yzFu4/MzCzjUDAzs4xD4dnW5F1AmVxnZbnOymmEGsF1jsnHFMzMLOORgpmZZRwKZmaWmXGhoMRPJZ1Z1PaXknokDUm6W9K9kr4t6Yh0/QJJ99aovidGLb9P0upRbXenU5UWtx0i6RJJD0i6U9LPi3/GataZPqc7JR2ZLh8rKSS9sajPgKRWSZ+W9D/StudIWi/p01WsMyT9R9Fyc1rLd9Ll96XLd0u6X9L7i/qeKak3bb9L0v+pVp0laj87rf8l6XL2dyjpLSM/Qx5GvVa+LmneqPaRr4vS9h9Lqvnpn5KekHRqUT27JD2U3t4w1vMo6cp0wq9a1Pd8SVdL+q2kOyTdLOlkSftGPY9dIz/PqPs/6//DVM24UIjkIMoHgC+k/5SeB/wDcAGwLyJeGRGnALvStroi6aUklxV/k6TnFq36e+BY4JSIeDVwNlBihuXKSp/T24DXpU2vB+5KvyPpxcBgRAyO3CedL+MbwB0R8ekqlvckcIqkkTkqF/PsiZuuiYhXAm8B/kHSMZJOAVYD74mIlwEdwPhzK1beUuCn6fd6U/xaOUDymipuH/m6JMcaAYiITSP1kMzP8vF0eVGedUkS8C3gxxHxooh4DfBJ4Bjgt6Oex+5a1TXjQgEgIu4Fvg18Avg7oDsifjuq289JpgiuN0uBfwe+DywBSN+lvR/4UETsB4iIRyPi2hrW9Z+kIZB+v5RnhsTPivo2A9cAD0TERTWo7WbgbentpcBVY3WKiB3Ab4ETgAuBz0XEr9J1QxHxlRrUCkD6ZuWNwPkkk03Vs58A5U8CbCNOB56KiH8daYiIe3jmPPU1NyNDIfUZ4F3AmcA/F6+Q1AScQT6zvs0tHjYCnx21/hzgapJ/bCPvIBcCv4uIx2pX5rP8jKdD4TSSd0AjU6u+niQ0RlwIHIiIlTWq7WrgXEnPAV4O3D5WJ0kvBF5IMiI4BbijRvWNZQnQExG/AQYlvSbHWkqS1EzyGtqUNs0dtdvjnBzLq3fj/Y29aNTz+KZaFVVXM6/VUkQ8Keka4ImRd9ekf9AkI4TNwPocStuXDnOBZJ8hya4L0n2yOyPid5L6gbWS6mUuwV8Cr0p3aR0SEU9IelDSQpJQKN4f/1Pg9ZJOTv/pVVVEbJS0gCREbx6jyznp8Y/9wH+PiF3JyD5XS4FV6e2r0+WK7jueopHXCiQjhSvS28/4+20Apc7Jz/tc/d8exPNY0VpnbCikhtOvEfsi4pXp7pjvkRxT+GIulY1tKfASSVvS5cOA/wZ8DThe0mF5jRYiYq+kB4C/Bu5Mm28D/hw4Gvh1UfdbgXXAdyW9MSK216DEG4F/ITlu0Dpq3TURsXxU233Aa4B7ql/aM6VB/1bgVElBcgwpgC/VupZxNNo//1IGgSNHtbVQmwvl3Qcc7AHtfZJmR8SBdLnitc7k3UclRcRe4MPAx9Lhce4kzQLeCZwaEQsiYgHJLoalab1XAKvSA7hIapP0lzUu8z+BlSTHY0i/rwBui1GfkoyIb5D8k+5RepZXla0FPhMRmybsmfg88ClJJ0Py/Ev6wAT3qZR3AP8eESekv+vjgId4enecVc4DwAvSEziQdALwCuDuGmz7h8AcSctGGiS9nPF/z7cA70n7ziX5n/CjShblUCghIu4CNlI/Z368CeiPiP9X1HYr8DJJxwL/CxgA7k9PW/wOUM1RwzxJ24q+PkpyXOGFPB0KdwLzeebxhEx64PZbwI3p/v6qiYhtEVH2qC8iNpIE3FWSNgP3kvxstbCU5Hkp9g2SM1P2P7t7XRl9TKH47KObiv5evp5bhUXSXcfvAf4t3R12HfA3EbGnBtsO4L8Ci9JTUu8D/hF4hGcfU/hwercVwF+ktd4GfD0ibq1kXb7MhVmDkLQEeHdEvDPvWmz6qotdI2Y2PkmfJdld+L6cS7FpziMFMzPL+JiCmZllHApmZpZxKJiZWcahYFZhKuOqummfdxUtd0iqpw9K2gzlUDDLxwKSa28BEBG9EfHh0t3NasOhYDNO+i79V5K+JmmzpOskzZN0hpJ5EzZJWitpTtp/i6R/Ttt/kV7P6VnX3R99rfuibf1EyRwXd0oauWjgJSSXP79b0kdUdF1/SS2Srpe0UdJt6adcUTIXxVolcxM8WPSBJrOKcSjYTPVi4MsR8VKST35/FLgSOCciTiX5DM8Hi/rvSdtXA5cdxHZ2AIvTOS7O4elraV0E/CS9Vv6lo+7zGeCuiHg58Cmg+Fr6LwH+jORKtBdLOuQgajGbkEPBZqqtETEyx8N/kFwq/aGiq7auA/6kqP9VRd9fR/kOAS6XtAn4OvCyMu7zRpI5M4iIHwKtkg5L190UEfsjYidJ4BxzELWYTcifaLaZavSnNnfz7Kunluo/crtA+sYqvWDh7DHu9xHgUZKLrM0C/jCJWosVX/toCL+GrcI8UrCZ6nhJI+/43wX0AgtGjhcA7yW5IuWIc4q+j1zwbwvJ5bUB3k4yKhjtcGB7RAynj9mUtj9O6elSfwK8G5K5mEnm0MhzAiWbQfwuw2aqXwMXSFoL3E9yqfTbgK+nl0v/JfCvRf2PlLSR5J36yJVzLwdukHQP0EMyH/RoXwa+oWTi9eI+G4Gh9L5XksxpPeLTJBMobQT2AudN7Uc1K5+vfWQzTjoL23fSSefL6b8F6Ej345tNa959ZGZmGY8UzMws45GCmZllHApmZpZxKJiZWcahYGZmGYeCmZll/j8x+84dFfS5AgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Look at lr-kallisto expression of these transcripts\n",
    "tid = 'transcript_345590'\n",
    "plot_t_exp(df, tid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7de03d85-96fc-4f64-a158-ff45aa5f05e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = expand(proc_cfg(config['lr']['kallisto']['quant']['merge_matrix_tpm_tsv'],od))[0]\n",
    "\n",
    "meta = load_meta()\n",
    "meta = meta.loc[meta.merged_run_mode==True]\n",
    "sample_d = dict([(entry.cell_line_id, entry['sample']) \\\n",
    "                 for ind, entry in meta.iterrows()])\n",
    "\n",
    "\n",
    "df = pd.read_csv(f, sep='\\t')\n",
    "df.columns = [d if d == 'transcript_id' else d.split('_')[0] for d in df.columns]\n",
    "df.rename({'transcript_id':'tid'}, axis=1, inplace=True)\n",
    "df.rename(sample_d, axis=1, inplace=True)\n",
    "# df.set_index('tid', inplace=True)\n",
    "df = df.melt(ignore_index=False, \n",
    "             id_vars=['tid'],\n",
    "             var_name='sample', \n",
    "             value_name='tpm')\n",
    "df = df.merge(meta[['sample', 'population']],\n",
    "              how='left',\n",
    "              on='sample')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "0fefc8a9-038c-43f0-90b7-ae5cf332ebd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>sample</th>\n",
       "      <th>tpm</th>\n",
       "      <th>population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000413811.3</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>YRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcript_6675</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>0.393218</td>\n",
       "      <td>YRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>transcript_6676</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>0.956060</td>\n",
       "      <td>YRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>transcript_6684</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>0.196609</td>\n",
       "      <td>YRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transcript_6687</td>\n",
       "      <td>YRI5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>YRI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tid sample       tpm population\n",
       "0  ENST00000413811.3   YRI5  0.000000        YRI\n",
       "1    transcript_6675   YRI5  0.393218        YRI\n",
       "2    transcript_6676   YRI5  0.956060        YRI\n",
       "3    transcript_6684   YRI5  0.196609        YRI\n",
       "4    transcript_6687   YRI5  0.000000        YRI"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "c7722637-dee5-4442-820c-64bf438c8d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_t_exp(df, tid):\n",
    "    c_dict, order = get_population_colors()\n",
    "    temp = df.loc[df.tid==tid].copy(deep=True)\n",
    "    ax = sns.boxplot(temp, y='tpm', x='population',\n",
    "                     palette=c_dict)\n",
    "    print(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daebb441-cf7c-4b87-bb0e-04c2a339191b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
