{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb1343de-f000-408f-a7b1-07f2a76f0b79",
   "metadata": {},
   "source": [
    "## Goal: Which long SIRVs do we detect? What does their coverage look like over the length of their bodies?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2beac5b7-1f14-461a-82c0-60ab5763f525",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import pyranges as pr\n",
    "from pyfaidx import Fasta\n",
    "from mizani.formatters import percent_format\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *\n",
    "\n",
    "from plotnine import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "25ff2a77-67d4-4f59-bb15-4701c02464f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def my_theme(base_size=11, w=4, h=3):\n",
    "    \"\"\"\n",
    "    Custom plotnine theme with:\n",
    "    - White background\n",
    "    - Clean styling\n",
    "    - Axes and ticks retained\n",
    "\n",
    "    Parameters:\n",
    "    - base_size: Base font size\n",
    "\n",
    "    Returns:\n",
    "    - plotnine.theme object\n",
    "    \"\"\"\n",
    "    return (\n",
    "        theme_minimal(base_size=base_size)\n",
    "        + theme(\n",
    "            # White background\n",
    "            panel_background=element_rect(fill='white', color=None),\n",
    "            plot_background=element_rect(fill='white', color=None),\n",
    "\n",
    "            # Remove grid lines\n",
    "            panel_grid_major=element_blank(),\n",
    "            panel_grid_minor=element_blank(),\n",
    "            panel_border=element_blank(),\n",
    "\n",
    "            # Keep axis lines & ticks (don't blank them)\n",
    "            axis_line=element_line(color='black'),\n",
    "            axis_ticks=element_line(color='black'),\n",
    "\n",
    "            plot_title=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_x=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_y=element_text(hjust=0.5, margin={'t':0, 'r':-2, 'b':0, 'l':0}, family='Helvetica'),\n",
    "            \n",
    "            # Styling text\n",
    "            legend_title=element_blank(),\n",
    "            axis_title=element_text(size=base_size + 1, family='Helvetica'),\n",
    "            legend_text=element_text(size=base_size-2, family='Helvetica'),\n",
    "            axis_text=element_text(size=base_size, color='black', family='Helvetica'),\n",
    "            figure_size=(w, h),  # Controls plot dimensions (width x height in inches)\n",
    "            plot_margin=0.05      # Shrinks surrounding white space\n",
    "        )\n",
    "    )\n",
    "\n",
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c748e1f9-6817-4fd9-848d-1658a2075e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "meta_df = load_meta()\n",
    "meta_df['lab_sample'] = meta_df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                        meta_df['lab_sampleid']+'_'+\\\n",
    "                        meta_df['cell_line_id']\n",
    "\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "54a80815-51d6-438d-97a0-7a345047dd8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ch/bc1lrgbd1fn_3gxz886mxjz00000gn/T/ipykernel_77584/1784241943.py:29: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'ERCC' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n"
     ]
    }
   ],
   "source": [
    "f = proc_cfg(expand(config['lr']['qc_sirvs']['sqanti_reads']['class_summary'],\n",
    "                    annot_completeness='C')[0], od)\n",
    "df = pd.read_csv(f, sep='\\t')\n",
    "\n",
    "# rename structural categories\n",
    "m = {'antisense': 'Antisense',\n",
    "     'full-splice_match': 'FSM',\n",
    "     'genic': 'Genic', \n",
    "     'incomplete-splice_match': 'ISM',\n",
    "     'intergenic': 'Intergenic',\n",
    "     'novel_in_catalog': 'NIC',\n",
    "     'novel_not_in_catalog': 'NNC'}\n",
    "df['structural_category'] = df.structural_category.map(m)\n",
    "\n",
    "# add antisense gene ids as assc_gene_2\n",
    "df['assc_gene_2'] = df.associated_gene\n",
    "inds = df.loc[df.structural_category=='Antisense'].index\n",
    "df.loc[inds, 'assc_gene_2'] = df.loc[inds, 'assc_gene_2'].str.split('_', expand=True)[1]\n",
    "\n",
    "# add relevant metadata\n",
    "df = df.merge(meta_df[['lab_sample', 'population']], \n",
    "              how='left', on='lab_sample')\n",
    "\n",
    "# add # mapped reads \n",
    "df['n_mapped_reads'] = df.groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "# add spike type\n",
    "df['spike_type'] = np.nan\n",
    "df.loc[df.chrom.str.contains('ERCC'), 'spike_type'] = 'ERCC'\n",
    "df.loc[df.chrom.str.contains('SIRV'), 'spike_type'] = 'SIRV'\n",
    "\n",
    "spliced_sirvs = ['SIRV1', 'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7']\n",
    "df.loc[df.chrom.isin(spliced_sirvs), 'spike_type_2'] = 'spliced_sirv'\n",
    "df.loc[~(df.chrom.isin(spliced_sirvs))&\\\n",
    "       (df.spike_type=='SIRV'), 'spike_type_2'] = 'long_sirv'\n",
    "\n",
    "\n",
    "\n",
    "# add # spliced sirv reads\n",
    "df['n_spliced_sirv_reads'] = df.loc[df.spike_type_2=='spliced_sirv'].groupby('lab_sample')['isoform'].transform('count')\n",
    "df['n_long_sirv_reads'] = df.loc[df.spike_type_2=='long_sirv'].groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "\n",
    "# splicing novelty\n",
    "df['splicing_novelty'] = 'Novel'\n",
    "df.loc[df.structural_category.isin(['FSM', 'ISM']), 'splicing_novelty'] = 'Known'\n",
    "\n",
    "# overal known vs. novel\n",
    "df['overall_nov'] = 'Novel'\n",
    "df.loc[df.structural_category == 'FSM', 'overall_nov'] = 'Known'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "50268dba-fe33-4864-9e22-614d3eab84b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "init_plot_settings()\n",
    "pop_c_dict, pop_order = get_population_colors()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e73202a-6d5f-43e8-b145-6cfcda1356e2",
   "metadata": {},
   "source": [
    "## Ground truth -- the sirv / ercc gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bd1d8a1f-30ae-475d-9fa4-0ca0d00e6fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = pr.read_gtf('../data/qc_sirvs/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf').df\n",
    "\n",
    "temp = gtf_df.loc[gtf_df.Chromosome.isin(spliced_sirvs)]\n",
    "n_spliced_sirvs_tot = len(temp.transcript_id.unique())\n",
    "\n",
    "# there are 69 spliced sirvs as ground truth\n",
    "\n",
    "# get the transcript length of each sirv transcript\n",
    "gtf_df['exon_len'] = gtf_df['End'] - gtf_df['Start']\n",
    "gtf_df['transcript_len'] = gtf_df.groupby('transcript_id')['exon_len'].transform('sum')\n",
    "\n",
    "gtf_df['mean_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('mean')\n",
    "gtf_df['med_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('median')\n",
    "\n",
    "# also get number of exons\n",
    "gtf_df['n_exons'] = gtf_df.groupby('transcript_id')['exon_assignment'].transform('count')\n",
    "gtf_df['mean_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('mean')\n",
    "gtf_df['med_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('median')\n",
    "\n",
    "# also get number of transcripts\n",
    "gtf_df['n_transcripts'] = gtf_df.groupby('gene_id')['transcript_id'].transform('nunique')\n",
    "\n",
    "# get # monoexonic transcripts\n",
    "temp = gtf_df[['gene_id', 'transcript_id', 'n_exons']].groupby(['gene_id', 'n_exons']).nunique().reset_index()\n",
    "temp = temp.loc[temp.n_exons==1]\n",
    "temp.drop('n_exons', axis=1, inplace=True)\n",
    "temp.rename({'transcript_id': 'n_monoexonic_transcripts'}, axis=1, inplace=True)\n",
    "temp.loc[temp.gene_id=='SIRV6']\n",
    "gtf_df = gtf_df.merge(temp,\n",
    "                      how='left',\n",
    "                      on='gene_id')\n",
    "\n",
    "# sort the long sirvs at least by length\n",
    "sirv_order = ['1', '2', '3', '4', '5', '6', '7',\n",
    "              '4001', '4002', '4003',\n",
    "              '6001', '6002', '6003', \n",
    "              '8001', '8002', '8003',\n",
    "              '10001', '10002', '10003',\n",
    "              '12001', '12002', '12003']\n",
    "sirv_order = [f'SIRV{s}' for s in sirv_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "28bbafdb-5dbb-46cb-80aa-7880da3d79a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7cb3a4b0-ea93-4928-9f09-ae568cf221f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only long sirvs\n",
    "temp = df.loc[df.spike_type_2 == 'long_sirv']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f81d7615-0157-4462-ad46-a2c9d5d93281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>chrom</th>\n",
       "      <th>length</th>\n",
       "      <th>diff_to_TSS</th>\n",
       "      <th>diff_to_TTS</th>\n",
       "      <th>diff_to_gene_TSS</th>\n",
       "      <th>diff_to_gene_TTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6856</th>\n",
       "      <td>35faaefb-8556-4c08-b4b3-8d463acf06a2:0</td>\n",
       "      <td>SIRV6001</td>\n",
       "      <td>1432</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6857</th>\n",
       "      <td>1597fa08-13f5-44fc-9a3e-8bc84b42fa35:0</td>\n",
       "      <td>SIRV6001</td>\n",
       "      <td>332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6858</th>\n",
       "      <td>25104590-1987-48b6-8536-eedd2e8e5806:0</td>\n",
       "      <td>SIRV6001</td>\n",
       "      <td>1419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859</th>\n",
       "      <td>a20319a3-52b2-4e33-acac-fc831726eb12:0</td>\n",
       "      <td>SIRV6001</td>\n",
       "      <td>537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6860</th>\n",
       "      <td>d97178f7-ef5d-420a-8b43-fc37e9cdcb7d:0</td>\n",
       "      <td>SIRV6001</td>\n",
       "      <td>857</td>\n",
       "      <td>-5141.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5141.0</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     isoform     chrom  length  diff_to_TSS  \\\n",
       "6856  35faaefb-8556-4c08-b4b3-8d463acf06a2:0  SIRV6001    1432          NaN   \n",
       "6857  1597fa08-13f5-44fc-9a3e-8bc84b42fa35:0  SIRV6001     332          NaN   \n",
       "6858  25104590-1987-48b6-8536-eedd2e8e5806:0  SIRV6001    1419          NaN   \n",
       "6859  a20319a3-52b2-4e33-acac-fc831726eb12:0  SIRV6001     537          NaN   \n",
       "6860  d97178f7-ef5d-420a-8b43-fc37e9cdcb7d:0  SIRV6001     857      -5141.0   \n",
       "\n",
       "      diff_to_TTS  diff_to_gene_TSS  diff_to_gene_TTS  \n",
       "6856          NaN               NaN               NaN  \n",
       "6857          NaN               NaN               NaN  \n",
       "6858          NaN               NaN               NaN  \n",
       "6859          NaN               NaN               NaN  \n",
       "6860         -2.0           -5141.0              -2.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[['isoform', 'chrom', 'length', 'diff_to_TSS', 'diff_to_TTS', 'diff_to_gene_TSS', 'diff_to_gene_TTS']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5698a51f-fbff-49ed-92c4-4d485011e963",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_assignment</th>\n",
       "      <th>exon_len</th>\n",
       "      <th>transcript_len</th>\n",
       "      <th>mean_transcript_len</th>\n",
       "      <th>med_transcript_len</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>mean_n_exons</th>\n",
       "      <th>med_n_exons</th>\n",
       "      <th>n_transcripts</th>\n",
       "      <th>n_monoexonic_transcripts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>SIRV10002</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>10001</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SIRV10002</td>\n",
       "      <td>SIRV10002</td>\n",
       "      <td>...</td>\n",
       "      <td>SIRV10002_0</td>\n",
       "      <td>10001</td>\n",
       "      <td>10001</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>SIRV10003</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>9970</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SIRV10003</td>\n",
       "      <td>SIRV10003</td>\n",
       "      <td>...</td>\n",
       "      <td>SIRV10003_0</td>\n",
       "      <td>9970</td>\n",
       "      <td>9970</td>\n",
       "      <td>9970.0</td>\n",
       "      <td>9970.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>SIRV12001</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>12029</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SIRV12001</td>\n",
       "      <td>SIRV12001</td>\n",
       "      <td>...</td>\n",
       "      <td>SIRV12001_0</td>\n",
       "      <td>12029</td>\n",
       "      <td>12029</td>\n",
       "      <td>12029.0</td>\n",
       "      <td>12029.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>SIRV12002</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>11999</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SIRV12002</td>\n",
       "      <td>SIRV12002</td>\n",
       "      <td>...</td>\n",
       "      <td>SIRV12002_0</td>\n",
       "      <td>11999</td>\n",
       "      <td>11999</td>\n",
       "      <td>11999.0</td>\n",
       "      <td>11999.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>SIRV12003</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>12000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SIRV12003</td>\n",
       "      <td>SIRV12003</td>\n",
       "      <td>...</td>\n",
       "      <td>SIRV12003_0</td>\n",
       "      <td>12000</td>\n",
       "      <td>12000</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>12000.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome           Source Feature  Start    End Score Strand Frame  \\\n",
       "459  SIRV10002  LexogenSIRVData    exon      0  10001     .      +     0   \n",
       "460  SIRV10003  LexogenSIRVData    exon      0   9970     .      +     0   \n",
       "461  SIRV12001  LexogenSIRVData    exon      0  12029     .      +     0   \n",
       "462  SIRV12002  LexogenSIRVData    exon      0  11999     .      +     0   \n",
       "463  SIRV12003  LexogenSIRVData    exon      0  12000     .      +     0   \n",
       "\n",
       "       gene_id transcript_id  ... exon_assignment exon_len  transcript_len  \\\n",
       "459  SIRV10002     SIRV10002  ...     SIRV10002_0    10001           10001   \n",
       "460  SIRV10003     SIRV10003  ...     SIRV10003_0     9970            9970   \n",
       "461  SIRV12001     SIRV12001  ...     SIRV12001_0    12029           12029   \n",
       "462  SIRV12002     SIRV12002  ...     SIRV12002_0    11999           11999   \n",
       "463  SIRV12003     SIRV12003  ...     SIRV12003_0    12000           12000   \n",
       "\n",
       "     mean_transcript_len  med_transcript_len  n_exons  mean_n_exons  \\\n",
       "459              10001.0             10001.0        1           1.0   \n",
       "460               9970.0              9970.0        1           1.0   \n",
       "461              12029.0             12029.0        1           1.0   \n",
       "462              11999.0             11999.0        1           1.0   \n",
       "463              12000.0             12000.0        1           1.0   \n",
       "\n",
       "     med_n_exons  n_transcripts  n_monoexonic_transcripts  \n",
       "459          1.0              1                       1.0  \n",
       "460          1.0              1                       1.0  \n",
       "461          1.0              1                       1.0  \n",
       "462          1.0              1                       1.0  \n",
       "463          1.0              1                       1.0  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c86e6325-30c4-4067-b25d-33423dddbe4f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
