{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9af7b21b-8662-48e5-b32a-b7a56a2a96bc",
   "metadata": {},
   "source": [
    "## Goal: We want to know to what novel exons appear because they have at least one highly-divegent variant position in them \n",
    "We'll limit this analyses to the contrasts that have CEU in them, because we know that the annotation is European-biased, and therefore the novelty of exons is dependent on this bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86a1e5c0-51cb-4904-958d-522e47da63eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "import math\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "58f7c437-966b-4fbd-b996-388fe0d9e589",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8eb463c1-a845-4e27-8dbd-a0ac5e1d8573",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2943235\n",
      "1719222\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(proc_cfg(config['lr']['exons_ss_in_pops_fsts_summary'], od),\n",
    "                 sep='\\t')\n",
    "df.rename({'pop2_der':'pop2_det'}, \n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# set all <0 fsts --> 0 https://www.biostars.org/p/132253/\n",
    "df.loc[df.fst<0, 'fst'] = 0\n",
    "\n",
    "# since this is biallelic snp level, remove all exons that didnt intersect\n",
    "print(len(df.index))\n",
    "df = df.loc[df.End_b!=-1]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "823b843a-0c68-41b2-a026-c672dd04bff0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>novelty</th>\n",
       "      <th>pop1_det</th>\n",
       "      <th>pop2_det</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>pop1</th>\n",
       "      <th>pop2</th>\n",
       "      <th>End_b</th>\n",
       "      <th>fst</th>\n",
       "      <th>Start_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chr1_+_100007033_100007156</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100007031</td>\n",
       "      <td>100007158</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>100007048</td>\n",
       "      <td>0.393014</td>\n",
       "      <td>100007047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chr1_+_100043072_100043229</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100043070</td>\n",
       "      <td>100043231</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>100043099</td>\n",
       "      <td>0.041198</td>\n",
       "      <td>100043098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr1_+_100043072_100043229</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100043070</td>\n",
       "      <td>100043231</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>100043213</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100043212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>chr1_+_100049908_100050004</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100049906</td>\n",
       "      <td>100050006</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>100049941</td>\n",
       "      <td>0.018298</td>\n",
       "      <td>100049940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>chr1_+_100140407_100141019</td>\n",
       "      <td>Known</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100140405</td>\n",
       "      <td>100141021</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>100140797</td>\n",
       "      <td>0.047965</td>\n",
       "      <td>100140796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            eid novelty  pop1_det  pop2_det Chromosome Strand  \\\n",
       "197  chr1_+_100007033_100007156   Known      True      True       chr1      +   \n",
       "198  chr1_+_100043072_100043229   Known      True      True       chr1      +   \n",
       "199  chr1_+_100043072_100043229   Known      True      True       chr1      +   \n",
       "200  chr1_+_100049908_100050004   Known      True      True       chr1      +   \n",
       "201  chr1_+_100140407_100141019   Known      True      True       chr1      +   \n",
       "\n",
       "         Start        End pop1 pop2      End_b       fst    Start_b  \n",
       "197  100007031  100007158  HAC  YRI  100007048  0.393014  100007047  \n",
       "198  100043070  100043231  HAC  YRI  100043099  0.041198  100043098  \n",
       "199  100043070  100043231  HAC  YRI  100043213  0.000000  100043212  \n",
       "200  100049906  100050006  HAC  YRI  100049941  0.018298  100049940  \n",
       "201  100140405  100141021  HAC  YRI  100140797  0.047965  100140796  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## For each exon / contrast, get the max fst\n",
    "df = df.sort_values(by=\n",
    "df = df.drop_duplicates(subset=['eid', 'pop1', 'pop2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83057277-2a00-4c44-88ac-c26f63cf83ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
