{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a2cf89b-e3b1-4b59-a898-e9579c497c25",
   "metadata": {},
   "source": [
    "## Goal: What are Fst scores per exon + ss (+-2bp)\n",
    "For exons that are \n",
    "* unique to one population or the other\n",
    "* shared between the population\n",
    "* this is on the exon level~!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf872b85-ea3a-44a1-ab69-584420bd80e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "36758246-51a9-41cc-8a8f-a4a8282b69fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "412f99a4-f85d-4990-9841-aac489265fff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2943235\n",
      "1719222\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>novelty</th>\n",
       "      <th>pop1_det</th>\n",
       "      <th>pop2_det</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>End_b</th>\n",
       "      <th>fst</th>\n",
       "      <th>Start_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chr1_+_100007033_100007156</td>\n",
       "      <td>Known</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100007031</td>\n",
       "      <td>100007158</td>\n",
       "      <td>100007048</td>\n",
       "      <td>0.393014</td>\n",
       "      <td>100007047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chr1_+_100043072_100043229</td>\n",
       "      <td>Known</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100043070</td>\n",
       "      <td>100043231</td>\n",
       "      <td>100043099</td>\n",
       "      <td>0.041198</td>\n",
       "      <td>100043098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr1_+_100043072_100043229</td>\n",
       "      <td>Known</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100043070</td>\n",
       "      <td>100043231</td>\n",
       "      <td>100043213</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100043212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>chr1_+_100049908_100050004</td>\n",
       "      <td>Known</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100049906</td>\n",
       "      <td>100050006</td>\n",
       "      <td>100049941</td>\n",
       "      <td>0.018298</td>\n",
       "      <td>100049940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>chr1_+_100140407_100141019</td>\n",
       "      <td>Known</td>\n",
       "      <td>HAC</td>\n",
       "      <td>YRI</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>100140405</td>\n",
       "      <td>100141021</td>\n",
       "      <td>100140797</td>\n",
       "      <td>0.047965</td>\n",
       "      <td>100140796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            eid novelty pop1_det pop2_det Chromosome Strand  \\\n",
       "197  chr1_+_100007033_100007156   Known      HAC      YRI       chr1      +   \n",
       "198  chr1_+_100043072_100043229   Known      HAC      YRI       chr1      +   \n",
       "199  chr1_+_100043072_100043229   Known      HAC      YRI       chr1      +   \n",
       "200  chr1_+_100049908_100050004   Known      HAC      YRI       chr1      +   \n",
       "201  chr1_+_100140407_100141019   Known      HAC      YRI       chr1      +   \n",
       "\n",
       "         Start        End      End_b       fst    Start_b  \n",
       "197  100007031  100007158  100007048  0.393014  100007047  \n",
       "198  100043070  100043231  100043099  0.041198  100043098  \n",
       "199  100043070  100043231  100043213  0.000000  100043212  \n",
       "200  100049906  100050006  100049941  0.018298  100049940  \n",
       "201  100140405  100141021  100140797  0.047965  100140796  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(proc_cfg(config['lr']['exons_ss_in_pops_fsts_summary'], od),\n",
    "                 sep='\\t')\n",
    "# df.rename({'ThickStart':'eid',\n",
    "#            'ThickEnd':'novelty',\n",
    "#            'ItemRGB':'pop1_det',\n",
    "#            'BlockCount':'pop2_det'}, \n",
    "#           axis=1, inplace=True)\n",
    "\n",
    "# # set all <0 fsts --> 0 https://www.biostars.org/p/132253/\n",
    "# df.loc[df.fst<0, 'fst'] = 0\n",
    "\n",
    "# # since this is biallelic snp level, remove all exons that didnt intersect\n",
    "# print(len(df.index))\n",
    "# df = df.loc[df.End_b!=-1]\n",
    "# print(len(df.index))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba32da25-1b94-452c-b5e3-eb4dec104f59",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['pop1', 'pop2'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-395fa8fdec02>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m# make sure there are no dupe eid + contrast + var. posiition\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgb_cols\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'End_b'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Start_b'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_duplicates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# oop there's a bug somewhere upstream w/ the vcf processing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'REMOVE ME WHEN FABIEN FIXES FST STUFF'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3462\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3463\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3464\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3465\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3466\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1312\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1313\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1314\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_read_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1315\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1316\u001b[0m         if needs_i8_conversion(ax.dtype) or isinstance(\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis)\u001b[0m\n\u001b[1;32m   1375\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1376\u001b[0m             \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmissing_mask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1377\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1378\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1379\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['pop1', 'pop2'] not in index\""
     ]
    }
   ],
   "source": [
    "gb_cols = ['eid', 'novelty', 'pop1_det', 'pop2_det',\n",
    "            'pop1', 'pop2']\n",
    "keep_cols = gb_cols+['fst']\n",
    "\n",
    "\n",
    "# make sure there are no dupe eid + contrast + var. posiition\n",
    "# assert len(df[gb_cols+['End_b', 'Start_b']].drop_duplicates().index) == len(df.index) # oop there's a bug somewhere upstream w/ the vcf processing\n",
    "\n",
    "print('REMOVE ME WHEN FABIEN FIXES FST STUFF')\n",
    "print(len(df.index))\n",
    "df = df.drop_duplicates(subset=gb_cols+['End_b', 'Start_b'], keep=False)\n",
    "print(len(df.index))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a56f8a1e-1068-44d1-ad54-c63af7a6b61f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop through each contrast\n",
    "stat_df = pd.DataFrame()\n",
    "for pop1, pop2 in zip(df[['pop1', 'pop2']].drop_duplicates()['pop1'].tolist(),\n",
    "                      df[['pop1', 'pop2']].drop_duplicates()['pop2'].tolist()):\n",
    "    temp = df.loc[(df.pop1==pop1)&(df.pop2==pop2)]\n",
    "    temp.rename({'pop1_det':pop1,\n",
    "             'pop2_det':pop2}, axis=1, inplace=True)\n",
    "\n",
    "    temp['med_per_group'] = temp.groupby([pop1, pop2])['fst'].transform('median')\n",
    "    temp['mean_per_group'] = temp.groupby([pop1, pop2])['fst'].transform('mean')\n",
    "\n",
    "    # print(temp[[pop1, pop2, 'med_per_group']].drop_duplicates())\n",
    "    # print(temp[[pop1, pop2, 'mean_per_group']].drop_duplicates())\n",
    "\n",
    "    temp['thing'] = ''\n",
    "    temp.loc[(temp[pop1]==True)&(temp[pop2]==True), 'thing'] = f'{pop1}_{pop2}'\n",
    "    temp.loc[(temp[pop1]==True)&(temp[pop2]==False), 'thing'] = f'{pop1}'\n",
    "    temp.loc[(temp[pop1]==False)&(temp[pop2]==True), 'thing'] = f'{pop2}'     \n",
    "\n",
    "    temp['thing2'] = 'unique'\n",
    "    temp.loc[(temp[pop1]==True)&(temp[pop2]==True), 'thing2'] = 'shared'\n",
    "    \n",
    "    x = temp.loc[temp.thing2=='unique','fst'].tolist() # hyp. --> greater\n",
    "    y = temp.loc[temp.thing2=='shared','fst'].tolist()\n",
    "    \n",
    "    print(f'{pop1} vs {pop2}')\n",
    "    print(temp.loc[temp.thing2=='unique','fst'].median())\n",
    "    print(temp.loc[temp.thing2=='shared','fst'].median())\n",
    "    print()\n",
    "\n",
    "    stat, pval = st.mannwhitneyu(x, y, alternative='greater')\n",
    "    temp3 = pd.DataFrame(data=[[pop1, pop2, stat, pval]], \n",
    "                         columns=['pop1', 'pop2', 'stat', 'pval'])\n",
    "    stat_df = pd.concat([stat_df, temp3], axis=0)\n",
    "\n",
    "# fdr correction\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "_, adj_p_vals, _, _ = multipletests(stat_df.pval.tolist(), method='fdr_bh')\n",
    "stat_df['adj_pval'] = adj_p_vals\n",
    "stat_df.head()\n",
    "\n",
    "# add sig. annotation\n",
    "# * <= 0.05\n",
    "# ** <= 0.01\n",
    "# *** <= 0.001\n",
    "stat_df['sig_annot'] = ''\n",
    "# stat_df.loc[stat_df.pval <= 0.05, 'sig_annot'] = '*'\n",
    "# stat_df.loc[stat_df.pval <= 0.01, 'sig_annot'] = '**'\n",
    "# stat_df.loc[stat_df.pval <= 0.001, 'sig_annot'] = '***'\n",
    "stat_df.loc[stat_df.adj_pval <= 0.05, 'sig_annot'] = '*'\n",
    "stat_df.loc[stat_df.adj_pval <= 0.01, 'sig_annot'] = '**'\n",
    "stat_df.loc[stat_df.adj_pval <= 0.001, 'sig_annot'] = '***'\n",
    "\n",
    "# pivot x2 to get a stat and annot df to plot\n",
    "stat_df1 = stat_df.pivot(index='pop1', columns='pop2', values='stat')\n",
    "annot_df = stat_df.pivot(index='pop1', columns='pop2', values='sig_annot')\n",
    "\n",
    "ax = sns.heatmap(stat_df1, cmap='Purples',\n",
    "            annot=annot_df,\n",
    "            fmt='',\n",
    "            cbar_kws={'label': 'Mann-Whitney U\\ntest statistic'})\n",
    "\n",
    "# labels need them\n",
    "ax.set(xlabel='Population 2',\n",
    "       ylabel='Popluation 1',\n",
    "       title='Fsts of exonic, biallelic SNPs based on\\n exon detection')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "164b31cf-067f-42d8-84dd-1c187af2400e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
