{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f11551d-36bc-49b2-b561-144d4bbd2c11",
   "metadata": {},
   "source": [
    "## Goal -- how does our precision, sensitivity, and accuracy get impacted when applying the PODER filters? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7eba1c60-3a06-4e91-8ebc-f1eee4cfedbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import pyranges as pr\n",
    "from pyfaidx import Fasta\n",
    "from mizani.formatters import percent_format\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *\n",
    "\n",
    "from plotnine import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a2318ad-9bc9-406f-bbc2-63d38f3f205c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def my_theme(base_size=11, w=4, h=3):\n",
    "    \"\"\"\n",
    "    Custom plotnine theme with:\n",
    "    - White background\n",
    "    - Clean styling\n",
    "    - Axes and ticks retained\n",
    "\n",
    "    Parameters:\n",
    "    - base_size: Base font size\n",
    "\n",
    "    Returns:\n",
    "    - plotnine.theme object\n",
    "    \"\"\"\n",
    "    return (\n",
    "        theme_minimal(base_size=base_size)\n",
    "        + theme(\n",
    "            # White background\n",
    "            panel_background=element_rect(fill='white', color=None),\n",
    "            plot_background=element_rect(fill='white', color=None),\n",
    "\n",
    "            # Remove grid lines\n",
    "            panel_grid_major=element_blank(),\n",
    "            panel_grid_minor=element_blank(),\n",
    "            panel_border=element_blank(),\n",
    "\n",
    "            # Keep axis lines & ticks (don't blank them)\n",
    "            axis_line=element_line(color='black'),\n",
    "            axis_ticks=element_line(color='black'),\n",
    "\n",
    "            plot_title=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_x=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_y=element_text(hjust=0.5, margin={'t':0, 'r':-2, 'b':0, 'l':0}, family='Helvetica'),\n",
    "            \n",
    "            # Styling text\n",
    "            legend_title=element_blank(),\n",
    "            axis_title=element_text(size=base_size + 1, family='Helvetica'),\n",
    "            legend_text=element_text(size=base_size-2, family='Helvetica'),\n",
    "            axis_text=element_text(size=base_size, color='black', family='Helvetica'),\n",
    "            figure_size=(w, h),  # Controls plot dimensions (width x height in inches)\n",
    "            plot_margin=0.05      # Shrinks surrounding white space\n",
    "        )\n",
    "    )\n",
    "\n",
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d07f3929-edfa-4e08-aaaf-52dd07202e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "meta_df = load_meta()\n",
    "meta_df['lab_sample'] = meta_df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                        meta_df['lab_sampleid']+'_'+\\\n",
    "                        meta_df['cell_line_id']\n",
    "\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry\n",
    "\n",
    "init_plot_settings()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81597fa8-fa95-4809-9ea4-82ca9045fa26",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ch/bc1lrgbd1fn_3gxz886mxjz00000gn/T/ipykernel_81330/2818831187.py:30: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'ERCC' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n"
     ]
    }
   ],
   "source": [
    "## SQANTI reads stuff\n",
    "f = proc_cfg(expand(config['lr']['qc_sirvs']['sqanti_reads']['class_summary'],\n",
    "                    annot_completeness='C')[0], od)\n",
    "df = pd.read_csv(f, sep='\\t')\n",
    "\n",
    "# rename structural categories\n",
    "m = {'antisense': 'Antisense',\n",
    "     'full-splice_match': 'FSM',\n",
    "     'genic': 'Genic', \n",
    "     'incomplete-splice_match': 'ISM',\n",
    "     'intergenic': 'Intergenic',\n",
    "     'novel_in_catalog': 'NIC',\n",
    "     'novel_not_in_catalog': 'NNC'}\n",
    "df['structural_category'] = df.structural_category.map(m)\n",
    "\n",
    "# add antisense gene ids as assc_gene_2\n",
    "df['assc_gene_2'] = df.associated_gene\n",
    "inds = df.loc[df.structural_category=='Antisense'].index\n",
    "df.loc[inds, 'assc_gene_2'] = df.loc[inds, 'assc_gene_2'].str.split('_', expand=True)[1]\n",
    "\n",
    "# add relevant metadata\n",
    "df = df.merge(meta_df[['lab_sample', 'population']], \n",
    "              how='left', on='lab_sample')\n",
    "\n",
    "# add # mapped reads \n",
    "df['n_mapped_reads'] = df.groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "# add spike type\n",
    "df['spike_type'] = np.nan\n",
    "df.loc[df.chrom.str.contains('ERCC'), 'spike_type'] = 'ERCC'\n",
    "df.loc[df.chrom.str.contains('SIRV'), 'spike_type'] = 'SIRV'\n",
    "\n",
    "spliced_sirvs = ['SIRV1', 'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7']\n",
    "df.loc[df.chrom.isin(spliced_sirvs), 'spike_type_2'] = 'spliced_sirv'\n",
    "\n",
    "# add # spliced sirv reads\n",
    "df['n_spliced_sirv_reads'] = df.loc[df.spike_type_2=='spliced_sirv'].groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "# splicing novelty\n",
    "df['splicing_novelty'] = 'Novel'\n",
    "df.loc[df.structural_category.isin(['FSM', 'ISM']), 'splicing_novelty'] = 'Known'\n",
    "\n",
    "# overal known vs. novel\n",
    "df['overall_nov'] = 'Novel'\n",
    "df.loc[df.structural_category == 'FSM', 'overall_nov'] = 'Known'\n",
    "\n",
    "## Ground truth sirv / ercc gtf\n",
    "gtf_df = pr.read_gtf('../data/qc_sirvs/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf').df\n",
    "\n",
    "temp = gtf_df.loc[gtf_df.Chromosome.isin(spliced_sirvs)]\n",
    "n_spliced_sirvs_tot = len(temp.transcript_id.unique())\n",
    "\n",
    "# there are 69 spliced sirvs as ground truth\n",
    "\n",
    "# get the transcript length of each sirv transcript\n",
    "gtf_df['exon_len'] = gtf_df['End'] - gtf_df['Start']\n",
    "gtf_df['transcript_len'] = gtf_df.groupby('transcript_id')['exon_len'].transform('sum')\n",
    "\n",
    "gtf_df['mean_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('mean')\n",
    "gtf_df['med_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('median')\n",
    "\n",
    "# also get number of exons\n",
    "gtf_df['n_exons'] = gtf_df.groupby('transcript_id')['exon_assignment'].transform('count')\n",
    "gtf_df['mean_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('mean')\n",
    "gtf_df['med_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('median')\n",
    "\n",
    "# also get number of transcripts\n",
    "gtf_df['n_transcripts'] = gtf_df.groupby('gene_id')['transcript_id'].transform('nunique')\n",
    "\n",
    "# get # monoexonic transcripts\n",
    "temp = gtf_df[['gene_id', 'transcript_id', 'n_exons']].groupby(['gene_id', 'n_exons']).nunique().reset_index()\n",
    "temp = temp.loc[temp.n_exons==1]\n",
    "temp.drop('n_exons', axis=1, inplace=True)\n",
    "temp.rename({'transcript_id': 'n_monoexonic_transcripts'}, axis=1, inplace=True)\n",
    "temp.loc[temp.gene_id=='SIRV6']\n",
    "gtf_df = gtf_df.merge(temp,\n",
    "                      how='left',\n",
    "                      on='gene_id')\n",
    "\n",
    "# sort the long sirvs at least by length\n",
    "sirv_order = ['1', '2', '3', '4', '5', '6', '7',\n",
    "              '4001', '4002', '4003',\n",
    "              '6001', '6002', '6003', \n",
    "              '8001', '8002', '8003',\n",
    "              '10001', '10002', '10003',\n",
    "              '12001', '12002', '12003']\n",
    "sirv_order = [f'SIRV{s}' for s in sirv_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc3dd3a9-447d-48ff-994a-1a6afc017ff3",
   "metadata": {},
   "source": [
    "## Apply filters one by one and create a label to say whether the transcript passed each filter \n",
    "* FSM? (pass_fsm_filter)\n",
    "* Monoexonic? (pass_monoexonic_filter)\n",
    "* Reproducibile for novel transcripts? (pass_reproducibility_filter)\n",
    "* Promoted ISM? (promoted_ism_filter)\n",
    "* ISM? (pass_ism_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b08a8700-7acd-40a3-96fb-bbf4358ca23c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, limit to spliced sirvs\n",
    "temp = df.loc[df.spike_type_2=='spliced_sirv']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "770c9534-dda6-48cd-aa1d-c4cca7f6cab5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>exons</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>ref_length</th>\n",
       "      <th>ref_exons</th>\n",
       "      <th>...</th>\n",
       "      <th>jxnHash</th>\n",
       "      <th>lab_sample</th>\n",
       "      <th>assc_gene_2</th>\n",
       "      <th>population</th>\n",
       "      <th>n_mapped_reads</th>\n",
       "      <th>spike_type</th>\n",
       "      <th>spike_type_2</th>\n",
       "      <th>n_spliced_sirv_reads</th>\n",
       "      <th>splicing_novelty</th>\n",
       "      <th>overall_nov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3904</th>\n",
       "      <td>0a5f1630-d746-49b6-abba-c6adfa5295d7:0</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>1362</td>\n",
       "      <td>6</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>SIRV103</td>\n",
       "      <td>1363.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>277f49c4a3633bf1591c1d77d02bc4c42e246ca47b54e8...</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>ITU</td>\n",
       "      <td>6922</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3907</th>\n",
       "      <td>0c85d074-a84a-4e9f-af13-098effd9fbbe:0</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>462</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>SIRV109</td>\n",
       "      <td>464.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7736df0b7910af28a57a785c0ef320241b6b369bbe5164...</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>ITU</td>\n",
       "      <td>6922</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3908</th>\n",
       "      <td>0c8473b2-7937-4b90-be63-201092c8eba9:0</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>1359</td>\n",
       "      <td>6</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>SIRV103</td>\n",
       "      <td>1363.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>277f49c4a3633bf1591c1d77d02bc4c42e246ca47b54e8...</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>ITU</td>\n",
       "      <td>6922</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3909</th>\n",
       "      <td>0dd9fbea-d401-4926-84f2-27527783e934:0</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>667</td>\n",
       "      <td>5</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>SIRV105</td>\n",
       "      <td>670.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3a1b1eddd311a31572338f49a0d58dcac043e8f71f1f11...</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>ITU</td>\n",
       "      <td>6922</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3912</th>\n",
       "      <td>1b0f17ac-35a9-445b-aed7-0d6dde421d3f:0</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>463</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>SIRV109</td>\n",
       "      <td>464.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7736df0b7910af28a57a785c0ef320241b6b369bbe5164...</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>SIRV1</td>\n",
       "      <td>ITU</td>\n",
       "      <td>6922</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427772</th>\n",
       "      <td>f4e4f36d-7d20-43fb-a440-640c62efc29a:0</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>-</td>\n",
       "      <td>412</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>SIRV704</td>\n",
       "      <td>428.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...</td>\n",
       "      <td>1_PY1_GM10492</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>MPC</td>\n",
       "      <td>6914</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427774</th>\n",
       "      <td>f2195ba7-81d3-41c4-83a9-f9f0b54b8949:0</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>-</td>\n",
       "      <td>423</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>SIRV704</td>\n",
       "      <td>428.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...</td>\n",
       "      <td>1_PY1_GM10492</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>MPC</td>\n",
       "      <td>6914</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427775</th>\n",
       "      <td>f800451c-e3d9-4e9c-bc1d-9a5d3186692f:0_2</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>-</td>\n",
       "      <td>426</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>SIRV704</td>\n",
       "      <td>428.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...</td>\n",
       "      <td>1_PY1_GM10492</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>MPC</td>\n",
       "      <td>6914</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427776</th>\n",
       "      <td>feb40ebf-8075-4884-9c5e-b594e56d9f48:0</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>-</td>\n",
       "      <td>412</td>\n",
       "      <td>3</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>SIRV704</td>\n",
       "      <td>428.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...</td>\n",
       "      <td>1_PY1_GM10492</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>MPC</td>\n",
       "      <td>6914</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427777</th>\n",
       "      <td>fec1346c-35c3-48f6-9ab8-2b467fc823f1:0</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>-</td>\n",
       "      <td>881</td>\n",
       "      <td>6</td>\n",
       "      <td>FSM</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>SIRV708</td>\n",
       "      <td>889.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8078adfe100760ab2c521a170f02b4642736a10122cbfc...</td>\n",
       "      <td>1_PY1_GM10492</td>\n",
       "      <td>SIRV7</td>\n",
       "      <td>MPC</td>\n",
       "      <td>6914</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144567 rows Ã— 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         isoform  chrom strand  length  exons  \\\n",
       "3904      0a5f1630-d746-49b6-abba-c6adfa5295d7:0  SIRV1      -    1362      6   \n",
       "3907      0c85d074-a84a-4e9f-af13-098effd9fbbe:0  SIRV1      +     462      3   \n",
       "3908      0c8473b2-7937-4b90-be63-201092c8eba9:0  SIRV1      -    1359      6   \n",
       "3909      0dd9fbea-d401-4926-84f2-27527783e934:0  SIRV1      -     667      5   \n",
       "3912      1b0f17ac-35a9-445b-aed7-0d6dde421d3f:0  SIRV1      +     463      3   \n",
       "...                                          ...    ...    ...     ...    ...   \n",
       "427772    f4e4f36d-7d20-43fb-a440-640c62efc29a:0  SIRV7      -     412      3   \n",
       "427774    f2195ba7-81d3-41c4-83a9-f9f0b54b8949:0  SIRV7      -     423      3   \n",
       "427775  f800451c-e3d9-4e9c-bc1d-9a5d3186692f:0_2  SIRV7      -     426      3   \n",
       "427776    feb40ebf-8075-4884-9c5e-b594e56d9f48:0  SIRV7      -     412      3   \n",
       "427777    fec1346c-35c3-48f6-9ab8-2b467fc823f1:0  SIRV7      -     881      6   \n",
       "\n",
       "       structural_category associated_gene associated_transcript  ref_length  \\\n",
       "3904                   FSM           SIRV1               SIRV103      1363.0   \n",
       "3907                   FSM           SIRV1               SIRV109       464.0   \n",
       "3908                   FSM           SIRV1               SIRV103      1363.0   \n",
       "3909                   FSM           SIRV1               SIRV105       670.0   \n",
       "3912                   FSM           SIRV1               SIRV109       464.0   \n",
       "...                    ...             ...                   ...         ...   \n",
       "427772                 FSM           SIRV7               SIRV704       428.0   \n",
       "427774                 FSM           SIRV7               SIRV704       428.0   \n",
       "427775                 FSM           SIRV7               SIRV704       428.0   \n",
       "427776                 FSM           SIRV7               SIRV704       428.0   \n",
       "427777                 FSM           SIRV7               SIRV708       889.0   \n",
       "\n",
       "        ref_exons  ...                                            jxnHash  \\\n",
       "3904          6.0  ...  277f49c4a3633bf1591c1d77d02bc4c42e246ca47b54e8...   \n",
       "3907          3.0  ...  7736df0b7910af28a57a785c0ef320241b6b369bbe5164...   \n",
       "3908          6.0  ...  277f49c4a3633bf1591c1d77d02bc4c42e246ca47b54e8...   \n",
       "3909          5.0  ...  3a1b1eddd311a31572338f49a0d58dcac043e8f71f1f11...   \n",
       "3912          3.0  ...  7736df0b7910af28a57a785c0ef320241b6b369bbe5164...   \n",
       "...           ...  ...                                                ...   \n",
       "427772        3.0  ...  7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...   \n",
       "427774        3.0  ...  7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...   \n",
       "427775        3.0  ...  7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...   \n",
       "427776        3.0  ...  7a0922befdaba1e23a5e78062a7794f5d5c5d892273509...   \n",
       "427777        6.0  ...  8078adfe100760ab2c521a170f02b4642736a10122cbfc...   \n",
       "\n",
       "            lab_sample  assc_gene_2  population n_mapped_reads  spike_type  \\\n",
       "3904    39_IN5_HG04217        SIRV1         ITU           6922        SIRV   \n",
       "3907    39_IN5_HG04217        SIRV1         ITU           6922        SIRV   \n",
       "3908    39_IN5_HG04217        SIRV1         ITU           6922        SIRV   \n",
       "3909    39_IN5_HG04217        SIRV1         ITU           6922        SIRV   \n",
       "3912    39_IN5_HG04217        SIRV1         ITU           6922        SIRV   \n",
       "...                ...          ...         ...            ...         ...   \n",
       "427772   1_PY1_GM10492        SIRV7         MPC           6914        SIRV   \n",
       "427774   1_PY1_GM10492        SIRV7         MPC           6914        SIRV   \n",
       "427775   1_PY1_GM10492        SIRV7         MPC           6914        SIRV   \n",
       "427776   1_PY1_GM10492        SIRV7         MPC           6914        SIRV   \n",
       "427777   1_PY1_GM10492        SIRV7         MPC           6914        SIRV   \n",
       "\n",
       "        spike_type_2  n_spliced_sirv_reads  splicing_novelty  overall_nov  \n",
       "3904    spliced_sirv                2952.0             Known        Known  \n",
       "3907    spliced_sirv                2952.0             Known        Known  \n",
       "3908    spliced_sirv                2952.0             Known        Known  \n",
       "3909    spliced_sirv                2952.0             Known        Known  \n",
       "3912    spliced_sirv                2952.0             Known        Known  \n",
       "...              ...                   ...               ...          ...  \n",
       "427772  spliced_sirv                3131.0             Known        Known  \n",
       "427774  spliced_sirv                3131.0             Known        Known  \n",
       "427775  spliced_sirv                3131.0             Known        Known  \n",
       "427776  spliced_sirv                3131.0             Known        Known  \n",
       "427777  spliced_sirv                3131.0             Known        Known  \n",
       "\n",
       "[144567 rows x 59 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fsm filter\n",
    "temp['pass_fsm_filter'] = False\n",
    "temp.loc[temp.structural_category=='FSM', 'pass_fsm_filter'] = True\n",
    "\n",
    "# monoexonic filter\n",
    "temp['pass_monoexonic_filter'] = True\n",
    "temp.loc[temp.exons==1, 'pass_monoexonic_filter'] = False\n",
    "\n",
    "# reproducibility filter -- only applies to potential novel transcripts\n",
    "# we would consider; ie FSM are immune; monoexonic and ISM are already out\n",
    "temp['pass_reproducibility_filter'] = False\n",
    "temp['sample_sharing'] = temp.groupby('jxnHash')['lab_sample'].transform('nunique')\n",
    "temp.loc[(temp.structural_category!='FSM')&\\\n",
    "         (temp.structural_cateogry!='ISM')&\\\n",
    "         (temp.exons!=1)&\\\n",
    "         (temp.sample_sharing>=2)], 'pass_reproducibility_filter'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea4377b-5c8c-490b-818b-04fc138abae5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
