{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f11551d-36bc-49b2-b561-144d4bbd2c11",
   "metadata": {},
   "source": [
    "## Goal -- how does our precision, sensitivity, and accuracy get impacted when applying the PODER filters? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7eba1c60-3a06-4e91-8ebc-f1eee4cfedbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import pyranges as pr\n",
    "from pyfaidx import Fasta\n",
    "from mizani.formatters import percent_format\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *\n",
    "\n",
    "from plotnine import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a2318ad-9bc9-406f-bbc2-63d38f3f205c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def my_theme(base_size=11, w=4, h=3):\n",
    "    \"\"\"\n",
    "    Custom plotnine theme with:\n",
    "    - White background\n",
    "    - Clean styling\n",
    "    - Axes and ticks retained\n",
    "\n",
    "    Parameters:\n",
    "    - base_size: Base font size\n",
    "\n",
    "    Returns:\n",
    "    - plotnine.theme object\n",
    "    \"\"\"\n",
    "    return (\n",
    "        theme_minimal(base_size=base_size)\n",
    "        + theme(\n",
    "            # White background\n",
    "            panel_background=element_rect(fill='white', color=None),\n",
    "            plot_background=element_rect(fill='white', color=None),\n",
    "\n",
    "            # Remove grid lines\n",
    "            panel_grid_major=element_blank(),\n",
    "            panel_grid_minor=element_blank(),\n",
    "            panel_border=element_blank(),\n",
    "\n",
    "            # Keep axis lines & ticks (don't blank them)\n",
    "            axis_line=element_line(color='black'),\n",
    "            axis_ticks=element_line(color='black'),\n",
    "\n",
    "            plot_title=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_x=element_text(hjust=0.5, family='Helvetica'),\n",
    "            axis_title_y=element_text(hjust=0.5, margin={'t':0, 'r':-2, 'b':0, 'l':0}, family='Helvetica'),\n",
    "            \n",
    "            # Styling text\n",
    "            legend_title=element_blank(),\n",
    "            axis_title=element_text(size=base_size + 1, family='Helvetica'),\n",
    "            legend_text=element_text(size=base_size-2, family='Helvetica'),\n",
    "            axis_text=element_text(size=base_size, color='black', family='Helvetica'),\n",
    "            figure_size=(w, h),  # Controls plot dimensions (width x height in inches)\n",
    "            plot_margin=0.05      # Shrinks surrounding white space\n",
    "        )\n",
    "    )\n",
    "\n",
    "def clean_figure(ax):\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.tick_params(axis=\"x\", rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d07f3929-edfa-4e08-aaaf-52dd07202e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "meta_df = load_meta()\n",
    "meta_df['lab_sample'] = meta_df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                        meta_df['lab_sampleid']+'_'+\\\n",
    "                        meta_df['cell_line_id']\n",
    "\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry\n",
    "\n",
    "init_plot_settings()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81597fa8-fa95-4809-9ea4-82ca9045fa26",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ch/bc1lrgbd1fn_3gxz886mxjz00000gn/T/ipykernel_81330/2818831187.py:30: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'ERCC' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n"
     ]
    }
   ],
   "source": [
    "## SQANTI reads stuff\n",
    "f = proc_cfg(expand(config['lr']['qc_sirvs']['sqanti_reads']['class_summary'],\n",
    "                    annot_completeness='C')[0], od)\n",
    "df = pd.read_csv(f, sep='\\t')\n",
    "\n",
    "# rename structural categories\n",
    "m = {'antisense': 'Antisense',\n",
    "     'full-splice_match': 'FSM',\n",
    "     'genic': 'Genic', \n",
    "     'incomplete-splice_match': 'ISM',\n",
    "     'intergenic': 'Intergenic',\n",
    "     'novel_in_catalog': 'NIC',\n",
    "     'novel_not_in_catalog': 'NNC'}\n",
    "df['structural_category'] = df.structural_category.map(m)\n",
    "\n",
    "# add antisense gene ids as assc_gene_2\n",
    "df['assc_gene_2'] = df.associated_gene\n",
    "inds = df.loc[df.structural_category=='Antisense'].index\n",
    "df.loc[inds, 'assc_gene_2'] = df.loc[inds, 'assc_gene_2'].str.split('_', expand=True)[1]\n",
    "\n",
    "# add relevant metadata\n",
    "df = df.merge(meta_df[['lab_sample', 'population']], \n",
    "              how='left', on='lab_sample')\n",
    "\n",
    "# add # mapped reads \n",
    "df['n_mapped_reads'] = df.groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "# add spike type\n",
    "df['spike_type'] = np.nan\n",
    "df.loc[df.chrom.str.contains('ERCC'), 'spike_type'] = 'ERCC'\n",
    "df.loc[df.chrom.str.contains('SIRV'), 'spike_type'] = 'SIRV'\n",
    "\n",
    "spliced_sirvs = ['SIRV1', 'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7']\n",
    "df.loc[df.chrom.isin(spliced_sirvs), 'spike_type_2'] = 'spliced_sirv'\n",
    "\n",
    "# add # spliced sirv reads\n",
    "df['n_spliced_sirv_reads'] = df.loc[df.spike_type_2=='spliced_sirv'].groupby('lab_sample')['isoform'].transform('count')\n",
    "\n",
    "# splicing novelty\n",
    "df['splicing_novelty'] = 'Novel'\n",
    "df.loc[df.structural_category.isin(['FSM', 'ISM']), 'splicing_novelty'] = 'Known'\n",
    "\n",
    "# overal known vs. novel\n",
    "df['overall_nov'] = 'Novel'\n",
    "df.loc[df.structural_category == 'FSM', 'overall_nov'] = 'Known'\n",
    "\n",
    "## Ground truth sirv / ercc gtf\n",
    "gtf_df = pr.read_gtf('../data/qc_sirvs/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf').df\n",
    "\n",
    "temp = gtf_df.loc[gtf_df.Chromosome.isin(spliced_sirvs)]\n",
    "n_spliced_sirvs_tot = len(temp.transcript_id.unique())\n",
    "\n",
    "# there are 69 spliced sirvs as ground truth\n",
    "\n",
    "# get the transcript length of each sirv transcript\n",
    "gtf_df['exon_len'] = gtf_df['End'] - gtf_df['Start']\n",
    "gtf_df['transcript_len'] = gtf_df.groupby('transcript_id')['exon_len'].transform('sum')\n",
    "\n",
    "gtf_df['mean_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('mean')\n",
    "gtf_df['med_transcript_len'] = gtf_df.groupby('gene_id')['transcript_len'].transform('median')\n",
    "\n",
    "# also get number of exons\n",
    "gtf_df['n_exons'] = gtf_df.groupby('transcript_id')['exon_assignment'].transform('count')\n",
    "gtf_df['mean_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('mean')\n",
    "gtf_df['med_n_exons'] = gtf_df.groupby('gene_id')['n_exons'].transform('median')\n",
    "\n",
    "# also get number of transcripts\n",
    "gtf_df['n_transcripts'] = gtf_df.groupby('gene_id')['transcript_id'].transform('nunique')\n",
    "\n",
    "# get # monoexonic transcripts\n",
    "temp = gtf_df[['gene_id', 'transcript_id', 'n_exons']].groupby(['gene_id', 'n_exons']).nunique().reset_index()\n",
    "temp = temp.loc[temp.n_exons==1]\n",
    "temp.drop('n_exons', axis=1, inplace=True)\n",
    "temp.rename({'transcript_id': 'n_monoexonic_transcripts'}, axis=1, inplace=True)\n",
    "temp.loc[temp.gene_id=='SIRV6']\n",
    "gtf_df = gtf_df.merge(temp,\n",
    "                      how='left',\n",
    "                      on='gene_id')\n",
    "\n",
    "# sort the long sirvs at least by length\n",
    "sirv_order = ['1', '2', '3', '4', '5', '6', '7',\n",
    "              '4001', '4002', '4003',\n",
    "              '6001', '6002', '6003', \n",
    "              '8001', '8002', '8003',\n",
    "              '10001', '10002', '10003',\n",
    "              '12001', '12002', '12003']\n",
    "sirv_order = [f'SIRV{s}' for s in sirv_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc3dd3a9-447d-48ff-994a-1a6afc017ff3",
   "metadata": {},
   "source": [
    "## Apply filters one by one and create a label to say whether the transcript passed each filter \n",
    "* FSM? (pass_fsm_filter)\n",
    "* Monoexonic? (pass_monoexonic_filter)\n",
    "* Reproducibile for novel transcripts? (pass_reproducibility_filter)\n",
    "* Promoted ISM? (promoted_ism_filter)\n",
    "* ISM? (pass_ism_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "770c9534-dda6-48cd-aa1d-c4cca7f6cab5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# filter 1 -- novel transcripts must be present in 2+ samples\n",
    "temp['sample_sharing'] = temp.groupby('jxnHash')['lab_sample'].transform('nunique')\n",
    "temp.loc[temp.sample_sharing>=2, 'filter_pass'] = True\n",
    "\n",
    "# filter 2 - for ISMs, if associated_transcript is not in original collection, add it and pass it\n",
    "\n",
    "# first, mark all ISMs as non-passing\n",
    "temp.loc[temp.structural_category=='ISM', 'filter_pass'] = False\n",
    "\n",
    "# get a random fsm entry to add for each uniq one\n",
    "fsm_temp = temp.loc[(temp.structural_category=='FSM')].drop_duplicates(subset='associated_transcript', keep='first')\n",
    "fsm_temp['lab_sample'] = np.nan\n",
    "assert len(fsm_temp.index) == n_spliced_sirvs_tot\n",
    "\n",
    "# now loop through lab samples and tack on undet. entries FSMs\n",
    "# for all ISMs\n",
    "add_df = pd.DataFrame()\n",
    "for s in temp.lab_sample.unique():\n",
    "    undet_fsms_from_isms = temp.loc[(temp.lab_sample==s)&\\\n",
    "                                    (temp.structural_category=='ISM')]\n",
    "    \n",
    "    temp2 = temp.loc[temp.lab_sample==s]\n",
    "\n",
    "    # these are the ids of transcripts that are detected via isms\n",
    "    ism_fsm_ids = temp2.loc[(temp2.structural_category=='ISM'),\n",
    "                           'associated_transcript'].unique().tolist()\n",
    "    \n",
    "    # these are the ids of transcripts that are detected via fsms\n",
    "    fsm_ids = temp2.loc[(temp2.structural_category=='FSM'),\n",
    "                        'associated_transcript'].unique().tolist()\n",
    "    \n",
    "    # the set difference will be the ones we need to \"promote\"\n",
    "    promote_fsm_ids = list(set(ism_fsm_ids)-set(fsm_ids))\n",
    "    promote_fsm_ids\n",
    "    \n",
    "    add_df_temp = fsm_temp.loc[fsm_temp.associated_transcript.isin(promote_fsm_ids)]\n",
    "    add_df_temp['lab_sample'] = s\n",
    "    add_df_temp['promoted_ism'] = True\n",
    "    \n",
    "    add_df = pd.concat([add_df, add_df_temp], axis=0)\n",
    "\n",
    "temp = pd.concat([temp, add_df], axis=0)\n",
    "\n",
    "# filter 0 -- all FSMs pass filtering\n",
    "temp.loc[temp.structural_category == 'FSM', 'filter_pass'] = True\n",
    "\n",
    "# filter 3 -- all monoexonics fail filtering \n",
    "temp.loc[temp.exons == 1, 'filter_pass'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fea4377b-5c8c-490b-818b-04fc138abae5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# now reduce to just uniq junction chains\n",
    "keep_cols = ['structural_category', 'associated_transcript', 'exons','jxnHash', 'assc_gene_2',\n",
    "             'lab_sample', 'population',\n",
    "             'n_mapped_reads', 'spike_type', 'spike_type_2', 'n_spliced_sirv_reads',\n",
    "             'splicing_novelty', 'overall_nov', 'filter_pass', 'sample_sharing', 'promoted_ism']\n",
    "temp = temp[keep_cols].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1243b9a6-c8df-41b9-ac16-b5344274d595",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9170\n",
      "8990\n"
     ]
    }
   ],
   "source": [
    "print(len(temp.index))\n",
    "print(len(temp[['jxnHash', 'filter_pass', 'lab_sample']].drop_duplicates()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "db153942-a07c-4653-8150-88968f0d3370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>exons</th>\n",
       "      <th>jxnHash</th>\n",
       "      <th>assc_gene_2</th>\n",
       "      <th>lab_sample</th>\n",
       "      <th>population</th>\n",
       "      <th>n_mapped_reads</th>\n",
       "      <th>spike_type</th>\n",
       "      <th>spike_type_2</th>\n",
       "      <th>n_spliced_sirv_reads</th>\n",
       "      <th>splicing_novelty</th>\n",
       "      <th>overall_nov</th>\n",
       "      <th>filter_pass</th>\n",
       "      <th>sample_sharing</th>\n",
       "      <th>promoted_ism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>209354</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV610</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>20_KE2_GM19328</td>\n",
       "      <td>LWK</td>\n",
       "      <td>6487</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3237.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395155</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV615</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>4_PY4_GM10495</td>\n",
       "      <td>MPC</td>\n",
       "      <td>29916</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>11068.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274481</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV610</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>15_CH3_GM18631</td>\n",
       "      <td>HAC</td>\n",
       "      <td>7671</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3573.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393335</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV610</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>4_PY4_GM10495</td>\n",
       "      <td>MPC</td>\n",
       "      <td>29916</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>11068.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176301</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV615</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>24_IS1_GM22234</td>\n",
       "      <td>AJI</td>\n",
       "      <td>7512</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3333.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       structural_category associated_transcript  exons  \\\n",
       "209354                 ISM               SIRV610      2   \n",
       "395155                 ISM               SIRV615      2   \n",
       "274481                 ISM               SIRV610      2   \n",
       "393335                 ISM               SIRV610      2   \n",
       "176301                 ISM               SIRV615      2   \n",
       "\n",
       "                                                  jxnHash assc_gene_2  \\\n",
       "209354  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "395155  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "274481  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "393335  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "176301  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "\n",
       "            lab_sample population  n_mapped_reads spike_type  spike_type_2  \\\n",
       "209354  20_KE2_GM19328        LWK            6487       SIRV  spliced_sirv   \n",
       "395155   4_PY4_GM10495        MPC           29916       SIRV  spliced_sirv   \n",
       "274481  15_CH3_GM18631        HAC            7671       SIRV  spliced_sirv   \n",
       "393335   4_PY4_GM10495        MPC           29916       SIRV  spliced_sirv   \n",
       "176301  24_IS1_GM22234        AJI            7512       SIRV  spliced_sirv   \n",
       "\n",
       "        n_spliced_sirv_reads splicing_novelty overall_nov filter_pass  \\\n",
       "209354                3237.0            Known       Novel       False   \n",
       "395155               11068.0            Known       Novel       False   \n",
       "274481                3573.0            Known       Novel       False   \n",
       "393335               11068.0            Known       Novel       False   \n",
       "176301                3333.0            Known       Novel       False   \n",
       "\n",
       "        sample_sharing promoted_ism  \n",
       "209354              37          NaN  \n",
       "395155              37          NaN  \n",
       "274481              37          NaN  \n",
       "393335              37          NaN  \n",
       "176301              37          NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# temp2 = temp.loc[temp.duplicated(subset=['jxnHash', 'filter_pass', 'lab_sample'], keep=False)].sort_values(by=['jxnHash'])\n",
    "# temp2.loc[temp2.structural_category == 'ISM'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d910affe-ba17-4c76-b470-5bbc60dbbc96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9170\n",
      "6348\n"
     ]
    }
   ],
   "source": [
    "temp['ic_id'] = temp.jxnHash\n",
    "temp.loc[temp.structural_category=='FSM', 'ic_id'] = 'associated_transcript'\n",
    "\n",
    "print(len(temp.index))\n",
    "print(len(temp[['ic_id', 'filter_pass', 'lab_sample']].drop_duplicates()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f1836760-172c-4525-9473-526987b9c291",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>exons</th>\n",
       "      <th>jxnHash</th>\n",
       "      <th>assc_gene_2</th>\n",
       "      <th>lab_sample</th>\n",
       "      <th>population</th>\n",
       "      <th>n_mapped_reads</th>\n",
       "      <th>spike_type</th>\n",
       "      <th>spike_type_2</th>\n",
       "      <th>n_spliced_sirv_reads</th>\n",
       "      <th>splicing_novelty</th>\n",
       "      <th>overall_nov</th>\n",
       "      <th>filter_pass</th>\n",
       "      <th>sample_sharing</th>\n",
       "      <th>promoted_ism</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>233872</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV615</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>11_NI6_GM19129</td>\n",
       "      <td>YRI</td>\n",
       "      <td>8911</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3983.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330385</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV615</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>43_EU3_GM12812</td>\n",
       "      <td>CEU</td>\n",
       "      <td>5457</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>2897.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356911</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV615</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>42_EU2_GM12778</td>\n",
       "      <td>CEU</td>\n",
       "      <td>30767</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>11012.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234565</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV610</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>11_NI6_GM19129</td>\n",
       "      <td>YRI</td>\n",
       "      <td>8911</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>3983.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316868</th>\n",
       "      <td>ISM</td>\n",
       "      <td>SIRV610</td>\n",
       "      <td>2</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "      <td>SIRV6</td>\n",
       "      <td>45_EU5_GM12878</td>\n",
       "      <td>CEU</td>\n",
       "      <td>15664</td>\n",
       "      <td>SIRV</td>\n",
       "      <td>spliced_sirv</td>\n",
       "      <td>6512.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Novel</td>\n",
       "      <td>False</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>006f2d0edb78027e4384277f846a34c50f44f014fa5eec...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       structural_category associated_transcript  exons  \\\n",
       "233872                 ISM               SIRV615      2   \n",
       "330385                 ISM               SIRV615      2   \n",
       "356911                 ISM               SIRV615      2   \n",
       "234565                 ISM               SIRV610      2   \n",
       "316868                 ISM               SIRV610      2   \n",
       "\n",
       "                                                  jxnHash assc_gene_2  \\\n",
       "233872  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "330385  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "356911  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "234565  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "316868  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...       SIRV6   \n",
       "\n",
       "            lab_sample population  n_mapped_reads spike_type  spike_type_2  \\\n",
       "233872  11_NI6_GM19129        YRI            8911       SIRV  spliced_sirv   \n",
       "330385  43_EU3_GM12812        CEU            5457       SIRV  spliced_sirv   \n",
       "356911  42_EU2_GM12778        CEU           30767       SIRV  spliced_sirv   \n",
       "234565  11_NI6_GM19129        YRI            8911       SIRV  spliced_sirv   \n",
       "316868  45_EU5_GM12878        CEU           15664       SIRV  spliced_sirv   \n",
       "\n",
       "        n_spliced_sirv_reads splicing_novelty overall_nov filter_pass  \\\n",
       "233872                3983.0            Known       Novel       False   \n",
       "330385                2897.0            Known       Novel       False   \n",
       "356911               11012.0            Known       Novel       False   \n",
       "234565                3983.0            Known       Novel       False   \n",
       "316868                6512.0            Known       Novel       False   \n",
       "\n",
       "        sample_sharing promoted_ism  \\\n",
       "233872              37          NaN   \n",
       "330385              37          NaN   \n",
       "356911              37          NaN   \n",
       "234565              37          NaN   \n",
       "316868              37          NaN   \n",
       "\n",
       "                                                    ic_id  \n",
       "233872  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...  \n",
       "330385  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...  \n",
       "356911  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...  \n",
       "234565  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...  \n",
       "316868  006f2d0edb78027e4384277f846a34c50f44f014fa5eec...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp2 = temp.loc[temp.duplicated(subset=['ic_id', 'filter_pass', 'lab_sample'], keep=False)].sort_values(by=['ic_id'])\n",
    "temp2.loc[temp2.structural_category == 'ISM'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01d9e76f-16c2-48cb-a37e-6414ab3e7325",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
