{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1d368bac-6b3d-4020-b807-674ef25fc14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8ef183d-78f8-47da-a35a-3ce434de5400",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3a631d6-594d-467d-a9e1-02cf8ab5ba82",
   "metadata": {},
   "source": [
    "## We want to get the dist. of distances between novel SSs and annotated SSs.\n",
    "\n",
    "Hopefully they are generally close\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3613df03-8676-4709-868b-acade41722aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df.columns\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID',\n",
    "          'map_reads_assemblymap','lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "# bad_samples = ['NA19328']\n",
    "# df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "d5e8acd8-f10c-459b-83cc-d6aba2ff6243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True\n",
    "\n",
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "432b7cd3-d9bf-438c-94e6-4cf77f8b5315",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000195.1_-_49120_86517</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>49118</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>49119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821508_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821506</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1821507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821512_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821510</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1821511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1823205_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1823203</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1823204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1825231_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1825229</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1825230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand                     sj_id start_site_novelty  \\\n",
       "0  GL000195.1      -  GL000195.1_-_49120_86517              novel   \n",
       "1       chr17      -   chr17_-_1821508_1827860              novel   \n",
       "2       chr17      -   chr17_-_1821512_1827860              novel   \n",
       "3       chr17      -   chr17_-_1823205_1827860              novel   \n",
       "4       chr17      -   chr17_-_1825231_1827860              novel   \n",
       "\n",
       "  end_site_category cell_line_id variable    Start sj_loc novelty_category  \\\n",
       "0             known      NA12273    Start    49118  start            novel   \n",
       "1             known      NA12273    Start  1821506  start            novel   \n",
       "2             known      NA12273    Start  1821510  start            novel   \n",
       "3             known      NA12273    Start  1823203  start            novel   \n",
       "4             known      NA12273    Start  1825229  start            novel   \n",
       "\n",
       "       End  \n",
       "0    49119  \n",
       "1  1821507  \n",
       "2  1821511  \n",
       "3  1823204  \n",
       "4  1825230  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nov_ss_df = temp.copy(deep=True)\n",
    "nov_ss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d1770e76-cc6e-4990-9a79-f97e6d7977b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp = pr.PyRanges(temp)\n",
    "# temp.to_bed('test_novel_ss.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4eafb303-4b31-4cd1-b449-d9b38cccb019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get annotated (v47) sss\n",
    "gtf_file = proc_cfg(config['ref']['gtf'], od)\n",
    "\n",
    "# read in ref. gtf and get SJs from there\n",
    "temp = pr.read_gtf(gtf_file)\n",
    "\n",
    "# get the introns\n",
    "temp = temp.features.introns()\n",
    "temp = temp.df\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# need to verify\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']+1\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+1\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "\n",
    "temp = pr.PyRanges(temp)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f27ac058-1660-496d-86a3-d1ae6ca4903f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.to_bed('test_v47_ss.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "36c9d817-8b08-4e1e-bab2-5cc47f4ffc89",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_ss_df = pr.PyRanges(nov_ss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "7ae7fb4e-c279-45ff-af30-241a2e0479cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get closest entry from novel sss to known sss\n",
    "df = nov_ss_df.nearest(temp, strandedness=False)\n",
    "df = df.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "8172dc20-7387-4e0e-b52e-ddc26f2edd8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8.17% of novel personalized haplotype SJs are w/i 12bp of an annotated one\n"
     ]
    }
   ],
   "source": [
    "# what % of these are w/i 12 bp of an annotated one\n",
    "assert len(df.loc[df.Distance<0]) == 0 # no neg. dists\n",
    "n_num = len(df.loc[df.Distance<=12].index)\n",
    "n = len(df.index)\n",
    "print(f'{(n_num/n)*100:.2f}% of novel personalized haplotype SJs are w/i 12bp of an annotated one')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b2e94ff8-cd98-4b5e-9309-9695484bede6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "360601"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# plot dist. \n",
    "sns.displot(df,\n",
    "            x='Distance',\n",
    "            kind='kde')\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "58768912-5029-4628-9740-7a1167035477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>variable_b</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>sj_loc_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000194.1_-_112851_114981</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12778</td>\n",
       "      <td>End</td>\n",
       "      <td>114981</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>114982</td>\n",
       "      <td>-</td>\n",
       "      <td>End</td>\n",
       "      <td>114985</td>\n",
       "      <td>end</td>\n",
       "      <td>114986</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000194.1_-_112851_114981</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG04216</td>\n",
       "      <td>End</td>\n",
       "      <td>114981</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>114982</td>\n",
       "      <td>-</td>\n",
       "      <td>End</td>\n",
       "      <td>114985</td>\n",
       "      <td>end</td>\n",
       "      <td>114986</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000194.1_-_112851_114981</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG01975</td>\n",
       "      <td>End</td>\n",
       "      <td>114981</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>114982</td>\n",
       "      <td>-</td>\n",
       "      <td>End</td>\n",
       "      <td>114985</td>\n",
       "      <td>end</td>\n",
       "      <td>114986</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000194.1_-_112851_114981</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>End</td>\n",
       "      <td>114981</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>114982</td>\n",
       "      <td>-</td>\n",
       "      <td>End</td>\n",
       "      <td>114985</td>\n",
       "      <td>end</td>\n",
       "      <td>114986</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000194.1_-_112851_114981</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>End</td>\n",
       "      <td>114981</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>114982</td>\n",
       "      <td>-</td>\n",
       "      <td>End</td>\n",
       "      <td>114985</td>\n",
       "      <td>end</td>\n",
       "      <td>114986</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360421</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20575888_20579616</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA19307</td>\n",
       "      <td>End</td>\n",
       "      <td>20579616</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>20579617</td>\n",
       "      <td>+</td>\n",
       "      <td>End</td>\n",
       "      <td>20579607</td>\n",
       "      <td>end</td>\n",
       "      <td>20579608</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360424</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20579692_20582591</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG00621</td>\n",
       "      <td>End</td>\n",
       "      <td>20582591</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>20582592</td>\n",
       "      <td>+</td>\n",
       "      <td>End</td>\n",
       "      <td>20582589</td>\n",
       "      <td>end</td>\n",
       "      <td>20582590</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360425</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20579692_20582591</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG01952</td>\n",
       "      <td>End</td>\n",
       "      <td>20582591</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>20582592</td>\n",
       "      <td>+</td>\n",
       "      <td>End</td>\n",
       "      <td>20582589</td>\n",
       "      <td>end</td>\n",
       "      <td>20582590</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360426</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20579692_20582591</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA18486</td>\n",
       "      <td>End</td>\n",
       "      <td>20582591</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>20582592</td>\n",
       "      <td>+</td>\n",
       "      <td>End</td>\n",
       "      <td>20582589</td>\n",
       "      <td>end</td>\n",
       "      <td>20582590</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360427</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20579692_20582591</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA19307</td>\n",
       "      <td>End</td>\n",
       "      <td>20582591</td>\n",
       "      <td>end</td>\n",
       "      <td>novel</td>\n",
       "      <td>20582592</td>\n",
       "      <td>+</td>\n",
       "      <td>End</td>\n",
       "      <td>20582589</td>\n",
       "      <td>end</td>\n",
       "      <td>20582590</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28173 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome Strand                       sj_id start_site_novelty  \\\n",
       "203     GL000194.1      -  GL000194.1_-_112851_114981              known   \n",
       "204     GL000194.1      -  GL000194.1_-_112851_114981              known   \n",
       "205     GL000194.1      -  GL000194.1_-_112851_114981              known   \n",
       "206     GL000194.1      -  GL000194.1_-_112851_114981              known   \n",
       "207     GL000194.1      -  GL000194.1_-_112851_114981              known   \n",
       "...            ...    ...                         ...                ...   \n",
       "360421        chrY      +    chrY_+_20575888_20579616              known   \n",
       "360424        chrY      +    chrY_+_20579692_20582591              known   \n",
       "360425        chrY      +    chrY_+_20579692_20582591              known   \n",
       "360426        chrY      +    chrY_+_20579692_20582591              known   \n",
       "360427        chrY      +    chrY_+_20579692_20582591              known   \n",
       "\n",
       "       end_site_category cell_line_id variable     Start sj_loc  \\\n",
       "203                novel      NA12778      End    114981    end   \n",
       "204                novel      HG04216      End    114981    end   \n",
       "205                novel      HG01975      End    114981    end   \n",
       "206                novel      HG03729      End    114981    end   \n",
       "207                novel      HG04217      End    114981    end   \n",
       "...                  ...          ...      ...       ...    ...   \n",
       "360421             novel      NA19307      End  20579616    end   \n",
       "360424             novel      HG00621      End  20582591    end   \n",
       "360425             novel      HG01952      End  20582591    end   \n",
       "360426             novel      NA18486      End  20582591    end   \n",
       "360427             novel      NA19307      End  20582591    end   \n",
       "\n",
       "       novelty_category       End Strand_b variable_b   Start_b sj_loc_b  \\\n",
       "203               novel    114982        -        End    114985      end   \n",
       "204               novel    114982        -        End    114985      end   \n",
       "205               novel    114982        -        End    114985      end   \n",
       "206               novel    114982        -        End    114985      end   \n",
       "207               novel    114982        -        End    114985      end   \n",
       "...                 ...       ...      ...        ...       ...      ...   \n",
       "360421            novel  20579617        +        End  20579607      end   \n",
       "360424            novel  20582592        +        End  20582589      end   \n",
       "360425            novel  20582592        +        End  20582589      end   \n",
       "360426            novel  20582592        +        End  20582589      end   \n",
       "360427            novel  20582592        +        End  20582589      end   \n",
       "\n",
       "           End_b  Distance  \n",
       "203       114986         4  \n",
       "204       114986         4  \n",
       "205       114986         4  \n",
       "206       114986         4  \n",
       "207       114986         4  \n",
       "...          ...       ...  \n",
       "360421  20579608         9  \n",
       "360424  20582590         2  \n",
       "360425  20582590         2  \n",
       "360426  20582590         2  \n",
       "360427  20582590         2  \n",
       "\n",
       "[28173 rows x 17 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Distance<12]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0682cd7b-c711-4f78-a28f-271f036033df",
   "metadata": {},
   "source": [
    "## Try specifically subsetting on the SJs that are exclusively found in hap1/hap2?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "67cae993-4cb9-4f15-81a5-1ef8a00e6c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "df00e916-a4ea-4b97-b7fc-c1309afdad68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24934518\n",
      "8311506\n"
     ]
    }
   ],
   "source": [
    "# dedupe w/ sqanti genome cause no one cares\n",
    "df = df.drop(['sqanti_genome', 'splice_motif', 'canonical'], axis=1)\n",
    "print(len(df.index))\n",
    "df = df.drop_duplicates()\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "a77b22bd-67a4-4846-a11f-52fdf59c35cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # try just w/ one cell line for now\n",
    "# df = df.loc[df.cell_line_id=='NA12273']\n",
    "# df_back = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "bee826a1-fa6f-40d6-8817-d9c88976f8b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>splice_motif</th>\n",
       "      <th>canonical</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KI270741.1_+_44470_47849</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KI270741.1_+_48008_52058</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KI270741.1_+_52137_130639</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KI270741.1_+_44470_47948</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000224.1_+_102592_105978</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sj_id splice_motif  canonical junction_novelty  \\\n",
       "0    KI270741.1_+_44470_47849         GTAG  canonical            known   \n",
       "1    KI270741.1_+_48008_52058         GTAG  canonical            known   \n",
       "2   KI270741.1_+_52137_130639         GTAG  canonical            known   \n",
       "3    KI270741.1_+_44470_47948         GTAG  canonical            known   \n",
       "4  GL000224.1_+_102592_105978         GTAG  canonical            known   \n",
       "\n",
       "  start_site_novelty end_site_category cell_line_id map_genome  \n",
       "0              known             known      NA12273       hg38  \n",
       "1              known             known      NA12273       hg38  \n",
       "2              known             known      NA12273       hg38  \n",
       "3              known             known      NA12273       hg38  \n",
       "4              known             known      NA12273       hg38  "
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_back.copy(deep=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "ce9c4ebf-e094-42ee-bf16-e8bac1539ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO put me in coach\n",
    "# # 1. filter for sjs w/ at least one novel ss \n",
    "# df = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "f0d69425-ad69-4c4e-8b01-d6c0e18d9078",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>splice_motif</th>\n",
       "      <th>canonical</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8784437</th>\n",
       "      <td>chr2_-_225040714_225041288</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>225040714</td>\n",
       "      <td>225041288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8784718</th>\n",
       "      <td>chr2_-_231513633_231514081</td>\n",
       "      <td>ATAG</td>\n",
       "      <td>non_canonical</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>231513633</td>\n",
       "      <td>231514081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8800830</th>\n",
       "      <td>chr10_-_132324374_132331650</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>132324374</td>\n",
       "      <td>132331650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8800831</th>\n",
       "      <td>chr10_-_132331529_132331650</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>132331529</td>\n",
       "      <td>132331650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8802477</th>\n",
       "      <td>chr16_+_29482592_29486512</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>29482592</td>\n",
       "      <td>29486512</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sj_id splice_motif      canonical  \\\n",
       "8784437   chr2_-_225040714_225041288         GTAG      canonical   \n",
       "8784718   chr2_-_231513633_231514081         ATAG  non_canonical   \n",
       "8800830  chr10_-_132324374_132331650         GTAG      canonical   \n",
       "8800831  chr10_-_132331529_132331650         GTAG      canonical   \n",
       "8802477    chr16_+_29482592_29486512         GTAG      canonical   \n",
       "\n",
       "        junction_novelty start_site_novelty end_site_category cell_line_id  \\\n",
       "8784437            novel              novel             novel      NA12273   \n",
       "8784718            known              known             known      NA12273   \n",
       "8800830            novel              known             novel      NA12273   \n",
       "8800831            novel              novel             novel      NA12273   \n",
       "8802477            novel              novel             known      NA12273   \n",
       "\n",
       "        map_genome Chromosome Strand      Start        End  \n",
       "8784437       hap1       chr2      -  225040714  225041288  \n",
       "8784718       hap1       chr2      -  231513633  231514081  \n",
       "8800830       hap1      chr10      -  132324374  132331650  \n",
       "8800831       hap1      chr10      -  132331529  132331650  \n",
       "8802477       hap1      chr16      +   29482592   29486512  "
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# melt to 5' and 3'\n",
    "df[['Chromosome', 'Strand', 'Start', 'End']] = df.sj_id.str.split('_', expand=True)\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "79ffa376-0556-4a1a-a8e1-f1dae4ceb650",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # TODO remove\n",
    "# df['ss_1_id'] = df.sj_id.str.rsplit('_', n=1, expand=True)[0]\n",
    "# df['ss_2_id'] = df.sj_id.str.rsplit('_', n=2, expand=True)[0]+'_'+df.sj_id.str.rsplit('_', n=1, expand=True)[1]\n",
    "\n",
    "# df['n_ss_1'] = df.groupby('ss_1_id')['sj_id'].transform('nunique')\n",
    "# df['n_ss_2'] = df.groupby('ss_2_id')['sj_id'].transform('nunique')\n",
    "# df.loc[df.n_ss_1>1].sort_values(by='ss_1_id').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "30b5d18d-e7b1-4a5f-9872-1f50147029a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.melt(id_vars=['Chromosome', 'Strand', 'map_genome',\n",
    "                      'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "df['sj_loc'] = ''\n",
    "df['detected'] = True\n",
    "\n",
    "\n",
    "df.loc[df.variable=='Start', 'sj_loc'] = 'start'\n",
    "df.loc[df.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "df['novelty_category'] = ''\n",
    "df.loc[df.variable=='Start', 'novelty_category'] = df.loc[df.variable=='Start', 'start_site_novelty']\n",
    "df.loc[df.variable=='End', 'novelty_category'] = df.loc[df.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "df = df.loc[df.novelty_category=='novel']\n",
    "\n",
    "df.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "\n",
    "# create id for each ss\n",
    "df['ss_id'] = 'ss_'+\\\n",
    "                df.sj_loc+'_'+\\\n",
    "                df.Chromosome+'_'+\\\n",
    "                df.Strand+'_'+\\\n",
    "                df.Start.astype(str)\n",
    "\n",
    "df = df[['ss_id', 'cell_line_id', 'map_genome', 'detected']]\n",
    "\n",
    "# check for sss that originated from different sjs\n",
    "assert len(df.loc[df.ss_id.duplicated()]) > 0\n",
    "\n",
    "# then dedupe and pivot\n",
    "df = df.drop_duplicates()\n",
    "df = df.pivot(index=['ss_id', 'cell_line_id'], \n",
    "                columns=['map_genome'],\n",
    "                values=['detected'])\n",
    "\n",
    "# flatten\n",
    "df.columns = df.columns.get_level_values(1)\n",
    "df.columns.name = None\n",
    "\n",
    "# reset index to make it a flat DataFrame\n",
    "df = df.reset_index()\n",
    "\n",
    "# fill missing values with False\n",
    "df = df.fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "350ad0bc-749b-4d67-99d5-04ef05331530",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b26ac0e9-e7e6-43d9-917c-703646ae3b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "0c040ad7-8327-42ff-83a5-42a8859354a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ss_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5156</th>\n",
       "      <td>ss_start_chr17_+_5426072</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ss_id cell_line_id  hap1  hap2  hg38\n",
       "5156  ss_start_chr17_+_5426072      NA12273  True  True  True"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.ss_id=='ss_start_chr17_+_5426072']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "bb9ba032-a468-4ff2-8625-21e6e50acf26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>detected</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>ss_id</th>\n",
       "      <th>ss_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>hg38</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>5426072</td>\n",
       "      <td>start</td>\n",
       "      <td>True</td>\n",
       "      <td>novel</td>\n",
       "      <td>ss_start_chr17_+_5426072</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>hg38</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>5426072</td>\n",
       "      <td>start</td>\n",
       "      <td>True</td>\n",
       "      <td>novel</td>\n",
       "      <td>ss_start_chr17_+_5426072</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>hg38</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>8174262</td>\n",
       "      <td>start</td>\n",
       "      <td>True</td>\n",
       "      <td>novel</td>\n",
       "      <td>ss_start_chr17_-_8174262</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>hg38</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>8174262</td>\n",
       "      <td>start</td>\n",
       "      <td>True</td>\n",
       "      <td>novel</td>\n",
       "      <td>ss_start_chr17_-_8174262</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>981</th>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>hg38</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>16381116</td>\n",
       "      <td>start</td>\n",
       "      <td>True</td>\n",
       "      <td>novel</td>\n",
       "      <td>ss_start_chr17_+_16381116</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome Strand map_genome start_site_novelty end_site_category  \\\n",
       "430      chr17      +       hg38              novel             known   \n",
       "431      chr17      +       hg38              novel             known   \n",
       "779      chr17      -       hg38              novel             known   \n",
       "781      chr17      -       hg38              novel             known   \n",
       "981      chr17      +       hg38              novel             novel   \n",
       "\n",
       "    cell_line_id variable     Start sj_loc  detected novelty_category  \\\n",
       "430      NA12273    Start   5426072  start      True            novel   \n",
       "431      NA12273    Start   5426072  start      True            novel   \n",
       "779      NA12273    Start   8174262  start      True            novel   \n",
       "781      NA12273    Start   8174262  start      True            novel   \n",
       "981      NA12273    Start  16381116  start      True            novel   \n",
       "\n",
       "                         ss_id  ss_count  \n",
       "430   ss_start_chr17_+_5426072         6  \n",
       "431   ss_start_chr17_+_5426072         6  \n",
       "779   ss_start_chr17_-_8174262         6  \n",
       "781   ss_start_chr17_-_8174262         6  \n",
       "981  ss_start_chr17_+_16381116        12  "
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['ss_count'] = df.groupby('ss_id')['map_genome'].transform('count')\n",
    "df.loc[df.ss_count>3].head()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "eaa2d37d-ab6f-4d7b-9229-ca0fe1cbc804",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-204-3bee9efaf0aa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# take max. of detection for the same ss id and map genome (in case one side of a splice junction was part of a different\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# sj in some other context)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mss_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mduplicated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mss_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mduplicated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ss_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'ss_id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'cell_line_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# # take max. of detection for the same ss id and map genome (in case one side of a splice junction was part of a different\n",
    "# # sj in some other context)\n",
    "# assert len(df.loc[df.ss_id.duplicated()]) > 0\n",
    "# print(df.loc[df.ss_id.duplicated(keep=False), 'ss_id'].head())\n",
    "# df = df.groupby(['ss_id', 'cell_line_id']).max().reset_index()\n",
    "# assert len(df.index) == len(df.drop_duplicates().index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "9405c08e-9439-4e00-a0ad-42d06092bd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. get those that are detected uniquely in hap1/2 \n",
    "# (ie those not in hg38)\n",
    "df = df.loc[(df.hg38==False)&\n",
    "                ((df.hap1+df.hap2)>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "9766836e-a576-4a45-bd67-67449c2ddeec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ss_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ss_id, cell_line_id, hap1, hap2, hg38]\n",
       "Index: []"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now check one of these guys in the og df\n",
    "df.head()\n",
    "df.loc[df.ss_id.str.contains('97738675')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "ffd13df9-2d6f-48cf-95aa-8e78a2041b9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>splice_motif</th>\n",
       "      <th>canonical</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8361277</th>\n",
       "      <td>chr10_+_97738675_97742567</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8625246</th>\n",
       "      <td>chr10_+_97738675_97742567</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>novel</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hap1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             sj_id splice_motif  canonical junction_novelty  \\\n",
       "8361277  chr10_+_97738675_97742567         GTAG  canonical            novel   \n",
       "8625246  chr10_+_97738675_97742567         GTAG  canonical            novel   \n",
       "\n",
       "        start_site_novelty end_site_category cell_line_id map_genome  \n",
       "8361277              known             novel      NA12273       hap2  \n",
       "8625246              known             novel      NA12273       hap1  "
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_back.loc[df_back.sj_id.str.contains('97742567')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26952d93-69e9-4f4b-ad0e-cb846cdf72c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb4e5573-a88d-4874-91a0-5ed97a5ee17b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58c70bed-135f-4b3f-b1ec-19c8d85629cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61069a03-8888-430d-9442-36decc5f5113",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "836900cf-c7c8-4777-809f-50cbf522971d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb8074a-cd8b-4296-b8d6-0608614c270a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14649d5-5ebd-4853-a5e7-3a63638cfcd6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "41b9ba61-3074-4795-9a85-9432423cce93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "7167ef4e-34b0-4ff6-9b84-68388d1c1487",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp_back = temp.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "9c8a7988-2065-4916-9db0-393786322200",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = temp_back.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "5ca104bf-1eb7-49de-89cd-d520c7a32316",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id',\n",
    "                          'map_genome', 'sqanti_genome'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp['detected'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "2849ed85-c675-47e4-803e-d754d6be6530",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "1680a7a2-eaa6-42dc-ba31-b0067daa681b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create id for each ss\n",
    "temp['ss_id'] = 'ss_'+\\\n",
    "                temp.sj_loc+'_'+\\\n",
    "                temp.Chromosome+'_'+\\\n",
    "                temp.Strand+'_'+\\\n",
    "                temp.Start.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "5b3717a1-628b-4492-97bb-63864f45198c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "a8a6d9d3-981c-4455-b9c2-a855f90510a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "temp = temp[['ss_id', 'sj_id', 'cell_line_id', 'map_genome', 'detected']]\n",
    "temp = temp.drop_duplicates()\n",
    "temp = temp.pivot(index=['ss_id', 'cell_line_id', 'sj_id'], \n",
    "                columns=['map_genome'],\n",
    "                values=['detected'])\n",
    "\n",
    "# flatten\n",
    "temp.columns = temp.columns.get_level_values(1)\n",
    "temp.columns.name = None\n",
    "\n",
    "# reset index to make it a flat DataFrame\n",
    "temp = temp.reset_index()\n",
    "\n",
    "# fill missing values with False\n",
    "temp = temp.fillna(False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "1386beaa-e5ab-4739-b6c1-277d40f922e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8333"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(temp.loc[temp[['hap1', 'hap2', 'hg38']].sum(axis=1)>1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "ecfe1930-ab6a-4fd2-b8df-e2649567678d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>splice_motif</th>\n",
       "      <th>canonical</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>sqanti_genome</th>\n",
       "      <th>detected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21436766</th>\n",
       "      <td>KI270726.1_+_26127_26229</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>hap1</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21540547</th>\n",
       "      <td>KI270726.1_+_26127_26229</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>hap1</td>\n",
       "      <td>hap2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21644328</th>\n",
       "      <td>KI270726.1_+_26127_26229</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>hap1</td>\n",
       "      <td>hap1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             sj_id splice_motif  canonical junction_novelty  \\\n",
       "21436766  KI270726.1_+_26127_26229         GTAG  canonical            novel   \n",
       "21540547  KI270726.1_+_26127_26229         GTAG  canonical            novel   \n",
       "21644328  KI270726.1_+_26127_26229         GTAG  canonical            novel   \n",
       "\n",
       "         start_site_novelty end_site_category cell_line_id map_genome  \\\n",
       "21436766              novel             novel      HG02261       hap1   \n",
       "21540547              novel             novel      HG02261       hap1   \n",
       "21644328              novel             novel      HG02261       hap1   \n",
       "\n",
       "         sqanti_genome  detected  \n",
       "21436766          hg38      True  \n",
       "21540547          hap2      True  \n",
       "21644328          hap1      True  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# kept in sj_id just to do some checking back in parent dataframe to make sure I'm not\n",
    "# doing anything insane\n",
    "temp_back.loc[(temp_back.sj_id=='KI270726.1_+_26127_26229')&(temp_back.cell_line_id=='HG02261')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "2a125a9a-564e-4efb-b84e-c0babcc057f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. get those that are detected uniquely in hap1/2 \n",
    "# (ie those not in hg38)\n",
    "temp = temp.loc[(temp.hg38==False)&\n",
    "                ((temp.hap1+temp.hap2)>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c4e5b09-b3f9-4556-bfef-50e7eabb23f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "e35ec275-1386-4200-aa44-0deb2f3a3f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "      <th>n_novel_sj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>GL000194.1_-_53893_55445</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>15825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>GL000194.1_-_53893_55445</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>15175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>GL000194.1_-_54956_55445</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>15175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>KI270726.1_+_26127_26229</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>19818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>chr10_+_100987611_100989643</td>\n",
       "      <td>NA19307</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>11472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sj_id cell_line_id   hap1   hap2   hg38  n_novel_sj\n",
       "67      GL000194.1_-_53893_55445      HG03729  False   True  False       15825\n",
       "68      GL000194.1_-_53893_55445      HG04217  False   True  False       15175\n",
       "73      GL000194.1_-_54956_55445      HG04217  False   True  False       15175\n",
       "470     KI270726.1_+_26127_26229      HG02261   True  False  False       19818\n",
       "515  chr10_+_100987611_100989643      NA19307   True   True  False       11472"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True\n",
    "\n",
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f6afa40-5691-434d-aa98-31e325ab7f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 1.5 only novel sss\n",
    "# temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# # only keep unique novel sjs\n",
    "# temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "# temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "#              'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "# temp.Start = temp.Start.astype(int)\n",
    "# temp.End = temp.End.astype(int)\n",
    "# assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# # melt to 5' and 3'\n",
    "# temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "#                           'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "#                  value_vars=['Start', 'End'])\n",
    "# temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "# temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "# temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "# temp['novelty_category'] = ''\n",
    "# temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "# temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# # only get the novel ones now\n",
    "# temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "# temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# # need to verify that this is working using like one motif or something make\n",
    "# # sure I don't have off-by-one errors\n",
    "# # verified\n",
    "# temp['Start'] = temp.Start-2\n",
    "# temp['End'] = temp.Start+1\n",
    "\n",
    "# # verified\n",
    "# temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "# temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "# temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "# temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
