{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a6cc2619-6af2-4314-9982-0c8359694d27",
   "metadata": {},
   "source": [
    "## Goal: look for exonic SNPs close to the SJs (+- 10 bp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "53c46ec9-3c91-4223-b015-e68c6dcb73cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "2260b6a7-5dfc-4d10-a779-49f42da334e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "4a726e50-381c-4d7c-af37-0868165caba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df.columns\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID',\n",
    "          'map_reads_assemblymap','lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "# bad_samples = ['NA19328']\n",
    "# df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39ee21e1-807b-492e-bc97-944fa9c4faa1",
   "metadata": {},
   "source": [
    "## We want to know what % of 1.5 NOVEL sjs that we 1. discover only in personal haplotypes that either have 2. noncanonical splice motifs in hg38 or 3. have snps +- 10 of splice junctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "71632d01-1258-495a-b386-0fa57f4199a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "816f88ad-16a5-4e99-b7a7-ade0fa6dd1a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.5 only novel sjs\n",
    "temp = df.loc[df.junction_novelty=='novel']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "15023a9f-d60b-44d0-816d-bb5530903497",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get only the detection from mapping\n",
    "temp = temp[['sj_id', 'cell_line_id', 'map_genome', 'detected']]\n",
    "temp = temp.drop_duplicates()\n",
    "temp = temp.pivot(index=['sj_id', 'cell_line_id'], \n",
    "                columns=['map_genome'],\n",
    "                values=['detected'])\n",
    "\n",
    "# flatten\n",
    "temp.columns = temp.columns.get_level_values(1)\n",
    "temp.columns.name = None\n",
    "\n",
    "# reset index to make it a flat DataFrame\n",
    "temp = temp.reset_index()\n",
    "\n",
    "# fill missing values with False\n",
    "temp = temp.fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "c0023ded-ff5c-41e7-a379-8f6869f1d637",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. get those that are detected uniquely in hap1/2 \n",
    "# (ie those not in hg38)\n",
    "temp = temp.loc[(temp.hg38==False)&\n",
    "                ((temp.hap1+temp.hap2)>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "b6a58a75-6509-47f5-8257-a5e259b801dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. get the hg38 canonicity of each of these sjs\n",
    "temp2 = df[['sj_id', 'splice_motif', 'canonical', 'sqanti_genome']].drop_duplicates()\n",
    "temp2 = temp2.loc[temp2.sqanti_genome=='hg38']\n",
    "assert len(temp2.loc[temp2.sj_id.duplicated()]) == 0\n",
    "temp = temp.merge(temp2, how='left', \n",
    "                  on='sj_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1abb25dc-1d73-4e0e-b5e3-b246c4218d8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. process snp content in each sj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8db8995b-c3a0-494d-bb59-d82ad31fb091",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55759306-3cd5-44e1-be45-702bbafb052d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb88895-26cf-4d9b-b633-b76eca505a37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c3d823-5b87-4062-b924-8553898ffceb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbed93bc-76a3-4adb-9ac8-10f67503080d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f644907d-d693-43dc-afff-054e4eff4554",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520ca7bc-9256-4d9e-a1dd-549afc5a4ab2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b302616-becb-42c7-8983-abcad9d0054f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25cff1b0-af3e-4e5c-bfdc-6a84b49be790",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9d6a5c3-4fc4-4464-8819-87f084fb9bd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a687bc16-e804-48e5-a0d3-90ff134e1c14",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ae6ad81-124a-4fa9-8212-d0644a36fcf6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "465b89e8-5a79-4698-8060-2a9b7d5ace7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49d84cd1-7599-4799-bbeb-28e357b2e7e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b78a9de-81d9-494f-9618-8934df9e50e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59566235-8820-457d-b822-28ed724ff8de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f239d40-d02a-4b71-b258-4a4d707c65e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c1aea0cf-f391-41e0-b257-d4fb378bf5cc",
   "metadata": {},
   "source": [
    "## dev for geting sjs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "58640fc9-87cf-49c3-9689-138b8aa570b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp_back = temp.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "9d0e441f-3a78-4525-8ddc-4a2c332f2ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = temp_back.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "9dad5fd2-f3f3-407d-a43f-9d04b2b4ef8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>map_genome</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_novelty</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2_+_155532_156720</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000008.2_+_83546_83926</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GL000008.2_+_83546_85566</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GL000008.2_+_84015_85566</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GL000008.2_+_85626_129984</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "map_genome                       sj_id junction_novelty start_site_novelty  \\\n",
       "0           GL000008.2_+_155532_156720            known              known   \n",
       "3             GL000008.2_+_83546_83926            known              known   \n",
       "7             GL000008.2_+_83546_85566            known              known   \n",
       "11            GL000008.2_+_84015_85566            known              known   \n",
       "18           GL000008.2_+_85626_129984            known              known   \n",
       "\n",
       "map_genome end_site_novelty cell_line_id  hap1  hap2  hg38  \n",
       "0                     known      HG03732  True  True  True  \n",
       "3                     known      HG03732  True  True  True  \n",
       "7                     known      HG03732  True  True  True  \n",
       "11                    known      HG03732  True  True  True  \n",
       "18                    known      HG03732  True  True  True  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # check w/ just 1 genome + cell line at first so I can \n",
    "# # dl GTF and check it\n",
    "# temp = temp.loc[(temp.cell_line_id=='HG03732')&\\\n",
    "#                 (temp.hap1==True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "44958fd9-72f4-480d-b9a3-48471644b76b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24934518\n",
      "8311506\n"
     ]
    }
   ],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "\n",
    "# remove sqanti genome and drop dupes\n",
    "# the sqanti genome / sqanti metrics SHOULD be irrelevant here\n",
    "df = df.drop(['sqanti_genome', 'canonical', 'splice_motif'], axis=1)\n",
    "print(len(df.index))\n",
    "df = df.drop_duplicates()\n",
    "print(len(df.index))\n",
    "\n",
    "# then make sure that there are no dupe. sj+sj nov+sample+map genome\n",
    "temp = df.loc[df[['sj_id', 'junction_novelty',\n",
    "                  'cell_line_id', 'map_genome', 'start_site_novelty',\n",
    "                  'end_site_category']].duplicated(keep=False)]\n",
    "assert len(temp.index) == 0\n",
    "del temp\n",
    "\n",
    "df.rename({'end_site_category': 'end_site_novelty'}, axis=1, inplace=True)\n",
    "\n",
    "# transform to be t/f for each ic per genome\n",
    "temp = pd.crosstab(index=[df.sj_id, df.junction_novelty,\n",
    "                                 df.start_site_novelty,\n",
    "                                 df.end_site_novelty, df.cell_line_id],\n",
    "                          columns=df.map_genome,\n",
    "                          values=df.map_genome,\n",
    "                          aggfunc=lambda x: True).fillna(False).reset_index()\n",
    "\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End',\n",
    "             'cell_line_id', 'sj_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'cell_line_id', 'sj_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp.loc[temp.map_genome=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.map_genome=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# add 10 to start, rm 10 from end\n",
    "# verified https://trello.com/c/qzMAZpAm\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-9\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+14\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "temp.drop(['cell_line_id', 'map_genome'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "4ac2009b-bbf4-425d-ae30-b69fe5753ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>GL000008.2_+_155532_156720</td>\n",
       "      <td>Start</td>\n",
       "      <td>155521</td>\n",
       "      <td>start</td>\n",
       "      <td>155531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786681</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>GL000008.2_+_155532_156720</td>\n",
       "      <td>End</td>\n",
       "      <td>156720</td>\n",
       "      <td>end</td>\n",
       "      <td>156733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome Strand cell_line_id                       sj_id  \\\n",
       "0        GL000008.2      +      HG03732  GL000008.2_+_155532_156720   \n",
       "2786681  GL000008.2      +      HG03732  GL000008.2_+_155532_156720   \n",
       "\n",
       "        map_genome   Start sj_loc     End  \n",
       "0            Start  155521  start  155531  \n",
       "2786681        End  156720    end  156733  "
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.sj_id=='GL000008.2_+_155532_156720']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "ce993cc8-1e07-47f9-9cfd-e7ffef2aed0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "dedbfee0-68e1-46b4-ac83-3ae4c509b962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000008.2_+_155532_156720</td>\n",
       "      <td>155521</td>\n",
       "      <td>start</td>\n",
       "      <td>155531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000008.2_+_83546_83926</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000008.2_+_83546_83926</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000008.2_+_83546_83926</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000008.2_+_83546_83926</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand                       sj_id   Start sj_loc     End\n",
       "0  GL000008.2      +  GL000008.2_+_155532_156720  155521  start  155531\n",
       "1  GL000008.2      +    GL000008.2_+_83546_83926   83535  start   83545\n",
       "2  GL000008.2      +    GL000008.2_+_83546_83926   83535  start   83545\n",
       "3  GL000008.2      +    GL000008.2_+_83546_83926   83535  start   83545\n",
       "4  GL000008.2      +    GL000008.2_+_83546_83926   83535  start   83545"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "94567f96-5a5f-445b-a678-5e3ee3bfc55b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>155521</td>\n",
       "      <td>start</td>\n",
       "      <td>155531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>83535</td>\n",
       "      <td>start</td>\n",
       "      <td>83545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>84004</td>\n",
       "      <td>start</td>\n",
       "      <td>84014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>85615</td>\n",
       "      <td>start</td>\n",
       "      <td>85625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>85615</td>\n",
       "      <td>start</td>\n",
       "      <td>85625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>85615</td>\n",
       "      <td>start</td>\n",
       "      <td>85625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>+</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>85615</td>\n",
       "      <td>start</td>\n",
       "      <td>85625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+------------+---+---------+-------+--------+-------+--------+\n",
       "| Chromosome\n",
       "(category)            | Strand\n",
       "(category)   | cell_line_id\n",
       "(object)         | map_genome\n",
       "(object)       |        Start\n",
       "(int64) | sj_loc\n",
       "(object)       |        End\n",
       "(int64) |\n",
       "|------------+---+---------+-------+--------+-------+--------|\n",
       "| GL000008.2 | + | HG03732 | Start | 155521 | start | 155531 |\n",
       "| GL000008.2 | + | HG03732 | Start |  83535 | start |  83545 |\n",
       "| GL000008.2 | + | HG03732 | Start |  83535 | start |  83545 |\n",
       "| GL000008.2 | + | HG03732 | Start |  84004 | start |  84014 |\n",
       "| GL000008.2 | + | HG03732 | Start |  85615 | start |  85625 |\n",
       "| GL000008.2 | + | HG03732 | Start |  85615 | start |  85625 |\n",
       "| GL000008.2 | + | HG03732 | Start |  85615 | start |  85625 |\n",
       "| GL000008.2 | + | HG03732 | Start |  85615 | start |  85625 |\n",
       "+------------+---+---------+-------+--------+-------+--------+\n",
       "Stranded PyRanges object has 8 rows and 7 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "d4f9c6a4-3351-4f46-a343-c72aec626694",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp.to_bed('test.bed')\n",
    "# temp.to_bed('test_2.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3764b21-4a67-40b0-830b-35789582f8a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3becdfd-0c68-4f5d-9883-7dae3b99b4d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a3d71a9-ea48-4eb1-9cb2-e77d191fbd83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c310b204-c096-45f9-8bc5-439cf720248d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c3e086f-8eb0-4434-a4ec-e2a91b7c9338",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a3bb136-c3b7-4304-9749-a611af8d3a38",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
