{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1d368bac-6b3d-4020-b807-674ef25fc14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8ef183d-78f8-47da-a35a-3ce434de5400",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3a631d6-594d-467d-a9e1-02cf8ab5ba82",
   "metadata": {},
   "source": [
    "## We want to get the dist. of distances between novel SSs and annotated SSs.\n",
    "\n",
    "Hopefully they are generally close\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3613df03-8676-4709-868b-acade41722aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df.columns\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID',\n",
    "          'map_reads_assemblymap','lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "# bad_samples = ['NA19328']\n",
    "# df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "d5e8acd8-f10c-459b-83cc-d6aba2ff6243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True\n",
    "\n",
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "432b7cd3-d9bf-438c-94e6-4cf77f8b5315",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000195.1_-_49120_86517</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>49118</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>49119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821508_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821506</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1821507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821512_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821510</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1821511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1823205_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1823203</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1823204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1825231_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1825229</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>1825230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand                     sj_id start_site_novelty  \\\n",
       "0  GL000195.1      -  GL000195.1_-_49120_86517              novel   \n",
       "1       chr17      -   chr17_-_1821508_1827860              novel   \n",
       "2       chr17      -   chr17_-_1821512_1827860              novel   \n",
       "3       chr17      -   chr17_-_1823205_1827860              novel   \n",
       "4       chr17      -   chr17_-_1825231_1827860              novel   \n",
       "\n",
       "  end_site_category cell_line_id variable    Start sj_loc novelty_category  \\\n",
       "0             known      NA12273    Start    49118  start            novel   \n",
       "1             known      NA12273    Start  1821506  start            novel   \n",
       "2             known      NA12273    Start  1821510  start            novel   \n",
       "3             known      NA12273    Start  1823203  start            novel   \n",
       "4             known      NA12273    Start  1825229  start            novel   \n",
       "\n",
       "       End  \n",
       "0    49119  \n",
       "1  1821507  \n",
       "2  1821511  \n",
       "3  1823204  \n",
       "4  1825230  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nov_ss_df = temp.copy(deep=True)\n",
    "nov_ss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d1770e76-cc6e-4990-9a79-f97e6d7977b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp = pr.PyRanges(temp)\n",
    "# temp.to_bed('test_novel_ss.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eafb303-4b31-4cd1-b449-d9b38cccb019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get annotated (v47) sss\n",
    "gtf_file = proc_cfg(config['ref']['gtf'], od)\n",
    "\n",
    "# read in ref. gtf and get SJs from there\n",
    "temp = pr.read_gtf(gtf_file)\n",
    "\n",
    "# get the introns\n",
    "temp = temp.features.introns()\n",
    "temp = temp.df\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# need to verify\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "\n",
    "temp = pr.PyRanges(temp)\n",
    "\n",
    "temp.to_bed(output.bed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27ac058-1660-496d-86a3-d1ae6ca4903f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
