{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cef4ea77-5ae5-4c18-9867-5d12d0edab68",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9887c238-03c2-4393-8fbf-26f744b41170",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8783956d-19ce-4a26-8fb9-439ccb114f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID', 'lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "bad_samples = ['NA19328']\n",
    "df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a826b885-9b7e-43dc-b0a0-d3b07b60b5fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>hapmap_DNA_ID</th>\n",
       "      <th>lab_sample</th>\n",
       "      <th>population</th>\n",
       "      <th>cell_line_id_1000g</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG04217</td>\n",
       "      <td>ITU5</td>\n",
       "      <td>no_hapmap</td>\n",
       "      <td>39_IN5_HG04217</td>\n",
       "      <td>ITU</td>\n",
       "      <td>HG04217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG04216</td>\n",
       "      <td>ITU4</td>\n",
       "      <td>no_hapmap</td>\n",
       "      <td>38_IN4_HG04216</td>\n",
       "      <td>ITU</td>\n",
       "      <td>HG04216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG03732</td>\n",
       "      <td>ITU3</td>\n",
       "      <td>no_hapmap</td>\n",
       "      <td>37_IN3_HG03732</td>\n",
       "      <td>ITU</td>\n",
       "      <td>HG03732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG03729</td>\n",
       "      <td>ITU2</td>\n",
       "      <td>no_hapmap</td>\n",
       "      <td>36_IN2_HG03729</td>\n",
       "      <td>ITU</td>\n",
       "      <td>HG03729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG03719</td>\n",
       "      <td>ITU1</td>\n",
       "      <td>no_hapmap</td>\n",
       "      <td>35_IN1_HG03719</td>\n",
       "      <td>ITU</td>\n",
       "      <td>HG03719</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_line_id sample hapmap_DNA_ID      lab_sample population  \\\n",
       "0      HG04217   ITU5     no_hapmap  39_IN5_HG04217        ITU   \n",
       "1      HG04216   ITU4     no_hapmap  38_IN4_HG04216        ITU   \n",
       "2      HG03732   ITU3     no_hapmap  37_IN3_HG03732        ITU   \n",
       "3      HG03729   ITU2     no_hapmap  36_IN2_HG03729        ITU   \n",
       "4      HG03719   ITU1     no_hapmap  35_IN1_HG03719        ITU   \n",
       "\n",
       "  cell_line_id_1000g  \n",
       "0            HG04217  \n",
       "1            HG04216  \n",
       "2            HG03732  \n",
       "3            HG03729  \n",
       "4            HG03719  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc43f850-0819-4385-bcfe-fc516b91a4ab",
   "metadata": {},
   "source": [
    "## How does IC sharing differ for the different genotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ee02b8f7-989c-45eb-896c-6ea7aa52e4da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4549058\n",
      "4549056\n"
     ]
    }
   ],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['cerb']['ic_summary'],od)\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "\n",
    "# remove chromosomes that aren't represented in normal hg38 alignmnts\n",
    "hg38_chrs = df.loc[df.genome=='hg38', 'Chromosome'].unique().tolist()\n",
    "print(len(df.index))\n",
    "df = df.loc[df.Chromosome.isin(hg38_chrs)]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2dd8bec3-2ffb-4cc2-8c84-d82ebb50ff26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Name</th>\n",
       "      <th>genome</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>86517-49092-48954-44923</td>\n",
       "      <td>ENSG00000276256_1</td>\n",
       "      <td>hg38</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>GL000195.1_-_86517-49092-48954-44923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>86517-49119-48954-44923</td>\n",
       "      <td>ENSG00000276256_2</td>\n",
       "      <td>hg38</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>GL000195.1_-_86517-49119-48954-44923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>86517-74180-74120-49119-48954-44923</td>\n",
       "      <td>ENSG00000303853_1</td>\n",
       "      <td>hg38</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>GL000195.1_-_86517-74180-74120-49119-48954-44923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>86517-83958</td>\n",
       "      <td>ENSG00000303853_2</td>\n",
       "      <td>hg38</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>GL000195.1_-_86517-83958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>86517-84116</td>\n",
       "      <td>ENSG00000303853_3</td>\n",
       "      <td>hg38</td>\n",
       "      <td>HG04217</td>\n",
       "      <td>GL000195.1_-_86517-84116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Chromosome Strand                          Coordinates  \\\n",
       "0           0  GL000195.1      -              86517-49092-48954-44923   \n",
       "1           1  GL000195.1      -              86517-49119-48954-44923   \n",
       "2           2  GL000195.1      -  86517-74180-74120-49119-48954-44923   \n",
       "3           3  GL000195.1      -                          86517-83958   \n",
       "4           4  GL000195.1      -                          86517-84116   \n",
       "\n",
       "                Name genome cell_line_id  \\\n",
       "0  ENSG00000276256_1   hg38      HG04217   \n",
       "1  ENSG00000276256_2   hg38      HG04217   \n",
       "2  ENSG00000303853_1   hg38      HG04217   \n",
       "3  ENSG00000303853_2   hg38      HG04217   \n",
       "4  ENSG00000303853_3   hg38      HG04217   \n",
       "\n",
       "                                              ic_id  \n",
       "0              GL000195.1_-_86517-49092-48954-44923  \n",
       "1              GL000195.1_-_86517-49119-48954-44923  \n",
       "2  GL000195.1_-_86517-74180-74120-49119-48954-44923  \n",
       "3                          GL000195.1_-_86517-83958  \n",
       "4                          GL000195.1_-_86517-84116  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['ic_id'] = df.Chromosome.astype(str)+'_'+\\\n",
    "              df.Strand.astype(str)+'_'+\\\n",
    "              df.Coordinates.astype(str)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "da7ac078-ce20-4613-9c56-ad21bdadb2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# what percentage of ics by cell line (personal haplotypes) \n",
    "# are found also by the reference genome?\n",
    "big_df = pd.DataFrame()\n",
    "for cl in df.cell_line_id.unique():\n",
    "    temp = df.loc[df.cell_line_id==cl]\n",
    "    temp = temp[['ic_id', 'genome']]\n",
    "    temp['hg38'] = temp['hap1'] = temp['hap2'] = False\n",
    "    temp.loc[temp.genome=='hg38', 'hg38'] = True\n",
    "    temp.loc[temp.genome=='hap1', 'hap1'] = True\n",
    "    temp.loc[temp.genome=='hap2', 'hap2'] = True\n",
    "\n",
    "    temp = temp[['ic_id', 'hg38', 'hap1', 'hap2']].groupby('ic_id').max().reset_index()\n",
    "    # temp['personal'] = temp.hap1|temp.hap2\n",
    "    temp = temp.groupby(['hg38', 'hap1', 'hap2']).nunique().reset_index()\n",
    "    temp.rename({'ic_id':'n_ic'}, axis=1, inplace=True)\n",
    "    temp['total'] = temp['n_ic'].sum()\n",
    "    temp['perc'] = (temp.n_ic/temp.total)*100\n",
    "    \n",
    "    # one haplotype at a time\n",
    "    temp2 = temp[['hg38', 'hap1', 'n_ic', 'total']]\n",
    "    temp2 = temp2.groupby(['hg38', 'hap1', 'total']).sum().reset_index()\n",
    "    temp2['personal_genotype'] = 'hap1'\n",
    "    temp2.rename({'hap1': 'personal_genotype'}, axis=1, inplace=True)\n",
    "    temp2\n",
    "\n",
    "    # one haplotype at a time\n",
    "    temp3 = temp[['hg38', 'hap2', 'n_ic', 'total']]\n",
    "    temp3 = temp3.groupby(['hg38', 'hap2', 'total']).sum().reset_index()\n",
    "    temp3['personal_genotype'] = 'hap2'\n",
    "    temp3.rename({'hap2': 'personal_genotype'}, axis=1, inplace=True)\n",
    "    temp3\n",
    "\n",
    "    temp2 = pd.concat([temp2, temp3], axis=0)\n",
    "    temp2['perc'] = (temp2['n_ic']/temp2['total'])*100\n",
    "    temp2\n",
    "    \n",
    "    temp2['cell_line_id'] = cl\n",
    "    big_df = pd.concat([temp2, big_df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b23c8284-6df2-4a4f-ad47-de6a4a9b2880",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hg38</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>total</th>\n",
       "      <th>perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>176</td>\n",
       "      <td>55711</td>\n",
       "      <td>0.315916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>160</td>\n",
       "      <td>55711</td>\n",
       "      <td>0.287196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2007</td>\n",
       "      <td>55711</td>\n",
       "      <td>3.602520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>88</td>\n",
       "      <td>55711</td>\n",
       "      <td>0.157958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>122</td>\n",
       "      <td>55711</td>\n",
       "      <td>0.218987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>108</td>\n",
       "      <td>55711</td>\n",
       "      <td>0.193858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>53050</td>\n",
       "      <td>55711</td>\n",
       "      <td>95.223564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    hg38   hap1   hap2   n_ic  total       perc\n",
       "0  False  False   True    176  55711   0.315916\n",
       "1  False   True  False    160  55711   0.287196\n",
       "2  False   True   True   2007  55711   3.602520\n",
       "3   True  False  False     88  55711   0.157958\n",
       "4   True  False   True    122  55711   0.218987\n",
       "5   True   True  False    108  55711   0.193858\n",
       "6   True   True   True  53050  55711  95.223564"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = temp[['ic_id', 'genome']]\n",
    "temp['hg38'] = temp['hap1'] = temp['hap2'] = False\n",
    "temp.loc[temp.genome=='hg38', 'hg38'] = True\n",
    "temp.loc[temp.genome=='hap1', 'hap1'] = True\n",
    "temp.loc[temp.genome=='hap2', 'hap2'] = True\n",
    "\n",
    "temp = temp[['ic_id', 'hg38', 'hap1', 'hap2']].groupby('ic_id').max().reset_index()\n",
    "# temp['personal'] = temp.hap1|temp.hap2\n",
    "temp = temp.groupby(['hg38', 'hap1', 'hap2']).nunique().reset_index()\n",
    "temp.rename({'ic_id':'n_ic'}, axis=1, inplace=True)\n",
    "temp['total'] = temp['n_ic'].sum()\n",
    "temp['perc'] = (temp.n_ic/temp.total)*100\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f330828c-9f34-4138-84af-cdb5c401a887",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hg38</th>\n",
       "      <th>personal_genotype</th>\n",
       "      <th>total</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>personal_genotype</th>\n",
       "      <th>perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55711</td>\n",
       "      <td>176</td>\n",
       "      <td>hap1</td>\n",
       "      <td>0.315916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>55711</td>\n",
       "      <td>2167</td>\n",
       "      <td>hap1</td>\n",
       "      <td>3.889717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55711</td>\n",
       "      <td>210</td>\n",
       "      <td>hap1</td>\n",
       "      <td>0.376945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>55711</td>\n",
       "      <td>53158</td>\n",
       "      <td>hap1</td>\n",
       "      <td>95.417422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55711</td>\n",
       "      <td>160</td>\n",
       "      <td>hap2</td>\n",
       "      <td>0.287196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>55711</td>\n",
       "      <td>2183</td>\n",
       "      <td>hap2</td>\n",
       "      <td>3.918436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55711</td>\n",
       "      <td>196</td>\n",
       "      <td>hap2</td>\n",
       "      <td>0.351816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>55711</td>\n",
       "      <td>53172</td>\n",
       "      <td>hap2</td>\n",
       "      <td>95.442552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    hg38 personal_genotype  total   n_ic personal_genotype       perc\n",
       "0  False             False  55711    176              hap1   0.315916\n",
       "1  False              True  55711   2167              hap1   3.889717\n",
       "2   True             False  55711    210              hap1   0.376945\n",
       "3   True              True  55711  53158              hap1  95.417422\n",
       "0  False             False  55711    160              hap2   0.287196\n",
       "1  False              True  55711   2183              hap2   3.918436\n",
       "2   True             False  55711    196              hap2   0.351816\n",
       "3   True              True  55711  53172              hap2  95.442552"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# one haplotype at a time\n",
    "temp2 = temp[['hg38', 'hap1', 'n_ic', 'total']]\n",
    "temp2 = temp2.groupby(['hg38', 'hap1', 'total']).sum().reset_index()\n",
    "temp2['personal_genotype'] = 'hap1'\n",
    "temp2.rename({'hap1': 'personal_genotype'}, axis=1, inplace=True)\n",
    "temp2\n",
    "\n",
    "# one haplotype at a time\n",
    "temp3 = temp[['hg38', 'hap2', 'n_ic', 'total']]\n",
    "temp3 = temp3.groupby(['hg38', 'hap2', 'total']).sum().reset_index()\n",
    "temp3['personal_genotype'] = 'hap2'\n",
    "temp3.rename({'hap2': 'personal_genotype'}, axis=1, inplace=True)\n",
    "temp3\n",
    "\n",
    "temp2 = pd.concat([temp2, temp3], axis=0)\n",
    "temp2['perc'] = (temp2['n_ic']/temp2['total'])*100\n",
    "temp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f84dccda-3dc3-4df0-9cd7-98cc17be0242",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68abbc2b-3179-46da-a30e-3bba1f199243",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5f4441-2e27-4031-8125-f273debcb30d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75b3662b-0b5f-4b4e-9c8d-161b8dd8bfce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "175758ef-5470-4415-8dbe-61db5709c653",
   "metadata": {},
   "source": [
    "## Do we get more NNC (novel splice site) when we look at the personal haploytpes versus the hg38?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f32bf56-9ed0-4b59-93cb-e57e94504f43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
