{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1d368bac-6b3d-4020-b807-674ef25fc14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8ef183d-78f8-47da-a35a-3ce434de5400",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3a631d6-594d-467d-a9e1-02cf8ab5ba82",
   "metadata": {},
   "source": [
    "## We want to get the dist. of distances between novel SSs and annotated SSs.\n",
    "\n",
    "Hopefully they are generally close\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3613df03-8676-4709-868b-acade41722aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df.columns\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID',\n",
    "          'map_reads_assemblymap','lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "# bad_samples = ['NA19328']\n",
    "# df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d5e8acd8-f10c-459b-83cc-d6aba2ff6243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "35418d13-a05f-4824-8c8a-51e7de738648",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5b911f23-c37a-4307-9886-8c7bc9f11e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp_back = temp.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a0420e97-bb7f-4663-a555-8cac75fc1dea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sj_id</th>\n",
       "      <th>splice_motif</th>\n",
       "      <th>canonical</th>\n",
       "      <th>junction_novelty</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>map_genome</th>\n",
       "      <th>sqanti_genome</th>\n",
       "      <th>detected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GL000195.1_-_49120_86517</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>chr17_-_1821508_1827860</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>chr17_-_1821512_1827860</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>chr17_-_1823205_1827860</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chr17_-_1825231_1827860</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>canonical</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>hg38</td>\n",
       "      <td>hg38</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sj_id splice_motif  canonical junction_novelty  \\\n",
       "23   GL000195.1_-_49120_86517         GTAG  canonical            novel   \n",
       "128   chr17_-_1821508_1827860         GTAG  canonical            novel   \n",
       "130   chr17_-_1821512_1827860         GTAG  canonical            novel   \n",
       "131   chr17_-_1823205_1827860         GTAG  canonical            novel   \n",
       "132   chr17_-_1825231_1827860         GTAG  canonical            novel   \n",
       "\n",
       "    start_site_novelty end_site_category cell_line_id map_genome  \\\n",
       "23               novel             known      NA12273       hg38   \n",
       "128              novel             known      NA12273       hg38   \n",
       "130              novel             known      NA12273       hg38   \n",
       "131              novel             known      NA12273       hg38   \n",
       "132              novel             known      NA12273       hg38   \n",
       "\n",
       "    sqanti_genome  detected  \n",
       "23           hg38      True  \n",
       "128          hg38      True  \n",
       "130          hg38      True  \n",
       "131          hg38      True  \n",
       "132          hg38      True  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = temp_back.copy(deep=True)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "06dffae1-f754-46c0-a495-d09068ea4fe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "      <th>sj_loc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000195.1_-_49120_86517</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>49120</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821508_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821508</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1821512_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1821512</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1823205_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1823205</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17_-_1825231_1827860</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>NA12273</td>\n",
       "      <td>Start</td>\n",
       "      <td>1825231</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand                     sj_id start_site_novelty  \\\n",
       "0  GL000195.1      -  GL000195.1_-_49120_86517              novel   \n",
       "1       chr17      -   chr17_-_1821508_1827860              novel   \n",
       "2       chr17      -   chr17_-_1821512_1827860              novel   \n",
       "3       chr17      -   chr17_-_1823205_1827860              novel   \n",
       "4       chr17      -   chr17_-_1825231_1827860              novel   \n",
       "\n",
       "  end_site_category cell_line_id variable    value sj_loc  \n",
       "0             known      NA12273    Start    49120         \n",
       "1             known      NA12273    Start  1821508         \n",
       "2             known      NA12273    Start  1821512         \n",
       "3             known      NA12273    Start  1823205         \n",
       "4             known      NA12273    Start  1825231         "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e1510243-145c-4284-90df-6c4e19a075f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # todo -- restricting just to one cell line so it's easier to viz.\n",
    "# print(len(temp.index))\n",
    "# temp = temp.loc[temp.cell_line_id=='HG03732']\n",
    "# print(len(temp.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "214f780f-9a94-4d29-afee-7e847d188305",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>263565</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_2854705_2866792</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>2854705</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263566</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY_-_3185396_3187827</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>3185396</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263567</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY_-_3187916_3188223</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>3187916</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263568</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY_-_3188385_3219255</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>3188385</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263569</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY_-_13477439_13479253</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>13477439</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Strand                     sj_id start_site_novelty  \\\n",
       "263565       chrY      +    chrY_+_2854705_2866792              novel   \n",
       "263566       chrY      -    chrY_-_3185396_3187827              novel   \n",
       "263567       chrY      -    chrY_-_3187916_3188223              novel   \n",
       "263568       chrY      -    chrY_-_3188385_3219255              novel   \n",
       "263569       chrY      -  chrY_-_13477439_13479253              novel   \n",
       "\n",
       "       end_site_category cell_line_id variable     value sj_loc  \\\n",
       "263565             known      HG03732    Start   2854705  start   \n",
       "263566             novel      HG03732    Start   3185396  start   \n",
       "263567             novel      HG03732    Start   3187916  start   \n",
       "263568             novel      HG03732    Start   3188385  start   \n",
       "263569             known      HG03732    Start  13477439  start   \n",
       "\n",
       "       novelty_category  \n",
       "263565            novel  \n",
       "263566            novel  \n",
       "263567            novel  \n",
       "263568            novel  \n",
       "263569            novel  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# temp.drop('sj_id', axis=1, inplace=True)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a8b44cc8-3ce3-4ef8-867f-63f8e2127c1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5f3cc9c0-c132-4538-b971-580a171500b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "dbbd902a-dd99-4cec-917d-ee2381fdae6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to verify\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d1770e76-cc6e-4990-9a79-f97e6d7977b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pr.PyRanges(temp)\n",
    "temp.to_bed('test_novel_ss.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "a910fe15-798e-4108-8532-c6b0b8d95f8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10067</th>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>chrX_+_56798818_56817319</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>56798816</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>56798817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10278</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_2854705_2866792</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>2854703</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>2854704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10279</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_19567405_19587209</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>19567403</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>19567404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10280</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_19567993_19587209</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>19567991</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>19567992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10281</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY_+_20576136_20579607</td>\n",
       "      <td>novel</td>\n",
       "      <td>known</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>Start</td>\n",
       "      <td>20576134</td>\n",
       "      <td>start</td>\n",
       "      <td>novel</td>\n",
       "      <td>20576135</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome Strand                     sj_id start_site_novelty  \\\n",
       "10067       chrX      +  chrX_+_56798818_56817319              novel   \n",
       "10278       chrY      +    chrY_+_2854705_2866792              novel   \n",
       "10279       chrY      +  chrY_+_19567405_19587209              novel   \n",
       "10280       chrY      +  chrY_+_19567993_19587209              novel   \n",
       "10281       chrY      +  chrY_+_20576136_20579607              novel   \n",
       "\n",
       "      end_site_category cell_line_id variable     Start sj_loc  \\\n",
       "10067             known      HG03732    Start  56798816  start   \n",
       "10278             known      HG03732    Start   2854703  start   \n",
       "10279             known      HG03732    Start  19567403  start   \n",
       "10280             known      HG03732    Start  19567991  start   \n",
       "10281             known      HG03732    Start  20576134  start   \n",
       "\n",
       "      novelty_category       End  \n",
       "10067            novel  56798817  \n",
       "10278            novel   2854704  \n",
       "10279            novel  19567404  \n",
       "10280            novel  19567992  \n",
       "10281            novel  20576135  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# temp = temp.df\n",
    "temp.loc[(temp.sj_loc=='start')&(temp.Strand=='+')].tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93befaa-25fe-4111-9b7c-86c2fbe435bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "79fb97c7-e02d-4689-92a5-5e43e4b71a7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>sj_id</th>\n",
       "      <th>start_site_novelty</th>\n",
       "      <th>end_site_category</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>Start</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>novelty_category</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Strand, sj_id, start_site_novelty, end_site_category, cell_line_id, variable, Start, sj_loc, novelty_category, End]\n",
       "Index: []"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = temp.df\n",
    "temp.loc[temp.End==56787072]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eafb303-4b31-4cd1-b449-d9b38cccb019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get annotated (v47) sss\n",
    "gtf_file = proc_cfg(config['ref']['gtf'], od)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
