{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1d368bac-6b3d-4020-b807-674ef25fc14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import cerberus\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.vcf_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8ef183d-78f8-47da-a35a-3ce434de5400",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../'\n",
    "\n",
    "def proc_cfg(entry, od):\n",
    "    entry = entry.replace('../../', '')\n",
    "    entry = od+entry\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3a631d6-594d-467d-a9e1-02cf8ab5ba82",
   "metadata": {},
   "source": [
    "## We want to get the dist. of distances between novel SSs and annotated SSs.\n",
    "\n",
    "Hopefully they are generally close\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3613df03-8676-4709-868b-acade41722aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_meta()\n",
    "df = df.loc[~df['sample'].str.contains('_')]\n",
    "df['lab_sample'] = df['lab_number_sample'].astype(str)+'_'+\\\n",
    "                      df['lab_sampleid'].astype(str)+'_'+\\\n",
    "                      df['cell_line_id'].astype(str)\n",
    "df.columns\n",
    "df = df[['cell_line_id', 'sample', 'hapmap_DNA_ID',\n",
    "          'map_reads_assemblymap','lab_sample', 'population']].drop_duplicates()\n",
    "\n",
    "temp_df = pd.read_csv('../snakemake/transcript_discovery_personal/cell_line_ids.txt', header=None, names=['cell_line_id'])\n",
    "\n",
    "# make a 1000g cell line id col\n",
    "df['cell_line_id_1000g'] = df.cell_line_id\n",
    "\n",
    "inds = df.loc[~df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())].index\n",
    "df.loc[inds, 'cell_line_id_1000g'] = df.loc[inds, 'hapmap_DNA_ID']\n",
    "len(df.index)\n",
    "\n",
    "# limit to just those in 1000g\n",
    "df = df.loc[df.cell_line_id_1000g.isin(temp_df.cell_line_id.tolist())]\n",
    "assert len(df.index) == 30\n",
    "\n",
    "# TODO bad sample that hasn't finished on espresso\n",
    "# bad_samples = ['NA19328']\n",
    "# df = df.loc[~df.cell_line_id_1000g.isin(bad_samples)]\n",
    "\n",
    "hap = ['hap1', 'hap2']\n",
    "\n",
    "meta_df = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5e8acd8-f10c-459b-83cc-d6aba2ff6243",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df['detected'] = True\n",
    "\n",
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category', 'cell_line_id']].drop_duplicates()\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "# nov_ss_df = temp.copy(deep=True)\n",
    "# nov_ss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "432b7cd3-d9bf-438c-94e6-4cf77f8b5315",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_ss_df = temp.copy(deep=True)\n",
    "nov_ss_df = pr.PyRanges(nov_ss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d1770e76-cc6e-4990-9a79-f97e6d7977b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_ss_df = nov_ss_df[['Chromosome', 'Strand', ' Start', 'End']]\n",
    "nov_ss_df.to_bed('person_hap_nov_ss.bed')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ea80ac7-2b41-4470-ac47-8d8c4478bd15",
   "metadata": {},
   "source": [
    "## Get annotated (v47) SSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eafb303-4b31-4cd1-b449-d9b38cccb019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get annotated (v47) sss\n",
    "gtf_file = proc_cfg(config['ref']['gtf'], od)\n",
    "\n",
    "# read in ref. gtf and get SJs from there\n",
    "temp = pr.read_gtf(gtf_file)\n",
    "\n",
    "# get the introns\n",
    "temp = temp.features.introns()\n",
    "temp = temp.df\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# need to verify\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']+1\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+1\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "\n",
    "temp = pr.PyRanges(temp)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f27ac058-1660-496d-86a3-d1ae6ca4903f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.to_bed('test_v47_ss.bed')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0682cd7b-c711-4f78-a28f-271f036033df",
   "metadata": {},
   "source": [
    "## Try specifically subsetting on the SJs that are exclusively found in hap1/hap2?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "67cae993-4cb9-4f15-81a5-1ef8a00e6c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = proc_cfg(config['lr']['td_personal']['sqanti']['sj_summary'],od)\n",
    "df = pd.read_csv(file)\n",
    "df.drop('Unnamed: 0', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "df00e916-a4ea-4b97-b7fc-c1309afdad68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24934518\n",
      "8311506\n"
     ]
    }
   ],
   "source": [
    "# dedupe w/ sqanti genome cause no one cares\n",
    "df = df.drop(['sqanti_genome', 'splice_motif', 'canonical'], axis=1)\n",
    "print(len(df.index))\n",
    "df = df.drop_duplicates()\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a77b22bd-67a4-4846-a11f-52fdf59c35cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # try just w/ one cell line for now\n",
    "# df = df.loc[df.cell_line_id=='HG03732']\n",
    "# df_back = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bee826a1-fa6f-40d6-8817-d9c88976f8b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = df_back.copy(deep=True)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ce9c4ebf-e094-42ee-bf16-e8bac1539ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. filter for sjs w/ at least one novel ss \n",
    "df = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "30b5d18d-e7b1-4a5f-9872-1f50147029a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# melt to 5' and 3'\n",
    "df[['Chromosome', 'Strand', 'Start', 'End']] = df.sj_id.str.split('_', expand=True)\n",
    "\n",
    "df = df.melt(id_vars=['Chromosome', 'Strand', 'map_genome',\n",
    "                      'start_site_novelty', 'end_site_category', 'cell_line_id'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "df['sj_loc'] = ''\n",
    "df['detected'] = True\n",
    "\n",
    "\n",
    "df.loc[df.variable=='Start', 'sj_loc'] = 'start'\n",
    "df.loc[df.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "df['novelty_category'] = ''\n",
    "df.loc[df.variable=='Start', 'novelty_category'] = df.loc[df.variable=='Start', 'start_site_novelty']\n",
    "df.loc[df.variable=='End', 'novelty_category'] = df.loc[df.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "df = df.loc[df.novelty_category=='novel']\n",
    "\n",
    "df.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "\n",
    "# create id for each ss\n",
    "df['ss_id'] = 'ss_'+\\\n",
    "                df.sj_loc+'_'+\\\n",
    "                df.Chromosome+'_'+\\\n",
    "                df.Strand+'_'+\\\n",
    "                df.Start.astype(str)\n",
    "\n",
    "df = df[['ss_id', 'cell_line_id', 'map_genome', 'detected']]\n",
    "\n",
    "# check for sss that originated from different sjs\n",
    "assert len(df.loc[df.ss_id.duplicated()]) > 0\n",
    "\n",
    "# then dedupe and pivot\n",
    "df = df.drop_duplicates()\n",
    "df = df.pivot(index=['ss_id', 'cell_line_id'], \n",
    "                columns=['map_genome'],\n",
    "                values=['detected'])\n",
    "\n",
    "# flatten\n",
    "df.columns = df.columns.get_level_values(1)\n",
    "df.columns.name = None\n",
    "\n",
    "# reset index to make it a flat DataFrame\n",
    "df = df.reset_index()\n",
    "\n",
    "# fill missing values with False\n",
    "df = df.fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9405c08e-9439-4e00-a0ad-42d06092bd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. get those that are detected uniquely in hap1/2 \n",
    "# (ie those not in hg38)\n",
    "df = df.loc[(df.hg38==False)&\n",
    "                ((df.hap1+df.hap2)>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "18331d05-cdc2-4246-9b65-ef4277977f1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ss_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ss_end_KI270726.1_+_26229</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>KI270726.1</td>\n",
       "      <td>+</td>\n",
       "      <td>26229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ss_end_chr10_+_101365093</td>\n",
       "      <td>NA12778</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101365093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>NA12778</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ss_id cell_line_id   hap1   hap2   hg38 sj_loc  \\\n",
       "92   ss_end_KI270726.1_+_26229      HG02261   True  False  False    end   \n",
       "128   ss_end_chr10_+_101365093      NA12778   True  False  False    end   \n",
       "129   ss_end_chr10_+_101374347      HG03729   True  False  False    end   \n",
       "130   ss_end_chr10_+_101374347      HG03732  False   True  False    end   \n",
       "131   ss_end_chr10_+_101374347      NA12778   True  False  False    end   \n",
       "\n",
       "     Chromosome Strand      Start  \n",
       "92   KI270726.1      +      26229  \n",
       "128       chr10      +  101365093  \n",
       "129       chr10      +  101374347  \n",
       "130       chr10      +  101374347  \n",
       "131       chr10      +  101374347  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['nothing', 'sj_loc', 'Chromosome', 'Strand', 'Start']] = df.ss_id.str.split('_', expand=True)\n",
    "df.drop('nothing', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "26952d93-69e9-4f4b-ad0e-cb846cdf72c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ss_id</th>\n",
       "      <th>cell_line_id</th>\n",
       "      <th>hap1</th>\n",
       "      <th>hap2</th>\n",
       "      <th>hg38</th>\n",
       "      <th>sj_loc</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ss_end_KI270726.1_+_26229</td>\n",
       "      <td>HG02261</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>KI270726.1</td>\n",
       "      <td>+</td>\n",
       "      <td>26229</td>\n",
       "      <td>26230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ss_end_chr10_+_101365093</td>\n",
       "      <td>NA12778</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101365093</td>\n",
       "      <td>101365094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>HG03729</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "      <td>101374348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>HG03732</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "      <td>101374348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>ss_end_chr10_+_101374347</td>\n",
       "      <td>NA12778</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>101374347</td>\n",
       "      <td>101374348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ss_id cell_line_id   hap1   hap2   hg38 sj_loc  \\\n",
       "92   ss_end_KI270726.1_+_26229      HG02261   True  False  False    end   \n",
       "128   ss_end_chr10_+_101365093      NA12778   True  False  False    end   \n",
       "129   ss_end_chr10_+_101374347      HG03729   True  False  False    end   \n",
       "130   ss_end_chr10_+_101374347      HG03732  False   True  False    end   \n",
       "131   ss_end_chr10_+_101374347      NA12778   True  False  False    end   \n",
       "\n",
       "     Chromosome Strand      Start        End  \n",
       "92   KI270726.1      +      26229      26230  \n",
       "128       chr10      +  101365093  101365094  \n",
       "129       chr10      +  101374347  101374348  \n",
       "130       chr10      +  101374347  101374348  \n",
       "131       chr10      +  101374347  101374348  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "df['Start'] = df.Start.astype(int)-2\n",
    "df['End'] = df.Start.astype(int)+1\n",
    "\n",
    "# verified\n",
    "df.loc[df.sj_loc=='start', 'Start'] = df.loc[df.sj_loc=='start', 'Start']-0\n",
    "df.loc[df.sj_loc=='start', 'End'] = df.loc[df.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "df.loc[df.sj_loc=='end', 'End'] = df.loc[df.sj_loc=='end', 'End']+2\n",
    "df.loc[df.sj_loc=='end', 'Start'] = df.loc[df.sj_loc=='end', 'Start']+2\n",
    "\n",
    "nov_ss_df = df.copy(deep=True)\n",
    "nov_ss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "eb4e5573-a88d-4874-91a0-5ed97a5ee17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_ss_df = pr.PyRanges(nov_ss_df)\n",
    "nov_ss_df.to_bed('person_hap_nov_ss_hg38_absent.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02baa556-0f47-45dc-8532-039639234539",
   "metadata": {},
   "source": [
    "## Ok now try w/ PODER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "cfb4e06b-e277-4476-9354-2a3c819f9e14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28544\n",
      "23244\n"
     ]
    }
   ],
   "source": [
    "file = proc_cfg(config['lr']['sjs'],od)\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "# df.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "df.rename({'start_site_category':'start_site_novelty'}, axis=1, inplace=True)\n",
    "df['detected'] = True\n",
    "\n",
    "df['sj_id'] = df['chrom'].astype(str) + '_' +\\\n",
    "              df['strand'].astype(str) + '_' +\\\n",
    "              df['genomic_start_coord'].astype(str) + '_' +\\\n",
    "              df['genomic_end_coord'].astype(str)\n",
    "\n",
    "# 1.5 only novel sss\n",
    "temp = df.loc[(df.start_site_novelty=='novel')|(df.end_site_category=='novel')]\n",
    "\n",
    "# only keep unique novel sjs\n",
    "temp[['Chromosome', 'Strand', 'Start', 'End']] = temp.sj_id.str.split('_', expand=True)\n",
    "print(len(temp.index))\n",
    "temp = temp[['Chromosome', 'Strand', 'Start', 'End', 'sj_id', \n",
    "             'start_site_novelty', 'end_site_category']].drop_duplicates()\n",
    "print(len(temp.index))\n",
    "\n",
    "temp.Start = temp.Start.astype(int)\n",
    "temp.End = temp.End.astype(int)\n",
    "assert len(temp.loc[temp.Start>temp.End])==0\n",
    "\n",
    "# melt to 5' and 3'\n",
    "temp = temp.melt(id_vars=['Chromosome', 'Strand', 'sj_id',\n",
    "                          'start_site_novelty', 'end_site_category'],\n",
    "                 value_vars=['Start', 'End'])\n",
    "temp['sj_loc'] = ''\n",
    "\n",
    "\n",
    "temp.loc[temp.variable=='Start', 'sj_loc'] = 'start'\n",
    "temp.loc[temp.variable=='End', 'sj_loc'] = 'end'\n",
    "\n",
    "temp['novelty_category'] = ''\n",
    "temp.loc[temp.variable=='Start', 'novelty_category'] = temp.loc[temp.variable=='Start', 'start_site_novelty']\n",
    "temp.loc[temp.variable=='End', 'novelty_category'] = temp.loc[temp.variable=='End', 'end_site_category']\n",
    "\n",
    "# only get the novel ones now\n",
    "temp = temp.loc[temp.novelty_category=='novel']\n",
    "\n",
    "temp.rename({'value':'Start'}, axis=1, inplace=True)\n",
    "# need to verify that this is working using like one motif or something make\n",
    "# sure I don't have off-by-one errors\n",
    "# verified\n",
    "temp['Start'] = temp.Start-2\n",
    "temp['End'] = temp.Start+1\n",
    "\n",
    "# verified\n",
    "temp.loc[temp.sj_loc=='start', 'Start'] = temp.loc[temp.sj_loc=='start', 'Start']-0\n",
    "temp.loc[temp.sj_loc=='start', 'End'] = temp.loc[temp.sj_loc=='start', 'End']+0\n",
    "\n",
    "\n",
    "temp.loc[temp.sj_loc=='end', 'End'] = temp.loc[temp.sj_loc=='end', 'End']+2\n",
    "temp.loc[temp.sj_loc=='end', 'Start'] = temp.loc[temp.sj_loc=='end', 'Start']+2\n",
    "\n",
    "nov_ss_df = temp.copy(deep=True)\n",
    "nov_ss_df = pr.PyRanges(nov_ss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "id": "957dc9b9-6532-4104-85a0-c49dc47059c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_ss_df.to_bed('poder_nov_ss.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415ca64e-3d50-4880-b100-f0d88caf2322",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
