{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3615cae1-5b4a-491d-803b-488041a5cf8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['transcript_id', 'gene_id', 'samples', 'orfanage_status', 'orfanage_duplicity', 'orfanage_template', 'orfanage_template_source']\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "# make GTF file that includes the ORF regions (as CDS features)\n",
    "# input - pacbio gtf ('jurkat.collapsed.gtf'), orf calls ('jurkat_refine_orf_calls.tsv')\n",
    "# output - pacbio gtf with added \"cds\" features (orfs)\n",
    "\n",
    "import argparse\n",
    "import copy\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "\n",
    "import gtfparse\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def string_to_boolean(string):\n",
    "    \"\"\"\n",
    "    Converts string to boolean\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    string :str\n",
    "    input string\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    result : bool\n",
    "    output boolean\n",
    "    \"\"\"\n",
    "    if isinstance(string, bool):\n",
    "        return str\n",
    "    if string.lower() in (\"yes\", \"true\", \"t\", \"y\", \"1\"):\n",
    "        return True\n",
    "    elif string.lower() in (\"no\", \"false\", \"f\", \"n\", \"0\"):\n",
    "        return False\n",
    "    else:\n",
    "        raise argparse.ArgumentTypeError(\"Boolean value expected.\")\n",
    "\n",
    "\n",
    "def get_first_block_index(orf_coord, cblens, pblens):\n",
    "    # get the index corresponding to the first block containing the orf start\n",
    "    # return index, and the dela (spacing upstream of end)\n",
    "    for i, cblen in enumerate(cblens):\n",
    "        if orf_coord <= cblen:\n",
    "            delta = cblen - orf_coord\n",
    "            return i, delta\n",
    "    return i, 0\n",
    "\n",
    "\n",
    "def make_cds_coords_positive_strand(i1, delta1, i2, delta2, coords):\n",
    "    orf_coords = copy.deepcopy(coords)\n",
    "    orf_coords = orf_coords[i1 : i2 + 1]\n",
    "    # trim ends to orf start/end\n",
    "    orf_coords[0][0] = orf_coords[0][1] - delta1\n",
    "    orf_coords[-1][1] = orf_coords[-1][1] - delta2\n",
    "    return orf_coords\n",
    "\n",
    "\n",
    "def make_cds_coords_negative_strand(i1, delta1, i2, delta2, coords):\n",
    "    orf_coords = copy.deepcopy(coords)\n",
    "    orf_coords = orf_coords[i1 : i2 + 1]\n",
    "    # trim ends to orf start/end\n",
    "    orf_coords[0][1] = orf_coords[0][0] + delta1\n",
    "    orf_coords[-1][0] = orf_coords[-1][0] + delta2\n",
    "    return orf_coords\n",
    "\n",
    "\n",
    "def get_min_and_max_coords_from_exon_chain(coords):\n",
    "    \"\"\"Gets the minumum and maximum coordinates from exon chain\n",
    "\n",
    "    Args:\n",
    "        coords ([[int, int]]): [[start,end]] exon coordinate chain\n",
    "\n",
    "    Returns:\n",
    "        (int, int): start and end coordinates of entire chain\n",
    "    \"\"\"\n",
    "\n",
    "    min_coord = min(map(min, coords))\n",
    "    max_coord = max(map(max, coords))\n",
    "    return min_coord, max_coord\n",
    "\n",
    "\n",
    "def make_pacbio_cds_gtf(sample_gtf, called_orfs, name):\n",
    "    \"\"\"Makes PacBio CDS and saves file with CDS\n",
    "\n",
    "    Args:\n",
    "        sample_gtf (filename): sample_gtf file\n",
    "        refined_orfs (filename): aggregate_orf info. from Refined DB\n",
    "        called_orfs (filename): orf calls from ORF_Calling\n",
    "        pb_gene (filename): PacBio gene name cross reference\n",
    "        name (string): name of sample\n",
    "    \"\"\"\n",
    "    # import gtf, only exon info.\n",
    "    # only move forward with representative pb isoform (for same-protein groups)\n",
    "    gtf = gtfparse.read_gtf(sample_gtf)\n",
    "    # gtf = gtf.to_pandas() # this needs to stay on for cluster version\n",
    "    gtf_gene_mapping = gtf.loc[gtf[\"feature\"] == \"transcript\"]\n",
    "    gtf_gene_mapping = gtf_gene_mapping[[\"transcript_id\", \"gene_id\"]]\n",
    "    gtf = gtf[\n",
    "        [\"seqname\", \"feature\", \"start\", \"end\", \"strand\", \"transcript_id\"]\n",
    "    ]\n",
    "    gtf = gtf.loc[gtf[\"feature\"] == \"exon\"]\n",
    "    gtf.columns = [\"chr\", \"feat\", \"start\", \"end\", \"strand\", \"acc\"]\n",
    "    # only move forward with \"base accession\" (representative pb)\n",
    "    # pb coords into dict\n",
    "    pbs = defaultdict(\n",
    "        lambda: [\"chr\", \"strand\", [], [], [], []]\n",
    "    )  # pb -> [chr, strand, [start, end], [block lengths],[cum. block lengths], [prior cumulative block lengths]]\n",
    "    # PB.1.1 -> ['chr1', '+', [[100,150], [200,270]], [50, 70], [50, 120], [150-200]]\n",
    "    for i, row in gtf.iterrows():\n",
    "        chr, feat, start, end, strand, acc = row\n",
    "        pbs[acc][0] = chr\n",
    "        pbs[acc][1] = strand\n",
    "        pbs[acc][2].append([int(start), int(end)])\n",
    "    # import pdb; pdb.set_trace()\n",
    "    # sort all coords, calc blocks\n",
    "    for acc, infos in pbs.items():\n",
    "        strand = infos[1]\n",
    "        if strand == \"+\":\n",
    "            infos[2] = sorted(infos[2])\n",
    "        elif strand == \"-\":\n",
    "            infos[2] = sorted(infos[2], reverse=True)\n",
    "        infos[3] = np.array([end - start + 1 for [start, end] in infos[2]])\n",
    "        infos[4] = np.cumsum(infos[3])\n",
    "        infos[5] = infos[4] - infos[3]\n",
    "    # import pdb; pdb.set_trace()\n",
    "\n",
    "    # read in the ranges of orf on pb transcripts\n",
    "    ranges = pd.read_table(called_orfs)[\n",
    "        [\"transcript_id\", \"ORF_start\", \"ORF_end\"]\n",
    "    ]\n",
    "    # print(gtf_gene_mapping)\n",
    "    # print(gtf_gene_mapping.gene_id)\n",
    "    # pb_gene = pd.Series(\n",
    "    #    gtf_gene_mapping[\"gene_id\"], index=gtf_gene_mapping[\"transcript_id\"]\n",
    "    # ).to_dict()\n",
    "    pb_gene = dict(\n",
    "        zip(gtf_gene_mapping[\"transcript_id\"], gtf_gene_mapping[\"gene_id\"])\n",
    "    )\n",
    "    with open(f\"{name}_cpat_with_cds.gtf\", \"w\") as ofile:\n",
    "        for i, row in ranges.iterrows():\n",
    "            acc, orf_start, orf_end = row\n",
    "            acc = acc.lower()\n",
    "            # remove stop exon\n",
    "            orf_end = orf_end - 3\n",
    "            if acc in pbs:\n",
    "                \n",
    "                if acc in pb_gene.keys():\n",
    "                    gene = pb_gene[acc]\n",
    "                else:\n",
    "                    raise ValueError\n",
    "                    gene = \"-\"\n",
    "                infos = pbs[acc]\n",
    "                chr, strand, coords, blens, cblens, pblens = infos\n",
    "\n",
    "                i1, delta1 = get_first_block_index(orf_start, cblens, pblens)\n",
    "                i2, delta2 = get_first_block_index(orf_end, cblens, pblens)\n",
    "                if strand == \"+\":\n",
    "                    orf_coords = make_cds_coords_positive_strand(\n",
    "                        i1, delta1, i2, delta2, coords\n",
    "                    )\n",
    "                elif strand == \"-\":\n",
    "                    orf_coords = make_cds_coords_negative_strand(\n",
    "                        i1, delta1, i2, delta2, coords\n",
    "                    )\n",
    "                # write out the coordinates\n",
    "                # out_acc = f'transcript_id \"{acc}\";'\n",
    "                # acc_w_gene_w_cpm = gene + \"|\" + acc\n",
    "                out_acc = f'transcript_id \"{acc}\"; gene_id \"{gene}\";'\n",
    "\n",
    "                out_acc_exon = f'transcript_id \"{acc}\";'\n",
    "\n",
    "                tstart, tend = get_min_and_max_coords_from_exon_chain(coords)\n",
    "                ofile.write(\n",
    "                    \"\\t\".join(\n",
    "                        [\n",
    "                            chr,\n",
    "                            \"hg38_canon\",\n",
    "                            \"transcript\",\n",
    "                            str(tstart),\n",
    "                            str(tend),\n",
    "                            \".\",\n",
    "                            strand,\n",
    "                            \".\",\n",
    "                            out_acc,\n",
    "                        ]\n",
    "                    )\n",
    "                    + \"\\n\"\n",
    "                )\n",
    "                for [start, end] in coords:\n",
    "                    ofile.write(\n",
    "                        \"\\t\".join(\n",
    "                            [\n",
    "                                chr,\n",
    "                                \"hg38_canon\",\n",
    "                                \"exon\",\n",
    "                                str(start),\n",
    "                                str(end),\n",
    "                                \".\",\n",
    "                                strand,\n",
    "                                \".\",\n",
    "                                out_acc_exon,\n",
    "                            ]\n",
    "                        )\n",
    "                        + \"\\n\"\n",
    "                    )\n",
    "                for [start, end] in orf_coords:\n",
    "                    ofile.write(\n",
    "                        \"\\t\".join(\n",
    "                            [\n",
    "                                chr,\n",
    "                                \"CPAT\",\n",
    "                                \"CDS\",\n",
    "                                str(start),\n",
    "                                str(end),\n",
    "                                \".\",\n",
    "                                strand,\n",
    "                                \".\",\n",
    "                                out_acc_exon,\n",
    "                            ]\n",
    "                        )\n",
    "                        + \"\\n\"\n",
    "                    )\n",
    "\n",
    "\n",
    "def main():\n",
    "#     parser = argparse.ArgumentParser(\n",
    "#         \"IO file locations for make pacbio cds gtf\"\n",
    "#     )\n",
    "#     parser.add_argument(\n",
    "#         \"--name\",\n",
    "#         action=\"store\",\n",
    "#         dest=\"name\",\n",
    "#         help=\"name of sample - used for output file name\",\n",
    "#     )\n",
    "#     parser.add_argument(\n",
    "#         \"--sample_gtf\",\n",
    "#         action=\"store\",\n",
    "#         dest=\"sample_gtf\",\n",
    "#         help=\"sample gtf, from sqanti3\",\n",
    "#     )\n",
    "#     parser.add_argument(\n",
    "#         \"--called_orfs\",\n",
    "#         action=\"store\",\n",
    "#         dest=\"called_orfs\",\n",
    "#         help=\"agg orf tsv file, from refined DB\",\n",
    "#     )\n",
    "\n",
    "#     results = parser.parse_args()\n",
    "    \n",
    "    warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "    make_pacbio_cds_gtf(\n",
    "        '../../data/protein/transcripts_cpat_cds_to_be_predicted.gtf',\n",
    "        '../../data/protein/transcripts.ORF_remaining.tsv',\n",
    "        'test',\n",
    "    )\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d97df33-39bb-46db-b1ce-f91d85e7f3f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd763042-7708-40ef-8dfc-f6c53bfb8a19",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
